---
title: "Xenium Analysis"
author: "Ryan Sheridan"
date: "`r Sys.Date()`"
output:
  html_document:
    toc:            true
    toc_float:      true
    toc_depth:      2
    df_print:       "paged"
    code_folding:   "hide"
    self_contained: true
params:
  res_dir:     "results/xenium"
  xen_dir:     "20240719__224527__Tamburini_Run_2_071924"
  so_dir:      "~/Dropbox/Ryan/Projects/antigen-exchange/results/xenium"
  ref_dir:     "ref"
  clst_rsln:   10
  slide_regex: "output-XETG00230__[0-9]{7}__(Group_[A-Z]-[0-9]|Mock_[0-9])__20240719__224653"
  samples:
    value:
      Group_A_5: "CHIKV WT"
      Group_B_6: "CHIKV WT"
      Group_A_7: "CHIKV K200R"
      Group_B_8: "CHIKV K200R"
      Group_A_3: "CHIKV 181/25"
      Group_B_4: "CHIKV 181/25"
      Group_D_4: "CHIKV WT + Ag"
      Group_D_5: "CHIKV WT + Ag"
      Group_D_6: "CHIKV WT + Ag"
      Group_C_1: "CHIKV 181/25 + Ag"
      Group_C_2: "CHIKV 181/25 + Ag"
      Group_C_3: "CHIKV 181/25 + Ag"
  ag_targets:
    value:
      Ag_tag: "GeoMx-Barcode01-ERCC00142"
  chikv_targets:
    value:
      CHIKV_sgRNA: "CHIKV-sgRNA-(7566-12036)"
      CHIKV_5:     "CHIKV-5-(0-7566)"
      CHIKV_neg:   "CHIKV-negative-(0-12036)"
  custom_targets:
    - "Ag-tag-A-(ERCC2)"
    - "Ag-tag-B-(ERCC3)"
    - "Ag-tag-C-(ERCC4)"
    - "T2-Sensitive-Ag-tag-(ERCC9)"
    - "T2-Resistant-Ag-tag-(ERCC12)"
    - "DNA-only-T2-Control-Ag-tag-(ERCC16)"
    - "GeoMx-Barcode01-ERCC00142"
    - "GeoMx-Barcode02-ERCC00144"
    - "CHIKV-5-(0-7566)"
    - "CHIKV-sgRNA-(7566-12036)"
    - "CHIKV-negative-(0-12036)"
    - "LCMV-(Armstrong)-Segment"
    - "Listeria-monocytogenes-10403S-gyrase-B-CP002002-1"
    - "Vaccinia-(Western-Reserve)-J6R"
    - "mRNA-vaccine-(3-UTR)"
    - "mRNA-vaccine-(ovalbumin)"
    - "mRNA-vaccine-(FLAG-3x)"
    - "mRNA-vaccine-(eGFP)"
editor_options: 
  chunk_output_type: inline
---

<br>

```{r "setup", include = FALSE}
# Default chunk options
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo    = FALSE,
  dpi     = 250
)

# Packages
library(Seurat)
library(SeuratObject)
library(tidyverse)
library(djvdj)
library(here)
library(qs)
library(clustifyr)
library(clustifyrdata)
library(patchwork)
library(cowplot)
library(scales)
library(colorblindr)
library(ggtrace)
library(sf)
library(sp)
library(reticulate)
library(broom)

use_condaenv("umap")

source(here("src/funs.R"))

sffx <- ".qs"

create_object <- !file.exists(here(params$so_dir, str_c("xen", sffx)))
```

```{r "functions"}
.plot_full_fov <- function(df_in, fov_in, color, trace = list(), clrs,
                           size = 0.3, trace_size = 1, trace_clr = "black") {
  # Format input data
  plt_theme <- umap_theme +
    theme(
      legend.title = element_blank(),
      legend.text  = element_text(size = 28)
    )
  
  dat <- df_in %>%
    filter(fov == fov_in)
  
  n_dat <- dat %>%
    group_by(!!sym(color)) %>%
    summarize(n = n(), .groups = "drop") %>%
    mutate(
      n_lab = str_c(!!sym(color), "\nn = ", label_comma()(n))
    )
  
  # Plot highlighted cells
  res <- trace %>%
    map(~ {
      n_dat <- n_dat %>%
        arrange(desc(!!sym(color) %in% .x), desc(n))
      
      n_labs <- set_names(n_dat$n_lab, n_dat[[color]])
      
      dat %>%
        mutate(!!sym(color) := fct_relevel(!!sym(color), names(n_labs))) %>%
        arrange(!!sym(color)) %>%
        ggplot(aes(y_cell, x_cell, fill = !!sym(color))) +
        geom_point_trace(
          trace_position    = !!sym(color) %in% .x,
          background_params = list(size = size, color = NA),
          size  = trace_size,
          color = trace_clr
        ) +
        scale_fill_manual(values = clrs, labels = n_labs) +
        guides(fill = guide_legend(override.aes = list(size = 10, color = NA))) +
        coord_fixed() +
        plt_theme
    })
  
  if (length(trace) == 1) res <- res[[1]]
  
  res
}

.process_xenium <- function(obj, dims = 1:40, resolution = c(1, 3, 5, 10)) {

  res <- obj %>%
    SCTransform(assay = "Xenium") %>%
    RunPCA(
      assay    = "SCT",
      npcs     = 50,
      features = rownames(.)
    ) %>%
    RunUMAP(
      assay       = "SCT",
      reduction   = "pca",
      dims        = dims,
      umap.method = "umap-learn"
    )
  
  res <- res %>%
    FindNeighbors(
      assay     = "SCT",
      reduction = "pca",
      dims      = dims
    ) %>%
    FindClusters(resolution = resolution)
  
  u_coords <- res %>%
    FetchData(c("UMAP_1", "UMAP_2"))
  
  res <- res %>%
    AddMetaData(u_coords)
  
  res
}

.classify_t_b <- function(obj, clmn) {
  res <- obj %>%
    classify_markers(
      Ptprc > 0.7,
      feats      = "Ptprc",
      type_label = "other CD45+",
      clst_col   = clmn,
      type_col   = "cell_type"
    ) %>%
    classify_markers(
      cell_type == "other CD45+" & Cd3d > 0.45,
      feats      = c("cell_type", "Cd3d"),
      type_label = "T cells",
      clst_col   = clmn,
      type_col   = "cell_type"
    ) %>%
    classify_markers(
      cell_type == "other CD45+" & Cd19 > 1.1,
      feats      = c("cell_type", "Cd19"),
      type_label = "B cells",
      clst_col   = clmn,
      type_col   = "cell_type"
    )
  
  res
}

.classify_dcs <- function(obj, clmn) {
  res <- obj %>%
    classify_markers(
      cell_type == "other CD45+" & Itgax > 0.15,
      feats      = c("cell_type", "Itgax"),
      type_label = "other DC",
      clst_col   = clmn,
      type_col   = "cell_type"
    ) %>% 
    classify_markers(
      cell_type == "other DC" & Sirpa > 0.15 & Itgam > 0.15,
      feats      = c("cell_type", "Sirpa", "Itgam"),
      type_label = "cDC2",
      clst_col   = clmn,
      type_col   = "cell_type"
    ) %>%
    classify_markers(
      cell_type == "other DC" & Xcr1 > 0.15,
      feats      = c("cell_type", "Xcr1"),
      type_label = "cDC1",
      clst_col   = clmn,
      type_col   = "cell_type"
    ) %>%
    classify_markers(
      cell_type %in% c("other DC", "other CD45+") & Siglech > 0.5,
      feats      = c("cell_type", "Siglech"),
      type_label = "Siglec-H DC",
      clst_col   = clmn,
      type_col   = "cell_type"
    ) 
  
  res
}

# Fraction of cells within each cell type that belong to a certain class
# e.g. Ag+ or CHIKV+
.create_class_figure <- function(bar_dat, plot_dat, class_clmn, class,
                                 celltype_clmn, clrs, top_types = NULL,
                                 n_cells = 30) {
  
  # FOVs to plot
  fovs_to_plot <- levels(bar_dat$fov) %||% unique(bar_dar$fov)
  
  # Set column to color cells by
  # based on cell class, e.g. Ag-low/-high
  other_class <- unique(bar_dat[[class_clmn]])
  other_class <- other_class[other_class != class]
  other_class <- str_c(other_class, collapse = "/")
  
  bar_dat <- bar_dat %>%
    mutate(
      color_clmn = ifelse(
        !!sym(class_clmn) == class,
        !!sym(celltype_clmn),
        other_class
      )
    )
  
  # Format data for bargraphs
  # Do not plot cell types with very few cells
  ag_dat <- bar_dat %>%
    group_by(sample, !!sym(celltype_clmn)) %>%
    filter(n() > n_cells) %>%
    
    group_by(sample, rep, !!sym(celltype_clmn)) %>%
    summarize(
      n       = n(),
      frac    = sum(!!sym(class_clmn) == class) / n,
      .groups = "drop"
    ) %>%
    mutate(
      !!sym(celltype_clmn) := fct_reorder(
        !!sym(celltype_clmn), frac,
        .fun = mean,
        .desc = TRUE
      )
    )
  
  # Create bargraphs
  ag_brs <- ag_dat %>%
    ggplot(aes(
      !!sym(celltype_clmn), frac,
      fill  = !!sym(celltype_clmn),
      alpha = as.character(rep)
    )) +
    geom_col(position = position_dodge()) +
    facet_wrap(~ sample, ncol = 1, strip.position = "right") +
    scale_alpha_manual(values = c("1" = 1, "2" = 0.75, "3" = 0.5, "4" = 0.25)) +
    scale_fill_manual(values = clrs) +
    scale_y_continuous(labels = label_percent(), breaks = pretty_breaks(n = 2)) +
    labs(y = str_c("% ", class, " cells")) +
    base_theme +
    theme(
      aspect.ratio    = 0.3,
      legend.position = "none",
      strip.placement = "outside",
      strip.clip      = "off",
      strip.text      = element_text(size = txt_pt2),
      axis.title.x    = element_blank(),
      axis.text.x     = element_text(angle = 45, hjust = 1, size = ttl_pt2),
      axis.title.y    = element_text(size = ttl_pt2),
      axis.text.y     = element_text(size = txt_pt1)
    )
  
  # Set column to color cells by
  # based on cell class, e.g. Ag-low/-high
  plot_dat <- plot_dat %>%
    mutate_meta(
      mutate,
      color_clmn = ifelse(
        !!sym(class_clmn) == class,
        as.character(!!sym(celltype_clmn)),
        other_class
      ),
      color_clmn = fct_relevel(color_clmn, levels(ag_dat[[celltype_clmn]])),
      color_clmn = fct_relevel(color_clmn, names(clrs)[!grepl("^other", names(clrs))]),
      color_clmn = fct_relevel(color_clmn, top_types)
    )
  
  # Set plot labels
  ag_labs <- bar_dat %>%
    group_by(color_clmn) %>%
    summarize(n = n(), .groups = "drop") %>%
    mutate(
      n_lab = str_c(color_clmn, "\nn = ", label_comma()(n))
    ) %>%
    arrange(color_clmn)
  
  ag_labs <- set_names(ag_labs$n_lab, ag_labs$color_clmn)
  
  # Plot tissue sections
  Idents(plot_dat) <- plot_dat$color_clmn
  
  plts <- fovs_to_plot %>%
    map(~ {
      p <- plot_dat %>%
        ImageDimPlot(
          fov  = .x,
          size = 2,
          crop = TRUE,
          dark.background = FALSE
        ) +
        guides(fill = guide_legend(override.aes = list(shape = 22, size = 5))) +
        scale_fill_manual(values = clrs, labels = ag_labs, drop = FALSE) +
        labs(tag = fov_lvls[.x]) +
        umap_theme +
        theme(
          legend.key.height = unit(35, "pt"),
          legend.title      = element_blank(),
          plot.tag.position = c(0.05, 0.95),
          plot.tag          = element_text(size = ttl_pt2)
        )
      
      plot_data <- p$data
      plot_data <- plot_data[order(plot_data$ident, decreasing = TRUE), ]
      p$data    <- plot_data
      
      p
    })
  
  # Create final figure
  fig <- plts %>%
    wrap_plots(
      nrow = 4,
      guides = "collect"
    ) &
    theme(
      legend.position = "bottom",
      legend.text     = element_text(size = ttl_pt2)
    )
  
  fig <- plot_grid(
    fig, ag_brs,
    nrow = 1,
    rel_widths = c(1, 0.45)
  )
  
  fig
}
```

```{r "references"}
# LEC reference
ref_lec <- ref_immgen[, grepl("^Endothelial cells", colnames(ref_immgen))]
ref_lec <- ref_lec[rownames(ref_lec) %in% rownames(ref_LEC_xiang), ]

colnames(ref_lec) <- colnames(ref_lec) %>%
  str_replace("Endothelial cells \\(BEC\\)", "BEC")

ref_LEC_xiang <- ref_LEC_xiang[rownames(ref_lec), ]

stopifnot(identical(rownames(ref_LEC_xiang), rownames(ref_lec)))

ref_lec <- cbind(ref_LEC_xiang, ref_lec)
ref_lec <- ref_lec[, !grepl("^Endothelial cells", colnames(ref_lec))]

# Fibroblast/stromal cell reference
ref_frc <- ref_immgen[, grepl("^Fibroblast", colnames(ref_immgen))]
ref_frc <- ref_frc[rownames(ref_frc) %in% rownames(ref_lymphnodestromal), ]

ref_lymphnodestromal <- ref_lymphnodestromal[rownames(ref_frc), ]

stopifnot(identical(rownames(ref_lymphnodestromal), rownames(ref_frc)))

ref_frc <- cbind(ref_lymphnodestromal, ref_frc)

# Combined Immgen/DC reference
rownames(ref_mousespleenDC) <- str_to_title(rownames(ref_mousespleenDC))

shared_gns    <- intersect(rownames(ref_immgen), rownames(ref_mousespleenDC))
ref_immgen_dc <- cbind(ref_immgen[shared_gns, ], ref_mousespleenDC[shared_gns, ])
```

```{r "create object", eval = create_object}
# Sample name key
sam_key <- set_names(
  as.character(unname(params$samples)),
  as.character(names(params$samples))
)

# Identify samples to load
xen_dirs <- dir(
  here(params$res_dir, params$xen_dir),
  pattern = params$slide_regex,
  full.names = TRUE
)

names(xen_dirs) <- xen_dirs %>%
  str_extract("(?<=__)[A-Za-z]+_[-A-Z0-9]+(?=__)") %>%
  str_replace_all("-", "_")

nms  <- names(xen_dirs)
dups <- duplicated(nms)

names(xen_dirs)[dups] <- str_c(nms[dups], "_1")

# Custom targets
ag_targets <- unname(params$ag_targets)
ag_clmns   <- names(params$ag_targets)

chikv_targets <- unname(params$chikv_targets)
chikv_clmns   <- names(params$chikv_targets)

custom_targets <- ag_targets %>%
  c(params$custom_targets) %>%
  unique() %>%
  unlist()

# Create Seurat object
# store custom targets as separate assay
xen <- xen_dirs %>%
  imap(~ {
    obj <- .x %>%
      LoadXenium(fov = .y)
    
    gns <- rownames(obj)
    gns <- gns[!gns %in% custom_targets]
    
    cstm <- obj %>%
      subset(features = custom_targets)
    
    obj <- obj %>%
      subset(features = gns)
    
    obj[["Xenium_custom"]] <- cstm@assays$Xenium
    
    obj$fov <- .y
    
    obj %>%
      subset(nCount_Xenium > 0)
  })

xen <- merge(xen[[1]], xen[-1], add.cell.ids = names(xen))
```

```{r "process object", include = FALSE, eval = create_object}
# Process data
# Add FOV labels
xen <- xen %>%
  .process_xenium(
    dims = 1:40,
    resolution = params$clst_rsln
  )

# Identify Ag+/CHIKV+ cells
ag_feats    <- str_c("xenium_custom_", ag_targets)
chikv_feats <- str_c("xenium_custom_", chikv_targets)

xen <- xen %>%
  AddMetaData(
    FetchData(., ag_feats),
    col.name = ag_clmns
  ) %>%
  AddMetaData(
    FetchData(., chikv_feats),
    col.name = chikv_clmns
  ) %>%
  mutate_meta(~ {
    .x %>%
      mutate(
        Ag_class = ifelse(
          rowSums(select(., all_of(ag_clmns)) > 0) > 0,
          "Ag-high",
          "Ag-low"
        ),
        chikv_class = ifelse(
          rowSums(select(., all_of(chikv_clmns)) > 0) > 0,
          "CHIKV-high",
          "CHIKV-low"
        )
      )
  })

# Identify Ag-high cells using km clustering
xen <- xen %>%
  cluster_signal(
    data_column  = "Ag_tag",
    clust_column = "Ag_km",
    clust_names  = c("Ag-low", "Ag-high"),
    method       = "km"
  )

# Add sample names
xen <- xen %>%
  mutate_meta(
    mutate,
    sample = sam_key[fov],
    sample = ifelse(grepl("^Mock", fov), "Mock", sample)
  )
```

```{r "cell type annotations", eval = create_object}
# Create reference using scRNA-seq data
ref_dir <- "~/Dropbox/Ryan/Projects/antigen-exchange/results/sobjs/2024-07-01"

ref_so <- c("neg", "pos") %>%
  map(~ {
    file.path(ref_dir, str_c("cd45", .x, "_so.qs")) %>%
      qread() %>%
      subset(mouse == "6wk")
  })

ref_so <- merge(ref_so[[1]], ref_so[[2]])

ref_so <- ref_so %>%
  SCTransform(vst.flavor = "v2")

ref_so <- ref_so %>%
  RunPCA(
    assay    = "SCT",
    npcs     = 50,
    features = rownames(.)
  ) %>%
  RunUMAP(
    assay       = "SCT",
    reduction   = "pca",
    dims        = dims,
    umap.method = "umap-learn"
  )

ref_so <- ref_so %>%
  AddMetaData(FetchData(., c("UMAP_1", "UMAP_2")))

ref_so %>%
  qsave(here(params$so_dir, str_c("ref_so", sffx)))

# Predict cell types for Xenium data
anchors <- FindTransferAnchors(
  ref_so,
  query = xen,
  normalization.method = "SCT"
)

preds <- TransferData(
  anchorset        = anchors,
  refdata          = ref_so$cell_type,
  prediction.assay = TRUE,
  weight.reduction = xen[["pca"]],
  dims             = 1:40
)

anchors %>%
  qsave(here(params$so_dir, str_c("ref_anchors", sffx)))

preds %>%
  qsave(here(params$so_dir, str_c("ref_preds", sffx)))

xen[["preds"]] <- preds

xen_types <- xen@assays$preds@data %>%
  as_tibble(rownames = "cell_type") %>%
  pivot_longer(-cell_type, values_to = "pred_conf") %>%
  filter(cell_type != "max") %>%
  group_by(name) %>%
  filter(pred_conf == max(pred_conf)) %>%
  mutate(cell_type = if_else(n() > 1, "unassigned", cell_type)) %>%
  ungroup() %>%
  column_to_rownames("name")

xen <- xen %>%
  AddMetaData(xen_types)

xen <- xen %>%
  mutate_meta(
    mutate,
    Ag_type    = ifelse(Ag_km == "Ag-high", cell_type, "Ag-low"),
    chikv_type = ifelse(chikv_class == "CHIKV-high", cell_type, "CHIKV-low")
  )
```

```{r "LEC reference object", eval = create_object}
# Create reference using scRNA-seq data
lec_ref_so <- qread(file.path(ref_dir, "lec_so.qs"))

lec_ref_so <- lec_ref_so %>%
  subset(sample != "d2") %>%
  SplitObject("mouse") %>%
  map(SCTransform, vst.flavor = "v2")

lec_int_feats <- lec_ref_so %>%
  SelectIntegrationFeatures()

lec_ref_so <- lec_ref_so %>%
  PrepSCTIntegration(anchor.features = lec_int_feats)

lec_ref_anchrs <- lec_ref_so %>%
  FindIntegrationAnchors(
    normalization.method = "SCT",
    anchor.features      = lec_int_feats,
    dims = 1:30
  )

lec_ref_so <- lec_ref_anchrs %>%
  IntegrateData(
    normalization.method = "SCT",
    dims = 1:30
  )

lec_ref_so <- lec_ref_so %>%
  RunPCA() %>%
  RunUMAP(
    dims = 1:30,
    umap.method = "umap-learn"
  )

lec_ref_so <- lec_ref_so %>%
  AddMetaData(FetchData(., c("UMAP_1", "UMAP_2")))

lec_ref_so %>%
  qsave(here(params$so_dir, str_c("lec_ref_so", sffx)))
```

```{r "cell types", eval = FALSE}
# Column containing clusters to use for cell type annotations
clst_clmn <- str_c("SCT_snn_res.", params$clst_rsln)

# Annotate CD45+ cells
xen <- xen %>%
  mutate_meta(mutate, cell_type = "unassigned") %>%
  .classify_t_b(clst_clmn) %>%
  .classify_dcs(clst_clmn)

# Annotate LECs
xen <- xen %>%
  classify_markers(
    Ptprc <= 0.7,
    feats      = "Ptprc",
    type_label = "other CD45-",
    clst_col   = clst_clmn,
    type_col   = "cell_type"
  ) %>%
  classify_markers(
    cell_type == "other CD45-" & Pdpn > 0.01 & Pecam1 > 0.1 & Prox1 > 0.1,
    feats      = c("cell_type", "Pdpn", "Pecam1", "Prox1"),
    type_label = "LEC",
    clst_col   = clst_clmn,
    type_col   = "cell_type"
  ) %>%
  classify_markers(
    cell_type == "other CD45-" & Pdpn < 0.01 & Pecam1 > 0.1,
    feats      = c("cell_type", "Pdpn", "Pecam1", "Prox1"),
    type_label = "BEC",
    clst_col   = clst_clmn,
    type_col   = "cell_type"
  )

xen <- xen %>%
  mutate_meta(
    mutate,
    # type_cluster = ifelse(cell_type == "unassigned", !!sym(clst_clmn), cell_type),
    Ag_type      = ifelse(Ag_class == "Ag-high", cell_type, "Ag-low"),
    chikv_type   = ifelse(chikv_class == "CHIKV-high", cell_type, "CHIKV-low")
  )



# # TEST PLOTS
# xen@meta.data %>%
#   # filter(sample != "Mock") %>%
#   filter(cell_type == "LEC") %>%
#   filter(sample %in% c("CHIKV 181/25 + Ag", "CHIKV WT + Ag")) %>%
#   plot_frequency(
#     "Ag_class",
#     cluster_col = "sample"
#   )
# 
# xen@meta.data %>%
#   filter(sample %in% c("CHIKV 181/25 + Ag", "CHIKV WT + Ag")) %>%
#   plot_violin(
#     "Ag_tag",
#     "cell_type",
#     group_col = "sample",
#     top       = Inf,
#     method    = "boxplot"
#   )
# 
# xen@meta.data %>%
#   # filter(Ag_tag > 0) %>%
#   filter(cell_type == "LEC") %>%
#   ggplot(aes(Ag_tag + 1, fill = fov)) +
#   geom_density() +
#   scale_x_log10() +
#   facet_wrap(~ sample, scales = "free_y")
# 
# xen@meta.data %>%
#   mutate(
#     nCount_CHIKV = CHIKV_sgRNA + CHIKV_5 + CHIKV_neg,
#     frac_CHIKV   = nCount_CHIKV / (nCount_CHIKV + nCount_Xenium)
#   ) %>%
#   # filter(Ag_tag > 0) %>%
#   filter(cell_type == "LEC") %>%
#   ggplot(aes(sample, Ag_tag, fill = fov)) +
#   geom_boxplot()
#   # scale_x_log10()
#   # facet_wrap(~ sample, scales = "free_y")
# 
# xen@meta.data %>%
#   pivot_longer(starts_with("CHIKV_", ignore.case = FALSE)) %>%
#   mutate(frac = value / (value + nCount_Xenium)) %>%
#   filter(
#     # sample %in% c("CHIKV 181/25 + Ag", "CHIKV WT + Ag"),
#     # sample != "Mock",
#     cell_type == "LEC"
#   ) %>%
# 
#   # filter(Ag_tag > 0) %>%
#   ggplot(aes(fov, value, fill = name)) +
#   geom_boxplot() +
#   # scale_x_log10()
#   facet_wrap(~ sample, scales = "free")
# 
# dat <- xen %>%
#   FetchData(c(
#     "fov", "sample", clst_clmn,
#     "UMAP_1", "UMAP_2", "cell_type",
#     "Pdpn", "Pecam1", "Prox1", "Ptprc", "Lyve1",
#     "Marco", "Madcam1", "Ackr4", "Kdr"
#   )) %>%
#   filter(fov == "Mock_1")
# 
# dat %>%
#   plot_scatter(
#     "Prox1",
#     group_col = "fov",
#     size = 0.1,
#     plot_colors = c("lightblue", "red")
#   )
# 
# dat %>%
#   plot_violin(
#     "Ptprc",
#     "cell_type",
#     "fov",
#     method = "boxplot"
#   )
# 
# xen %>%
#   plot_violin(
#     "Lyve1",
#     "cell_type",
#     "fov",
#     method = "boxplot"
#   )
```

```{r "create lec objects", include = FALSE, eval = create_object}
# Process data
type_key <- c(
  lec = "LEC"
  # frc = "Fibroblasts"
)

cd45neg <- type_key %>%
  imap(~ {
    obj <- xen %>%
      subset(cell_type == .x) %>%
      .process_xenium(
        dims = 1:30,
        resolution = c(1, 2, 3, 5)
      )
    
    obj
  })

lec <- cd45neg$lec

# Predict LEC subsets for Xenium data
anchors <- FindTransferAnchors(
  lec_ref_so,
  query = lec,
  normalization.method = "SCT"
)

preds <- TransferData(
  anchorset        = anchors,
  refdata          = lec_ref_so$subtype,
  prediction.assay = TRUE,
  weight.reduction = lec[["pca"]],
  dims             = 1:40
)

anchors %>%
  qsave(here(params$so_dir, str_c("lec_ref_anchors", sffx)))

preds %>%
  qsave(here(params$so_dir, str_c("lec_ref_preds", sffx)))

lec[["preds"]] <- preds

xen_lec_types <- lec@assays$preds@data %>%
  as_tibble(rownames = "lec_type") %>%
  pivot_longer(-lec_type, values_to = "pred_conf") %>%
  filter(lec_type != "max") %>%
  group_by(name) %>%
  filter(pred_conf == max(pred_conf)) %>%
  mutate(lec_type = if_else(n() > 1, "unassigned", lec_type)) %>%
  ungroup() %>%
  column_to_rownames("name")

# Refine LEC subsets
# ann <- clustify(
#   input       = lec@assays$SCT@data,
#   ref_mat     = ref_LEC_xiang,
#   metadata    = lec@meta.data,
#   cluster_col = "SCT_snn_res.5",
#   vec_out     = TRUE
# )
# 
# names(ann) <- colnames(lec)
# 
# lec <- lec %>%
#   AddMetaData(ann, col.name = "lec_type") %>%
#   mutate_meta(
#     mutate,
#     lec_type = str_replace_all(lec_type, "_", "-")
#   ) %>%
#   classify_markers(
#     Madcam1 > 0.1,
#     feats      = "Madcam1",
#     type_label = "fLEC",
#     clst_col   = "SCT_snn_res.5",
#     type_col   = "lec_type"
#   )
lec <- lec %>%
  AddMetaData(xen_lec_types) %>%
  classify_markers(
    Marco > 0.25,
    feats      = "Marco",
    type_label = "Marco-LEC",
    clst_col   = "SCT_snn_res.5",
    type_col   = "lec_type"
  ) %>%
  mutate_meta(
    mutate,
    Ag_type    = ifelse(Ag_km == "Ag-high", lec_type, "Ag-low"),
    chikv_type = ifelse(chikv_class == "CHIKV-high", lec_type, "CHIKV-low")
  )

# Add LEC annotations to main object
xen <- xen %>%
  AddMetaData(lec$lec_type,   col.name = "lec_type") %>%
  AddMetaData(lec$Ag_type,    col.name = "lec_Ag_type") %>%
  AddMetaData(lec$chikv_type, col.name = "lec_chikv_type") %>%
  mutate_meta(
    mutate,
    across(
      c(lec_type, lec_Ag_type, lec_chikv_type),
      ~ replace_na(.x, "other")
    )
  )

# Save LEC object
lec %>%
  qsave(here(params$so_dir, str_c("lec", sffx)))
```

```{r "refine cell types", include = FALSE, eval = FALSE}
# Adjust LEC annotations based on marker genes
# clustifyr annotation exclude fLECs despite expression of fLEC markers
lec_clst_clmn <- "SCT_snn_res.3"

cd45neg <- cd45neg %>%
  map(~ {
    obj <- .x %>%
      .classify_t_b(lec_clst_clmn) %>%
      .classify_dcs(lec_clst_clmn)

    obj
  })

# Update cell type annotations
cd45neg %>%
  walk(~ {
    xen <<- xen %>%
      AddMetaData(.x$cell_type, col.name = "new_type") %>%
      mutate_meta(~ {
        .x %>%
          mutate(
            cell_type    = ifelse(is.na(new_type), cell_type, new_type),
            type_cluster = ifelse(cell_type == "unassigned", !!sym(clst_clmn), cell_type),
            Ag_type      = ifelse(Ag_class == "Ag-high", cell_type, "Ag-low"),
            chikv_type   = ifelse(chikv_class == "CHIKV-high", cell_type, "CHIKV-low")
          ) %>%
          dplyr::select(-new_type)
      })
  })

# Recluster filtered LECs/FRCs
cd45neg <- cd45neg %>%
  imap(~ {
    obj <- .x %>%
      subset(cell_type == type_key[.y]) %>%
      .process_xenium(
        dims = 1:30,
        resolution = c(1, 2, 3, 5)
      )
    
    obj
  })

# Split list
lec <- cd45neg$lec
frc <- cd45neg$frc

rm(cd45neg)
gc()
```

```{r "calculate distance from Ag+ LEC", eval = create_object}
# Calculate distance from closest Ag+ LEC
fovs <- sort(unique(xen$fov))

dists <- fovs %>%
  map_dfr(~ {
    # Pull cell centroid coordinates for plotting
    cell_coords <- xen@images[[.x]]@boundaries$centroids
    
    cell_coords <- cell_coords@coords %>%
      as_tibble() %>%
      mutate(.cell_id = cell_coords@cells)
    
    cell_meta <- xen@meta.data %>%
      filter(fov == .x) %>%
      as_tibble(rownames = ".cell_id") %>%
      dplyr::select(.cell_id, cell_type, Ag_class) %>%
      left_join(cell_coords, by = ".cell_id")
    
    # Convert to sf object (sp package is deprecated)
    # first convert to sp::SpatialPolygonsDataFrame
    # mol <- obj@images$A@molecules$molecules$GeoMx_Barcode01_ERCC00142
    # Ensure geometries are valid
    cells <- xen@images[[.x]]@boundaries$segmentation@polygons
    plys  <- SpatialPolygons(cells)
    
    plys <- SpatialPolygonsDataFrame(
      plys,
      data = data.frame(
        id        = seq_along(cells),
        cell_type = cell_meta$cell_type,
        Ag_class  = cell_meta$Ag_class,
        row.names = map_chr(cells, ~ .x@ID)
      )
    )
    
    plys <- plys %>%
      st_as_sf() %>%
      st_make_valid()
    
    # Pull cell types
    lecs <- cell_meta %>%
      filter(cell_type == "LEC" & Ag_class == "Ag-high")
    
    # Only FOVs that received Ag will have distance calculations
    if (nrow(lecs) == 0) {
      res <- cell_meta %>%
        rename(
          other = .cell_id,
          x_cell = x,
          y_cell = y
        ) %>%
        mutate(fov = .x)
      
      return(res)
    }
    
    # Calculate distances between each LEC and all other cells
    dists <- st_distance(plys[lecs$.cell_id, ], plys)
    
    rownames(dists) <- lecs$.cell_id
    colnames(dists) <- rownames(plys)
    
    dists <- dists %>%
      as_tibble(rownames = "lec_id") %>%
      pivot_longer(-lec_id, names_to = "other", values_to = "lec_dist") %>%
      left_join(cell_meta, by = c(other = ".cell_id"))
    
    dists <- dists %>%
      left_join(
        cell_coords,
        by     = c(lec_id = ".cell_id"),
        suffix = c("_cell", "_lec")
      )
    
    # Determine closest LEC
    res <- dists %>%
      group_by(other) %>%
      mutate(
        closest = lec_dist == min(lec_dist),
        fov = .x
      ) %>%
      ungroup()

    res
  })

dist_meta <- dists %>%
  filter(is.na(closest) | closest) %>%
  group_by(other) %>%
  slice(1) %>%
  ungroup() %>%
  dplyr::select(other, lec_id, lec_dist, matches("^[xy]_")) %>%
  column_to_rownames("other")

xen <- xen %>%
  AddMetaData(dist_meta)
```

```{r "lec types", eval = FALSE}
# Annotate LEC types
ann <- clustify(
  input       = lec@assays$SCT@data,
  ref_mat     = ref_LEC_xiang,
  metadata    = lec@meta.data,
  cluster_col = lec_clst_clmn,
  vec_out     = TRUE,
)

names(ann) <- colnames(lec)

lec <- lec %>%
  AddMetaData(ann, col.name = "lec_type")

# Refine fLEC annotations
lec <- lec %>%
  classify_markers(
    Madcam1 > 0.25,
    feats      = "Madcam1",
    type_label = "fLEC",
    clst_col   = lec_clst_clmn,
    type_col   = "lec_type"
  )

# Add annotations to meta.data
lec <- lec %>%
  mutate_meta(
    mutate,
    Ag_type = ifelse(Ag_class == "Ag-high", lec_type, "Ag-low")
  )

xen <- xen %>%
  AddMetaData(lec$lec_type, col.name = "lec_type") %>%
  AddMetaData(lec$Ag_type, col.name = "lec_Ag_type") %>%
  mutate_meta(
    mutate,
    lec_type = replace_na(lec_type, "other")
  )

# Save objects
lec %>%
  qsave(here(params$so_dir, str_c("lec", sffx)))

frc %>%
  qsave(here(params$so_dir, str_c("frc", sffx)))
```

```{r "FRC TYPES", eval = FALSE}
# As a test combine lec and frc references
ref_new <- rownames(ref_lec)
ref_new <- ref_new[ref_new %in% rownames(ref_frc)]
ref_new <- bind_cols(ref_frc[ref_new, ], ref_lec[ref_new, ])

# Annotate FRC types
ann <- clustify(
  input       = frc@assays$SCT@data,
  ref_mat     = ref_new,
  metadata    = frc@meta.data,
  cluster_col = frc_clst_clmn,
  vec_out     = TRUE
)

names(ann) <- colnames(frc)

frc <- frc %>%
  AddMetaData(ann, col.name = "frc_type")

frc %>%
  plot_scatter(
    "frc_type",
    group_col = "fov",
    size = 0.2
  )

# # FRC markers
# clst_clmn <- "SCT_snn_res.2"
# 
# dat <- xen %>%
#   subset(cell_type == "other CD45-") %>%
#   classify_markers(
#     Ptprc > 1,
#     feats      = "Ptprc",
#     type_label = "other CD45+",
#     clst_col   = clst_clmn,
#     type_col   = "cell_type"
#   ) %>%
#   subset(cell_type == "other CD45-")

# Plot markers
gns <- c(
  "Ptprc", "Pdpn", "Pecam1",
  "Prox1", "Lyve1",
  "Cr2",
  # "Cxcl13",
  # "Pthlh",
  # "Tmem119",
  "Sox9",
  "Cr2",
  "Acta2",
  # "Pdgfrb",
  "Itga7",
  "Cd34",
  # "Il7",
  # "Ccl21a",
  # "Ccl19",
  # "Tnfsf13b",
  "Cxcl9",
  "Cxcl10",
  # "Tnfsf11",
  "Ch25h",
  # "Cxcl13",
  "Nr4a1",
  "Inmt",
  # "Cdh1",
  "Krt8",
  "Muc1"
  # "Ocln",
  
  # "Ptprc",
  # # "Cd14",
  # "Itgam",
  # # "Fcer1g",
  # "Cx3cr1"
  # # "Fcgr3",
  # # "Ly6g",
  # # "Ceacam2",
  # # "Fut4",
  # # "Fcgr3"
)

marker_plt <- gns %>%
  map(~ {
    frc %>%
      plot_violin(
        .x,
        cluster_col   = "frc_type",
        # plot_colors   = typ_clrs,
        method        = "boxplot",
        outlier.size  = 0.5,
        outlier.alpha = 1,
        top           = Inf
        # n_label       = "none"
      )
  }) %>%
  plot_grid(
    plotlist = .,
    nrow     = 3,
    align    = "vh",
    axis     = "trbl"
  )

marker_plt
```

```{r "small object", eval = create_object}
# Format Ag labels
dcs <- xen$cell_type %>%
  unique() %>%
  grep("DC", ., value = TRUE) %>%
  sort()

lecs <- xen$lec_type %>%
  table() %>%
  sort() %>%
  rev() %>%
  names()

lecs <- lecs[lecs != "other"]

# top_types <- c("LEC", dcs)
top_types <- "LEC"

# xen <- xen %>%
#   mutate_meta(
#     mutate,
#     lec_Ag_type    = ifelse(Ag_class == "Ag-high", as.character(lec_type), "Ag-low"),
#     lec_Ag_type    = fct_relevel(lec_Ag_type, lecs),
#     lec_dc_type    = ifelse(cell_type %in% top_types, as.character(cell_type), "other"),
#     lec_dc_type    = fct_relevel(lec_dc_type, top_types),
#     lec_dc_Ag_type = ifelse(Ag_class == "Ag-high", as.character(lec_dc_type), "Ag-low"),
#     lec_dc_Ag_type = fct_relevel(lec_dc_Ag_type, top_types)
#   )

# Select subset of cells to plot
# too many points to plot, downsample when plotting sections
set.seed(42)

plt_cells <- sample(colnames(xen), 50000)

plt_cells <- xen@meta.data %>%
  filter(cell_type %in% top_types | Ag_class == "Ag-high" | chikv_class == "CHIKV-high") %>%
  rownames() %>%
  c(plt_cells) %>%
  unique()

plt_cells <- xen@meta.data[plt_cells, ] %>%
  mutate(cell_type = fct_relevel(cell_type, top_types)) %>%
  arrange(cell_type) %>%
  rownames()

small <- xen %>%
  subset(cells = plt_cells)

# Save objects
xen %>%
  qsave(here(params$so_dir, str_c("xen", sffx)))

small %>%
  qsave(here(params$so_dir, str_c("small", sffx)))
```

```{r "load objects"}
# Load objects
xen   <- qread(here(params$so_dir, str_c("xen", sffx)))
small <- qread(here(params$so_dir, str_c("small", sffx)))
lec   <- qread(here(params$so_dir, str_c("lec", sffx)))

# Top cell types to plot
dcs <- xen$cell_type %>%
  unique() %>%
  grep("DC", ., value = TRUE) %>%
  sort()

ag_dcs <- xen@meta.data %>%
  filter(cell_type %in% dcs) %>%
  group_by(cell_type) %>%
  filter(all(table(Ag_class) > 15)) %>%
  pull(cell_type) %>%
  unique()

# top_types <- c("LEC", dcs)
top_types <- "LEC"

# Cells to plot for sections
plt_cells <- colnames(small)

# Top FOVs to plot
top_fov   <- "Group_D_4"
top_fov_2 <- "Group_A_5"
```

```{r "format data"}
# Sample levels
sam_lvls <- unname(params$samples) %>%
  unlist() %>%
  unique()

# Format plotting data
chikv_feats <- c(
  "CHIKV_sgRNA",
  "CHIKV_5",
  "CHIKV_neg"
)

clmns <- c(
  ".cell_id", "sample", "fov", "cell_type",
  "nCount_Xenium", "nCount_Xenium_custom",
  "Ag_class", "Ag_km", "Ag_type", "Ag_tag",
  "lec_type", "lec_Ag_type", "lec_chikv_type",
  # "lec_dc_type", "lec_dc_Ag_type",
  "chikv_type", "chikv_class", chikv_feats
  # "lec_id", "lec_dist"
)
# "cell_type", "lec_type", "sample", "fov",
# chikv_feats, "Ag_tag",
# "nCount_Xenium", "nCount_Xenium_custom",
# "Ag_class", "chikv_class"

xen_dat <- xen@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  dplyr::select(all_of(clmns), matches("^[xy]_")) %>%
  mutate(
    across(
      all_of(chikv_feats),
      ~ .x / (nCount_Xenium + nCount_Xenium_custom),
      .names = "frac_{.col}"
    ),
    tot_CHIKV  = CHIKV_sgRNA + CHIKV_5 + CHIKV_neg,
    frac_CHIKV = tot_CHIKV / (nCount_Xenium + nCount_Xenium_custom)
  ) %>%
  group_by(sample) %>%
  mutate(rep = as.numeric(factor(fov))) %>%
  ungroup() %>%
  mutate(
    sample  = fct_relevel(sample, sam_lvls),
    fov_lab = str_remove(fov, "^Group_|ock_"),
    fov_lab = str_c(sample, "\n", fov_lab)
  )

fov_lvls <- xen_dat %>%
  arrange(sample) %>%
  distinct(fov, fov_lab)

fov_lvls <- set_names(fov_lvls$fov_lab, fov_lvls$fov)

fovs <- names(fov_lvls)

xen_dat <- xen_dat %>%
  mutate(fov = fct_relevel(fov, names(fov_lvls))) %>%
  arrange(fov) %>%
  mutate(fov_lab = fct_inorder(fov_lab))

small_dat <- small@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  dplyr::select(all_of(clmns), matches("^[xy]_")) %>%
  mutate(Ag_class = fct_relevel(Ag_class, c("Ag-low", "Ag-high"))) %>%
  arrange(Ag_class)
```

```{r "theme"}
# ggplot2 themes
txt_pt1  <- 10
txt_pt2  <- 12
ttl_pt1  <- 14
ttl_pt2  <- 16
ln_pt    <- 0.5
ln_col   <- "grey85"
cir_size <- 3.5
sqr_size <- 4

txt_theme_1 <- theme(
  strip.text   = element_text(size = ttl_pt2),
  legend.text  = element_text(size = ttl_pt1),
  axis.title   = element_text(size = ttl_pt2),
  axis.text    = element_text(size = txt_pt2, color = "black"),
  legend.title = element_text(size = ttl_pt2),
)

line_theme <- theme(
  axis.line.x  = element_line(linewidth = ln_pt, color = ln_col),
  axis.line.y  = element_line(linewidth = ln_pt, color = ln_col),
  axis.ticks.x = element_line(linewidth = ln_pt, color = ln_col),
  axis.ticks.y = element_line(linewidth = ln_pt, color = ln_col)
)

base_theme <- djvdj_theme() +
  theme(
    plot.title       = element_text(face = "plain", size = ttl_pt2),
    strip.background = element_blank(),
    strip.text       = element_text(face = "plain")
  ) +
  txt_theme_1

umap_theme <- base_theme +
  theme(
    panel.border = element_rect(color = NA, fill = NA),
    axis.title   = element_blank(),
    axis.line.x  = element_blank(),
    axis.line.y  = element_blank(),
    axis.ticks.x = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text    = element_blank()
  )

umap_theme_2 <- umap_theme %+replace%
  theme(panel.border = element_rect(colour = ln_col, linewidth = ln_pt, fill = NA))

fade_0 <- "#FAFAFA"
fade_1 <- "#F0F0F0"
fade_2 <- "#D9D9D9"

# Colors
# Okabe Ito color palettes
ito_cols <- c(
  palette_OkabeIto[c(1:3, 5)], "#D7301F",
  palette_OkabeIto[c(4, 6)],   "#6A51A3",
  palette_OkabeIto[7],         "#875C04",
  "#065D43", "#F7AF34", "#00446E"
)

ito_cols <- ito_cols[3:length(ito_cols)] %>%
  darken(0.2) %>%
  c(ito_cols, ., "#686868", "#000000")

clrs <- ito_cols

# Cell type colors
typs <- unique(xen$cell_type)

typ_clrs <- set_names(
  clrs[seq_along(typs)],
  typs
)

typ_clrs[c("other CD45+", "other CD45-", "other", "Ag-low", "CHIKV-low")] <- "grey85"

# typ_clrs[c("LEC", "BEC", "cDC1")] <- c("#D7301F", "#0072B2", "#6A51A3")
# typ_clrs[c("LEC", "BEC", "cDC1")] <- c("#56B4E9", "#D7301F", "#6A51A3")
typ_clrs[c("LEC", "cDC1", "Monocytes", "B cells")] <- c("#0072B2", "#6A51A3", "#D55E00", "#56B4E9")

top_typ_clrs <- typ_clrs
top_typ_clrs[!names(top_typ_clrs) %in% top_types] <- "grey85"

top_typ_clrs_2 <- top_typ_clrs
top_typ_clrs_2[!names(top_typ_clrs_2) %in% top_types] <- "grey99"
top_typ_clrs_2[names(top_typ_clrs_2) %in% dcs] <- "#D7301F"

# Subtype colors
lec_clrs <- lec@meta.data %>%
  pull(lec_type) %>%
  unique()

lec_clrs <- set_names(
  clrs[seq_along(lec_clrs)],
  lec_clrs
)

# lec_clrs["cLEC"]      <- "#56B4E9"
lec_clrs["fLEC"]      <- "#D7301F"
lec_clrs["Ptx3-LEC"]  <- "#E69F00"
lec_clrs["tzLEC"]     <- "black"
lec_clrs["Marco-LEC"] <- "#6A51A3"

lec_clrs[c("Ag-low", "CHIKV-low", "other")] <- "grey85"

lec_lvls <- c(
  "cLEC", "fLEC", "Marco-LEC",
  "tzLEC", "Collecting", "Ptx3-LEC",
  "Valve", "unassigned", "BEC",
  "other"
)

# Sample colors
sam_clrs <- set_names(
  c("#D7301F", "#E69F00", "#009E73", "#0072B2", "#56B4E9", "grey85"),
  c(sam_lvls, "Mock")
)
```

## Summary

* Ag signals are sparse (most Ag+ cells have 1 count).
  This is at least partly due to the short length of the Ag tags, which only
  accommodate a single Xenium probe.
  We should consider re-designing and lengthening the Ag tags to allow binding
  of 2-3 probes.
* Due to the low number of Ag+ cells, Ag+ cells were identified as those with at
  least 1 count for either the 3 week or 6 week tag.
* LECs (specifically cLECs) are the top cell type enriched for Ag+ cells.
* When compared with Ag- DCs, Ag+ DCs tend to be in closer proximity to
  Ag+ LECs.
  This supports our previous GeoMx data and is in line with published work
  showing Ag is passed to DCs by LECs.
* Analysis of intracellular localization of Ag molecules for LECs interacting
  with a DC was inconclusive. This was due to the very few LEC-DC interactions,
  and the low number of Ag molecules.

<br>

<br>

---

```{r, eval = FALSE}
xen_dat %>%
  ggplot(aes(fov, nCount_Xenium, fill = sample)) +
  geom_boxplot() +
  facet_wrap(~ sample, scales = "free_x", nrow = 1) +
  scale_y_log10() +
  base_theme

xen_dat %>%
  filter(
    cell_type == "LEC"
    # tot_CHIKV > 0
  ) %>%
  ggplot(aes(fov, CHIKV_sgRNA + 1, fill = sample)) +
  geom_boxplot() +
  facet_wrap(~ sample, scales = "free_x", nrow = 1) +
  scale_y_log10() +
  base_theme
```

## Cell types

Annotated cell types are shown for LN sections.
Cells were downsampled to allow for visualization of all cell types.

```{r "SLIDE 1: cell type sections", fig.width = 15, fig.height = 16}
# Legend labels
cell_n <- xen@meta.data %>%
  group_by(cell_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    n_lab = str_c(cell_type, "\nn = ", label_comma()(n))
  ) %>%
  arrange(desc(n))

lgnd_labs <- set_names(cell_n$n_lab, cell_n$cell_type)

# Create section plots
Idents(small) <- small$cell_type

plts <- fovs %>%
  map(~ {
    p <- small %>%
      ImageDimPlot(
        fov  = .x,
        size = 1.5,
        crop = TRUE,
        dark.background = FALSE
      ) +
      guides(fill = guide_legend(override.aes = list(shape = 22, size = 4))) +
      scale_fill_manual(values = typ_clrs, labels = lgnd_labs) +
      labs(tag = fov_lvls[.x]) +
      coord_fixed() +
      umap_theme +
      theme(
        # legend.key.height = unit(35, "pt"),
        # legend.title      = element_blank(),
        # legend.text       = element_text(size = ttl_pt1),
        legend.position   = "none",
        plot.tag.position = c(0.15, 0.95),
        plot.tag          = element_text(size = ttl_pt2)
      )
    
    plot_data <- p$data
    plot_data <- plot_data[order(plot_data$ident, decreasing = TRUE), ]
    p$data    <- plot_data
    
    p
  })

plts <- plts %>%
  wrap_plots(
    nrow = 4,
    guides = "collect"
  )

# Create bargraphs
# be sure to NOT use downsampled data
brs <- xen@meta.data %>%
  mutate(sample = fct_relevel(sample, sam_lvls)) %>%
  plot_frequency(
    "cell_type",
    cluster_col = "fov",
    plot_colors = typ_clrs,
    plot_lvls   = names(lgnd_labs),
    alpha       = 1
  ) +
  facet_grid(~ sample, scales = "free_x", space = "free_x") +
  base_theme +
  scale_y_continuous(expand = expansion(c(0, 0))) +
  theme(
    # aspect.ratio    = 0.7,
    plot.margin     = margin(0, 0, 0, 0),
    panel.border    = element_blank(),
    legend.position = "right",
    legend.title    = element_blank(),
    strip.clip      = "off",
    axis.title      = element_blank(),
    axis.ticks      = element_blank(),
    axis.text.x     = element_text(angle = 45, hjust = 1),
    axis.text.y     = element_blank()
  )

fig <- wrap_plots(
  plts, brs,
  ncol = 1,
  heights = c(1, 0.25)
)
  # plot_annotation(caption = "* cells downsampled for visualization") &
  # theme(plot.caption = element_text(hjust = 0.02, vjust = 30, face = "italic"))

fig
```

### {.tabset .tabset-pills}

```{r "SLIDE 3: LEC DC sections", fig.width = 13, fig.height = 16, eval = FALSE}
# Create section plots
Idents(small) <- small$cell_type

plts <- fovs %>%
  map(~ {
    p <- small %>%
      ImageDimPlot(
        fov  = .x,
        size = 1.5,
        crop = TRUE,
        dark.background = FALSE
      ) +
      guides(fill = guide_legend(override.aes = list(shape = 22, size = 4))) +
      scale_fill_manual(values = top_typ_clrs, labels = lgnd_labs) +
      labs(tag = .x) +
      umap_theme +
      theme(
        legend.position   = "none",
        plot.tag.position = c(0.15, 0.95),
        plot.tag          = element_text(size = ttl_pt2, face = "bold")
      )
    
    plot_data <- p$data
    plot_data <- plot_data[order(plot_data$ident, decreasing = TRUE), ]
    p$data    <- plot_data
    
    p
  })

plts <- plts %>%
  wrap_plots(
    nrow = 4,
    guides = "collect"
  )

# Create bargraphs
br_dat <- xen@meta.data %>%
  group_by(fov, sample) %>%
  mutate(total_cells = n()) %>%
  group_by(fov, sample, total_cells, cell_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(frac = n / total_cells) %>%
  filter(cell_type %in% top_types) %>%
  mutate(
    cell_type = fct_relevel(cell_type, rev(top_types)),
    sample    = fct_relevel(sample, sam_lvls)
  )

x_labs <- br_dat %>%
  group_by(fov) %>%
  summarize(n = sum(n), .groups = "drop") %>%
  mutate(
    n_lab = str_c(fov, "\nn = ", label_comma()(n))
  )

x_labs <- set_names(x_labs$n_lab, x_labs$fov)

brs <- br_dat %>%
  ggplot(aes(fov, frac, fill = cell_type)) +
  geom_col(position = "stack") +
  facet_grid(~ sample, scales = "free_x", space = "free_x") +
  scale_fill_manual(values = typ_clrs) +
  scale_x_discrete(labels = x_labs) +
  scale_y_continuous(expand = expansion(c(0, 0.05)), labels = label_percent()) +
  labs(y = "% of cells") +
  base_theme +
  theme(
    legend.position = "none",
    strip.clip      = "off",
    axis.line.y     = element_line(color = ln_col, linewidth = ln_pt),
    axis.title.x    = element_blank(),
    axis.ticks.x    = element_blank(),
    axis.text.x     = element_text(angle = 45, hjust = 1)
  )

# Create final figure
fig <- wrap_plots(
  plts, brs,
  ncol = 1,
  heights = c(1, 0.25)
)

fig
```

```{r, fig.width = 45, fig.height = 10, results = "asis", out.height = "70%"}
fovs %>%
  walk(~ {
    cat("\n\n####", .x, "\n\n")
    
    fig <- xen_dat %>%
      .plot_full_fov(
        fov   = .x,
        color = "cell_type",
        trace = list("NONE", "LEC"),
        clrs  = typ_clrs
      ) %>%
      plot_grid(
        plotlist = .,
        nrow  = 1,
        align = "h",
        axis  = "tb"    
      )
    
    print(fig)
    cat("\n\n<br>\n\n<br>\n\n")
  })
```

###

The expression of key marker genes is shown below.

```{r "cell type markers", fig.width = 20, fig.height = 9}
gns <- c(
  "Ptprc", "Pdpn", "Pecam1",
  "Cd3d", "Cd19",
  "Prox1", "Lyve1",
  "Itgax",           # DCs
  "Xcr1",            # cDC1
  "Sirpa", "Itgam",  # cDC2
  "Siglech",         # Siglech DCs
  "Ccr7"
)

marker_plt <- gns %>%
  map(~ {
    xen %>%
      plot_violin(
        .x,
        # cluster_col   = "cell_type",
        cluster_col   = "cell_type",
        plot_colors   = typ_clrs,
        method        = "boxplot",
        outlier.size  = 0.5,
        outlier.alpha = 1
      )
  }) %>%
  plot_grid(
    plotlist = .,
    ncol     = 4,
    align    = "vh",
    axis     = "trbl"
  )

marker_plt
```

---

<br>

<br>

## LEC subsets

UMAP projections show annotated LEC subsets and marker gene expression. 

```{r "SLIDE 4: LEC UMAPs", fig.width = 13, fig.height = 7}
# LEC annotations UMAPs
lec_u <- lec %>%
  plot_scatter(
    "lec_type",
    size         = 1.2,
    plot_colors  = lec_clrs,
    label_params = list(size = ttl_pt1)
  ) +
  base_theme +
  theme(
    aspect.ratio    = 1,
    legend.position = "none",
    axis.title      = element_text(size = txt_pt2),
    axis.text       = element_blank(),
    axis.ticks      = element_blank()
  )

# LEC bargraphs
lec_brs <- lec@meta.data %>%
  mutate(sample = fct_relevel(sample, sam_lvls)) %>%
  
  ggplot(aes(fov, fill = lec_type)) +
  geom_bar(position = "fill") +
  facet_grid(sample ~ ., scales = "free", space = "free", switch = "y") +
  scale_fill_manual(values = lec_clrs) +
  scale_y_continuous(expand = expansion(c(0.01, 0))) +
  labs(x = "section", y = "fraction of cells") +
  coord_flip() +
  base_theme +
  theme(
    panel.border = element_blank(),
    legend.title = element_blank(),
    strip.text.y.left = element_text(angle = 0, hjust = 1),
    axis.ticks   = element_blank(),
    axis.title   = element_blank(),
    axis.text    = element_blank()
  )

lec_u <- plot_grid(
  lec_u, lec_brs,
  rel_heights = c(1, 0.7),
  ncol  = 1
  # align = "v",
  # axis  = "rl"
)

# LEC markers
gns <- c(
  cLEC      = "Ackr4",
  fLEC      = "Madcam1",
  `Marco-LEC` = "Marco",
  `Ptx3-LEC`  = "Ptx3"
)

gn_u <- gns %>%
  imap(~ {
    lec %>%
      plot_scatter(
        .x,
        plot_colors = c("white", lec_clrs[.y]),
        n_label = "none",
        outline = TRUE
      ) +
      guides(color = guide_colorbar(ticks = FALSE)) +
      umap_theme_2 +
      theme(
        aspect.ratio = 1,
        legend.title = element_text(angle = 270, hjust = 0.5),
        legend.text  = element_text(size = txt_pt1),
        legend.key.height = unit(25, "pt"),
        legend.key.width = unit(5, "pt")
      )
  }) %>%
  plot_grid(
    plotlist = .,
    nrow = 2,
    align = "vh",
    axis  = "trbl"
  )

fig <- plot_grid(
  lec_u, gn_u,
  nrow = 1,
  rel_widths = c(0.8, 1)
)

fig
```

<br>

The expression of key LEC marker genes is shown below.

```{r "LEC type markers", fig.width = 10, fig.height = 4.5}
gns <- c(
  "Pdpn", "Pecam1",
  "Prox1",
  "Lyve1", "Mrc1",
  "Marco",
  "Madcam1", "Icam1",
  "Ackr4",
  "Ptx3"
)

marker_plt <- gns %>%
  map(~ {
    lec %>%
      plot_violin(
        .x,
        cluster_col   = "lec_type",
        plot_colors   = lec_clrs,
        method        = "boxplot",
        outlier.size  = 0.5,
        outlier.alpha = 1,
        n_label = "none"
      ) +
      theme(
        axis.text.x = element_text(angle = 45, hjust = 1)
      )
  }) %>%
  plot_grid(
    plotlist = .,
    nrow     = 2,
    align    = "vh",
    axis     = "trbl"
  )

marker_plt
```

<br>

<br>

### {.tabset .tabset-pills}

Annotated LEC subsets are shown for LN sections.
High-resolution images showing all cells are included at the bottom

```{r "SLIDE 5: LEC type sections", fig.width = 15, fig.height = 15}
# Legend labels
lec_labs <- xen@meta.data %>%
  group_by(lec_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    n_lab    = str_c(lec_type, "\nn = ", label_comma()(n)),
    lec_type = fct_relevel(lec_type, lec_lvls)
  ) %>%
  arrange(lec_type)

lec_labs <- set_names(lec_labs$n_lab, lec_labs$lec_type)

# Create section plots
small <- small %>%
  mutate_meta(
    mutate, 
    lec_type = fct_relevel(lec_type, lec_lvls)
  )

Idents(small) <- small$lec_type

plts <- fovs %>%
  map(~ {
    p <- small %>%
      ImageDimPlot(
        fov  = .x,
        size = 2,
        crop = TRUE,
        dark.background = FALSE
      ) +
      guides(fill = guide_legend(override.aes = list(shape = 22, size = 4))) +
      scale_fill_manual(values = lec_clrs, labels = lec_labs, drop = FALSE) +
      labs(tag = fov_lvls[.x]) +
      umap_theme +
      theme(
        legend.key.height = unit(35, "pt"),
        legend.title      = element_blank(),
        legend.text       = element_text(size = ttl_pt2),
        plot.tag.position = c(0.05, 0.95),
        plot.tag          = element_text(size = ttl_pt2)
      )
    
    plot_data <- p$data
    plot_data <- plot_data[order(plot_data$ident, decreasing = TRUE), ]
    p$data    <- plot_data
    
    p
  })

fig <- plts %>%
  wrap_plots(
    nrow = 4,
    guides = "collect"
  )
  # plot_annotation(caption = "* cells downsampled for visualization") &
  # theme(plot.caption = element_text(hjust = 0.02, face = "italic"))

fig
```

```{r, fig.width = 20, fig.height = 10, results = "asis", out.height = "50%"}
ag_types <- xen_dat$lec_type %>%
  unique() %>%
  grep("other|Ag-low", ., invert = TRUE, value = TRUE)

fovs %>%
  walk(~ {
    cat("\n\n####", .x, "\n\n")
    
    fig <- xen_dat %>%
      .plot_full_fov(
        fov        = .x,
        color      = "lec_type",
        clrs       = lec_clrs,
        trace      = list(ag_types),
        trace_clr  = NA,
        trace_size = 2
      )
    
    print(fig)
    cat("\n\n---\n\n<br>\n\n<br>\n\n")
  })
```

## CHIKV infection

### LECs are the primary cell type harboring CHIKV RNA {.tabset .tabset-pills}

CHIKV+ cells are shown for LN sections.
High-resolution images showing all cells are included at the bottom

```{r "CHIKV+ cell types", fig.width = 17, fig.height = 11}
.create_class_figure(
  bar_dat       = xen_dat,
  plot_dat      = small,
  class_clmn    = "chikv_class",
  class         = "CHIKV-high",
  celltype_clmn = "cell_type",
  clrs          = typ_clrs,
  top_types     = top_types
)
```

```{r, fig.width = 20, fig.height = 10, results = "asis", out.height = "50%"}
ag_types <- xen_dat$cell_type %>%
  unique() %>%
  grep("other|CHIKV-low", ., invert = TRUE, value = TRUE)

fovs %>%
  walk(~ {
    cat("\n\n####", .x, "\n\n")
    
    fig <- xen_dat %>%
      .plot_full_fov(
        fov        = .x,
        color      = "chikv_type",
        clrs       = typ_clrs,
        trace      = list(ag_types),
        trace_clr  = NA,
        trace_size = 1.5
      )
    
    print(fig)
    cat("\n\n<br>\n\n<br>\n\n")
  })
```

###

The fraction of total counts originating from CHIKV probes is shown below.
Replicates are plotted separately.
Cell types with <20 cells are not shown.
The number of cells identified for each cell type is shown above each boxplot.

```{r "CHIKV SIGNAL CELL TYPES", fig.width = 15, fig.height = 4}
# Format CHIKV cell type data
type_dat <- xen_dat %>%
  group_by(sample, cell_type) %>%
  filter(n() > 20) %>%
  ungroup() %>%
  mutate(
    cell_type = fct_reorder(cell_type, frac_CHIKV, mean, .desc = TRUE),
    sample = fct_relevel(sample, sam_lvls)
  )

# Format N labels
n_dat <- type_dat %>%
  group_by(sample, cell_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    n_lab = if_else(
      n > 1000,
      str_c(round(n / 1000, 1), "k"),
      label_comma()(n)
    )
  )

# Fraction CHIKV reads boxplots
pseudo <- 0.01

type_dat %>%
  ggplot(aes(cell_type, frac_CHIKV + pseudo, fill = cell_type, alpha = as.character(rep))) +
  geom_boxplot(
    outlier.size = 0.1,
    outlier.alpha = 1
  ) +
  geom_text(
    aes(y = Inf, label = n_lab, alpha = NULL, fill = NULL),
    data = n_dat,
    size = txt_pt1 / .pt,
    vjust = 1.25
  ) +
  facet_wrap(~ sample) +
  scale_y_log10(expand = expansion(c(0.05, 0.2)), labels = label_percent()) +
  scale_fill_manual(values = typ_clrs) +
  scale_alpha_manual(values = c("1" = 1, "2" = 0.75, "3" = 0.5, "4" = 0.25)) +
  guides(fill = "none", alpha = "none") +
  labs(y = str_c("% CHIKV counts + ", pseudo * 100)) +
  base_theme +
  theme(
    panel.background = element_rect(fill = NA, color = "grey75", linewidth = ln_pt),
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1)
  )
```

<br>

<br>

### CHIKV WT samples show highest levels of viral RNA

Total CHIKV counts are shown below for CHIKV+ LECs.
The number of CHIKV+ LECs and total LECs is shown for each sample.
Replicates were pooled together.

```{r, fig.width = 20, fig.height = 3}
# Format N labels
n_dat <- type_dat %>%
  group_by(sample, cell_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    n_lab = if_else(
      n > 1000,
      str_c(round(n / 1000, 1), "k"),
      label_comma()(n)
    )
  )

# Format histogram data
hist_dat <- type_dat %>%
  filter(cell_type == "LEC")

# Format N labels
n_dat <- hist_dat %>%
  group_by(cell_type, sample) %>%
  summarize(n = n(), n_chikv = sum(chikv_class == "CHIKV-high"), .groups = "drop") %>%
  mutate(n_lab = str_c("n = ", label_comma()(n_chikv), " / ", label_comma()(n)))

# Create histogram
hist_dat %>%
  filter(chikv_class == "CHIKV-high") %>%
  
  ggplot(aes(tot_CHIKV, fill = sample)) +
  geom_histogram(bins = 30) +
  geom_text(
    aes(1, Inf, label = n_lab, alpha = NULL, fill = NULL),
    data = n_dat,
    color = "black",
    hjust = 0,
    vjust = 1.3
  ) +
  facet_wrap(~ sample, nrow = 1) +
  scale_fill_manual(values = sam_clrs) +
  # scale_alpha_manual(values = c("1" = 1, "2" = 0.75, "3" = 0.5, "4" = 0.25)) +
  scale_x_log10() +
  scale_y_continuous(expand = expansion(c(0.05, 0.1))) +
  labs(x = "total CHIKV counts", y = "number of CHIKV+ LECs") +
  base_theme +
  theme(
    legend.position = "none"
  )
```

<br>

Total CHIKV counts are shown for all LECs (left) or CHIKV+ LECs (right).
The number of cells plotted is shown above each boxplot.

```{r "CHIKV SIGNAL SAMPLES", fig.width = 10, fig.height = 4, out.height = "40%"}
args <- list(
  y      = c(rep("tot_CHIKV", 2), rep("frac_CHIKV", 2)),
  ttl    = c("LECs", "CHIKV+ LECs", "LECs", "CHIKV+ LECs"),
  y_ttl  = c(rep("total CHIKV counts", 2), rep("% CHIKV counts", 2)),
  pseudo = c(rep(1, 2), rep(0.01, 2)),
  filt   = c(FALSE, TRUE, FALSE, TRUE),
  pct    = c(rep(FALSE, 2), rep(TRUE, 2))
)

plts <- args %>%
  pmap(~ {
    args <- list(...)
    
    dat <- type_dat %>%
      filter(cell_type == "LEC")
    
    if (args$filt) {
      dat <- dat %>%
        filter(chikv_class == "CHIKV-high")
      
      y_ttl <- args$y_ttl
    } else {
      y_ttl <- str_c(args$y_ttl, " + ", args$pseudo)
      
      if (args$pct) y_ttl <- str_c(args$y_ttl, " + ", pseudo * 100)
    }
    
    labs_fn <- waiver
    
    if (args$pct) labs_fn <- label_percent
    
    n_dat <- dat %>%
      group_by(sample) %>%
      summarize(n = n(), .groups = "drop") %>%
      mutate(n_lab = label_comma()(n))
    
    dat %>%
      ggplot(aes(sample, !!sym(args$y) + args$pseudo, fill = sample)) +
      geom_boxplot(
        outlier.size = 0.1,
        outlier.alpha = 1
      ) +
      geom_text(
        aes(y = Inf, label = n_lab, fill = NULL),
        data  = n_dat,
        color = "black",
        vjust = 1.25
      ) +
      scale_y_log10(expand = expansion(c(0.05, 0.2)), labels = labs_fn()) +
      scale_fill_manual(values = sam_clrs) +
      guides(fill = "none", alpha = "none") +
      labs(title = args$ttl, y = y_ttl) +
      base_theme +
      theme(
        aspect.ratio     = 1,
        panel.background = element_rect(fill = NA, color = "grey75", linewidth = ln_pt),
        axis.title.x     = element_blank(),
        axis.text.x      = element_text(angle = 45, hjust = 1)
      )
  })

plot_grid(
  plotlist = plts[1:2],
  nrow  = 1,
  align = "h",
  axis  = "tb"
)
```

<br>

The fraction of total counts assigned to CHIKV probes is shown for
all LECs (left) or CHIKV+ LECs (right).
The number of cells plotted is shown above each boxplot.

```{r "CHIKV SIGNAL SAMPLES 2", fig.width = 10, fig.height = 4, out.height = "40%"}
plot_grid(
  plotlist = plts[3:4],
  nrow  = 1,
  align = "h",
  axis  = "tb"
)
```

<br>

<br>

### CHIKV signals are higher for sgRNA probes

Total counts assigned to sgRNA, 5' (nsP genes), and
negative strand CHIKV probes is shown for all LECs (left) or
CHIKV+ LECs (right).
The number of cells plotted is shown above each boxplot.

```{r "CHIKV SIGNAL TYPES", fig.width = 15, fig.height = 4}
# CHIKV signal colors
key_nms <- set_names(
  chikv_feats,
  str_c("frac_", chikv_feats)
)

chikv_clrs <- set_names(c("#D7301F", "#0072B2", "#009E73"), names(key_nms))

args <- list(
  pattern = c(rep("CHIKV_", 2), rep("frac_CHIKV_", 2)),
  ttl    = c("LECs", "CHIKV+ LECs", "LECs", "CHIKV+ LECs"),
  y_ttl  = c(rep("CHIKV counts", 2), rep("% CHIKV counts", 2)),
  pseudo = c(rep(1, 2), rep(0.01, 2)),
  filt   = c(FALSE, TRUE, FALSE, TRUE),
  pct    = c(rep(FALSE, 2), rep(TRUE, 2))
)

plts <- args %>%
  pmap(~ {
    args <- list(...)
    
    dat <- type_dat %>%
      filter(cell_type == "LEC") %>%
      pivot_longer(starts_with(args$pattern, ignore.case = FALSE))
    
    lvls <- chikv_feats
    
    if (grepl("^frac_", args$pattern)) lvls <- str_c("frac_", chikv_feats)
    
    clrs <- set_names(
      unname(chikv_clrs),
      lvls
    )
    
    dat <- dat %>%
      mutate(name = fct_relevel(name, lvls))
    
    if (args$filt) {
      dat <- dat %>%
        filter(chikv_class == "CHIKV-high")
      
      y_ttl <- args$y_ttl
    } else {
      y_ttl <- str_c(args$y_ttl, " + ", args$pseudo)
      
      if (args$pct) y_ttl <- str_c(args$y_ttl, " + ", pseudo * 100)
    }
    
    labs_fn <- waiver
    
    if (args$pct) labs_fn <- label_percent
    
    n_dat <- dat %>%
      group_by(sample) %>%
      summarize(n = n_distinct(.cell_id), .groups = "drop") %>%
      mutate(n_lab = label_comma()(n))
    
    dat %>%
      ggplot(aes(sample, value + args$pseudo, fill = name)) +
      geom_boxplot(
        outlier.size = 0.1,
        outlier.alpha = 1
      ) +
      geom_text(
        aes(y = Inf, label = n_lab, fill = NULL),
        data  = n_dat,
        color = "black",
        vjust = 1.25
      ) +
      scale_y_log10(expand = expansion(c(0.05, 0.2)), labels = labs_fn()) +
      scale_fill_manual(values = clrs) +
      labs(title = args$ttl, y = y_ttl) +
      base_theme +
      theme(
        aspect.ratio     = 0.5,
        plot.margin      = margin(15, 15, 15, 30, "pt"),
        panel.background = element_rect(fill = NA, color = "grey75", linewidth = ln_pt),
        legend.title     = element_blank(),
        axis.title.x     = element_blank(),
        axis.text.x      = element_text(angle = 45, hjust = 1)
      )
  })

wrap_plots(
  plts[1:2],
  guides = "collect",
  nrow = 1
)
```

<br>

The fraction of total counts assigned to sgRNA, 5' (nsP genes), and
negative strand CHIKV probes is shown below for all LECs (left) or
CHIKV+ LECs (right).
The number of cells plotted is shown above each boxplot.

```{r "CHIKV SIGNAL TYPES 2", fig.width = 15, fig.height = 4}
wrap_plots(
  plts[3:4],
  guides = "collect",
  nrow = 1
)
```

<br>

The ratio of sgRNA counts / 5' counts is shown below for LECs with at least 1
5' count.
The number of cells plotted and p values are shown above each boxplot.
P values comparing the CHIKV WT sample with all other samples were 
calculated using a wilcoxon rank sum test with Benjamini-Hochberg correction.

```{r "CHIKV sgRNA RATIO", fig.width = 15, fig.height = 4}
# Format data for plotting
dat <- type_dat %>%
  filter(cell_type == "LEC", CHIKV_5 > 0) %>%
  mutate(sgRNA_ratio = (CHIKV_sgRNA) / (CHIKV_5))

# Format N labels
n_dat <- dat %>%
  group_by(sample) %>%
  summarize(n = n_distinct(.cell_id), .groups = "drop") %>%
  mutate(n_lab = str_c("n = ", label_comma()(n)))

# Calculate p-values
p_dat <- pairwise.wilcox.test(
  dat$sgRNA_ratio, dat$sample,
  p.adjust.method = "BH",
  alternative     = "less",
  exact           = FALSE
) %>%
  tidy() %>%
  filter(group2 == sam_lvls[1]) %>%
  rowwise() %>%
  mutate(
    p_lab = .format_pvalue(p.value)
    # p_lab = str_c("italic(p) == ", p_lab)
  ) %>%
  ungroup() %>%
  rename(sample = group1)

# Create boxplots
dat %>%
  ggplot(aes(sample, sgRNA_ratio, fill = sample)) +
  geom_boxplot(outlier.size = 0.65) +
  geom_text(
    aes(y = Inf, label = n_lab, fill = NULL),
    data  = n_dat,
    color = "black",
    vjust = 1.25,
    size  = 9 / .pt
  ) +
  geom_text(
    aes(y = Inf, label = p_lab, fill = NULL),
    data  = p_dat,
    parse = TRUE,
    color = "black",
    vjust = 3,
    size  = 9 / .pt
  ) +
  geom_hline(yintercept = 1, linetype = 2) +
  scale_fill_manual(values = sam_clrs) +
  scale_y_log10(expand = expansion(c(0.05, 0.2))) +
  labs(y = "sgRNA / 5' counts") +
  base_theme +
  theme(
    aspect.ratio    = 1,
    legend.position = "none",
    axis.text.x     = element_text(angle = 45, hjust = 1),
    axis.title.x    = element_blank()
  )
```


```{r "CHIKV+ CELL TYPES", fig.width = 15, fig.height = 4, eval = FALSE}
# Fraction CHIKV reads boxplots
pseudo <- 0.01

type_dat %>%
  group_by(cell_type, sample, rep) %>%
  summarize(
    frac = sum(chikv_class == "CHIKV-high") / n(),
    .groups = "drop"
  ) %>%
  ggplot(aes(cell_type, frac, fill = cell_type, alpha = as.character(rep))) +
  geom_text(
    aes(y = Inf, label = n_lab, alpha = NULL, fill = NULL),
    data = n_dat,
    size = txt_pt1 / .pt,
    vjust = 1.25
  ) +
  geom_col(position = "dodge") +
  facet_wrap(~ sample) +
  scale_y_continuous(expand = expansion(c(0.05, 0.2)), labels = label_percent()) +
  scale_fill_manual(values = typ_clrs) +
  scale_alpha_manual(values = c("1" = 1, "2" = 0.75, "3" = 0.5, "4" = 0.25)) +
  guides(fill = "none", alpha = "none") +
  labs(y = str_c("% CHIKV+ cells")) +
  base_theme +
  theme(
    panel.background = element_rect(fill = NA, color = "grey75", linewidth = ln_pt),
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1)
  )
```

```{r "CHIKV SIGNAL LEC TYPES", fig.width = 15, fig.height = 4, eval = FALSE}
lec_dat <- plt_dat %>%
  filter(lec_type != "other") %>%
  group_by(sample) %>%
  mutate(rep = as.numeric(factor(fov))) %>%
  ungroup() %>%
  mutate(
    lec_type = fct_reorder(lec_type, frac_CHIKV, mean, .desc = TRUE),
    sample = fct_relevel(sample, sam_lvls)
  )

n_dat <- lec_dat %>%
  group_by(sample, lec_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    n_lab = if_else(
      n > 1000,
      str_c(round(n / 1000, 1), "k"),
      label_comma()(n)
    )
  )

# Fraction CHIKV reads boxplots
pseudo <- 0.01

lec_dat %>%
  ggplot(aes(lec_type, frac_CHIKV + pseudo, fill = lec_type, alpha = as.character(rep))) +
  geom_boxplot(
    outlier.size = 0.1,
    outlier.alpha = 1
  ) +
  geom_text(
    aes(y = Inf, label = n_lab, alpha = NULL, fill = NULL),
    data = n_dat,
    size = txt_pt1 / .pt,
    vjust = 1.25
  ) +
  facet_wrap(~ sample) +
  scale_y_log10(expand = expansion(c(0.05, 0.2)), labels = label_percent()) +
  scale_fill_manual(values = lec_clrs) +
  scale_alpha_manual(values = c("1" = 1, "2" = 0.75, "3" = 0.5, "4" = 0.25)) +
  guides(fill = "none", alpha = "none") +
  labs(y = str_c("% CHIKV counts + ", pseudo * 100)) +
  base_theme +
  theme(
    panel.background = element_rect(fill = NA, color = "grey75", linewidth = ln_pt),
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1)
  )
```

```{r "CHIKV+ LEC TYPES", fig.width = 15, fig.height = 4, eval = FALSE}
# Fraction CHIKV reads boxplots
pseudo <- 0.01

lec_dat %>%
  group_by(lec_type, sample, rep) %>%
  summarize(
    frac = sum(chikv_class == "CHIKV-high") / n(),
    .groups = "drop"
  ) %>%
  ggplot(aes(lec_type, frac, fill = lec_type, alpha = as.character(rep))) +
  geom_text(
    aes(y = Inf, label = n_lab, alpha = NULL, fill = NULL),
    data = n_dat,
    size = txt_pt1 / .pt,
    vjust = 1.25
  ) +
  geom_col(position = position_dodge2(preserve = "single", width = 1, padding = 0)) +
  facet_wrap(~ sample) +
  scale_y_continuous(expand = expansion(c(0.05, 0.2)), labels = label_percent()) +
  scale_fill_manual(values = lec_clrs) +
  scale_alpha_manual(values = c("1" = 1, "2" = 0.75, "3" = 0.5, "4" = 0.25)) +
  guides(fill = "none", alpha = "none") +
  labs(y = str_c("% CHIKV+ cells")) +
  base_theme +
  theme(
    panel.background = element_rect(fill = NA, color = "grey75", linewidth = ln_pt),
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1)
  )
```

<br>

<br>

### CHIKV+ LECs {.tabset .tabset-pills}

CHIKV+ LECs are shown for LN sections.
High-resolution images showing all cells are included at the bottom

```{r "CHIKV+ LEC types", fig.width = 17, fig.height = 11}
fig <- .create_class_figure(
  bar_dat       = xen_dat,
  plot_dat      = small,
  class_clmn    = "chikv_class",
  class         = "CHIKV-high",
  celltype_clmn = "lec_type",
  clrs          = lec_clrs
)

fig
```

```{r, fig.width = 20, fig.height = 10, results = "asis", out.height = "50%"}
ag_types <- xen_dat$lec_Ag_type %>%
  unique() %>%
  grep("other|CHIKV-low", ., invert = TRUE, value = TRUE)

fovs %>%
  walk(~ {
    cat("\n\n####", .x, "\n\n")
    
    fig <- xen_dat %>%
      .plot_full_fov(
        fov        = .x,
        color      = "lec_chikv_type",
        clrs       = lec_clrs,
        trace      = list(ag_types),
        trace_clr  = NA,
        trace_size = 2
      )
    
    print(fig)
    cat("\n\n---\n\n<br>\n\n<br>\n\n")
  })
```

## Antigen archiving

### LECs show strongest enrichment of Ag-high cells {.tabset .tabset-pills}

Ag counts are shown below for Ag+ LECs.
The number of Ag+ LECs and total LECs is shown for each sample.
Replicates were pooled together.

```{r, fig.width = 20, fig.height = 4}
# Format N labels
n_dat <- type_dat %>%
  group_by(sample, cell_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    n_lab = if_else(
      n > 1000,
      str_c(round(n / 1000, 1), "k"),
      label_comma()(n)
    )
  )

# Format histogram data
hist_dat <- type_dat %>%
  filter(cell_type == "LEC")

# Format N labels
n_dat <- hist_dat %>%
  group_by(cell_type, sample) %>%
  summarize(
    n       = n(),
    n_ag    = sum(Ag_km == "Ag-high"),
    .groups = "drop"
  ) %>%
  mutate(n_lab = str_c(
    "n = ", label_comma()(n_ag),
    " / ", label_comma()(n)
  ))

# Create histogram
hist_dat %>%
  filter(Ag_tag > 0) %>%
  
  ggplot(aes(Ag_tag, fill = Ag_km)) +
  geom_histogram(bins = 30) +
  geom_text(
    aes(1, Inf, label = n_lab, alpha = NULL, fill = NULL),
    data = n_dat,
    color = "black",
    hjust = 0,
    vjust = 1.3
  ) +
  facet_wrap(~ sample, nrow = 1) +
  scale_fill_manual(values = c(`Ag-low` = "#56B4E9", `Ag-high` = "#D7301F")) +
  scale_x_log10() +
  scale_y_continuous(expand = expansion(c(0.05, 0.1))) +
  labs(x = "Ag counts", y = "number of Ag+ LECs") +
  base_theme +
  theme(
    legend.position = "bottom",
    legend.title    = element_blank()
  )
```

<br>

<br>

Ag-high cells are shown for LN sections.
High-resolution images showing all cells are included at the bottom

```{r "SLIDE 2: Ag+ cell types", fig.width = 17, fig.height = 11}
.create_class_figure(
  bar_dat       = xen_dat,
  plot_dat      = small,
  class_clmn    = "Ag_km",
  class         = "Ag-high",
  celltype_clmn = "cell_type",
  clrs          = typ_clrs,
  top_types     = top_types
)
```

```{r, fig.width = 20, fig.height = 10, results = "asis", out.height = "50%"}
ag_types <- xen_dat$cell_type %>%
  unique() %>%
  grep("other|Ag-low", ., invert = TRUE, value = TRUE)

fovs %>%
  walk(~ {
    cat("\n\n####", .x, "\n\n")
    
    fig <- xen_dat %>%
      .plot_full_fov(
        fov        = .x,
        color      = "Ag_type",
        clrs       = typ_clrs,
        trace      = list(ag_types),
        trace_clr  = NA,
        trace_size = 1.5
      )
    
    print(fig)
    cat("\n\n<br>\n\n<br>\n\n")
  })
```

### Ag signal is higher in CHIKV 181/25 samples

The fraction of total counts assigned to Ag-tag probes (Ag-tag counts / total counts)
for each sample is shown on the left.
The fraction of total cells that are Ag-high is shown on the right.
Replicates are plotted separately.

```{r}
dat <- xen_dat %>%
  group_by(fov, rep, sample) %>%
  summarize(
    n_ag = sum(Ag_tag),
    n_tot = sum(nCount_Xenium + nCount_Xenium_custom),
    .groups = "drop"
  ) %>%
  mutate(
    frac_ag = n_ag / n_tot,
    sample  = fct_relevel(sample, sam_lvls)
  )

n_dat <- dat %>%
  group_by(sample) %>%
  summarize(
    n_tot = sum(n_tot),
    n_ag  = sum(n_ag),
    .groups = "drop"
  ) %>%
  mutate(
    n_tot = case_when(
      n_tot > 1000000 ~ str_c(round(n_tot / 1000000, 1), "M"),
      n_tot > 1000    ~ str_c(round(n_tot / 1000, 1), "k"),
      TRUE            ~ as.character(n_tot)
    ),
    n_ag = case_when(
      n_ag > 1000000 ~ str_c(round(n_ag / 1000000, 1), "M"),
      n_ag > 1000    ~ str_c(round(n_ag / 1000, 1), "k"),
      TRUE           ~ as.character(n_ag)
    ),
    n_lab = str_c(n_ag, " / ", n_tot)
  )

frac_ag_counts <- dat %>%
  ggplot(aes(sample, frac_ag, fill = sample, alpha = as.character(rep))) +
  geom_col(position = position_dodge2(preserve = "single")) +
  geom_text(
    aes(y = Inf, label = n_lab, fill = NULL, alpha = NULL),
    data = n_dat,
    color = "black",
    vjust = 1.25
  ) +
  scale_fill_manual(values = sam_clrs) +
  scale_alpha_manual(values = c("1" = 1, "2" = 0.75, "3" = 0.5, "4" = 0.25)) +
  scale_y_continuous(label = label_percent(), expand = expansion(c(0.05, 0.15))) +
  guides(alpha = "none") +
  labs(y = "% Ag-tag counts") +
  base_theme +
  theme(
    aspect.ratio    = 0.35,
    legend.position = "none",
    axis.title.x    = element_blank(),
    axis.text.x     = element_text(angle = 45, hjust = 1)
  )
```

```{r}
dat <- xen_dat %>%
  group_by(fov, rep, sample) %>%
  summarize(
    n_ag  = sum(Ag_km == "Ag-high"),
    n_tot = n(),
    .groups = "drop"
  ) %>%
  mutate(
    frac_ag = n_ag / n_tot,
    sample  = fct_relevel(sample, sam_lvls)
  ) %>%
  group_by(sample) %>%
  filter(sum(frac_ag) > 0) %>%
  ungroup()

n_dat <- dat %>%
  group_by(sample) %>%
  summarize(
    n_tot = sum(n_tot),
    n_ag  = sum(n_ag),
    .groups = "drop"
  ) %>%
  mutate(
    n_tot = case_when(
      n_tot > 1000000 ~ str_c(round(n_tot / 1000000, 1), "M"),
      n_tot > 1000    ~ str_c(round(n_tot / 1000, 1), "k"),
      TRUE            ~ as.character(n_tot)
    ),
    n_ag = case_when(
      n_ag > 1000000 ~ str_c(round(n_ag / 1000000, 1), "M"),
      n_ag > 1000    ~ str_c(round(n_ag / 1000, 1), "k"),
      TRUE           ~ as.character(n_ag)
    ),
    n_lab = str_c(n_ag, " / ", n_tot)
  )

frac_ag_cells <- dat %>%
  ggplot(aes(sample, frac_ag, fill = sample, alpha = as.character(rep))) +
  geom_col(position = position_dodge2(preserve = "single")) +
  geom_text(
    aes(y = Inf, label = n_lab, fill = NULL, alpha = NULL),
    data = n_dat,
    color = "black",
    vjust = 1.25
  ) +
  scale_fill_manual(values = sam_clrs) +
  scale_alpha_manual(values = c("1" = 1, "2" = 0.75, "3" = 0.5, "4" = 0.25)) +
  scale_y_continuous(label = label_percent(), expand = expansion(c(0.05, 0.15))) +
  guides(alpha = "none") +
  labs(y = "% Ag-high cells") +
  base_theme +
  theme(
    aspect.ratio    = 1,
    legend.position = "none",
    axis.title.x    = element_blank(),
    axis.text.x     = element_text(angle = 45, hjust = 1)
  )
```

```{r "TOTAL AG COUNTS", fig.width = 15, fig.height = 4}
plot_grid(
  frac_ag_counts, frac_ag_cells,
  nrow  = 1,
  align = "h",
  axis  = "tb",
  rel_widths = c(1, 0.4)
)
```

<br>

The fraction of LECs identified as Ag-high is shown on the left.
Ag counts are shown on the right.
The number of cells plotted is shown for each panel.

```{r}
dat <- xen_dat %>%
  filter(cell_type == "LEC") %>%
  group_by(fov, rep, sample, cell_type) %>%
  summarize(
    n_ag    = sum(Ag_km == "Ag-high"),
    n_tot   = n(),
    .groups = "drop"
  ) %>%
  mutate(
    frac_ag = n_ag / n_tot
  ) %>%
  group_by(sample) %>%
  filter(sum(frac_ag) > 0) %>%
  ungroup()

n_dat <- dat %>%
  group_by(sample) %>%
  summarize(
    n_tot = sum(n_tot),
    n_ag  = sum(n_ag),
    .groups = "drop"
  ) %>%
  mutate(
    n_tot = case_when(
      n_tot > 1000000 ~ str_c(round(n_tot / 1000000, 1), "M"),
      n_tot > 1000    ~ str_c(round(n_tot / 1000, 1), "k"),
      TRUE            ~ as.character(n_tot)
    ),
    n_ag = case_when(
      n_ag > 1000000 ~ str_c(round(n_ag / 1000000, 1), "M"),
      n_ag > 1000    ~ str_c(round(n_ag / 1000, 1), "k"),
      TRUE           ~ as.character(n_ag)
    ),
    n_lab = str_c(n_ag, " / ", n_tot)
  )

frac_ag_lecs <- dat %>%
  ggplot(aes(sample, frac_ag, fill = sample, alpha = as.character(rep))) +
  geom_col(position = position_dodge2(preserve = "single")) +
  geom_text(
    aes(y = Inf, label = n_lab, fill = NULL, alpha = NULL),
    data = n_dat,
    color = "black",
    vjust = 1.25
  ) +
  scale_fill_manual(values = sam_clrs) +
  scale_alpha_manual(values = c("1" = 1, "2" = 0.75, "3" = 0.5, "4" = 0.25)) +
  scale_y_continuous(label = label_percent(), expand = expansion(c(0.05, 0.15))) +
  guides(alpha = "none") +
  labs(y = "% Ag-high LECs") +
  base_theme +
  theme(
    aspect.ratio    = 1,
    legend.position = "none",
    axis.title.x    = element_blank(),
    axis.text.x     = element_text(angle = 45, hjust = 1)
  )
```

```{r "Ag SIGNAL SAMPLES", fig.width = 10, fig.height = 4, out.height = "40%"}
args <- list(
  y      = rep("Ag_tag", 2),
  ttl    = c("LECs", "Ag+ LECs"),
  y_ttl  = rep("Ag counts", 2),
  pseudo = rep(1, 2),
  filt   = c(FALSE, TRUE),
  pct    = c(FALSE, FALSE)
)

plts <- args %>%
  pmap(~ {
    args <- list(...)
    
    dat <- type_dat %>%
      filter(cell_type == "LEC") %>%
      group_by(sample) %>%
      filter(sum(Ag_tag) > 0) %>%
      ungroup()
    
    if (args$filt) {
      dat <- dat %>%
        filter(Ag_class == "Ag-high")
      
      y_ttl <- args$y_ttl
    } else {
      y_ttl <- str_c(args$y_ttl, " + ", args$pseudo)
      
      if (args$pct) y_ttl <- str_c(args$y_ttl, " + ", pseudo * 100)
    }
    
    labs_fn <- waiver
    
    if (args$pct) labs_fn <- label_percent
    
    n_dat <- dat %>%
      group_by(sample) %>%
      summarize(n = n(), .groups = "drop") %>%
      mutate(n_lab = label_comma()(n))
    
    dat %>%
      ggplot(aes(sample, !!sym(args$y) + args$pseudo, fill = sample, alpha = as.character(rep))) +
      geom_boxplot(
        outlier.size = 0.1,
        outlier.alpha = 1
      ) +
      geom_text(
        aes(y = Inf, label = n_lab, fill = NULL, alpha = NULL),
        data  = n_dat,
        color = "black",
        vjust = 1.25
      ) +
      scale_y_log10(expand = expansion(c(0.05, 0.2)), labels = labs_fn()) +
      scale_fill_manual(values = sam_clrs) +
      scale_alpha_manual(values = c(1, 0.75, 0.5, 0.25)) +
      guides(fill = "none", alpha = "none") +
      labs(title = args$ttl, y = y_ttl) +
      base_theme +
      theme(
        aspect.ratio     = 0.75,
        panel.background = element_rect(fill = NA, color = "grey75", linewidth = ln_pt),
        axis.title.x     = element_blank(),
        axis.text.x      = element_text(angle = 45, hjust = 1)
      )
  })

# Create final figure
plot_grid(
  frac_ag_lecs, plts[[1]] + theme(plot.title = element_blank()),
  nrow  = 1,
  align = "h",
  axis  = "tb"
)
```

<br>

<br>

### Ag-high LECs {.tabset .tabset-pills}

Ag-high LECs are shown for LN sections.
High-resolution images showing all cells are included at the bottom

```{r "SLIDE 6: Ag+ LEC types", fig.width = 17, fig.height = 11}
.create_class_figure(
  bar_dat       = xen_dat,
  plot_dat      = small,
  class_clmn    = "Ag_km",
  class         = "Ag-high",
  celltype_clmn = "lec_type",
  clrs          = lec_clrs
)
```

```{r, fig.width = 20, fig.height = 10, results = "asis", out.height = "50%"}
ag_types <- xen_dat$lec_Ag_type %>%
  unique() %>%
  grep("other|Ag-low", ., invert = TRUE, value = TRUE)

fovs %>%
  walk(~ {
    cat("\n\n####", .x, "\n\n")
    
    fig <- xen_dat %>%
      .plot_full_fov(
        fov        = .x,
        color      = "lec_Ag_type",
        clrs       = lec_clrs,
        trace      = list(ag_types),
        trace_clr  = NA,
        trace_size = 2
      )
    
    print(fig)
    cat("\n\n---\n\n<br>\n\n<br>\n\n")
  })
```

## Session info

```{r}
sessionInfo()
```



```{r "LEC-DC INTERACTIONS", eval = FALSE}
fov_id <- "D"

# Pull cell coordinates from object
lec_dc_obj <- xen %>%
  subset(cell_type %in% c("LEC", dcs))

cell_coords <- lec_dc_obj@images[[fov_id]]@boundaries$segmentation@polygons %>%
  imap(~ {
    meta <- xen_dat %>%
      filter(.cell_id == .y) %>%
      dplyr::select(.cell_id, cell_type, Ag_class)
    
    .x@Polygons[[1]]@coords %>%
      as_tibble() %>%
      mutate(.cell_id = .y, .before = x) %>%
      left_join(meta, by = ".cell_id")
  })

lecs <- xen_dat %>%
  filter(cell_type == "LEC" & Ag_class == "Ag-high") %>%
  pull(.cell_id)

# Compare LEC-DC pairs identified using centroid coordinates
lec_dc_pairs <- set_names(xen_dat$lec_id, xen_dat$.cell_id)

mem_dist <- cell_coords %>%
  imap_dfr(~ {
    x <- cell_coords[[.y]] %>%
      mutate(id = as.character(row_number()))
    
    y <- cell_coords[[lec_dc_pairs[[.y]]]] %>%
      mutate(id = as.character(row_number()))
    
    res <- expand_grid(other = x$id, lec = y$id) %>%
      left_join(x, by = c(other = "id")) %>%
      left_join(y, by = c(lec = "id"), suffix = c("", "_lec")) %>%
      mutate(
        mem_dist = sqrt(((x - x_lec)^2) + ((y - y_lec)^2))
      ) %>%
      group_by(.cell_id, lec_id = .cell_id_lec) %>%
      summarize(mem_dist = min(mem_dist), .groups = "drop")
    
    res
  })

# Summarize interacting cells
dist_dat <- xen_dat %>%
  filter(fov == "D", cell_type %in% c("LEC", dcs)) %>%
  left_join(mem_dist, by = c(".cell_id", "lec_id")) %>%
  mutate(LEC_interacting = mem_dist < 0.5)

dist_dat %>%
  group_by(fov, cell_type, Ag_class) %>%
  summarize(
    n         = n(),
    n_intr    = sum(LEC_interacting),
    frac_intr = n_intr / n,
    .groups   = "drop"
  )
```

```{r "PROXIMITY CELL PAIRS", fig.width = 14, fig.height = 7, eval = FALSE}
# Scatter plot with lines connecting closest Ag-high LEC
prox_dat %>%
  ggplot(aes(x_cell, y_cell)) +
  geom_point(size = 1, color = "grey85") +
  geom_point(
    aes(x_cell, y_cell, color = Ag_class),
    data = ~ filter(.x, cell_type == "Siglec-H DC")
  ) +
  geom_point(
    data = ~ filter(.x, cell_type == "LEC" & Ag_class == "Ag-high"),
    color = "black"
  ) +
  geom_segment(
    aes(x = x_cell, xend = x_lec, y = y_cell, yend = y_lec),
    data = ~ filter(.x, cell_type == "Siglec-H DC" & Ag_class == "Ag-high"),
    color = "black"
  ) +
  coord_fixed() +
  base_theme
```

```{r "Ag signal", eval = FALSE}
lec <- lec %>%
  mutate_meta(
    mutate,
    Ag_class_1 = Ag_tag_1 > 0,
    Ag_class_2 = Ag_tag_2 > 0
  )

lec@meta.data %>%
  # ggplot(aes(lec_type, fill = Ag_tag_1, group = as.character(Ag_tag_1))) +
  ggplot(aes(lec_type, fill = Ag_class)) +
  geom_bar(position = "fill") +
  facet_wrap(~ fov)

lec@meta.data %>%
  ggplot(aes(fov, fill = lec_type)) +
  geom_bar(position = "fill")

xen@meta.data %>%
  # ggplot(aes(lec_type, fill = Ag_tag_1, group = as.character(Ag_tag_1))) +
  ggplot(aes(cell_type, fill = Ag_class)) +
  geom_bar(position = "fill") +
  facet_wrap(~ fov)

xen@meta.data %>%
  filter(Ag_class == "Ag-high") %>%
  dplyr::select(fov, cell_type, Ag_class, starts_with("Ag_tag_")) %>%
  pivot_longer(starts_with("Ag_tag_")) %>%
  
  ggplot(aes(value + 1, fill = Ag_class)) + 
  geom_histogram() +
  facet_grid(name ~ cell_type) +
  scale_x_log10()


# lec %>%
#   mutate_meta(
#     mutate,
#     Ag_class = ifelse(Ag_tag_1 > 1, "Ag-high", "Ag-low"),
#   ) %>%
#   plot_violin(
#     "Cavin1",
#     cluster_col = "Ag_class",
#     group_col   = "lec_type",
#     method = "boxplot",
#     notch = TRUE
#   )
```

```{r, eval = FALSE}

# Plot cell types
xen %>%
  plot_scatter("SCT_snn_res.0.5", size = 0.0001, top = c("15", "8", "7", "12"))

xen %>%
  plot_scatter("Lyve1", size = 0.0001, plot_colors = c("lightblue", "white", "red"))

lec %>%
  plot_scatter("SCT_snn_res.3", size = 2, plot_colors = c("lightblue", "red"))

lec %>%
  plot_scatter("lec_type", size = 3, plot_colors = c("lightblue", "red"))

lec %>%
  plot_scatter("Mrc1", size = 3, plot_colors = c("lightblue", "red"))

xen %>%
  plot_scatter("cell_type", group_col = "fov", size = 0.0001)

xen %>%
  plot_scatter(
    # "xenium_custom_mRNA-vaccine-(3-UTR)",
    "xenium_custom_GeoMx-Barcode01-ERCC00142",
    # "xenium_custom_GeoMx-Barcode02-ERCC00144",
    size = 0.00001,
    plot_colors = c("lightblue", "red")
  )

Idents(xen) <- xen$cell_type

clrs <- unique(xen$cell_type)

clrs <- set_names(
  rep("white", length(clrs)),
  clrs
)

clrs["LEC"] <- "red"
clrs["BEC"] <- "lightblue"
clrs[c("cDC1", "cDC2")] <- "green4"

unique(xen$fov) %>%
  map(~ {
    xen %>%
      ImageDimPlot(
        fov = .x,
        size = 2,
        cols = clrs
      )
  }) %>%
  plot_grid(plotlist = .)

gns <- c(
  # "Madcam1",
  # "Marco", "Ackr4",
  # "Zbtb46", "Xcr1", "Itgax",
  "Cavin1", "Ctsd", "Psap", "Grn", "Tgfa",
  "GeoMx-Barcode01-ERCC00142",
  "GeoMx-Barcode02-ERCC00144"
)

xen %>%
  ImageFeaturePlot(
    fov = "A",
    features = gns,
    max.cutoff = "q1",
    size = 0.75,
    cols = c("white", "red")
  )

xen %>%
  ImageDimPlot(
    fov = "A",
    molecules = c("Ptprc", "Madcam1", "Marco", "Ackr4", "Ccr7"),
    nmols = 20000
  )

gns <- c(
  "Marco",
  "Madcam1", "Icam1", "Cd274",
  "Ackr4",
  "Ptx3", "Mrc1",
  "Fgl2"
)

gns %>%
  map(~ {
    lec %>%
      plot_violin(
        .x,
        cluster_col = "SCT_snn_res.3",
        group_col = "lec_type",
        method      = "boxplot",
        top = Inf
      )
  }) %>%
  plot_grid(plotlist = .)


# gns <- c(
#   "Ptprc", "Cd3d", "Cd19",
#   "Pecam1", "Pdpn",
#   "Prox1", "Lyve1",
#   "Ccr7", "Sirpa", "Xcr1",
#   "Itgax", "Itgam", "Siglech"
# )
# gns %>%
#   map(~ {
#     xen %>%
#       plot_violin(
#         .x,
#         cluster_col = clst_clmn,
#         group_col   = "cell_type",
#         method      = "boxplot",
#         top = Inf
#       )
#   }) %>%
#   plot_grid(plotlist = .)
# gns %>%
#   map(~ {
#     xen %>%
#       plot_violin(
#         .x,
#         cluster_col = "cell_type",
#         method      = "boxplot",
#         top = Inf
#       )
#   }) %>%
#   plot_grid(plotlist = .)
```

```{r "SLIDE 2: LEC type sections OLD", fig.width = 26, fig.height = 5, eval = FALSE}
new_clrs <- typ_clrs
new_clrs[!names(new_clrs) %in% c("LEC", "cDC2", "cDC1")] <- "grey85"

plts <- fovs %>%
  map(~ {
    p <- small %>%
      ImageDimPlot(
        fov = .x,
        size = 1.5,
        cols = new_clrs,
        crop = TRUE,
        dark.background = FALSE
      ) +
      labs(tag = .x) +
      base_theme +
      theme_void() +
      theme(
        panel.border = element_rect(color = ln_col, fill = NA, linewidth = ln_pt),
        legend.title = element_blank(),
        plot.tag.position = c(0.05, 0.95),
        plot.tag = element_text(size = ttl_pt2, face = "bold")
      )
    
    plot_data <- p$data
    plot_data <- plot_data[order(plot_data$ident, decreasing = TRUE), ]
    p$data    <- plot_data
    
    brs <- xen@meta.data %>%
      filter(fov == .x) %>%
      mutate(cell_type = fct_infreq(cell_type)) %>%
      ggplot(aes(y = fov, fill = cell_type)) +
      geom_bar(position = "fill") +
      scale_fill_manual(values = new_clrs) +
      base_theme +
      theme_void() +
      theme(
        legend.position = "none",
        axis.title      = element_blank(),
        axis.text       = element_blank()
      )
    
    list(p, brs)
  })

fig <- map(plts, pluck, 1) %>%
  wrap_plots(
    nrow = 1,
    guides = "collect"
  ) &
  theme(legend.direction = "horizontal")

fig
```
