---
output:
  html_document:
    toc:            true
    toc_float:      true
    toc_depth:      4
    df_print:       "paged"
    code_folding:   "hide"
    self_contained: true
    highlight:      "kate"
editor_options: 
  chunk_output_type: inline
---

```{r "setup reviews"}
# Default chunk options
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo    = FALSE,
  dpi     = 300
)

generate_plots <- TRUE
```

## Reviewer 1

In this manuscript, the authors aim to characterize the role of Lymphatic Endothelial Cell (LEC) populations in immunization, with a focus on their capacity to capture and store antigens for extended periods (“antigen archiving”). By utilizing DNA-barcoded antigens, single-cell sequencing, and machine learning models, they claim to have identified predictive features of LECs particularly competent in antigen archiving. Additionally, they demonstrate that LECs previously exposed to an antigen are more adept at acquiring and storing another antigen upon secondary immunization.
While the use of DNA-barcoded antigens combined with scRNA-Seq is an innovative approach for studying antigen archiving, there are several methodological concerns and limitations in the interpretation of the results that should be addressed.

<br>

### Comment 1 (a) {#R1C1}

[Figure 2]

<details>
  <summary>full comment</summary>
The main underlined result consists of the identification of a gene signature,
that can predict Ag-archiving capacity of a cell and correlate it with the
duration of antigen retention [236-239].
However, this prediction system has several limitations:
    
* The manuscript lacks a statistical test (e.g. importance plot) to validate
  the correlation between predicted responses and observed antigen archiving,
  which could have been done using the data shown in Fig. 2B.
* It is not clear what the LECs identified as “Ag-competent” are.
  In line 226-227 they are defined as having “low amounts of antigen but is
  predicted to have high antigen levels by our classification models”.
  In Fig. 2B, when comparing the results of the predictive model with the
  observed Ag retention, the discrepancy (Ag-high vs Ag-competent) increases
  significantly as time passes from the 14 days (the time point of the
  samples that trained the algorithm).
  This can have different explanations:
      
  * The prediction dramatically drop its accuracy as we move from the
    training conditions.
    So these cells are misclassified by the algorithm
  * In Fig. 2C it is shown that this cells better correlate with the
    Ag-high module score, however, in the manuscript there are not
    functional evidence that support the hypothesis that this cells are
    really better in Ag-archiving.
    It is plausible that this cells are good in capturing the Ag but not in
    retaining it.
    This could explain why their number increases with time (Fig. 2B).
  * To summarize this point, the authors exploit an algorithm to predict
    which cells are better in Ag-archiving,
    however, in this manuscript it is not shown that the identified cells
    are really better in capturing and retain the Ags.
</details>

<br>

* The manuscript lacks a statistical test (e.g. importance plot) to validate
  the correlation between predicted responses and observed antigen archiving.
  
  * **P values indicating overall feature (gene) importance are included in
    [Table S2]**

* The authors use an algorithm to predict which cells are better at archiving
  Ag, however, it is not shown that the identified cells are really better at
  capturing and retaining the antigen.
  
  * **Include a bargraph similar to [Figure 2]B that includes dual immunized
    mice, to show that predicted Ag-competent cells are taking up Ag after the
    2nd immunization.**
  * **Need to explain better why there is increased dual antigen positive LECs
    and that we think it is because the increased inflammatory response in LECs
    results in increased archiving. Re-work the text around this figure.**

<br>

[Figure 2]B was modified to include the dual vaccinated sample.
The fraction of cells predicted to be Ag-competent is shown for each LEC subset.
For dual-immunized mice, Ag-high refers to cells that took up both antigens.

```{r "dual vax bars", fig.width = 6, fig.height = 4, out.width = "50%", eval = generate_plots}
# Format data
lvls <- c("Ag-low", "Ag-competent", "double-high", "Ag-high")

new_clrs <- pred_clrs
new_clrs["double-high"]  <- pred_clrs["Ag-competent"]
new_clrs["Ag-competent"] <- "grey75"

dat <- ex_mod_bx_dat %>%
  filter(subtype %in% rf_typs) %>%
  mutate(
    pred_grp = Ag_class_dual,
    pred_grp = recode(pred_grp, `single-high` = "Ag-high"),
    tm = "dual"
  ) %>%
  bind_rows(rf_bar_dat) %>%
  mutate(pred_grp = fct_relevel(pred_grp, rev(lvls)))

# Create bargraphs
dat %>%
  ggplot(aes(tm, fill = pred_grp)) +
  geom_bar(position = "fill", alpha = 0.85, key_glyph = draw_key_point) +
  geom_bar(position = "fill", key_glyph = draw_key_point) +
  facet_grid(~ subtype) +
  scale_fill_manual(values = new_clrs) +
  scale_y_continuous(breaks = c(0, 0.5, 1)) +
  guides(fill = guide_legend(override.aes = list(shape = 22, size = 5))) +
  labs(y = "fraction of cells", x = "days post vaccination") +
  bar_thm +
  theme(
    aspect.ratio = 1.2,
    legend.title = element_blank(),
    legend.position = "bottom"
  )
```

<br>

### Comment 2 (a) {#R1C2}

[Figure 2]

In [Figure 2]C, results indicate that the predicted antigen-competent cells
exhibit higher expression of the antigen-high module.
However, as the random forest algorithm was trained on the antigen score at day
14, for this time point the outcome results strongly biased by the input data.
For this reason, it is not clear why this comparison is shown.

* **Not all day 14 LECs were used to train the models.
  The cells used for training have been removed from the [Figure 2] plots.**

<br>

### Comment 3 (t)

In line 348 sentence “Taken together, our spatial transcriptomics results
(Figure 4A-D, S6, S7)” is repeated.

* **Remove repeated line.**

<br>

### Comment 4 (e)

[Figure 5]

<details>
  <summary>full comment</summary>
    In Fig. 5G-H it is shown how CHIKV infection reduce the amount of Ag.
    In line 396 it is written “To experimentally test the accuracy of these
    predictions”.
    However, this result only shows that CHIKV reduces Ag uptake by LECs,
    since the readout is limited at 2 days post-immunization.
    Moreover, it is not shown how different populations are differently
    perturbed by the infection, which is one of the main results of the
    prediction.
</details>

<br>

For [Figure 5], the experiment uses a 2 day timepoint to assess archiving,
but this timepoint will only show differences in Ag uptake.
It is not shown how Ag archiving is impacted for different populations during
CHIKV infection.
  
* **Perform CHIKV flow experiment using a better panel to look at
  archiving in different LEC subsets.**
* **CHIKV infection for 24 hours followed by ova488 immunization for 2 days, or
  2 weeks with OT1 transfer to look at antigen presentation.**

<br>

### Comment 5 (t)

[Figure S9]

In lines 428-430 it is reported “Overall these results illustrate the utility of
our approach to investigate antigen archiving in human samples and other systems
where inclusion of antigen-DNA conjugates is not feasible”.
However, this conclusion is not supported by the data (Fig. S9).
The reported results only show a transcriptional overlap between Ag-competent
cLECs and fLECs (but not for collecting LECs) between human and mouse.
It is therefore not possible to state that these cells are really able to
archive the Ag.
To do so, it is necessary to validate that, in human, these cells are capable of
retaining Ags for long time (e.g. ex vivo assay).
Otherwise, the statement needs to be strongly resized.

* **Reword conclusions about the human samples and archiving to be more
  suggestive rather than conclusive.**
* **Add in existing human LN data showing that RBD protein is present in lung
  LECs and not pancreatic LECs from human LNs.**

<br>

### Comment 6 (a, t) {#R1C6}

[Figure 4] [Figure S4]

<details>
  <summary>full comment</summary>
In lines 509-513 the authors claim: “We found higher antigen levels in both LEC
subsets and DC subsets, CCR7hi DC and cDC2 Tbet-, upon a second antigen exposure
when compared to single-immunized mice (Figure 3D,G, S4A-C).
This supports our published data demonstrating that migratory DCs and cDC1s are
required for cross-presentation of archived antigens during the time frame of
LEC apoptosis.
cDC1 cross-presentation of cell-associated antigens is partly attributed to
their ability to internalize apoptotic cells or cell debris”.
However, in this manuscript any evidence was shown, regarding the role of cDC1
in antigen exchange with LECs. Data was instead provided for CCR7hi DCs and
Tbet- DCs.
</details>

<br>

Authors claim higher antigen levels in CCR7hi DCs and cDC2 Tbet- upon a second
antigen exposure, which supports their published data demonstrating that
migratory DCs and cDC1s are required for cross-presentation of archived antigens
during the time frame of LEC apoptosis.
However, no evidence is shown regarding the role of cDC1s in antigen exchange
with LECs.

* **CCR7hi DCs might include some cDC1s, but since we do not distinguish between
  cDC1 and cDC2 in this subset we can clarify the text.
  We can also further investigate the CCR7hi DCs to see if we can
  identify cDC1s in this populations (working on this).**

<br>

### Comment 7 (e, t)

Generally, the study would benefit from functional assays to validate the
findings from scRNA-Seq and spatial transcriptomics. For example, perturbing
genes associated with the high-archiving signature in specific LEC subsets could
validate their supposed roles in antigen acquisition or retention.
The analysis presented without additional functional studies does not
demonstrate claims on the function of the cells identified by RNA sequencing
analysis.

* **Downplay the language and include in vitro studies, described in more detail
  for reviewer 2.**
* **Cite new experiments with ova488 and CHIKV looking at LEC subsets by flow
  cytometry (comment 4).**

---

<br>

<br>

## Reviewer 2

This article describes the duration and maintenance of antigen archives by
different subsets of LEC and the associated genetic programs.
The topic is novel, and the primary experimental approach (DNA tagging of Ag for
later detection through scRNASeq) is innovative, robust, and elegant.
However, the reviewer has some concerns regarding the process's relevance, the
model's translatability, and the robustness of some key manuscript conclusions.
As detailed below, these doubts could be addressed through additional
experiments.

<br>

### Comment 1 (e, t)

[Figure 2]

In line 236, could the observed signatures be associated with other relevant
features unrelated to the archive antigen, such as the activation status of the
cell?
This needs to be clarified and further discussed.
Could the authors demonstrate in vitro that the KO of some of the described
genes/pathways is affecting the antigen archived process?

**Connect in vitro experiments with Ag archiving gene signatures.
Relevant in vitro data includes:**

  * **Nystatin - Ag lower**
  * **Clathrin inhibitor - Ag lower**
  * **EIPA (micropinocytosis inhibitor) - Ag higher**
  * **cholesterol - Ag higher. [Haidar et al](https://www.jbc.org/article/S0021-9258(20)76831-6/fulltext)
    find that inactivation of CTSD results in accumulation of cholesterol in
    late endosomes/lysosomes, potentially linking the in vitro cholesterol data
    with the Ag-high gene module**
    
<br>

### Comment 2 (t)

[Figure 3]

It is unclear why the authors decided to evaluate the effect of repeated
immunization on day 42.
This effect is most likely affected by the previous inflammatory status, a
process well characterized in other vaccine models.
We suggest repeating this experiment with a longer time frame to discard an
effect associated with inflammation.

* **Discuss previous data looking at LEC expansion and contraction and Autumn’s
  paper.**

<br>

### Comment 3 (e, t)

[Figure 4]

<details>
  <summary>full comment</summary>
As the authors previously demonstrated that DCs acquire Ag following LECs
apoptosis and consequent Ag release, how can they exclude the possibility that
the Ag in DCs is phagocyted early following vaccination (either at the injection
site or upon drainage in the LN) without involving LECs? Thus, how relevant are
LECs for DCs Ag uptake and immune activation?
Could the authors, for example, test this by vaccinating irradiated mice and
causing Ag accumulation only in the LECs but not in the DCs and later measure
immune activation (at DCs and T level) following reconstitution with
hematopoietic cells not previously exposed to the Ag by vaccination?
Or they could measure the functional immune responses (T cell activation or
proliferation) to the vaccine in the CHIKV model, where LECs fail to archive the
Ag?
</details>

<br>

* How can the authors exclude the possibility that the Ag in DCs is phagocyted
  early following vaccination (either at the injection site or upon drainage in
  the LN) without involving LECs?
  
  * **Discuss the duration of antigen and half-life of DCs and LEC/DC antigen
    presentation studies.**

* Could the authors measure the functional immune responses (T cell activation
  or proliferation) to the vaccine in the CHIKV model, where LECs fail to
  archive the Ag?
  
  * **Plot antigen levels compared to antigen presentation over time or
    reference paper showing this?**

<br>

### Comment 4 (e, t)

[Figure 1]

The statement that “LECs are the predominant cell population that acquires and
archives antigens in the draining lymph node” (line 137) seems too general.
Most experiments measure the DNA tag rather than OVA (with only one early
timepoint measurement).
To what extent is the observed process due to the DNA nature of the Ag?
Do LECs express DNA receptors such as TLRs? Would the results differ for protein
Ag?
Is OVA cleaved from DNA, and if so, when?
Is OVA degraded over time? Is archiving associated with an OVA-specific immune
response?

* **Reference other papers demonstrating different antigens and degradation of ova
  and DNA tag.
  Show antigen duration by flow and OT1 transfer (Beth has these data).**

<br>

### Comment 5 (t)

[Figure 1]

[Figure 1]E shows a general decrease in Ag across populations over time.
Does this mean that only a small group of LECs retain the Ag while the others
lose or degrade it?
Is the amount of archived Ag positively correlated with memory or effector
immunity?

* **Presumably, reference Autumns paper showing that beyond the time frame of
  antigen archiving protection is lost.**

<br>

### Comment 6 (t)

[Figure S3]

<details>
  <summary>full comment</summary>
In Fig S3C-D, Ag-high specific genes are found at similar levels in Ag-high from
vaccinated mice and naïve LECs.
Were the Ag-high LECs compared to total naïve LEC or to a cluster of
Ag-high-similar LEC identified also in naïve mice?
If the former, why do the Ag-high genes distinguish Ag-high and Ag-low LECs, but
not Ag-high and naïve cells?
If the latter, how were Ag-high LECs identified in naïve mice?
Additionally, for fLEC, some Ag-high genes increase over Ag stimulation (day 2
vs. day 0), differently from what is stressed in the text.
This seems in contrast with the conclusion of Figure 2, that the archiving
capacity is already present in naïve cells at a transcriptomic level.
</details>

<br>

* In [Figure S3]D-E, were the Ag-high LECs compared to total naïve LEC or to a
  cluster of Ag-high-similar LEC identified also in naïve mice?
  
  * **The reviewer is confused about the figure.
    The LECs were not divided into Ag-high/low, these plots include all cells
    for the indicated LEC subset. Clarify text.**
  
* For fLECs, some Ag-high genes increase after immunization,
  this seems in contrast to the conclusion that the archiving capacity is
  already present in naïve cells.
  
  * **These plots show that the Ag-high gene module is expressed in cells from
    naive mice, suggesting that some of these cells are predisposed to archive
    antigen.
    However, there is still some upregulation that occurs in response to
    vaccination.
    It is possible that these Ag-competent LECs are predisposed to archive
    antigen, but this ability is further induced during vaccination, so both of
    these models can be true.**

<br>

### Comment 7 (a, t) {#R2C7}

[Figure 4]

<details>
  <summary>full comment</summary>
Fig 4 shows LN sections containing only the LN cortex (SCS and follicles) and a
small portion of the T cell zone, excluding the medulla.
This is important because antigens are typically processed and transferred to
the medulla, which is not considered here.
Additionally, macrophages (Fig S6A) should be present in SCS but not detected.
Fig S7 also raises doubts about the LN structure analysis, as low magnifications
(Fig S7A-B) hinder LN region discrimination.
In addition, regions were highlighted (and analyzed?) inconsistently across
groups.
For example, the SCS is shown only in the “dual” group.
Finally, it is also unclear why the signals from the same regions from single
and dual-immunized mice were pooled (lines 342-343).
</details>

<br>

* Concerns about the structure observed in Xenium sections.
  Sections contain only the LN cortex (SCS and follicles) and a small portion of
  the T cell zone, excluding the medulla.
  Additionally, macrophages ([Figure S6]A) should be present in SCS but not
  detected.
  
  * **There are some macrophages shown in [Figure S6]A.
    The colors have been adjusted to make this more apparent.
    We could also include higher resolution images in the paper
    (high resolution images are on GEO).**
  * **Antigen in the medulla is shown in the GeoMx analysis ([Figure S7]).**
  * **Look at other LN sections for the Xenium data to see if there is
    antigen detected in the medulla.**

* Concerns about the regions selected for the GeoMx analysis.
  Low magnifications ([Figure S7]A-B) hinder LN region discrimination.
  In addition, regions were highlighted inconsistently across groups.
  For example, the SCS is shown only in the “dual” group.
  
  * **There are sinus regions included for the the day 21 and day 42 samples.
    The colors have been adjusted to make this more apparent.**
  
* It is also unclear why the signals from the same regions from single and
  dual-immunized mice were pooled ([Figure S7]C).

  * **The signals were pooled due to the small number of regions available for
    analysis.**

<br>

### Comment 8 (a) {#R2C8}

[Figure S9]

While human equivalent genes split cLECs and fLECs into fractions like mouse
Ag-high and Ag-low populations, this experiment does not confirm the presence of
archived Ag, its alteration in disease (FL vs. CHIKV),
or increased archiving gene expression due to a lack of negative controls
(no antigen-stimulated, as metastasis free are also antigen-stimulated in
the primary tumor).
This proof of concept is appreciated, but a validation using proper controls
(e.g., no Ag exposure or disrupted LEC Ag archiving as in CHIKV) would
strengthen these data.

* **For a non-antigen-stimulated control we could use scRNA-seq data from original
  paper describing LEC subsets (Zhang?)**

<br>

### Minor comments {#R2-minor}

1. Could the authors explain why the 4:4:1:1 ratio in the scSeq exp in
   [Figure 1]?
2. The results section often integrates discussion and references (of other
   works and previous figures), making the readability nonlinear.
   Additionally, some methodological details (e.g., lines 311–313) disrupt the
   flow and could fit better in Methods.
   A more apparent structure might improve readability.
3. In [Figure 4]E, CCR7+ DC shows the highest Ag at day 21 with single
   immunization, while cDC2 Tbet- shows the highest values in dual immunization.
   Could this reflect the migratory nature of CCR7+ DCs, which might migrate to
   the LN within the first 21 days and then downregulate CCR7 in later stages
   and differentiate into cDC2 Tbet-?
   More discussion about the implications of these DCs subset-specific findings
   would increase the interest in the manuscript, especially considering that
   the authors showed before that the DCs uptake of archived Ag is CCR7
   dependent.
4. In [Figure 1] and [Figure S1], the colors of the populations are similar and
   not distinguishable, e.g., two similar shades of blue for different
   populations.
5. [Figure 1]D: Why is n significantly lower on day 2 compared to the other
   groups?
6. [Figure 5]: why was it not the same classification as before, including Ag
   high, low, and competent?

   **Mice were not vaccinated with ova-psDNA, so we can only predict
   Ag-competent cells.**
   
7. A few typos need to be corrected:

   * Line 213 findings vs finding.
   * Line 336 is probably Fig S7A and not S6A.
   * Line 340 should not be Fig 4B. Maybe S7C? But CD4 is missing in the figure.
   * Line 344, a past tense, is missing
   * Line 348 shows a repetition
   * Line 384 might be Fig S8A and not S7A
   * Line 487 is Tbet-

---

<br>

<br>

## Reviewer 3

Tamburini and colleagues investigated the temporal kinetics of lymphatic
endothelial cell (LEC) antigen uptake and storage using a previously developed
antigen detection method that attaches DNA tags to ovalbumin.
Coupling oligonucleotides to this model antigen facilitated the detection of
captured protein by scRNA-seq.
The authors found that the antigen was retained (or attached) by various cells
in mouse lymph nodes after immunization with the foot pad.
It appears that ovalbumin was attached to and/or taken up by both hematopoietic
and stromal cells, including LECs, neutrophils, fibroblasts, and dendritic
cells.
Interestingly, different LEC subsets such as cLECs and fLECs showed high antigen
retention even at 21 and 42 days post vaccination.
The authors used a combination of computational frameworks to calculate antigen
retention scores and to identify gene signature programs associated with antigen
archiving in LECs.
These calculated and inferred gene modules included gene sets of the
endosomal-lysosomal pathway.
In addition, the authors applied machine learning algorithms to predict antigen
archiving ability of different cell types/subsets.
The predictive model identified low-antigen LECs as the major cell population
involved in antigen storage.
Since antigen-presenting dendritic cells also appeared to harbor late time
points (mainly day 42), the authors concluded that antigen is exchanged between
LECs and DCs leading to presentation to memory CD8+ T cells.

The proximity and potential interaction of antigen-positive LECs and DCs was
assessed using the spatial transcriptomics methods Xenium (10x Genomics) and
GeoMx DSP (Nanostring).
High antigen levels at later time points (days 21 and 42 post vaccination) were
detected in CCR7hi DC and cDC2 Tbet subsets.
The authors provide additional evidence for the depicted LEC-DC interaction
scenario by perturbing the lymph node activation state by infection with CHIKV
and by processing lymph node LEC gene sets from human patients.

Overall, the authors present an interesting study that provides additional
evidence that lymph node LECs can capture and retain antigen.
However, there are a number of concerns that need to be addressed to improve the
quality of the manuscript and further strengthen the main conclusions.
One of the major challenges in interpreting the data is that short-term antigen
retention/storage does not appear to be a unique property of LECs.
The authors should be encouraged to further determine the extent and mechanisms
of the apparent antigen presence in BECs and in fibroblasts.

<br>

### Comment 1 (t)

Introduction, line 91: It is not clear whether references 18 and 19 show
CHIKV RNA in murine lymph node stromal cells, including FRCs.

* **Check these references.**

<br>

### Comment 2 (a, e) {#R3C2}

[Figure 1]

The determination of the antigen (Ag) score in the scRNA-seq data should be
compared and validated with at least one additional method, such as injection of
fluorescence-labeled OVA followed by FACS/confocal microscopy analysis.
To what extent is the reduction in LEC Ag-score from 6 on day 2 to approximately
2 on day 42 related to the Ag-score of 2 in dendritic cells on day 2?

* **We will include some of this with new experiments, but this has been validated
  in elife paper.**
* **It sounds like they want a direct comparison of the calculated Ag-scores,
  I wonder if we could make a correlation plot showing mean ova488 for
  LECs/FRCs/DCs vs mean Ag-score from the scRNA-seq data.**

<br>

### Comment 3 (a) {#R3C3}

[Figure 1]

The cell type annotation needs to be validated and checked against additional
published data. Lymph nodes do not contain epithelial cells, although some Cdh1
and Krt8 expression could be assigned.

* **Revise cell type annotations.
  The population assigned as epithelial cells express epithelial cell markers
  ([Cell annotations]).**

<br>

### Comment 4

[Figure 3]

[Figure 3]H shows the correlation between 21-day and 42-day Ag-scores for LEC
subsets and BEC.
The same analysis is repeated in Supplementary Figure 4D for a biological
replicate.
Both figures are consistent in showing a very strong correlation of Ag levels at
21 and 42 days after multiple vaccinations for LEC subsets.
However, in the case of a biological replicate, the correlation is stronger for
BECs, which contradicts the results in [Figure 3]H.

* **The reviewer is mistaken, the correlation is not stronger for BECs vs other
  subsets for either replicate.**

<br>

### Comment 5 (a) {.tabset .tabset-pills #R3C5}

[Figure S3]

The specificity of the proposed gene signatures is not clear.
Which genes/modules related to antigen storage in cLECs are expressed in other
stromal cells or dendritic cells in lymph nodes?
The authors should provide further details on specific genes associated with
antigen-storing LECs and confirm differences in gene expression by an
independent method such as RT-qPCR of sorted cells.
Again, a comparison of LECs (subsets) and dendritic cells would be most
informative.

* **State and cite some papers demonstrating that sequencing is a better measure
  than RT-PCR and that this does not add value to the existing data.**
* **Compare expression of the Ag modules for LECs vs other stromal cells and
  DCs.**

<br>

Expression of the Ag-high and -low genes modules is shown below for LEC subsets,
DC subsets, and BECs, for each timepoint.
Data from the CHIKV 24 hpi mock sample is included as the "0" timepoint.
To assess specificity of expression, pairwise p values were calculated by
comparing the subset used to identify the module to each of the other cell
types.
P values were calculated using a two-sided Wilcoxon rank sum test and adjusted
using Benjamini-Hochberg correction.
*** p < 0.0001, ** p < 0.001, * p < 0.05.

```{r "module exp data", include = FALSE, eval = generate_plots}
gc()

# Load 24 hpi mock-infected data
# * merge mock-infected with Ag LEC object
# * to compare timepoints, need all cells used for comparison to be present when
#   scores are calculated
# * set correct subtype column used for Lucas et al
chikv_naive <- chikv_so %>%
  subset(tm == "24hpi" & treatment == "mock") %>%
  mutate_meta(
    mutate,
    tm = 0,
    subtype = lec_subtype
  )

# Merge with lec and dc objects
naive_so <- chikv_naive %>%
  merge(lec_so)

rm(lec_so)
gc()

naive_so <- naive_so %>%
  merge(dc_so) %>%
  mutate_meta(mutate, tm = ifelse(is.na(tm), "dual", as.character(tm)))

rm(dc_so)
gc()

# Add module scores
# * need to recalculate scores since we are now including the mock-infected data
rf_feats %>%
  iwalk(~ {
    clmn_nms <- str_c(.y, "_", names(.x))

    clmn_nms <- set_names(
      str_c(clmn_nms, seq_along(clmn_nms)),
      clmn_nms
    )

    naive_so <<- naive_so %>%
      mutate_meta(dplyr::select, -any_of(names(clmn_nms))) %>%
      AddModuleScore(
        features = .x,
        name     = names(clmn_nms)
      ) %>%
      mutate_meta(dplyr::rename, !!!syms(clmn_nms))
  })

naive_dat <- naive_so@meta.data %>%
  as_tibble(rownames = ".cell_id")

rm(naive_so)
gc()
```

```{r "module exp plots", fig.width = 10, fig.height = 6, results = "asis", eval = generate_plots}
# Format n labels
dat <- naive_dat %>%
  pivot_longer(all_of(names(module_nms)), names_to = "module") %>%
  mutate(tm_type = str_c(tm, "_", subtype))

n_dat <- dat %>%
  group_by(tm, subtype, tm_type) %>%
  summarize(n = n_distinct(.cell_id), .groups = "drop") %>%
  mutate(
    n_lab = ifelse(n > 1000, str_c(round(n / 1000, 1), "k"), as.character(n)),
    n_lab = str_c(subtype, " (", n_lab, ")")
  )

n_labs <- set_names(n_dat$n_lab, n_dat$tm_type)

# Calculate p values
p_dat <- dat %>%
  group_by(tm, module) %>%
  nest() %>%
  mutate(
    p = map(data, ~ {
      pairwise.wilcox.test(
        .x$value, .x$subtype,
        p.adjust.method = "none"
      )
    }),
    p = map2(p, module, ~ {
      typ <- str_extract(.y, "^[^_]+")
      
      .x %>%
        tidy() %>%
        filter(group1 == typ | group2 == typ) %>%
        mutate(module_type = typ)
    })
  ) %>%
  select(-data) %>%
  unnest(p) %>%
  ungroup()

# Adjust p values
p_dat <- p_dat %>%
  mutate(
    p_adj = p.adjust(
      p.value,
      method = "BH"
    )
  ) %>%
  rowwise() %>%
  mutate(
    p_lab = case_when(
      p_adj < 0.0001  ~ "***",
      p_adj < 0.001   ~ "**",
      p_adj < 0.05    ~ "*"
    ),
    subtype = c(group1, group2)[!c(group1, group2) %in% module_type]
  ) %>%
  ungroup() %>%
  filter(!is.na(p_lab)) %>%
  select(-group1, -group2) %>%
  mutate(tm_type = str_c(tm, "_", subtype))

# Plot module scores
plts <- names(module_nms) %>%
  set_names() %>%
  map(~ {
    ttl  <- str_c(.x, " module expression")
    typ  <- str_extract(.x, "^[^_]+")
    lvls <- c("0", tms, "dual")
    
    dat <- naive_dat %>%
      mutate(
        tm_type = str_c(tm, "_", subtype),
        tm_type = fct_reorder(tm_type, !!sym(.x), .desc = TRUE),
        tm      = fct_relevel(tm, lvls)
      )
    
    mx_dat <- dat %>%
      group_by(tm, subtype, tm_type) %>%
      summarize(max = max(!!sym(.x)), .groups = "drop") %>%
      mutate(max = max(max)) %>%
      filter(subtype == typ)
    
    p <- p_dat %>%
      filter(module == .x) %>%
      mutate(
        max = max(mx_dat$max),
        tm  = fct_relevel(tm, lvls)
      )
    
    dat %>%
      ggplot(aes(x = tm_type, y = !!sym(.x), fill = subtype)) +
      geom_violin(color = "black", scale = "width", alpha = 0.5) +
      stat_summary(geom = "point", fun = median) +
      facet_wrap(~ tm, scales = "free_x") +
      scale_fill_manual(values = c(lec_clrs, dc_clrs)) +
      geom_text(
        aes(x = tm_type, y = max, label = p_lab),
        data  = p,
        color = "black",
        vjust = 0,
        size  = 12 / .pt
      ) +
      geom_point(
        aes(x = tm_type, y = max, fill = NULL),
        data     = mx_dat,
        position = position_nudge(y = 0.2),
        color    = "black",
        size     = 3,
        shape    = 25,
        stroke   = 1
      ) +
      scale_y_continuous(expand = expansion(c(0.05, 0.07))) +
      scale_x_discrete(labels = n_labs) +
      labs(title = ttl, y = ttl) +
      djvdj_theme() +
      theme(
        legend.position = "none",
        axis.title.x    = element_blank(),
        axis.text.x     = element_text(angle = 70, hjust = 1, color = "black")
      )
  })

plts %>%
  iwalk(~ {
    cat("\n\n####", .y, "\n\n")
    print(.x)
    cat("\n\n<br>\n\n")
  })
```

### Comment 6 (a, t) {#R3C6}

[Figure 3]

In the double vaccination experiment, the authors conclude that they have
detected gene programs that promote antigen archiving during serial
immunization, similar to Ag-high and Ag-low gene programs at 21 and 42 days for
LEC subsets.
They base their results on the existence of a significant correlation between Ag
scores at 21 and 42 days and the Ag-high and Ag-low module gene score programs
([Figure S4]E-F).
The interpretation based on the p-values shown should be re-evaluated.
Alternatively, the authors could use different metrics for measuring correlation
or, in general, other similarity/dissimilarity measures to support their claim.

* **It is unclear why there is an issue with the interpretation of the provided
  p values.
  The reviewer may not have realized that the gene programs shown in [Figure S4]
  are the same as the ones described in [Figure 2].
  Clarify this in the text.**
* **Also report Pearson correlation to further support conclusions.**

<br>

### Comment 7 (a) {#R3C7}

[Figure 2]

The authors have developed a random forest classifier to predict Ag-competent
LECs.
Can these predictions be confirmed using another machine learning classifier?

* **Train additional models, could try AutoML approach.**

<br>

### Comment 8 (a, t) {#R3C8}

[Figure 4]

Although a most likely scenario, the presented evidence for direct LEC-dendritic
cell interaction and antigen transfer remains week.
The spatial transcriptomic analysis is based on only a few lymph node sections
(not clear which time point in 4A).
To improve the robustness of the data and conclusions, the authors should
consider applying the following analyses:
(i) demonstrate co-localization and membrane interactions of antigen-positive
LECs with specific dendritic cell subsets using confocal microscopy,
(ii) apply CellChat-based interactome analysis,
(iii) construct spatial networks using the k-nearest neighbor algorithm and
calculate cell proximity enrichments in the specific cell neighborhoods.

* **Refer to Beth's previous 2017 paper.**
* **Perform additional Xenium analysis.
  Construct spatial networks and calculate cell proximity enrichment in the
  specific cell neighborhoods**

<br>

### Comment 9 (t)

[Figure S9]

Human data: Although an important validation step, the analysis of the human
data (Abe et al.) appears rather incomplete to justify major conclusions on the
suggested LEC-dendritic cell interaction in human lymph nodes.
The authors critically discuss the validity and robustness of the
“antigen-retention” gene signatures in human LECs.
Again, co-localization and direct interaction of LECs and particular dendritic
cell subsets in human lymph nodes needs to be demonstrated.
Without additional and more convincing data from human lymph nodes, statements
about unique transcriptional programs in both human and mouse lymph nodes need
to be toned down throughout the manuscript.

* **Tone down statements about human LN data.**

---

<br>

<br>
