---
title: "Antigen Tracking Figures"
author: "Ryan Sheridan"
date: "`r Sys.Date()`"
output:
  html_document:
    toc:            true
    toc_float:      true
    toc_depth:      4
    df_print:       "paged"
    code_folding:   "hide"
    self_contained: true
    highlight:      "kate"
params:
  template_dir: "src"                                                                 # Rmd templates
  so_dir:       "~/Dropbox/Ryan/Projects/antigen-exchange/results/sobjs/2024-07-01"   # Seurat objects
  mod_dir:      "~/Dropbox/Ryan/Projects/antigen-exchange/results/models/2024-07-17"  # RF models
  abe_dir:      "~/Dropbox/Ryan/Projects/antigen-exchange/results/abe"                # Abe et al objects
  geomx_dir:    "results/geomx"                                                       # GeoMx objects
  chikv_dir:    "~/Dropbox/Ryan/Projects/morrison-lnsc/results/sobjs"                 # CHIKV objects
  xen_dir:      "~/Dropbox/Ryan/Projects/antigen-exchange/results/xenium/2024-03-22"  # Xenium objects
  ref_dir:      "ref"                                                                 # clustifyr references
  table_dir:    "results/tables/2024-07-17"                                           # tables
  sample_info:  "sample_info.xlsx"                                                    # samples names/info
  res_dir:
    value:
      exp-0: "~/Projects/antigen-tracking/results"
      exp-1: "~/Projects/tamburini-antigen-tracking/results/2022-03-11"
      exp-2: "~/Projects/tamburini-antigen-tracking/results/2022-10-28"
  xen_res_dir: "results/xenium/20240322__194506__032224_Tamburini_Run_1"                     # Directory with Xenium data
  xen_regex:   ["output-XETG00230__0022624__[A-Z]__", "output-XETG00230__0022841__[A-Z]__"]  # Regex to pull samples for Xenium slides
  xen_samples: 
    value:
      A: ["6wk", 1, 2]      # Sample identity and 6wk and 3wk barcode identity (in this order)
      B: ["6wk", 2, 1]      # e.g. `1` indicates GeoMx-Barcode01
      C: ["6wk-3wk", 1, 2]
      D: ["6wk-3wk", 2, 1]
      E: ["3wk", 2, 1]
      F: ["3wk", 1, 2]
  xen_ag_targets:
    value:
      Ag_tag_1: "GeoMx-Barcode01-ERCC00142"
      Ag_tag_2: "GeoMx-Barcode02-ERCC00144"
  xen_custom_targets:
    - "Ag-tag-A-(ERCC2)"
    - "Ag-tag-B-(ERCC3)"
    - "Ag-tag-C-(ERCC4)"
    - "T2-Sensitive-Ag-tag-(ERCC9)"
    - "T2-Resistant-Ag-tag-(ERCC12)"
    - "DNA-only-T2-Control-Ag-tag-(ERCC16)"
    - "GeoMx-Barcode01-ERCC00142"
    - "GeoMx-Barcode02-ERCC00144"
    - "CHIKV-5-(0-7566)"
    - "CHIKV-sgRNA-(7566-12036)"
    - "CHIKV-negative-(0-12036)"
    - "LCMV-(Armstrong)-Segment"
    - "Listeria-monocytogenes-10403S-gyrase-B-CP002002-1"
    - "Vaccinia-(Western-Reserve)-J6R"
    - "mRNA-vaccine-(3-UTR)"
    - "mRNA-vaccine-(ovalbumin)"
    - "mRNA-vaccine-(FLAG-3x)"
    - "mRNA-vaccine-(eGFP)"
editor_options: 
  chunk_output_type: inline
---

```{r "setup", include = FALSE}
# Default chunk options
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo    = FALSE,
  dpi     = 300
)

options(java.parameters = "-Xmx12000m")

# Set environment for running UMAP
reticulate::use_condaenv("umap", conda = "~/.local/bin/micromamba")

# Main setup
knitr::knit(here::here(params$template_dir, "setup.Rmd"), "")

# Process Xenium slides 1 and 2
slides <- seq_along(params$xen_regex)

chunks <- slides %>%
  map_chr(~ knit_expand(here(params$template_dir, "setup-xenium.Rmd")))

knit(text = chunks, output = "")

# Load Xenium data
xen <- slides %>%
  map(~ {
    read_tsv(here(params$xen_dir, str_c("xen_s", .x, ".tsv.gz"))) %>%
      mutate(slide = .x)
  }) %>%
  bind_rows()

# Process other datasets
knit(here(params$template_dir, "setup-ml.Rmd"), "")
knit(here(params$template_dir, "setup-geomx.Rmd"), "")
knit(here(params$template_dir, "setup-abe.Rmd"), "")

# Plot parameters
min_cells <- 5
rf_typs   <- c("cLEC", "fLEC", "Collecting")
rf_tms    <- c(14, 21, 42)

ag_modules <- rf_typs %>%
  map(str_c, c("_high", "_low")) %>%
  unlist()

# Keys for naming labels
ag_clmns <- c("Ag_3wk_score", "Ag_6wk_score")

ag_labs <- set_names(
  c("Ag-score (21 day)", "Ag-score (42 day)"),
  ag_clmns
)

ag_labs_2 <- set_names(
  c("Ag-score\n21 day", "Ag-score\n42 day"),
  ag_clmns
)

# Mouse key
m_key <- c(
  d2        = "2 day",
  d14       = "14 day",
  `3wk`     = "21 day",
  `6wk`     = "42 day",
  `6wk-3wk` = "dual"
)

# Experiment to show in main figures
ag_exps <- list(LEC = "exp-1", DC = "exp-2")

# LEC/DC subset colors
subtype_clrs <- list(LEC = lec_clrs, DC = dc_clrs)
```

```{r "load naive data", include = FALSE}
# Load 24 hpi mock-infected data
chikv_naive <- qread(here(params$chikv_dir, "so_lec.qs"))

chikv_naive <- chikv_naive %>%
  subset(tm == "24hpi" & treatment == "mock") %>%
  mutate_meta(mutate, tm = 0)

naive_so <- lec_so %>%
  subset(mouse != "6wk-3wk") %>%
  merge(chikv_naive)

# Add module scores
# need to recalculate
rf_feats %>%
  iwalk(~ {
    clmn_nms <- str_c(.y, "_", names(.x))

    clmn_nms <- set_names(
      str_c(clmn_nms, seq_along(clmn_nms)),
      clmn_nms
    )

    naive_so <<- naive_so %>%
      mutate_meta(dplyr::select, -any_of(names(clmn_nms))) %>%
      AddModuleScore(
        features = .x,
        name     = names(clmn_nms)
      ) %>%
      mutate_meta(dplyr::rename, !!!syms(clmn_nms))
  })

naive_dat <- naive_so@meta.data %>%
  as_tibble(rownames = ".cell_id")

rm(chikv_naive, naive_so)
gc()
```

```{r "format lec data"}
# Objects to use for each cell type
ag_objs <- list(LEC = lec_so, DC = dc_so)

# Format prediction labels
# * make sure the correct predictions and module scores are being used
# * once the models are finalized, this information can be saved in the
#   meta.data
# * rf_pred will be NA if cell type was not used for training
# * pred_grp includes Ag-competent groups based on Ag predictions
lec_so <- lec_so %>%
  AddMetaData(rf_preds, col.name = str_c("rf_", colnames(rf_preds)))

lec_so <- lec_so %>%
  mutate_meta(
    mutate,
    pred_grp = ifelse(
      rf_pred == "high" & !rf_correct,
      "Ag-competent", str_c("Ag-", !!sym(rf_dat_clmn))
    ),
    rf_pred = str_c("Ag-", rf_pred),
    mouse   = fct_relevel(mouse, mice)
  )

# Add module scores
rf_feats %>%
  iwalk(~ {
    clmn_nms <- str_c(.y, "_", names(.x))

    clmn_nms <- set_names(
      str_c(clmn_nms, seq_along(clmn_nms)),
      clmn_nms
    )

    lec_so <<- lec_so %>%
      mutate_meta(dplyr::select, -any_of(names(clmn_nms))) %>%
      AddModuleScore(
        features = .x,
        name     = names(clmn_nms)
      ) %>%
      mutate_meta(dplyr::rename, !!!syms(clmn_nms))
  })
```

---

<br>

## Figure 1

*Antigen persists in discrete cell populations within the lymph node*

A.  Experimental design
B.  Mean Ag-score is shown for each cell type for 2, 14, 21, and 42 days post
    vaccination.
C.  UMAP projection shows LEC subsets.
D.  UMAP projections show Ag-scores for LEC subsets for each timepoint.
E.  Mean Ag-score is shown for LEC subsets for each timepoint.
F.  Ag-scores are shown for each timepoint for each LEC subset.
G.  UMAP projection shows DC subsets.
H.  UMAP projections show Ag-scores for DC subsets for each timepoint.
I.  Mean Ag-score is shown for DC subsets for each timepoint.
J.  Ag-scores are shown for each timepoint for each DC subset.

<u>Notes</u>

* Only experiment 1 is shown for the LEC 21 and 42 day timepoints (including
  for the UMAPs). Experiment 2 data include very few LECs for the 21 day
  timepoint.
* Experiment 2 is shown for the DC 21 and 42 day timepoints. Experiment 1 data
  are lower quality (sparse signal, mostly zeros).
* Subsets are only shown if there are >= `r min_cells` cells for each timepoint
* CD45+ and CD45- cell types are only shown for the CD45+ and CD45- datasets,
  respectively.

```{r "type Ag heat", fig.width = 16, fig.height = 5}
# Format Ag data for broad cell types
# exclude CD45+ cell types from CD45- datasets and vice versa
cd45neg_typs <- c("LEC", "BEC", "Fibroblasts", "Epithelial cells")

typ_ag_dat <- all_meta %>%
  filter(
    vaccination == "single",
    cell_type != "unassigned"
  ) %>%
  group_by(cell_sort) %>%
  mutate(
    tm        = fct_relevel(as.character(tm), tms),
    cell_type = fct_infreq(cell_type),
    cell_type = fct_relevel(cell_type, c("LEC", "Fibroblasts", "DC"))
  ) %>%
  ungroup() %>%
  filter(
    (cell_sort == "CD45neg" & cell_type %in% cd45neg_typs) |
      (cell_sort == "CD45pos" & !cell_type %in% cd45neg_typs)
  )

# Format data for heatmap
# exclude unassigned cells since these likely include multiple cell types
heat_dat <- typ_ag_dat %>%
  group_by(tm, cell_type) %>%
  summarize(
    Ag_score = mean(Ag_score), .groups = "drop",
    n        = n()
  ) %>%
  group_by(cell_type) %>%
  mutate(cell_type = str_c(cell_type, "\nn = ", label_comma()(sum(n)))) %>%
  ungroup() %>%
  mutate(
    tm      = fct_relevel(as.character(tm), rev(as.character(tms))),
    cell_type = fct_reorder(cell_type, Ag_score, mean, .desc = TRUE)
  )

# Create heatmap summarizing Ag scores
typ_heat <- heat_dat %>%
  ggplot(aes(cell_type, tm, fill = Ag_score)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("white", "#D55E00"), na.value = "black") +
  guides(fill = guide_colorbar(
    title.position = "top", ticks = FALSE, title = "Ag-score",
    barwidth = unit(150, "pt"), barheight = unit(10, "pt")
  )) +
  labs(y = "days post vaccination") +
  base_theme +
  theme(
    plot.margin  = margin(0, 15, 0, 15),
    legend.position = "top",
    legend.justification = 0,
    legend.text  = element_text(size = txt_pt1),
    aspect.ratio = n_distinct(heat_dat$tm) / n_distinct(heat_dat$cell_type) * 0.9,
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1),
  )
```

```{r "subtype Ag UMAPs"}
# Data for UMAPs
# show exp1 for LECs and exp2 for DCs
fig1_dat <- ag_objs %>%
  imap(~ {
    .x@meta.data %>%
      as_tibble(rownames = ".cell_id") %>%
      filter(
        exp %in% c("exp-0", ag_exps[[.y]]),
        vaccination == "single"
      ) %>%
      mutate(tm_lab = factor(as.character(tm), tms))    
  })

# Subtype UMAPs
fig1_typ_u <- fig1_dat %>%
  imap(~ {
    typ_lvls <- .x %>%
      pull(subtype) %>%
      unique() %>%
      as.character()
    
    res <- .x %>%
      plot_scatter(
        "subtype", "hUMAP_1", "hUMAP_2",
        size         = 0.1,
        plot_colors  = c(lec_clrs, dc_clrs),
        plot_lvls    = rev(typ_lvls),
        label_params = list(size = ttl_pt1),
        panel_nrow   = 2
      ) +
      guides(color = guide_legend(ncol = 1, override.aes = list(size = 4))) +
      umap_theme +
      theme(
        aspect.ratio = 0.9,
        legend.position = "right",
        legend.title = element_blank(),
        legend.text  = element_text(size = txt_pt2)
      )
    
    res
  })

# Ag-score UMAPs
u_clrs <- list(
  LEC = c("lightblue", "white", "#D7301F"),
  DC  = c("lightblue", "white", "#6A51A3", "#6A51A3")
  # DC  = c("white", "#6A51A3")
)

fig1_ag_u <- fig1_dat %>%
  imap(~ {
    .x %>%
      filter(tm_lab %in% as.character(tms)) %>%
      plot_scatter(
        "Ag_score", "hUMAP_1", "hUMAP_2",
        plot_colors  = u_clrs[[.y]],
        label_params = list(size = ttl_pt1),
        group_col    = "tm_lab",
        panel_nrow   = 1,
        outline      = TRUE,
        size         = 0.2,
        stroke       = 0.7
      ) +
      facet_wrap(
        ~ tm_lab, nrow = 1,
        labeller = as_labeller(function(x) str_c("day ", x))
      ) +
      guides(fill = guide_colorbar(
        title.position = "right", ticks = FALSE, title = "Ag-score",
        barheight = unit(120, "pt"), barwidth = unit(10, "pt")
      )) +
      umap_theme +
      theme(
        aspect.ratio         = 0.9,
        panel.border         = element_rect(color = ln_col, linewidth = ln_pt, fill = NA),
        legend.position      = "right",
        legend.justification = 0,
        legend.title         = element_text(angle = 270, hjust = 0.5),
        strip.text.y         = element_blank()
      )
  })
```

```{r "subtype Ag boxplots", fig.width = 8, fig.height = 2}
# Format data for boxplots
ag_bx_dat <- fig1_dat %>%
  map(~ {
    .x %>%
      filter(subtype != "unassigned") %>%
      mutate(mouse = as.character(mouse)) %>%
      group_by(subtype) %>%
      filter(all(table(mouse) >= min_cells)) %>%
      ungroup() %>%
      mutate(subtype = fct_reorder(subtype, Ag_score, mean, .desc = TRUE))
  })
  
# Create boxplots
ag_bxs <- ag_bx_dat %>%
  imap(~ create_ag_boxes(.x, clrs = subtype_clrs[[.y]]))

# Pull subsets used for main figures
ag_plt_typs <- ag_bx_dat %>%
  map(~ levels(.x$subtype))
```

```{r "subtype Ag heat"}
# Create heatmaps
heat_clrs <- c(LEC = "#D7301F", DC = "#6A51A3")

ag_heats <- ag_bx_dat %>%
  imap(~ {
    dat <- .x %>%
      group_by(mouse, tm, subtype) %>%
      summarize(Ag_score = mean(Ag_score), .groups = "drop") %>%
      mutate(
        subtype = fct_reorder(subtype, Ag_score, mean),
        tm      = fct_relevel(as.character(tm), as.character(tms)),
        ttl     = "Ag-score"
      )

    rat <- n_distinct(dat$subtype) / n_distinct(dat$tm)

    dat %>%
      ggplot(aes(tm, subtype, fill = Ag_score)) +
      geom_tile() +
      facet_wrap(~ ttl) +
      labs(x = "days post vaccination") +
      scale_fill_gradientn(colours = c("lightblue", "white", heat_clrs[[.y]])) +
      guides(fill = guide_colorbar(
        ticks = FALSE, title = "Ag-score",
        barheight = unit(105, "pt"), barwidth = unit(10, "pt")
      )) +
      base_theme +
      theme(
        legend.title = element_blank(),
        aspect.ratio = rat,
        axis.title.y = element_blank(),
        plot.margin  = margin(0, 15, 0, 15),
        plot.title   = element_text(hjust = 0.5, size = ttl_pt2)
      )
  })
```

```{r "Fig 1 diagram"}
img <- readPNG(here("results/figures/experiment_design/fig1_exp_design.png"))
img <- rasterGrob(img, interpolate = TRUE)

# Adjust plot limits if necessary
img_lims <- tibble(x = c(0, 1), y = c(0, 1))

# Create an empty plot for placing the image, adjusted to fill the area
fig1_dsgn <- ggplot(img_lims, aes(x, y)) + 
  annotation_custom(
    grob = img,
    xmin = -Inf, xmax = Inf,
    ymin = -Inf, ymax = Inf
  ) +
  xlim(c(-Inf, Inf)) +
  ylim(c(-Inf, Inf)) +
  theme_void()
```

```{r "Fig 1", fig.width = 16, fig.height = 15.5}
# Create final figure
top <- plot_grid(
  fig1_dsgn, typ_heat,
  labels     = c("A", "B"),
  label_size = ttl_pt2 * 1.5,
  nrow       = 1,
  hjust      = 1.1
)

bot <- list(
  "C" = fig1_typ_u$LEC,
  "D" = fig1_ag_u$LEC,
  "E" = ag_heats[[1]],
  "F" = ag_bxs$LEC,
  "G" = fig1_typ_u$DC,
  "H" = fig1_ag_u$DC,
  "I" = ag_heats[[2]],
  "J" = ag_bxs$DC
)

bot <- bot %>%
  imap(~ .x + labs(tag = .y))

fig1 <- top /
  (bot[[1]] + bot[[2]]) /
  (bot[[3]] + bot[[4]]) /
  (bot[[5]] + bot[[6]]) /
  (bot[[7]] + bot[[8]]) +
  plot_layout(heights = c(1.4, 1, 0.7, 1, 0.7)) &
  theme(
    plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"),
    plot.tag.position = c(0, 1)
  )

fig1
```

<br>

<br>

### Figure S1

A.  UMAP projections show cell types identified for CD45- and CD45+ samples.
B.  UMAP projections show LEC subsets identified for each timepoint.
C.  UMAP projections show DC subsets identified for each timepoint.
D.  The expression of key marker genes is shown for LEC subsets.
E.  The expression of key marker genes is shown for DC subsets.

```{r "type UMAPs"}
# Format data
typ_dat <- all_meta %>%
  mutate(
    vaccination = str_c(vaccination, "\nvaccination"),
    tm          = ifelse(is.na(tm), vaccination, str_c("day ", tm)),
    tm          = fct_relevel(tm, str_c("day ", tms)),
    cell_type   = fct_infreq(cell_type),
    cell_sort   = recode(cell_sort, CD45neg = "CD45-", CD45pos = "CD45+")
  )

n_dat <- typ_dat %>%
  group_by(tm, cell_sort) %>%
  summarize(n = str_c("n = ", label_comma()(n())), .groups = "drop")

# Create UMAP projections
typ_u <- typ_dat %>%
  plot_scatter(
    "cell_type",
    "hUMAP_1", "hUMAP_2",
    plot_colors  = typ_clrs,
    plot_lvls    = c("BEC", "LEC", "Epithelial cells", "Monocytes"),
    size         = 0.001,
    n_label      = "legend"
  ) +
  geom_text(
    aes(-Inf, Inf, label = n, color = NULL), data = n_dat,
    hjust = -0.1, vjust = 1.5, size = 10 / .pt
  ) +
  facet_grid(cell_sort ~ tm) +
  guides(color = guide_legend(nrow = 2, override.aes = list(size = 4))) +
  umap_theme_2 +
  theme(
    legend.position = "bottom",
    legend.title    = element_blank()
  )
```

```{r "subtype UMAPs"}
# Marker genes to plot
typ_gns <- list(
  LEC = c(
    "Lyve1", "Marco", "Madcam1",
    "Vcam1", "Ptx3",  "Ackr4"
  ),
  DC = c(
    "Xcr1", "Itgax", "Sirpa",
    "Lyz2", "Ccr7",  "Siglech"
  )
)

# Format data for subtype plots
subtype_dat <- ag_objs %>%
  imap(~ {
    clmns <- c("subtype", "tm", "vaccination", "hUMAP_1", "hUMAP_2")
    
    .x %>%
      FetchData(c(clmns, typ_gns[[.y]])) %>%
      as_tibble(rownames = ".cell_id") %>%
      mutate(
        vaccination = str_c(vaccination, "\nvaccination"),
        tm          = ifelse(is.na(tm), vaccination, str_c("day ", tm)),
        tm          = fct_relevel(tm, str_c("day ", tms)),
        subtype     = fct_relevel(subtype, levels(markers$annotation))
      )
  })

# Create UMAPs
subtype_u <- subtype_dat %>%
  imap(~ {
    .x %>%
      plot_scatter(
        "subtype", "hUMAP_1", "hUMAP_2",
        group_col    = "tm", panel_nrow = 2,
        n_label      = "corner",
        label_params = list(size = 14),
        size         = 0.1,
        plot_colors  = subtype_clrs[[.y]],
        plot_lvls    = levels(.x$subtype)
      ) +
      umap_theme_2 +
      theme(
        legend.position  = c(0.85, 0.25),
        legend.title     = element_blank()
      )
  })
```

```{r "subtype boxplots"}
# Create LEC/DC subtype boxplots
subtype_bxs <- subtype_dat %>%
  imap(~ {
    dat <- .x %>%
      pivot_longer(all_of(typ_gns[[.y]])) %>%
      mutate(
        subtype_tm = str_c(subtype, "_", tm),
        name       = fct_relevel(name, typ_gns[[.y]])
      ) %>%
      arrange(subtype) %>%
      mutate(subtype_tm = fct_inorder(subtype_tm))
    
    # Data for n labels
    n_labs <- dat %>%
      group_by(subtype, tm, subtype_tm) %>%
      summarize(
        n = label_comma()(n_distinct(.cell_id)),
        .groups = "drop"
      ) %>%
      mutate(
        n_lab = str_replace(subtype, "_", " "),
        n_lab = str_c(n_lab, " (", n, ")")
      )
    
    n_labs <- set_names(n_labs$n_lab, n_labs$subtype_tm)

    # Create boxplots
    n_typs <- n_distinct(dat$subtype)

    dat %>%
      ggplot(aes(subtype_tm, value, fill = subtype)) +
      geom_boxplot(outlier.size = 0.1, key_glyph = draw_key_point, alpha = 0.9) +
      facet_grid(name ~ tm, scales = "free") +
      guides(fill = guide_legend(override.aes = list(shape = 22, size = 4))) +
      scale_x_discrete(labels = n_labs) +
      scale_fill_manual(values = subtype_clrs[[.y]]) +
      base_theme +
      theme(
        aspect.ratio    = 4 / n_typs,
        legend.position = "none",
        strip.clip      = "off",
        axis.title      = element_blank(),
        axis.text.x     = element_text(angle = 90, hjust = 1, vjust = 0.5)
      )
  })
```

```{r "Fig S1", fig.width = 16, fig.height = 20}
# plts <- list(
#   "A" = typ_u,
#   "B" = 
# )

figS1 <- typ_u /
  (subtype_u$LEC + subtype_u$DC) /
  
  wrap_plots(subtype_bxs$LEC, subtype_bxs$DC, widths = c(1, 0.63)) +
  
  plot_layout(heights = c(0.7, 0.7, 1)) +
  plot_annotation(tag_levels = "A") &
  theme(plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"))

figS1
```

<br>

<br>

### Figure S2

A.  Ag-scores are shown for each timepoint for CD45- cell types. Two biological
    replicates are shown for the day 21 and day 42 timepoints. The number of
    cells identified for each cell type is shown above each timepoint.
B.  Ag-scores are shown for each timepoint for CD45+ cell types,
    as described in A.
C.  Ag-scores are shown for each timepoint for LEC subsets, as described in A.
D.  Ag-scores are shown for each timepoint for DC subsets, as described in A.

```{r "type Ag boxplots"}
# Format data for boxplots
dat <- c("exp-1", "exp-2") %>%
  map_dfr(~ {
    typ_ag_dat %>%
      filter(exp %in% c("exp-0", .x)) %>%
      mutate(
        exp       = str_replace(.x, "exp-", "r"),
        tm        = as.numeric(as.character(tm)),
        cell_sort = recode(cell_sort, CD45neg = "CD45-", CD45pos = "CD45+")
      )
  })

# Create boxplots
typ_ag_bxs <- c("CD45-", "CD45+") %>%
  set_names() %>%
  map(~ {
    dat <- dat %>%
      filter(cell_sort == .x)
      
    n_dat <- dat %>%
      group_by(cell_sort, exp, tm, cell_type) %>%
      summarize(n = label_comma()(n()), .groups = "drop")
    
    dat %>%
      create_ag_boxes(clrs = typ_clrs, facet_vars = c("exp", "cell_type")) +
      geom_text_repel(
        aes(y = Inf, label = n), data = n_dat,
        seed = 42, force = 0.005,
        size = 8 / .pt, vjust = 1.2
      ) +
      scale_y_continuous(expand = expansion(c(0.05, 0.15))) +
      theme(plot.margin = margin(15, 15, 30, 15))
  })
```

```{r "all subtype Ag boxplots"}
# Format data for boxplots
bx_dat <- ag_objs %>%
  map(~ {
    .x@meta.data %>%
      as_tibble(rownames = ".cell_id") %>%
      filter(vaccination == "single") %>%
      group_by(mouse, subtype, exp) %>%
      ungroup() %>%
      mutate(
        mouse   = as.character(mouse),
        tm_lab  = factor(as.character(tm), tms),
        subtype = fct_reorder(subtype, Ag_score, mean, .desc = TRUE)
      )
  })

# Create boxplots showing all LEC/DC subsets
all_ag_bxs <- bx_dat %>%
  imap(~ {
    dat <- .x
    
    dat <- c("exp-1", "exp-2") %>%
      map_dfr(~ {
        dat %>%
          filter(exp %in% c("exp-0", .x)) %>%
          mutate(exp = str_replace(.x, "exp-", "r"))
      })
    
    n_dat <- dat %>%
      group_by(tm, subtype, exp) %>%
      summarize(n = label_comma()(n()), .groups = "drop")
    
    dat %>%
      create_ag_boxes(clrs = subtype_clrs[[.y]], facet_vars = c("exp", "subtype")) +
      geom_text_repel(
        aes(y = Inf, label = n), data = n_dat,
        seed = 42, force = 0.005,
        size = 8 / .pt, vjust = 1.2
      ) +
      scale_y_continuous(expand = expansion(c(0.05, 0.15))) +
      theme(plot.margin = margin(15, 15, 30, 15))
  })
```

```{r "Fig S2", fig.width = 15, fig.height = 17.5}
# Create final figure
figS2 <- plot_grid(
  typ_ag_bxs$`CD45-`, typ_ag_bxs$`CD45+`,
  all_ag_bxs$LEC, all_ag_bxs$DC,
  labels = c("A", "B", "C", "D"),
  label_size = ttl_pt2 * 1.5,
  # rel_heights = c(1.1, 1.2, 1, 1.2),
  align = "v", axis = "rl",
  ncol = 1
)

figS2
```

---

<br>

<br>

## Figure 2

*Identification of an antigen archiving gene signature*

A.  Ag-score is shown for day 14 Ag-high and -low cells identified for LEC
    subsets with the highest Ag-score.
B.  The fraction of cells predicted to be Ag-competent is shown for each LEC
    subset 14, 21, and 42 days post vaccination.
C.  Ag-high module score is shown for Ag-low, Ag-high, and predicted
    Ag-competent LECs.
    The Spearman correlation between Ag classes and Ag-high module score is
    shown for each timepoint.
    One-sided p values were calculated and adjusted using Benjamini-Hochberg
    correction.
D.  UMAP projections show cLEC Ag-high module scores for each timepoint.
E.  The expression of select genes from the Ag-high (top four) and Ag-low
    (bottom four) gene modules is shown for cLECs. Genes identified as top
    predictors for multiple LEC subsets are shown. Triangles indicate the gene
    is differentially expressed when compared to Ag-low cells. P values were
    calculated using a one-sided Wilcoxon rank sum test with Benjamini-Hochberg
    correction.

<u>Notes</u>

* Ag-low and -high cells were identified by separately clustering each d14 LEC
  subset for each sample into two groups based on Ag-score.
* The cutoffs identified for the d14 dataset were then used to identify Ag-high
  cells for the other timepoints.
* With this approach there is a different cutoff for each LEC subset, but the
  cutoff is consistent across timepoints.

```{r "RF Ag signal", fig.width = 12, fig.height = 3.25}
# Format plotting data
plt_dat <- lec_so@meta.data %>%
  filter(
    subtype %in% rf_typs,
    mouse %in% rf_mouse
  ) %>%
  mutate(
    mouse = as.character(mouse),  # to remove factor
    !!sym(rf_dat_clmn) := fct_relevel(!!sym(rf_dat_clmn), c("low", "high"))
  ) %>%
  arrange(!!sym(rf_dat_clmn))

# Format plot labels
plt_labs <- plt_dat %>%
  group_by(subtype, mouse, tm, !!sym(rf_dat_clmn)) %>%
  summarize(
    n = label_comma()(n()),
    .groups = "drop"
  ) %>%
  mutate(n = str_c(!!sym(rf_dat_clmn), ": ", n, " cells")) %>%
  group_by(subtype, mouse, tm) %>%
  arrange(!!sym(rf_dat_clmn)) %>%
  mutate(n = str_c(n, collapse = "\n"))

plt_clrs <- lec_clrs

# Create bargraphs
brs <- rf_typs %>%
  imap(~ {
    dat <- plt_dat %>%
      filter(subtype == .x) %>%
      mutate(!!sym(rf_dat_clmn) := fct_rev(!!sym(rf_dat_clmn)))
    
    dat %>%
      ggplot(aes(y = mouse, fill = subtype, alpha = !!sym(rf_dat_clmn))) +
      geom_bar() +
      facet_wrap(~ subtype, nrow = 1, scales = "free") +
      scale_alpha_manual(values = c(low = 0.35, high = 1)) +
      scale_fill_manual(values = plt_clrs) +
      scale_y_discrete(expand = expansion(c(0, 0))) +
      scale_x_continuous(expand = expansion(c(0, 0))) +
      theme_void() +
      theme(
        legend.position = "none",
        plot.margin     = margin(t = 5),
        strip.text      = element_text(size = ttl_pt2)
      )
  })

hsts <- rf_typs %>%
  imap(~ {
    dat <- plt_dat %>%
      filter(subtype == .x)
    
    fst <- .y == 1
    lst <- .y == length(rf_typs)
    
    # Create histograms
    hst <- dat %>%
      ggplot(aes(Ag_score, fill = subtype, alpha = !!sym(rf_dat_clmn))) +
      geom_histogram(bins = 30, key_glyph = draw_key_point) +
      geom_text(
        aes(-Inf, Inf, label = n),
        data = filter(plt_labs, subtype == .x),
        check_overlap = TRUE,
        alpha = 1,
        hjust = -0.25,
        vjust = 1.1,
        size  = txt_pt2 / .pt
      ) +
      facet_wrap(~ subtype, scales = "free", nrow = 1) +
      guides(fill = "none") +
      scale_alpha_manual(
        values = c(low = 0.35, high = 1),
        labels = function(x) str_c("Ag-", x)
      ) +
      scale_fill_manual(values = plt_clrs) +
      scale_y_continuous(expand = expansion(c(0.05, 0.25))) +
      guides(alpha = guide_legend(override.aes = list(shape = 22, size = 5))) +
      labs(x = "Ag-score", y = "number of cells") +
      base_theme +
      theme(
        aspect.ratio    = 0.6,
        strip.text      = element_blank(),
        legend.position = "none",
        legend.title    = element_blank(),
        plot.margin     = margin(t = 0, r = 5, b = 0, l = 5)
      )
    
    if (!fst && !lst) {
      hst <- hst +
        theme(legend.position = "bottom")
    }
    
    if (!fst) {
      hst <- hst +
        theme(axis.title.y = element_blank())
    }
    
    if (fst || lst) {
      hst <- hst +
        theme(axis.title.x = element_blank())
    }
    
    hst
  })

# Create final figure
hsts <- wrap_plots(
  append(brs, hsts),
  ncol = length(rf_tms),
  heights = c(0.1, 1)
)
```

```{r "RF Ag-competent bars", fig.width = 8, fig.height = 5}
# Format data for bargraphs
# * for main figure include cells used for training
# * exclude training data for supp figure
rf_bar_dat <- lec_so@meta.data %>%
  filter(
    tm %in% rf_tms,
    subtype %in% rf_typs
  ) %>%
  mutate(
    subtype    = fct_relevel(subtype, rf_typs),
    rf_correct = fct_relevel(as.character(rf_correct), c("TRUE", "FALSE")),
    pred_grp   = fct_relevel(pred_grp, rev(names(pred_clrs))),
    tm         = as.character(tm)
  )

bar_thm <- base_theme +
  theme(aspect.ratio = 1.4)

rf_bar_1 <- rf_bar_dat %>%
  ggplot(aes(tm, fill = pred_grp)) +
  geom_bar(position = "fill", alpha = 0.85, key_glyph = draw_key_point) +
  facet_grid(~ subtype) +
  scale_fill_manual(values = pred_clrs) +
  scale_y_continuous(breaks = c(0, 0.5, 1)) +
  guides(fill = guide_legend(override.aes = list(shape = 22, size = 5))) +
  labs(y = "fraction of cells", x = "days post vaccination") +
  bar_thm +
  theme(
    legend.title = element_blank(),
    legend.position = "bottom"
  )
```

```{r "RF module boxplots", fig.width = 10, fig.height = 4}
# Format plot data
# exclude dual vaccination data
clmns <- c(
  "tm", "subtype", "pred_grp",
  "Ag_score", ag_modules
)

rf_bx_dat <- lec_so %>%
  FetchData(clmns) %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(tm %in% rf_tms)

rf_mod_args <- list(
  mod_typ = c("high", "low"),
  p_alt   = c("greater", "less"),
  p_hjust = c(0.7, 0.55),
  n_label = c(FALSE, TRUE)
)

mod_bxs <- rf_mod_args %>%
  pmap(~ {
    args <- list(...)
    
    module_clmns <- str_c(rf_typs, "_", args$mod_typ)
    module_title <- str_c("Ag-", args$mod_typ, " module score")
    
    rf_bx_dat %>%
      create_rf_module_boxes(
        module_clmns = module_clmns,
        module_title = module_title,
        p_alt        = args$p_alt,
        p_hjust      = c(args$p_hjust, 0.5),
        p_vjust      = c(1.2, 2),
        clrs         = pred_clrs,
        n_label      = args$n_label
      )
  })
```

```{r "RF cLEC module UMAPs", fig.width = 8, fig.height = 6.5}
# Format plotting data
u_dat <- lec_so@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(tm %in% rf_tms) %>%
  mutate(Ag_class_ml = str_c("Ag-", Ag_class_ml))

# Create UMAPs
u_args <- list(
  typ = set_names(rf_typs),
  pt_size = list(1, c(0.5, 0.1), c(0.5, 0.1)),
  ag_clrs = list(
    c("lightblue", "lightblue", "white", "#D7301F"),
    c("lightblue", "white", "#D7301F"),
    c("lightblue", "white", "#D7301F")
  )
)

mod_u <- u_args %>%
  pmap(~ {
    args <- list(...)
    
    u_dat %>%
      create_rf_module_umaps(
        typ           = args$typ,
        ag_class_clmn = "Ag_class_ml",
        module_clmn   = str_c(args$typ, "_high"),
        module_title  = str_c(args$typ, " Ag-high module"),
        equal_scales  = FALSE,
        pt_size       = args$pt_size,
        ag_clrs       = args$ag_clrs
      )
  })
```

```{r "RF top features examples", fig.height = 8, fig.width = 8}
# Calculate p-values
# * use all features included in models, adjust based on this number of genes
# * adjust p-values separately for each tm and pred_grp
p_vals <- rf_feats[rf_typs] %>%
  imap_dfr(~ {
    p_alt <- c(high = "greater", low = "less")
    
    typ <- .y
    
    gns <- .x %>%
      imap(~ set_names(rep(.y, length(.x)), .x)) %>%
      reduce(c)
    
    dat <- lec_so %>%
      subset(subtype == typ & tm %in% rf_tms) %>%
      FetchData(c("pred_grp", "tm", names(gns))) %>%
      as_tibble(rownames = ".cell_id") %>%
      pivot_longer(all_of(names(gns)), names_to = "gene") %>%
      mutate(
        class = gns[gene],
        p_alt = p_alt[class]
      ) %>%
      split(.$gene)
      
    dat %>%
      map_dfr(~ {
        p_grps <- names(pred_clrs)[-1]
        p_con  <- names(pred_clrs)[1]
        
        dat <- .x
        
        p_grps %>%
          map_dfr(~ {
            dat %>%
              filter(pred_grp %in% c(p_con, .x)) %>%
              group_by(tm) %>%
              summarize(
                p = wilcox.test(
                  value[pred_grp == .x], value[pred_grp == p_con],
                  alternative = unique(p_alt),
                  exact       = FALSE
                )$p.value,
                med_diff = abs(median(value[pred_grp == .x]) - median(value[pred_grp == p_con])),
                n        = length(value[pred_grp == .x]),
                gene     = unique(gene),
                class    = unique(class),
                p_alt    = unique(p_alt),
                pred_grp = .x,
                .groups  = "drop"
              )
          })
      }) %>%
      mutate(subtype = typ)
  })

# Adjust p-values for multiple testing
# * adjust each tm and pred_grp separately
p_vals <- p_vals %>%
  group_by(tm, pred_grp, subtype) %>%
  mutate(
    p_adj   = p.adjust(p, method = "BH"),
    n_genes = n()
  ) %>%
  ungroup()

# Identify genes shared between multiple cell types
shared_feats <- rf_feats %>%
  imap_dfr(~ {
    typ <- .y
    
    .x %>%
      imap_dfr(~ {
        tibble(
          subtype = typ,
          class   = .y,
          gene    = .x,
        )
      })
  }) %>%
  group_by(class, gene) %>%
  filter(n_distinct(subtype) > 1) %>%
  ungroup()

p_dat <- p_vals %>%
  filter(
    subtype == rf_cell_type,
    p_adj < 0.05
  ) %>%
  group_by(gene) %>%
  filter(any(med_diff > 0)) %>%
  mutate(n_sig = n()) %>%
  ungroup() %>%
  arrange(desc(n_sig))

# Plot example genes
gene_examples <- lec_so %>%
  subset(subtype == rf_cell_type) %>%
  subset(tm %in% rf_tms) %>%
  create_gn_plots(
    p_data = p_dat,
    top_gns = c(
      "Ctsd", "Lgmn", "Psap", "Grn",
      "Col4a2", "Tmsb4x",  # cell migration, angiogenesis
      "Ctla2a", "Ltbp4"    # immune response
    ),
    n_gns     = 4,
    plt_clrs  = pred_clrs,
    x_lvls    = names(pred_clrs),
    draw_line = FALSE,
    pt_size   = 2.5
  )
```

```{r "Fig 2", fig.width = 15, fig.height = 16.1}
left <- plot_grid(
  plot_grid(hsts, nrow = 2, rel_heights = c(1, 0.1)),
  mod_bxs[[1]],
  mod_u$cLEC[[1]],
  mod_u$cLEC[[2]],
  ncol = 1,
  rel_heights = c(1, 1, 1.11, 1.1),
  labels = c("A", "C", "D"),
  label_size = ttl_pt2 * 1.5,
  label_y = c(1, 1.1, 1)
)

right <- plot_grid(
  plotlist = gene_examples,
  ncol     = 1,
  align    = "v",
  axis     = "rl",
  rel_heights = c(1, 1.225)
)

right <- plot_grid(
  rf_bar_1, right,
  ncol = 1,
  rel_heights = c(1, 3),
  labels = c("B", "E"),
  label_size = ttl_pt2 * 1.5
)

fig2 <- plot_grid(
  left, right,
  rel_widths = c(1, 0.5)
)

fig2
```

<br>

<br>

### Figure S3

A.  The top 15 gene ontology terms (cellular component) are shown for Ag-high gene modules
    for each cLECs, fLECs, and collecting LECs.
B.  Top gene ontology terms are shown for Ag-low gene modules as described in A.
C.  The expression of each Ag-high gene module is shown for naive LECs
    (0 days post vaccination) along with each time point.
D.  The expression of each Ag-low gene module is shown as described in C.
E.  The fraction of cells predicted to be Ag-low and Ag-high is shown for each
    LEC subset.
F.  Ag-low module score is shown for Ag-low, Ag-high, and predicted
    Ag-competent LECs.
    The Spearman correlation between Ag classes and Ag-high module score is
    shown for each timepoint.
    One-sided p values were calculated and adjusted using Benjamini-Hochberg
    correction.
    The number of cells plotted is shown below each boxplot.
G.  UMAP projections show fLEC Ag-high module scores for each timepoint.
H.  UMAP projections show collecting LEC Ag-high module scores for each
    timepoint.

```{r "RF GO"}
# Term similarity for simplifying terms
dbs <- set_names(c("BP", "CC", "MF"))

go_sim_data <- dbs %>%
  map(~ {
    godata(
      OrgDb   = org.Mm.eg.db,
      keytype = "SYMBOL",
      ont     = .x
    )
  })

# Get enriched GO terms
rf_grps <- rf_feats %>%
  map(names) %>%
  reduce(unique)

go_res <- rf_grps %>%
  map_dfr(~ {
    grp <- .x
    
    go_feats <- rf_feats[rf_typs] %>%
      map(pluck, grp)
    
    go_sim_data %>%
      imap_dfr(~ {
        file <- str_c("ag-", grp, "_", .y, "_go")
        file <- here(params$table_dir, file)
        
        res <- go_feats %>%
          get_go(
            lec_so,
            type_clmn      = "subtype",
            max_term_size  = 500,
            min_term_size  = 10,
            db             = .y,
            simplify_terms = TRUE,
            sim_data       = .x,
            file           = file
          )
        
        res@compareClusterResult %>%
          as_tibble() %>%
          mutate(db = .y, pred_grp = grp)
      })
  })

# Filter GO terms
go_sig <- go_res %>%
  filter(p.adjust < 0.05) %>%
  arrange(p.adjust)

# Plot top CC GO terms
go_fig <- go_sig %>%
  filter(db == "CC") %>%
  group_by(Cluster, pred_grp) %>%
  slice(1:15) %>%
  split(.$pred_grp) %>%
  imap(~ {
    dat <- .x %>%
      mutate(
        y_var = str_c(Cluster, "-", Description),
        y_var = fct_reorder(y_var, p.adjust, min, .desc = TRUE),
      )
      
    y_labs <- dat %>%
      distinct(Description, y_var)
    
    y_labs <- set_names(y_labs$Description, y_labs$y_var)
    
    dat %>%
      ggplot(aes(-log10(p.adjust), y_var, fill = Cluster)) +
      geom_col() +
      geom_vline(xintercept = -log10(0.05), color = "black", linetype = 2) +
      facet_grid(space = "free", scales = "free_y", rows = vars(Cluster)) +
      labs(title = str_c("Ag-", .y), x = "-log10(p-value)") +
      scale_y_discrete(labels = y_labs) +
      scale_fill_manual(values = lec_clrs) +
      base_theme +
      theme(
        legend.position = "none",
        plot.title      = element_text(size = ttl_pt2 * 1.5, hjust = 0.25),
        strip.clip      = "off",
        axis.title.y    = element_blank()
      )
  }) %>%
  plot_grid(
    plotlist = .,
    nrow  = 1,
    align = "vh",
    axis  = "trbl",
    labels     = c("A", "B"),
    label_size = ttl_pt2 * 1.5
  )
```

```{r "RF naive expression", fig.width = 10, fig.height = 4}
naive_dat <- naive_dat %>%
  filter(subtype %in% rf_typs) %>%
  pivot_longer(all_of(ag_modules), names_to = "module_type") %>%
  separate(module_type, sep = "_", into = c("module_type", "module_ag")) %>%
  
  filter(subtype == module_type) %>%
  
  mutate(
    module_type = str_c(module_type, " Ag-", module_ag),
    tm = fct_relevel(as.character(tm), c("0", tms))
  )

naive_fig <- naive_dat %>%
  split(.$module_ag) %>%
  imap(~ {
    .x %>%
      ggplot(aes(tm, value, fill = subtype)) +
      geom_boxplot(outlier.size = 0.25, width = 0.5) +
      facet_grid(~ module_type) +
      scale_fill_manual(values = lec_clrs) +
      labs(x = "days post vaccination", y = "module score") +
      base_theme +
      theme(
        aspect.ratio = 1.1,
        legend.position = "none"
      )
  }) %>%
  plot_grid(
    plotlist = .,
    nrow  = 1,
    align = "h",
    axis  = "tb",
    labels = c("C", "D"),
    label_size = ttl_pt2 * 1.5
  )
```

```{r "RF predicted Ag class", fig.width = 8, fig.height = 5}
# Plot fraction predicted groups
rf_bar_2 <- rf_bar_dat %>%
  filter(!rf_training_data) %>%
  ggplot(aes(tm, fill = rf_pred, alpha = rf_correct)) +
  geom_bar(position = "fill") +
  scale_alpha_manual(values = c(`TRUE` = 1, `FALSE` = 0.5)) +
  scale_fill_manual(values = pred_clrs) +
  scale_y_continuous(breaks = c(0, 0.5, 1)) +
  facet_grid(~ subtype) +
  labs(y = "fraction of cells", x = "days post vaccination") +
  guides(
    fill = guide_legend(order = 1, title = "prediction"),
    alpha = guide_legend(title.theme = element_blank())
  ) +
  bar_thm
```

```{r "Fig S3", fig.width = 20, fig.height = 22}
top <- plot_grid(
  rf_bar_2, mod_bxs[[2]],
  rel_widths = c(0.75, 1),
  labels = c("E", "F"),
  label_size = ttl_pt2 * 1.5,
  align = "h",
  axis  = "tb"
)

left <- plot_grid(
  plotlist = mod_u$fLEC,
  ncol  = 1,
  align = "v",
  axis  = "rl"
)

right <- plot_grid(
  plotlist = mod_u$Collecting,
  ncol  = 1,
  align = "v",
  axis  = "rl"
)

bot <- plot_grid(
  left, right,
  labels = c("G", "H"),
  label_size = ttl_pt2 * 1.5,
  nrow  = 1,
  align = "h",
  axis  = "tb"
)

figS3 <- plot_grid(
  go_fig, naive_fig, top, bot,
  ncol = 1,
  rel_heights = c(0.5, 0.5, 0.5, 1)
)

figS3
```

<br>

<br>

### Table S1

Stats are included for each random forest model.

```{r}
rf_clmns <- c(
  "F1", "Balanced Accuracy",
  "Precision", "Recall",
  "mtry", "min.node.size", "num.trees"
)

tbl_s1 <- rf_mods_best %>%
  filter(subtype %in% rf_typs) %>%
  select(cell_type = subtype, all_of(rf_clmns)) %>%
  mutate(cell_type = fct_relevel(cell_type, rf_typs)) %>%
  arrange(cell_type)

tbl_s1 %>%
  write_tsv(here(params$table_dir, "table_s1.tsv"))

tbl_s1
```

<br>

<br>

### Table S2

Genes used to train random forest models are included along with p-values
indicating overall feature importance.

```{r}
tbl_s2 <- rf_mods_best %>%
  filter(subtype %in% rf_typs) %>%
  mutate(subtype = fct_relevel(subtype, rf_typs)) %>%
  arrange(subtype) %>%
  pmap_dfr(~ {
    args <- list(...)
    typ  <- as.character(args$subtype)
    
    gns <- args$feats_p %>%
      names() %>%
      str_replace_all("_", "-") %>%  # names were adjusted when training models
      str_remove("^x")               # fix names here
    
    tibble(
      cell_type = args$subtype,
      gene      = gns,
      p_value   = args$feats_p
    ) %>%
      mutate(
        Ag_class = case_when(
          gene %in% rf_feats[[typ]]$high ~ "Ag-high",
          gene %in% rf_feats[[typ]]$low  ~ "Ag-low"
        )
      )
  })

tbl_s2 %>%
  write_tsv(here(params$table_dir, "table_s2.tsv"))

tbl_s2
```

<br>

<br>

### Table S3

Biological function and cellular component gene ontology terms are shown for
Ag-low and Ag-high cells.

```{r}
tbl_s3 <- go_sig %>%
  rename(Ag_class = pred_grp, cell_type = Cluster) %>%
  mutate(Ag_class = str_c("Ag-", Ag_class)) %>%
  arrange(cell_type, Ag_class, p.adjust) %>%
  select(-bgID) %>%
  relocate(Ag_class, .after = cell_type)
  
tbl_s3 %>%
  write_tsv(here(params$table_dir, "table_s3.tsv"))

tbl_s3
```

---

<br>

<br>

## Figure 3

*Antigen archiving is enhanced by sequential vaccinations*

A.  Experimental design
B.  Mean 21 day Ag-score is shown for LECs from mice that received a single
    vaccination (21 day or 42 day) or dual vaccination (21 day and 42 day).
C.  UMAP projections show 21 day Ag-scores for LEC subsets for single and dual
    vaccinations.
D.  Prior vaccination enhances antigen archiving.
    Ag-score is shown for single and dual vaccinations for the 21 day timepoint
    for each LEC subset. Other timepoints are shown in grey. P values
    were calculated using a one-sided Wilcoxon rank sum test with
    Benjamini-Hochberg correction.
E.  Mean 42 day Ag-score is shown as described in A.
F.  UMAP projections show 42 day Ag-scores for LEC subsets as described in B.
G.  Successive vaccinations enhances retention of previously archived antigen.
    Ag-score is shown for the 42 day timepoint as described in C.
H.  21 day and 42 day Ag-score is compared for LEC subsets.
I.  Ag-high module scores described in Figure 2 are shown for LECs that archived
    antigens from both vaccinations (double-high), from only one vaccination
    (single-high), or have low levels of both antigens (Ag-low). Module scores
    are shown for the corresponding LEC subset. P values were calculated using
    a one-sided Wilcoxon rank sum test with Benjamini-Hochberg correction.
J.  Ag-low module scores are shown for LEC subset as described in H.

```{r "exchange UMAPs", fig.width = 16, fig.height = 4}
# Format Ag exchange plot data
# use experiment 1 since, experiment 2 is missing 3wk LECs
ex_dat <- lec_so@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(exp %in% c("exp-0", ag_exps$LEC)) %>%
  mutate(
    mouse = m_key[as.character(mouse)],
    mouse = fct_relevel(mouse, unname(m_key))
  )

# Create UMAPs
ex_heat_args <- list(
  ag_clmns = ag_clmns,
  clr      = tm_clrs[c("21", "42")]
)

ex_u <- ex_heat_args %>%
  pmap(~ {
    dat <- ex_dat %>%
      filter(exp == ag_exps$LEC)
    
    # mx_sig <- max(dat[ex_heat_args$ag_clmns])
    
    dat %>%
      plot_scatter(
        ..1,
        x           = "hUMAP_1",
        y           = "hUMAP_2",
        group_col   = "mouse",
        panel_nrow  = 1,
        outline     = TRUE,
        size        = 0.6,
        stroke      = 0.7,
        plot_colors = c("lightblue", "white", ..2),
        label_params = list(size = ttl_pt1)
      ) +
      # scale_fill_gradientn(colours = c("lightblue", "white", ..2), limits = c(0, mx_sig)) +
      guides(fill = guide_colorbar(
        ticks = FALSE,
        title = ag_labs[.x],
        barheight = unit(150, "pt"), barwidth = unit(10, "pt")
      )) +
      umap_theme +
      theme(
        legend.position = "right",
        panel.border    = element_rect(color = ln_col, linewidth = ln_pt)
      )
  })
```

```{r "exchange heatmaps"}
# Format data for heatmaps
heat_dat <- ex_dat %>%
  filter(
    exp == ag_exps$LEC,
    subtype %in% ag_plt_typs$LEC
  ) %>%
  group_by(mouse, subtype) %>%
  summarize(across(all_of(ag_clmns), mean), .groups = "drop")

# Create heatmaps
ex_heat <- ex_heat_args %>%
  pmap(~ {
    heat_dat %>%
      create_exchange_heats(
        ag_clmn = .x,
        ttl     = ag_labs_2[[.x]],
        clr     = .y
      )
  })
```

```{r "exchange boxes"}
# Effect of successive vaccinations on antigen uptake/retention
# use LEC subsets shown in Fig 1
ex_bx_args <- list(
  tm      = c("21", "42"),
  tm_clmn = c("tm_3", "tm_6"),
  ag_clmn = c("Ag_score_3", "Ag_score_6"),
  p_hjust = c(0.5, 0.8)
)

ex_bxs <- ex_bx_args %>%
  pmap(~ {
    args <- list(...)
    
    ex_dat %>%
      create_exchange_boxes(
        tm        = args$tm,
        tm_clmn   = args$tm_clmn,
        ag_clmn   = args$ag_clmn,
        typs      = ag_plt_typs$LEC,
        clrs      = tm_clrs,
        p_cutoff = Inf,
        p_hjust   = args$p_hjust,
        p_vjust   = 0.7
      )
  })
```

```{r "exchange correlation", fig.width = 10, fig.height = 4}
# Format scatter plot data
scat_dat <- lec_so@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(
    vaccination == "dual",
    subtype %in% ag_plt_typs$LEC,
  ) %>%
  mutate(subtype = fct_relevel(subtype, ag_plt_typs$LEC))

ex_scat <- scat_dat %>%
  split(.$exp) %>%
  map(create_exchange_scatter)
```

```{r "exchange Ag-high module scores", fig.width = 9, fig.height = 5.5}
# Format plot data
# module scores for biological replicates are similar
# pool all cells together to increase power
clmns <- c(
  "exp", "subtype", "mouse", "vaccination",
  "rf_pred", "Ag_class_dual", ag_modules
)

ex_mod_bx_dat <- lec_so %>%
  FetchData(clmns) %>%
  filter(vaccination == "dual") %>%
  mutate(Ag_class_dual = recode(Ag_class_dual, low = "Ag-low")) %>%
  
  mutate(Ag_class_dual = ifelse(
    Ag_class_dual == "Ag-low" & rf_pred == "Ag-high",
    "Ag-competent",
    Ag_class_dual)
  )

# Create boxplots
bx_lvls <- c("Ag-low", "Ag-competent", "single-high", "double-high")

rf_mod_args$p_x <- c(1, 4)
rf_mod_args$p_hjust <- c(0.2, 0.8)

ex_mod_bxs <- rf_mod_args %>%
  pmap(~ {
    args <- list(...)

    module_clmns <- str_c(rf_typs, "_", args$mod_typ)
    module_title <- str_c("Ag-", args$mod_typ, " module score")

    ex_mod_bx_dat %>%
      create_module_boxes(
        module_clmns = module_clmns,
        module_title = module_title,
        lvls         = bx_lvls,
        p_alt        = args$p_alt,
        p_x          = args$p_x,
        p_hjust      = args$p_hjust
      ) +
      theme(plot.margin = margin(15, 15, 5, 15))
  })
```

```{r "Fig 3 diagram", fig.width = 15.5, fig.height = 10}
img <- "results/figures/experiment_design/fig3_exp_design_landscape-2.png"
img <- readPNG(here(img))
img <- rasterGrob(img, interpolate = TRUE)

# Adjust plot limits if necessary
img_lims <- tibble(x = c(0, 1), y = c(0, 1))

# Create an empty plot for placing the image, adjusted to fill the area
fig3_dsgn <- ggplot(img_lims, aes(x, y)) + 
  annotation_custom(
    grob = img,
    xmin = -Inf, xmax = Inf,
    ymin = -Inf, ymax = Inf
  ) +
  xlim(c(-Inf, Inf)) +
  ylim(c(-Inf, Inf)) +
  theme_void() +
  theme(plot.margin = margin(0, 0, 30, 0))

fig3_dsgn <- plot_grid(
  fig3_dsgn, NULL,
  rel_widths = c(1, 0.4)
)
```

```{r "Fig 3", fig.width = 15.5, fig.height = 23}
fig3 <- fig3_dsgn /
  (ex_heat[[1]] + ex_u[[1]]) / ex_bxs[[1]] /
  (ex_heat[[2]] + ex_u[[2]]) / ex_bxs[[2]] /
  ex_scat[[ag_exps$LEC]] /
  (ex_mod_bxs[[1]] + ex_mod_bxs[[2]]) +
  plot_layout(heights = c(1.1, 1, 0.55, 1, 0.55, 0.8, 0.95)) +
  plot_annotation(tag_levels = "A") &
  theme(
    plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"),
    plot.tag.position = c(0, 1)
  )

fig3
```

<br>

<br>

### Figure S4

A.  Ag-score is shown for single and dual vaccinations for the 21 day timepoint
    for DC subsets from a biological replicate.
    This analysis was not performed for a biological replicate for LEC subsets
    due to the low number of LECs recovered (<40 total cells).
    Other timepoints are shown in grey.
    P values were calculated using a one-sided Wilcoxon rank sum test with
    Benjamini-Hochberg correction.
B.  Ag-score is shown for the 42 day timepoint for LEC subsets from a
    biological replicate, as described in A.
C.  Ag-score is shown for the 42 day timepoint for DC subsets from a biological
    replicate, as described in A.
D.  21 day and 42 day Ag-score is compared for LEC subsets from a biological
    replicate.

```{r "rep exchange boxes"}
# Set input data
# plot replicates not shown in main figure
exps        <- c("exp-1", "exp-2")
plt_lec_exp <- exps[exps != ag_exps$LEC]
plt_dc_exp  <- exps[exps != ag_exps$DC]

ex_bx_args$exps <-list(
  list(LEC = NULL,        DC = plt_dc_exp),
  list(LEC = plt_lec_exp, DC = plt_dc_exp)
)

# Create boxplots for biological replicates
rep_ex_bxs <- ex_bx_args %>%
  pmap(~ {
    args <- list(...)
    
    objs <- purrr::discard(args$exps, is.null)
    objs <- ag_objs[names(objs)]
    
    objs %>%
      imap(~ {
        dat <- .x@meta.data
        
        dat <- args$exps[[.y]] %>%
          map_dfr(~ {
            dat %>%
              filter(exp %in% c("exp-0", .x))
          })
        
        dat %>%
          create_exchange_boxes(
            tm         = args$tm,
            tm_clmn    = args$tm_clmn,
            ag_clmn    = args$ag_clmn,
            clrs       = tm_clrs,
            typs       = ag_plt_typs[[.y]],
            p_hjust    = args$p_hjust,
            facet_vars = "subtype",
            p_cutoff   = Inf
          ) +
          scale_y_continuous(expand = expansion(c(0.05, 0.15))) +
          theme(legend.position = "bottom")
      })
  }) %>%
  flatten()
```

```{r "RF module exchange correlation", fig.height = 8}
# Set plot colors
ex_ag_clrs <- c(
  `double-high` = "#D7301F",
  `single-high` = "black",
  `Ag-low`      = "#56B4E9"
  # `single-high` = "#6A51A3",
)

# Format plotting data
scat_dat <- lec_so@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(vaccination == "dual") %>%
  mutate(Ag_class_dual = recode(Ag_class_dual, low = "Ag-low"))

# Create scatter plots
scat_args <- list(
  mod_typ     = c("high", "low"),
  label_x     = c(-Inf, Inf),
  label_hjust = c(-0.1, 1.1)
)

ex_mod_scat <- scat_args %>%
  pmap(~ {
    args <- list(...)
    
    module_clmns <- str_c(rf_typs, "_", args$mod_typ)
    module_title <- str_c("Ag-", args$mod_typ, " module score")
    
    scat_dat %>%
      create_module_scatter(
        ag_clmns     = ag_clmns,
        module_clmns = module_clmns,
        clrs         = ex_ag_clrs,
        module_title = module_title,
        label_x      = args$label_x,
        label_hjust  = args$label_hjust
      )
  })
```

```{r "Fig S4", fig.width = 13, fig.height = 15}
# Create final figure
figS4 <- wrap_plots(rep_ex_bxs, ncol = 1) /
  ex_scat$`exp-2` /
  (ex_mod_scat[[1]] + ex_mod_scat[[2]]) +
  plot_layout(heights = c(1, 1, 1, 1.5, 3)) +
  plot_annotation(tag_levels = "A") &
  theme(
    plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"),
    plot.tag.position = c(0, 1)
  )

figS4
```

---

<br>

<br>

## Figure 4

*Antigen exchange with DCs correlates with levels of antigen archiving*

A.  Annotated regions used for antigen quantification are shown for a
    representative lymph node.
B.  Segment type is shown as described in A.
C.  Antigen signal is shown as described in A.
D.  Relative antigen tag signal is shown for each sample and each
    region segmented based on Lyve1 (LECs) and CD11C (DCs). Normalized 21 day
    antigen signal was scaled across all regions from 21 day mice and dual
    immunized mice, 42 day antigen signal was scaled across all regions from
    42 day mice and dual immunized mice. The number of plotted segments is shown
    above each boxplot.
E.  Normalized antigen tag signal is shown for regions that included both Lyve1+
    and CD11C+ segments.
    21 day antigen signal is shown for 21 day and dual immunized mice,
    42 day signal is shown for 42 day and dual immunized mice.
    The Spearman correlation coefficient is shown.
F.  Mean 21 day Ag-score is shown for DCs from mice that received a single
    vaccination (21 day or 42 day) or dual vaccination (21 day and 42 day).
G.  Ag-score is shown for single and dual vaccinations for the 21 day timepoint
    for DC subsets with highest Ag-score. Other timepoints are shown in grey.
    P values were calculated using a one-sided Wilcoxon rank sum test with
    Benjamini-Hochberg correction.
H.  Mean 42 day Ag-score is shown for DCs from mice that received single or
    dual vaccination as described in F.
I.  Ag-score is shown for single and dual vaccinations for the 42 day timepoint
    as described in G.

```{r "geomx cropped images", fig.width = 14, fig.height = 7}
plot_geomx_image <- function(obj_in, color, clrs = NULL, labs = NULL,
                             legend_values = names(labs) %||% names(clrs),
                             ttl = NULL, line_size = 0.5, scale_size = 8,
                             scale_dist = 1.5, ...) {
  
  labs <- labs %||% waiver()
  ttl  <- ttl %||% color
  
  res <- obj_in %>%
    plotSpatialOverlay(
      colorBy          = color,
      scaleBar         = TRUE,
      corner           = "bottomleft",
      textDistance     = scale_dist,
      scaleBarLineSize = line_size,
      scaleBarFontSize = scale_size / .pt,
      ...
    ) +
    ggtitle(ttl) +
    theme(
      plot.margin  = margin(r = 15, l = 15),
      plot.title   = element_text(hjust = 0.5, size = ttl_pt2 * 1.5),
      legend.title = element_blank()
    )
  
  if (!is.null(legend_values)) {
    all_sams_df <- tibble(x = 1, y = 1, colorBy = legend_values)
    
    res <- res +
      geom_blank(aes(x, y, fill = colorBy), data = all_sams_df)
  }
  
  if (!is.null(clrs)) {
    if (is.numeric(obj_in@plottingFactors[[color]])) {
      res <- res +
        scale_fill_gradientn(colours = clrs) +
        guides(fill = guide_colorbar(ticks = FALSE)) +
        theme(
          legend.position   = "bottom",
          legend.key.width  = unit(50, "pt"),
          legend.key.height = unit(7, "pt"),
          legend.text       = element_text(size = txt_pt2)
        )
    } else {
      res <- res +
        scale_fill_manual(values = clrs, labels = labs) +
        theme(legend.text = element_text(size = ttl_pt2))
    }
  }
  
  res
}

# Select LN to plot
sams <- geomx_dat %>%
  filter(
    LN == "ova-GeoMX+VV 6week and 3week LN1",
    Segment %in% c("Lyve1", "CD11C")
  ) %>%
  pull(Sample_ID)

img_cropped <- geomx_imgs$`Slide 1` %>%
  cropSamples(sampleIDs = sams, sampsOnly = TRUE)

img_args <- list(
  color = c("Sample", "Region", "Segment", "Ag-score"),
  clrs  = list(m_clrs, reg_clrs, seg_clrs, c("white", "#D7301F")),
  labs  = list(geo_m_key, NULL, geo_seg_key, NULL),
  ttl   = list(NULL, NULL, "Segment type", "Ag signal")
)

# Plot cropped image
img_fig <- img_args %>%
  pmap(~ {
    args <- list(...)
    
    img_cropped %>%
      plot_geomx_image(
        color = args$color,
        clrs  = args$clrs,
        ttl   = args$ttl,
        labs  = args$labs,
        scale_size = ttl_pt1,
        scale_dist = 3,
        legend_values = NULL,
      ) +
      theme(legend.position = "bottom")
  })

img_fig <- plot_grid(
  plotlist = img_fig[-1],
  nrow  = 1,
  align = "h",
  axis  = "tb",
  hjust = 0.05,
  labels     = c("A", "B", "C"),
  label_size = ttl_pt2 * 1.5
)
```

```{r "geomx Ag boxplots", fig.height = 4}
# Signals to plot
geo_clmns <- c(
  Ag_3wk_score = "Ag signal (21 day)",
  Ag_6wk_score = "Ag signal (42 day)"
)

# Function for strip labels
geo_lab_fn <- function(x) {
  keys <- list(geo_m_key, geo_seg_key)
  
  keys %>%
    walk(~ {
      idx    <- x %in% names(.x)
      x[idx] <- .x[x[idx]]
      x      <<- x
    })
  
  x
}

# Calculate relative Ag signal
# want to plot both Ag tags on same plot
# this is not useful without scaling values since Ag-3wk signal is much higher
# than Ag-6wk signal
bx_dat <- geo_clmns %>%
  imap_dfr(~ {
    sam <- str_extract(.y, "(?<=Ag_)[0-9]+wk")
    
    geomx_dat %>%
      filter(
        sample %in% c(sam, "6wk-3wk"),
        Segment %in% c("Lyve1", "CD11C")
      ) %>%
      mutate(
        key = .y,
        Ag_rel = as.numeric(scale(!!sym(.y))),
        Segment = fct_relevel(Segment, names(seg_clrs))
      )
  })

# Calculate p-values
p_dat <- bx_dat %>%
  split(.$Segment) %>%
  imap_dfr(~ {
    pairwise.wilcox.test(
      .x$Ag_rel, .x$region_3,
      p.adjust.method = "none"
    ) %>%
      tidy() %>%
      mutate(Segment = .y)
  }) %>%
  mutate(p_adj = p.adjust(p.value, method = "BH"))

# Calculate n values
n_dat <- bx_dat %>%
  group_by(region_3, Segment) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(n = str_c("n = ", n))

# Create boxplots
geo_bxs <- bx_dat %>%  
  ggplot(aes(region_3, Ag_rel, fill = region_3)) +
  geom_boxplot(
    linewidth = ln_pt * 1.5, alpha = 1, outlier.color = NA, width = 0.65,
    key_glyph = draw_key_point
  ) +
  geom_jitter(size = 2, width = 0.1) +
  geom_text(
    aes(region_3, Inf, label = n, fill = NULL),
    data  = n_dat,
    vjust = 1.4,
    size  = txt_pt2 / .pt
  ) +
  
  facet_grid(~ Segment, scales = "free_y", labeller = as_labeller(geo_lab_fn)) +
  scale_fill_manual(values = reg_clrs) +
  scale_y_continuous(expand = expansion(c(0.05, 0.2))) +
  labs(y = "Relative Ag signal") +
  base_theme +
  theme(
    aspect.ratio    = 1.4,
    legend.position = "none",
    axis.text.x     = element_text(size = ttl_pt2, angle = 45, hjust = 1),
    axis.title.y    = element_text(size = ttl_pt2),
    axis.title.x    = element_blank()
  )
```

```{r "geomx Ag scatter"}
# Format data
# want to plot both Ag tags on same plot
sct_dat <- geo_clmns %>%
  imap_dfr(~ {
    sam <- str_extract(.y, "(?<=Ag_)[0-9]+wk")
    
    geomx_dat %>%
      filter(
        sample %in% c(sam, "6wk-3wk"),
        Segment %in% c("Lyve1", "CD11C")
      ) %>%
      mutate(
        key      = .y,
        Ag_score = !!sym(.y)
      )
  })

sct_dat <- sct_dat %>%
  dplyr::select(
    `ROI Coordinate X`, `ROI Coordinate Y`,
    region_3, Segment, sample, key,
    Ag_score
  ) %>%
  group_by(key, `ROI Coordinate X`, `ROI Coordinate Y`) %>%
  filter(all(c("Lyve1", "CD11C") %in% Segment)) %>%
  ungroup() %>%
  mutate(Segment = str_c(Segment, "+ Ag signal")) %>%
  pivot_wider(names_from = Segment, values_from = Ag_score) %>%
  mutate(
    tm = str_extract(key, "[0-9+]wk"),
    tm = as.character(tm_key[tm])
  )

# Legend labels
sct_labs <- sct_dat %>%
  group_by(key, tm) %>%
  summarize(n = str_c("n = ", n()), .groups = "drop") %>%
  mutate(n = str_c(geo_clmns[key], "\n", n))

sct_labs <- set_names(sct_labs$n, sct_labs$tm)

# Calculate correlation
cor <- cor(
  sct_dat$`CD11C+ Ag signal`,
  sct_dat$`Lyve1+ Ag signal`,
  method = "spearman"
)
cor <- round(cor, digits = 2)
cor = str_c("italic(r[s]) == ", cor)

geo_sct <- sct_dat %>%
  ggplot(aes(`CD11C+ Ag signal`, `Lyve1+ Ag signal`, color = tm)) +
  geom_smooth(method = "lm", color = "black", alpha = 0.1, linewidth = 0.5, linetype = 2) +
  geom_point(size = 3) +
  geom_text(
    aes(x = 0, y = max(`Lyve1+ Ag signal`), label = cor),
    parse = TRUE,
    hjust = -0.3,
    vjust = 0.5,
    color = "black",
    size  = ttl_pt1 / .pt,
    check_overlap = TRUE
  ) +
  scale_color_manual(values = tm_clrs, labels = sct_labs) +
  scale_x_continuous(trans = "log10") +
  scale_y_continuous(trans = "log10") +
  guides(color = guide_legend(override.aes = list(size = 4))) +
  base_theme +
  theme(
    aspect.ratio = 1,
    legend.title = element_blank(),
    legend.text  = element_text(size = ttl_pt2),
    axis.title   = element_text(size = ttl_pt2)
  )
```

```{r "geomx figure", fig.width = 14, fig.height = 10}
bot <- plot_grid(
  geo_bxs, geo_sct,
  nrow   = 1,
  align  = "h",
  axis   = "tb",
  labels = c("D", "E"),
  label_size = ttl_pt2 * 1.5
)

geo_fig <- plot_grid(
  img_fig, bot,
  ncol = 1,
  rel_heights = c(1, 0.75)
)
```

```{r "DC exchange heatmaps", fig.width = 16, fig.height = 4}
# Format data for DC exchange plots
dc_ex_dat <- dc_so@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(
    exp %in% c("exp-0", ag_exps$DC),
    subtype %in% ag_plt_typs$DC
  ) %>%
  mutate(
    mouse = m_key[as.character(mouse)],
    mouse = fct_relevel(mouse, unname(m_key))
  )

# Format data for heatmaps
dc_heat_dat <- dc_ex_dat %>%
  filter(exp == ag_exps$DC) %>%
  group_by(mouse, subtype) %>%
  summarize(across(all_of(ag_clmns), mean), .groups = "drop")

# Create heatmaps
dc_ex_heat <- ex_heat_args %>%
  pmap(~ {
    dc_heat_dat %>%
      create_exchange_heats(
        ag_clmn = .x,
        ttl     = ag_labs_2[[.x]],
        clr     = .y
      )
  })
```

```{r "DC exchange boxes", fig.width = 10, fig.height = 2}
# Effect of successive vaccinations on antigen uptake/retention
# use DC subsets shown in figure 1
dc_ex_bxs <- ex_bx_args %>%
  pmap(~ {
    args <- list(...)
    
    dc_ex_dat %>%
      create_exchange_boxes(
        tm        = args$tm,
        tm_clmn   = args$tm_clmn,
        ag_clmn   = args$ag_clmn,
        typs      = ag_plt_typs$DC[1:3],
        clrs      = tm_clrs,
        p_size    = 11,
        p_hjust   = args$p_hjust,
        p_vjust   = 0.7
      ) +
      theme(legend.position = "bottom")
  })
```

```{r "Fig 4", fig.width = 14, fig.height = 18}
bot <- list(
  `F` = dc_ex_heat[[1]],
  G   = dc_ex_bxs[[1]],
  H   = dc_ex_heat[[2]],
  I   = dc_ex_bxs[[2]]
)

bot <- bot %>%
  imap(~ .x + labs(tag = .y))

bot <- wrap_plots(
  bot,
  nrow = 2,
  widths = c(0.25, 1)
) &
  theme(
    plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"),
    plot.tag.position = c(0, 1)
  )

fig4 <- plot_grid(
  geo_fig, NULL, bot,
  ncol = 1,
  rel_heights = c(1, 0.1, 0.7)
)

fig4
```

<br>

<br>

### Xenium figures

This figure could replace the GeoMx data shown in Figure 4A-E.

```{r "XENIUM DATA"}
# Top cell types to plot
dcs <- xen$cell_type %>%
  unique() %>%
  grep("DC", ., value = TRUE) %>%
  sort()

ag_dcs <- xen %>%
  filter(cell_type %in% dcs) %>%
  group_by(cell_type) %>%
  filter(all(table(Ag_class) > 15)) %>%
  pull(cell_type) %>%
  unique()

top_types <- c("LEC", dcs)

# FOVs to plot
fovs      <- sort(unique(xen$fov))
top_fov   <- "D"
top_fov_2 <- "B"

# Format plotting data
xen <- xen %>%
  mutate(Ag_class = fct_relevel(Ag_class, c("Ag-low", "Ag-high"))) %>%
  arrange(Ag_class)

# Xenium colors
xen_clrs <- typ_clrs

xen_clrs["unassigned"] <- "grey65"

xen_clrs <- xen_clrs %>%
  spruce_up_colors(
    adjust_colors = "Neutrophils",
    filter        = c("deutan", "protan", "tritan"),
    difference    = 10,
    property      = "lightness",
    range         = c(10, 85)
  )

xen_clrs <- xen_clrs[names(xen_clrs) %in% xen$cell_type]
```

```{r "XENIUM AG CLASS"}
# Format plot data
dat <- xen %>%
  filter(fov %in% top_fov) %>%
  mutate(
    broad_type = if_else(cell_type %in% dcs, "DC", cell_type),
    broad_type = if_else(
      broad_type %in% c("LEC", "DC"),
      broad_type,
      "other"
    ),
    broad_type = fct_relevel(broad_type, c("LEC", "DC", "other")),
    
    Ag_class = recode(Ag_class, `Ag-high` = "Ag+", `Ag-low` = "Ag-"),
    Ag_class = fct_relevel(Ag_class, c("Ag+", "Ag-")),
    
    type_ag = if_else(
      broad_type %in% c("LEC", "DC"),
      str_c(broad_type, " (", Ag_class, ")"),
      broad_type
    ),
    slide = str_c("r", slide)
  ) %>%
  arrange(broad_type, Ag_class) %>%
  mutate(type_ag = fct_inorder(type_ag))

# Adjust color palette
new_clrs <- c("#E7301F", "#56B4E9", "white") %>%
  expand_colors(
    n             = c(2, 2, 1),
    keep_original = TRUE,
    filter        = c("deutan", "protan", "tritan"),
    difference    = Inf,
    direction     = -1,
    range         = c(10, 90),
    maxit         = 2000
  )

names(new_clrs) <- c(
  "LEC (Ag-)", "LEC (Ag+)",
  "DC (Ag-)", "DC (Ag+)",
  "other"
)

# Format legend labels
lec_dc_ag_labs <- dat %>%
  group_by(type_ag) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    n_lab = str_c(type_ag, "\nn = ", label_comma()(n))
  ) %>%
  arrange(type_ag)

lec_dc_ag_labs <- set_names(
  lec_dc_ag_labs$n_lab,
  lec_dc_ag_labs$type_ag
)

# Plot Ag+ cells
ag_pos_fig <- dat %>%
  arrange(desc(type_ag)) %>%
  mutate(type_ag = fct_inorder(type_ag)) %>%
  
  ggplot(aes(y_cell, x_cell, fill = type_ag)) +
  geom_point_trace(
    size = 0.5,
    color = NA,
    trace_position = type_ag != "other",
    background_params = list(color = "black", size = 0.25)
  ) +
  facet_wrap(~ slide) +
  guides(fill = guide_legend(
    nrow = 2,
    title.position = "top",
    override.aes = list(size = 4, color = "black")
  )) +
  scale_fill_manual(
    values = new_clrs,
    breaks = names(new_clrs),
    labels = lec_dc_ag_labs,
    drop   = FALSE
  ) +
  coord_fixed() +
  umap_theme_2 +
  theme(
    legend.position   = "bottom",
    legend.key.height = unit(35, "pt"),
    legend.title      = element_blank(),
    legend.text       = element_text(size = ttl_pt2),
    strip.text        = element_text(size = ttl_pt2)
  )
```

```{r "XENIUM AG SIGNAL"}
# Plot Ag+ cells
ag_sig_fig <- dat %>%
  arrange(Ag_counts) %>%
  
  ggplot(aes(y_cell, x_cell, fill = Ag_counts + 1)) +
  geom_point_trace(size = 0.5, trace_position = "bottom") +
  facet_wrap(~ slide) +
  scale_fill_gradientn(colours = c("white", "#E7301F"), trans = "log10") +

  guides(fill = guide_colorbar(
    ticks = FALSE,
    title.position = "top",
    title = "Ag counts + 1"
  )) +
  coord_fixed() +
  umap_theme_2 +
  theme(
    legend.position   = "bottom",
    legend.key.width  = unit(35, "pt"),
    legend.key.height = unit(7, "pt"),
    legend.title      = element_text(size = ttl_pt2, hjust = 0.5),
    legend.text       = element_text(size = txt_pt2),
    strip.text        = element_text(size = ttl_pt2)
  )
```

```{r "XENIUM LEC PROXIMITY", fig.width = 18, fig.height = 13}
# Ag class labels
ag_classes <- c(
  `Ag-low`  = "Ag-",
  `Ag-high` = "Ag+"
)

# Format data for histograms
# only include cell types that have both Ag-low/high cells
hist_dat <- xen %>%
  dplyr::select(cell_id, slide, fov, cell_type, lec_type, Ag_class, lec_id, lec_dist) %>%
  filter(
    fov %in% top_fov,
    cell_type == "DC"
  ) %>%
  group_by(slide, fov, cell_type) %>%
  filter(
    all(names(ag_classes) %in% Ag_class),
    all(table(Ag_class) > 5)
  ) %>%
  ungroup() %>%
  mutate(slide = str_c("r", slide))
  
# Format data for median lines
ln_dat <- hist_dat %>%
  group_by(slide, fov, cell_type, Ag_class) %>%
  summarize(
    lec_dist = median(lec_dist),
    n        = n(),
    .groups  = "drop"
  )

# Format data for n labels
n_dat <- ln_dat %>%
  group_by(slide, fov, cell_type) %>%
  mutate(
    n_lab = str_c(ag_classes[Ag_class], " ", label_comma()(n), " cells")
  ) %>%
  summarize(
    n_lab = str_c(n_lab, collapse = "\n"),
    .groups = "drop"
  )

# Format data for p-values
p_dat <- hist_dat %>%
  group_by(slide, fov, cell_type) %>%
  summarize(
    p = wilcox.test(
      lec_dist[Ag_class == "Ag-high"],
      lec_dist[Ag_class == "Ag-low"],
      alternative = "less"
    )$p.value,
    .groups = "drop"
  ) %>%
  mutate(p_adj = p.adjust(p, method = "bonferroni")) %>%
  rowwise() %>%
  mutate(
    p_lab = .format_pvalue(p_adj),
    p_lab = ifelse(
      p_adj != 0,
      str_c("italic(p) == ", p_lab),
      p_lab
    )
  ) %>%
  ungroup()

# Plot distance from nearest Ag+ LEC
hist <- hist_dat %>%
  mutate(lec_dist = lec_dist + 1) %>%  # ADD 1
  ggplot(aes(lec_dist, fill = Ag_class, alpha = Ag_class)) +
  
  geom_density(aes(color = Ag_class)) +
  geom_vline(
    aes(xintercept = lec_dist, color = Ag_class, alpha = NULL),
    data      = ln_dat,
    linetype  = 2,
    linewidth = 0.75,
    show.legend = FALSE
  ) +
  
  geom_text(
    aes(x = 1, y = Inf, label = n_lab, fill = NULL, alpha = NULL),
    data = n_dat,
    hjust = 0,
    vjust = 1.2,
    size  = txt_pt2 / .pt,
    show.legend = FALSE
  ) +
  
  geom_text(
    aes(x = Inf, y = Inf, label = p_lab, fill = NULL, alpha = NULL),
    data  = p_dat,
    parse = TRUE,
    hjust = 1.1,
    vjust = 1.2,
    size  = txt_pt2 / .pt,
    show.legend = FALSE
  ) +
  
  facet_wrap(~ slide, nrow = 1) +
  labs(x = "distance from nearest Ag+ LEC", y = "density") +
  scale_fill_manual(values = c(`Ag-low` = "grey85", `Ag-high` = "#D7301F"), labels = ag_classes) +
  scale_color_manual(values = c(`Ag-low` = "black", `Ag-high` = "#D7301F"), labels = ag_classes) +
  scale_alpha_manual(values = c(`Ag-low` = 0.5, `Ag-high` = 0.85), labels = ag_classes) +
  scale_x_log10() +
  scale_y_continuous(expand = expansion(c(0.05, 0.2))) +
  base_theme +
  theme(
    aspect.ratio    = 0.5,
    legend.position = "right",
    legend.title    = element_blank(),
    strip.text      = element_text(size = ttl_pt2),
    legend.text     = element_text(size = ttl_pt2),
    axis.title      = element_text(size = ttl_pt2)
  )
```

```{r "XENIUM MAIN FIGURE", fig.width = 14, fig.height = 8.5}
top <- wrap_plots(
  ag_sig_fig + labs(tag = "A"),
  ag_pos_fig + labs(tag = "B"),
  nrow = 1
)

bot <- wrap_plots(
  plot_spacer(), hist + labs(tag = "C"), plot_spacer(),
  nrow = 1,
  widths = c(0.2, 1, 0.2)
)

xen_fig <- wrap_plots(
  top, bot,
  ncol = 1,
  heights = c(1, 0.65)
) &
  theme(
    plot.tag = element_text(
      size  = ttl_pt2 * 1.5,
      face  = "bold",
      hjust = 2,
      vjust = -0.2
    )
  )

xen_fig
```

<br>

Ag-tag barcodes used for each mouse are shown below.

```{r}
params$xen_samples %>%
  imap_dfr(~ {
    tibble(
      fov       = .y,
      mouse     = .x[[1]],
      `6wk tag` = params$xen_ag_targets[[.x[[2]]]],
      `3wk tag` = params$xen_ag_targets[[.x[[3]]]]
    ) %>%
      mutate(
        across(ends_with(" tag"), ~ {
          .x %>%
            str_remove("-ERCC[0-9]+$") %>%
            str_remove("^GeoMx-")
        })
      )
  })
```

<br>

The fraction of total counts assigned to each custom target is shown below for
each tissue section.
The GeoMx-Barcode02 probes produced very sparse signals regardless of sample
identity.
Values are labeled for targets assigned >0.01% of total counts.

```{r "fraction custom counts", fig.height = 6, fig.width = 12}
# Format plotting data
feats <- c(
  "mouse", "fov", "cell_type",
  "nCount_Xenium", "nCount_Xenium_custom"
)

# NEED TO REPROCESS SLIDE 2
qc_dat <- seq_along(slides)[1] %>%
  map(~ {
    s <- qread(here(params$xen_dir, str_c("xen_s", .x, ".qs")))
    
    DefaultAssay(s) <- "Xenium_custom"
    
    s %>%
      FetchData(c(feats, params$xen_custom_targets), slot = "counts") %>%
      as_tibble(rownames = ".cell_id") %>%
      pivot_longer(all_of(params$xen_custom_targets)) %>%
      mutate(
        slide  = .x,
        sample = str_c(mouse, "-", fov)
      )
  })

qc_dat <- bind_rows(qc_dat)

# Format heatmap data
heat_dat <- qc_dat %>%
  group_by(slide, fov, mouse, name) %>%
  mutate(
    tot_counts = sum(nCount_Xenium + nCount_Xenium_custom)
  ) %>%
  group_by(name, slide, fov, mouse, sample, tot_counts) %>%
  summarize(
    value   = sum(value),
    .groups = "drop"
  ) %>%
  mutate(
    frac   = value / tot_counts,
    fov    = fct_relevel(fov, fovs),
    name   = fct_reorder(name, value),
    mouse  = fct_relevel(mouse, c("6wk", "3wk", "6wk-3wk"))
  )

# Tiles to outline
min_frac <- 0.0001

top_dat <- heat_dat %>%
  filter(frac >= min_frac)

# Tiles to label
lab_dat <- heat_dat %>%
  filter(frac >= min_frac) %>%
  mutate(
    lab = round(frac * 100, 2),
    lab = str_c(lab, "%")
  )

# Create heatmap
custom_counts_heat <- heat_dat %>%
  ggplot(aes(fov, name, fill = frac + min_frac)) +
  geom_tile() +
  geom_tile(
    data = top_dat,
    color = "black"
  ) +
  geom_text(
    aes(color = NULL, fill = NULL, label = lab),
    data = lab_dat
  ) +
  facet_grid(~ mouse, scales = "free") +
  scale_fill_gradientn(
    colours = c("white", "#D7301F"),
    labels  = label_percent(),
    trans   = "log10",
    name    = str_c("fraction of\ntotal counts\n(+", min_frac * 100, "%)")
  ) +
  base_theme +
  theme(
    aspect.ratio       = 5,
    strip.clip         = "off",
    strip.text.y.right = element_text(angle = 0, hjust = 0),
    axis.title.y       = element_blank()
  )

custom_counts_heat
```

<br>

Cells were identified as Ag+ if they have at least 1 Ag-tag count.
The fraction of Ag+ cells is shown below for each Ag-tag for
each tissue section.

```{r "fraction 3wk/6wk Ag+ cells", fig.height = 3.5, fig.width = 6}
bar_dat <- xen %>%
  group_by(fov, mouse) %>%
  summarize(
    `3wk tag` = (sum(Ag_3wk_counts > 0) / n()) * 100,
    `6wk tag` = (sum(Ag_6wk_counts > 0) / n()) * 100,
    .groups   = "drop"
  )

bar_dat %>%
  pivot_longer(ends_with("wk tag")) %>%
  mutate(mouse = fct_relevel(mouse, c("6wk", "3wk", "6wk-3wk"))) %>%
  
  ggplot(aes(fov, value, fill = name)) +
  geom_col(position = position_dodge2()) +
  facet_wrap(~ mouse, nrow = 1, scales = "free_x") +
  scale_fill_brewer() +
  scale_fill_spruce(difference = 30, range = c(25, 75)) +
  
  labs(y = "% Ag+ cells") +
  base_theme +
  theme(
    aspect.ratio = 1.5,
    legend.title = element_blank(),
    axis.title.x = element_blank()
  )
```



```{r "XENIUM AG TYPE HEAT ALL SUPPLEMENTAL", fig.width = 14, fig.height = 6, eval = FALSE}
bar_dat <- xen %>%
  group_by(fov, slide, cell_type) %>%
  summarize(
    n = n(),
    pct_ag_pos = (sum(Ag_class == "Ag-high") / n) * 100,
    .groups = "drop"
  ) %>%
  filter(n > 100)

# Filter to exclude cell types with < 100 cells
bar_dat %>%
  mutate(cell_type = fct_reorder(cell_type, pct_ag_pos)) %>%
  
  group_by(fov, slide) %>%
  mutate(pct_ag_pos = as.numeric(scale(pct_ag_pos))) %>%
  
  ggplot(aes(fov, cell_type, fill = pct_ag_pos)) +
  geom_tile() +
  facet_wrap(~ slide) +
  scale_fill_gradientn(colours = c("white", "#E7301F"), name = "relative %\nAg+ cells") +
  base_theme +
  theme(
    aspect.ratio = n_distinct(bar_dat$cell_type) / n_distinct(bar_dat$fov),
    axis.title.y = element_blank()
  )
```

```{r "XENIUM 3WK 6WK SIGNAL", eval = FALSE}

bar_dat <- xen@meta.data %>%
  group_by(fov, cell_type) %>%
  summarize(
    ag_pct_1 = (sum(Ag_tag_1 > 0) / n()) * 100,
    ag_pct_2 = (sum(Ag_tag_2 > 0) / n()) * 100,
    .groups  = "drop"
  )

# Filter to exclude cell types with < 100 cells
bar_dat %>%
  mutate(cell_type = fct_reorder(cell_type, pct_ag_pos)) %>%
  
  group_by(fov) %>%
  mutate(pct_ag_pos = as.numeric(scale(pct_ag_pos))) %>%
  
  ggplot(aes(fov, cell_type, fill = pct_ag_pos)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("white", "#E7301F")) +
  base_theme +
  theme(
    aspect.ratio = n_distinct(bar_dat$cell_type) / n_distinct(bar_dat$fov)
  )


```

```{r "XENIUM UMAPS", eval = FALSE}
top <- c("LEC", "other CD45-", "BEC")

xen@meta.data %>%
  plot_scatter(
    "cell_type",
    group_col = "fov",
    size      = 0.00001
    # top       = top
  ) +
  scale_color_spruce(difference = 20) +
  theme(aspect.ratio = 0.9)

xen@meta.data %>%
  mutate(Ag_counts = Ag_tag_1 + Ag_tag_2 + 1) %>%
  plot_scatter(
    "Ag_counts",
    outline     = TRUE,
    plot_colors = c("white", "red"),
    group_col   = "fov",
    size        = 0.00001,
    trans       = "log10"
  ) +
  theme(aspect.ratio = 0.9)

xen@meta.data %>%
  filter(cell_type %in% top) %>%
  plot_scatter(
    "SCT_snn_res.10",
    group_col = "fov",
    size      = 0.00001
  ) +
  scale_color_spruce(difference = 20) +
  theme(aspect.ratio = 0.9)

# Boxplots
gns <- c(
  "Ptprc", "Pdpn", "Pecam1",
  "Cd3d", "Cd19",
  "Prox1", "Lyve1",
  "Itgax",           # DCs
  "Xcr1",            # cDC1
  "Sirpa", "Itgam",  # cDC2
  "Siglech",         # Siglech DCs
  "Ccr7"
)

dat <- xen %>%
  FetchData(c(
    "fov", "cell_type",
    "UMAP_1", "UMAP_2",
    "SCT_snn_res.10", gns
  ))

gns %>%
  map(~ {
    dat %>%
      # filter(cell_type %in% top) %>%
      plot_violin(
        .x,
        cluster_col = "SCT_snn_res.10",
        method = "boxplot",
        top = 20
      ) +
      scale_fill_spruce() +
      scale_color_spruce()
  }) %>%
  plot_grid(plotlist = .)

dat %>%
  filter(cell_type %in% top) %>%
  plot_violin(
    "Ptprc",
    cluster_col = "SCT_snn_res.10",
    method = "boxplot",
    # plot_colors = c("lightblue", "red"),
  )

dat %>%
  plot_scatter(
    "Ptprc",
    group_col = "fov",
    size = 0.0001,
    plot_colors = c("lightblue", "red")
  )

dat %>%
  ggplot(aes(Ptprc)) +
  geom_histogram() +
  facet_wrap(~ fov)

```

```{r "XENIUM AG TYPE HEAT 1", fig.width = 14, fig.height = 6, eval = FALSE}
bar_dat <- xen %>%
  filter(fov %in% c("C", "D")) %>%
  group_by(cell_type, fov, slide) %>%
  filter(n() > 100) %>%

  group_by(cell_type) %>%
  # group_by(slide, fov, cell_type) %>%
  summarize(
    n = n(),
    pct_ag_pos = (sum(Ag_class == "Ag-high") / n) * 100,
    .groups = "drop"
  )
    
# Filter to exclude cell types with < 100 cells
bar_dat %>%
  mutate(cell_type = fct_reorder(cell_type, pct_ag_pos)) %>%
  
  # group_by(fov) %>%
  # mutate(pct_ag_pos = as.numeric(scale(pct_ag_pos))) %>%
  
  ggplot(aes(pct_ag_pos, cell_type, fill = cell_type)) +
  geom_col() +
  # facet_wrap(~ slide, nrow = 1) +
  # scale_fill_gradientn(colours = c("white", "#E7301F")) +
  scale_fill_manual(values = xen_clrs) +
  base_theme
  # theme(
  #   aspect.ratio = n_distinct(bar_dat$cell_type) / n_distinct(bar_dat$fov)
  # )
```

<br>

<br>

### Figure S6

A.  Sample identity is shown for all regions analyzed using the GeoMx DSP
    platform.
B.  Annotated regions are shown as described in A.
C.  Normalized 21 day antigen tag signal is shown for each sample and each
    region segmented based on Lyve1 (LECs), CD11C (DCs), and CD4 (T cells). The
    number of plotted segments is shown above each boxplot.
D.  Normalized 42 day antigen tag signal is shown as described in C.

```{r "geomx uncropped images", fig.width = 10, fig.height = 8}
all_img_fig <- img_args %>%
  pmap(~ {
    args <- list(...)
    
    geomx_imgs %>%
      imap(~ {
        .x %>%
          plot_geomx_image(
            color = args$color,
            clrs  = args$clrs,
            labs  = args$labs,
            ttl   = .y
          )
      }) %>%
      wrap_plots(
        guides = "collect",
        nrow = 1
      ) &
      theme(legend.position = "bottom")
  })

all_img_fig <- plot_grid(
  plotlist = all_img_fig[1:2],
  nrow     = 1,
  align    = "h",
  axis     = "tb",
  labels   = c("A", "B"),
  label_size = ttl_pt2 * 1.5
)
```

```{r "geomx all Ag boxplots", fig.width = 10, fig.height = 6.5}
# Create boxplot
all_geo_bxs <- geo_clmns %>%
  imap(~ {
    dat <- geomx_dat %>%
      filter(sample != "naive", Segment != "Other") %>%
      mutate(Segment = fct_relevel(Segment, names(seg_clrs)))
    
    n_dat <- dat %>%
      group_by(region_3, Segment, sample) %>%
      summarize(n = n(), .groups = "drop") %>%
      mutate(n = str_c("n = ", n))
    
    dat %>%  
      ggplot(aes(region_3, !!sym(.y), fill = region_3)) +
      geom_boxplot(key_glyph = draw_key_point, alpha = 1, outlier.size = 0.25, width = 0.65) +
      geom_text(
        aes(region_3, Inf, label = n, fill = NULL),
        data  = n_dat,
        vjust = 1.4,
        size  = 8 / .pt
      ) +
      
      facet_grid(Segment ~ sample, scales = "free_y", labeller = as_labeller(geo_lab_fn)) +
      scale_fill_manual(values = reg_clrs) +
      scale_y_continuous(expand = expansion(c(0.05, 0.2))) +
      labs(y = .x) +
      base_theme +
      theme(
        aspect.ratio    = 1.4,
        legend.position = "none",
        axis.text.x     = element_text(angle = 45, hjust = 1),
        axis.title.x    = element_blank()
      )
  }) %>%
  plot_grid(
    plotlist = .,
    nrow  = 1,
    align = "h",
    axis  = "tb",
    labels     = c("C", "D"),
    label_size = ttl_pt2 * 1.5
  )
```

```{r "Fig S5", fig.width = 14, fig.height = 12}
figS5 <- plot_grid(
  all_img_fig, NULL, all_geo_bxs,
  ncol = 1,
  rel_heights = c(1, 0.1, 1)
)

figS5
```

---

<br>

<br>

## Figure 5

*Antigen archiving is impaired during CHIKV infection*

A.  UMAP projections show LEC subsets for mock and CHIKV-infected mice.
B.  UMAP projections show predicted Ag-competent cLECs for mock and
    CHIKV-infected mice.
C.  The fraction of predicted Ag-competent cLECs is shown for mock and
    CHIKV-infected mice for each biological replicate. P values were calculated
    using Fisher's exact test with Benjamini-Hochberg correction.
D.  UMAP projections show cLEC Ag-high module scores for mock and CHIKV-infected
    mice.
E.  cLEC Ag-high module scores are shown for mock and CHIKV-infected mice for
    each biological replicate. P values were calculated
    using a two-sided Wilcoxon rank sum test with Benjamini-Hochberg correction.
F.  Expression of the cLEC Ag-high gene module is shown for mock and CHIKV-infected
    mice for cLECs from each biological replicate.

```{r "CHIKV data"}
# CHIKV theme
chikv_clrs <- c(
  `Ag-low`       = "#56B4E9",
  `Ag-competent` = "#D7301F",
  other          = "white"
)

treat_clrs <- set_names(chikv_clrs[1:2], c("mock", "CHIKV"))

chikv_u_theme <- umap_theme_2 +
  theme(aspect.ratio = 0.7)

# Load CHIKV data
rm(lec_so, dc_so)

chikv_so <- qread(here(params$chikv_dir, "so_lec.qs"))

chikv_so <- chikv_so %>%
  subset(tm == "24hpi") %>%
  AddMetaData(chikv_preds, col.name = str_c("rf_", colnames(chikv_preds)))

chikv_so <- chikv_so %>%
  mutate_meta(
    mutate,
    pred_grp  = dplyr::recode(rf_pred, high = "Ag-competent", low = "Ag-low"),
    pred_grp  = fct_relevel(pred_grp, rev(names(chikv_clrs))),
    treatment = fct_relevel(treatment, c("mock", "CHIKV")),
    subtype   = lec_subtype  # add subtype column to keep consistent with other
  )                          # objects

chikv_so <- chikv_so %>%
  AddMetaData(FetchData(., c("hUMAP_1", "hUMAP_2")))

# Add module scores
rf_feats %>%
  iwalk(~ {
    clmn_nms <- str_c(.y, "_", names(.x))

    clmn_nms <- set_names(
      str_c(clmn_nms, seq_along(clmn_nms)),
      clmn_nms
    )

    chikv_so <<- chikv_so %>%
      AddModuleScore(
        features = .x,
        name     = names(clmn_nms)
      ) %>%
      mutate_meta(dplyr::rename, !!!syms(clmn_nms))
  })

# Data for plotting
chikv_dat <- chikv_so@meta.data %>%
  as_tibble(rownames = ".cell_id")
```

```{r "CHIKV subsets"}
# LEC subsets
top <- chikv_so %>%
  plot_scatter(
    "lec_subtype",
    x            = "hUMAP_1",
    y            = "hUMAP_2",
    group_col    = "treatment",
    plot_colors  = lec_clrs,
    size         = 0.75,
    label_params = list(size = ttl_pt2)
  ) +
  chikv_u_theme +
  theme(
    legend.title = element_blank(),
    legend.text  = element_text(size = 11.5),
    panel.border = element_blank()
  )
```

```{r "CHIKV Ag class UMAPs"}
# Create UMAPs
u_args <- list(
  typ = set_names(rf_typs),
  pt_size = list(1, c(0.5, 0.1), c(0.5, 0.1))
)

chikv_u <- u_args %>%
  pmap(~ {
    chikv_dat %>%
      create_rf_module_umaps(
        typ           = .x,
        ag_class_clmn = "pred_grp",
        module_clmn   = str_c(.x, "_high"),
        module_title  = str_c(.x, " Ag-high module"),
        facet_var     = "treatment",
        lab_fn        = "label_value",
        class_clrs    = chikv_clrs,
        ag_clrs       = c("lightblue", "white", "#D7301F"),
        pt_size       = 0.75,
        circle_cells  = FALSE,
        show_scale_title = TRUE
      )
  })
```

```{r "CHIKV Ag class bargraphs"}
# Format data for bargraphs
chikv_bar_dat <- chikv_dat %>%
  filter(subtype %in% rf_typs) %>%
  mutate(subtype = fct_relevel(subtype, rf_typs))

# Create Ag class bargraphs
chikv_ag_bars <- chikv_bar_dat %>%
  filter(subtype == rf_cell_type) %>%
  create_chikv_class_bars(
    x   = "treatment",
    n_p = n_distinct(chikv_bar_dat$subtype),
    ttl = str_c("fraction Ag-competent ", rf_cell_type, "s")
  ) +
  theme(strip.text = element_blank())
```

```{r "CHIKV Ag module score boxplots"}
# Format data for plotting
chikv_bx_dat <- chikv_dat %>%
  filter(subtype %in% rf_typs) %>%
  mutate(subtype = fct_relevel(subtype, rf_typs))

# Plot Ag module scores
chikv_ag_bxs <- chikv_bx_dat %>%
  filter(subtype == rf_cell_type) %>%
  create_chikv_module_boxes(
    module_clmns = "cLEC_high",
    ttl          = "cLEC Ag-high module",
    n_p          = length(ag_modules)
  ) +
  theme(strip.text = element_blank())
```

```{r "CHIKV Ag-high heatmap"}
# Ag-high heatmap data
# * two genes are not in the CHIKV object
heat_feats <- rf_feats$cLEC$high
heat_feats <- heat_feats[heat_feats %in% rownames(chikv_so)]

heat_dat <- chikv_so %>%
  FetchData(c("treatment", "rep", "lec_subtype", heat_feats)) %>%
  filter(lec_subtype %in% rf_cell_type) %>%
  pivot_longer(all_of(heat_feats)) %>%
  group_by(lec_subtype, treatment, rep, name) %>%
  summarize(value = mean(value), .groups = "drop")

gn_lvls <- heat_dat %>%
  pivot_wider(names_from = treatment, values_from = value) %>%
  mutate(value = CHIKV / mock, .keep = "unused") %>%
  arrange(desc(value)) %>%
  pull(name) %>%
  unique()
  
heat_dat <- heat_dat %>%
  group_by(name) %>%
  mutate(value = as.numeric(scale(value))) %>%
  ungroup() %>%
  mutate(
    treatment = fct_relevel(treatment, c("mock", "CHIKV")),
    name      = fct_relevel(name, rev(gn_lvls))
  )

# Ag heatmap
heat <- heat_dat %>%
  ggplot(aes(name, rep, fill = value)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("#56B4E9", "white", "#D7301F")) +
  facet_wrap(~ treatment, ncol = 1, strip.position = "left") +
  guides(fill = guide_colorbar(
    barwidth = unit(150, "pt"),
    barheight = unit(6, "pt")
  )) +
  base_theme +
  theme(
    aspect.ratio    = length(rf_typs) / length(heat_feats),
    strip.placement = "outside",
    strip.clip      = "off",
    legend.position = "bottom",
    legend.title    = element_blank(),
    axis.title      = element_blank(),
    axis.text.x     = element_text(angle = 90, hjust = 1, vjust = 0.5)
  )
```

```{r "Cormac's figure"}
img_g <- readPNG(here("results/figures/morrison_chikv/fig5g.png"))
img_h <- readPNG(here("results/figures/morrison_chikv/fig5h.png"))

fig5gh <- list(img_g, img_h) %>%
  map(~ {
    img <- rasterGrob(.x, interpolate = TRUE)
    
    # Adjust plot limits if necessary
    img_lims <- tibble(x = c(0, 1), y = c(0, 1))
    
    # Create an empty plot for placing the image, adjusted to fill the area
    ggplot(img_lims, aes(x, y)) + 
      annotation_custom(
        grob = img,
        xmin = -Inf, xmax = Inf,
        ymin = -Inf, ymax = Inf
      ) +
      xlim(c(-Inf, Inf)) +
      ylim(c(-Inf, Inf)) +
      theme_void()
  }) %>%
  plot_grid(
    plotlist = .,
    nrow     = 1,
    align    = "h",
    axis     = "tb",
    labels   = c("G", "H"),
    label_size = ttl_pt2 * 1.5,
    rel_widths = c(1, 0.7)
  )
```

```{r "Cormac's data", eval = FALSE}
# Format data for plot
tbl_nm <- "ova+ LECs (%)"

pzfx_dat <- read_pzfx(
  here("results/figures/morrison_chikv/TEM2306 results.pzfx"),
  table = tbl_nm
)

pzfx_lvls <- colnames(pzfx_dat)

pzfx_clrs <- set_names(
  c("#56B4E9", "#E69F00", "#D7301F"),
  pzfx_lvls
)

pzfx_dat <- pzfx_dat %>%
  as_tibble() %>%
  filter(if_any(everything(), ~ !is.na(.x))) %>%
  mutate(mouse = row_number()) %>%
  pivot_longer(-mouse) %>%
  filter(!is.na(value)) %>%
  mutate(name = fct_relevel(name, pzfx_lvls))

# Add significance labels based on Cormac's original plot
pzfx_sig <- list(
  `CHIKV 181/25` = "*",
  `polyI:C/ova`  = "**"
)

p_dat <- pzfx_dat %>%
  pull(name) %>%
  unique() %>%
  expand.grid(., .) %>%
  filter(
    Var1 != Var2,
    Var1 == last(pzfx_lvls),
  ) %>%
  mutate(Var2 = fct_relevel(Var2, pzfx_lvls)) %>%
  arrange(desc(Var2)) %>%
  mutate(
    sig  = pzfx_sig[as.character(Var2)],
    y    = 110 + (10 * (row_number() - 1)),
    x    = match(Var2, pzfx_lvls),
    x    = x + ((length(pzfx_lvls) - x) / 2)
  )

# Format data for bargraphs
bar_dat <- pzfx_dat %>%
  group_by(name) %>%
  summarize(value = median(value), .groups = "drop")

# Create bargraphs
pzfx_plt <- pzfx_dat %>%
  ggplot(aes(name, value, fill = name)) +
  geom_col(data = bar_dat, width = 0.75) +
  geom_point(size = 3, show.legend = FALSE) +
  
  geom_segment(
    aes(x = Var1, xend = Var2, y = y, yend = y, fill = NULL),
    data = p_dat
  ) +
  geom_text(
    aes(x = x, y = y, label = sig, fill = NULL),
    data = p_dat,
    size = ttl_pt2 * 1.5 / .pt
  ) +
  
  scale_fill_manual(values = pzfx_clrs) +
  scale_y_continuous(expand = expansion(c(0.01, 0.05))) +
  labs(y = tbl_nm) +
  base_theme +
  theme(
    plot.margin     = margin(20, 0, 0, 0),
    aspect.ratio    = 1.3,
    legend.position = "none",
    axis.title.x    = element_blank(),
    axis.text.x     = element_text(size = ttl_pt2, angle = 30, hjust = 1)
  )
```

```{r "Fig 5", fig.width = 12, fig.height = 20}
# Combine and annotate center panels
c("D", "B") %>%
  iwalk(~ {
    chikv_u$cLEC[[.y]][[1]] <<- chikv_u$cLEC[[.y]][[1]] + labs(tag = .x)
  })

chikv_u$cLEC[[1]] <- chikv_u$cLEC[[1]] &
  theme(
    legend.position   = "left",
    legend.title      = element_text(size = ttl_pt2, hjust = 0.5, angle = 90),
    legend.key.height = unit(30, "pt"),
    legend.key.width  = unit(6, "pt")
  )

mid <- wrap_plots(
  chikv_u$cLEC[[2]],
  chikv_ag_bars + labs(tag = "C"),
  chikv_u$cLEC[[1]],
  chikv_ag_bxs + labs(tag = "E"),
  widths = c(1, 0.3)
) &
  theme(plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"))

# Final figure
fig5 <- plot_grid(
  top, mid, heat, fig5gh,
  ncol       = 1,
  labels     = c("A", "", "F", ""),
  label_size = ttl_pt2 * 1.5,
  hjust      = c(0, 0, -0.5),
  vjust      = c(1.5, 1, 0),
  rel_heights = c(1, 2.5, 1, 1)
)

fig5
```

<br>

<br>

### Figure S7

A.  Ag-high and Ag-low module scores are shown for the predicted Ag classes.
    P values were calculated using a two-sided Wilcoxon rank sum test with
    Benjamini-Hochberg correction.
B.  Ag-high and Ag-low module scores are shown for mock and CHIKV-infected mice
    for each biological replicate. P values were calculated using a two-sided
    Wilcoxon rank sum test with Benjamini-Hochberg correction.
C.  The total fraction of cLECs, fLECs, and collecting LECs predicted to be
    Ag-competent is shown for mock and CHIKV-infected mice for each biological
    replicate.
    P values were calculated using Fisher's exact test with Benjamini-Hochberg
    correction.
D.  The fraction of fLECs and collecting LECs predicted to be Ag-competent is
    shown as described in C.

```{r "Fig S7", fig.width = 13.5, fig.height = 13}
# Plot Ag module scores for predicted classes
all_chikv_bxs_1 <- chikv_bx_dat %>%
  create_chikv_module_boxes(
    module_clmns = ag_modules,
    x            = "pred_grp",
    clrs         = pred_clrs,
    ttl          = "module score"
  ) +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))

# Plot Ag module scores for treatment groups for all cell types
all_chikv_bxs_2 <- chikv_bx_dat %>%
  create_chikv_module_boxes(
    module_clmns = ag_modules,
    ttl = "module score"
  )

# Bargraphs for combined LEC subsets
all_chikv_bars <- chikv_bar_dat %>%
  mutate(subtype = "cLEC, fLEC, Collecting") %>%
  create_chikv_class_bars(
    x   = "treatment",
    n_p = n_distinct(chikv_bar_dat$subtype),
    ttl = "fraction Ag-competent"
  )

# Bargraphs for other LEC subtypes
other_chikv_bars <- chikv_bar_dat %>%
  filter(subtype != rf_cell_type) %>%
  create_chikv_class_bars(
    x   = "treatment",
    n_p = n_distinct(chikv_bar_dat$subtype),
    ttl = "fraction Ag-competent"
  )

# Create final figure
figS6 <- plot_grid(
  all_chikv_bxs_1, all_chikv_bxs_2,
  all_chikv_bars,  other_chikv_bars,
  labels = c("A", "B", "C", "D"),
  label_size = ttl_pt2 * 1.5,
  nrow  = 2,
  align = "vh",
  axis  = "trbl",
  rel_heights = c(1, 0.6)
)

figS6
```

<br>

<br>

### Figure S8

A.  The fraction of cLECs predicted to be Ag-competent is shown for
    metastasis-free lymph node (MFLN) and follicular lymphoma (FL) samples
    from Abe et al.
B.  The total fraction of cLECs predicated to be Ag-competent is shown for
    pooled MFLN and FL samples shown in A.
    P values were calculated using Fisher's exact test.

```{r "Fig S8", fig.width = 10, fig.height = 4.5, out.width = "75%"}
# Plot fraction Ag-competent cLECs
# * cLECs are the LEC1 population
# * only include samples with at least 20 cLECs
abe_bar_dat <- abe_lec_so@meta.data %>%
  rownames_to_column(".cell_id") %>%
  filter(type == "LEC1") %>%
  mutate(
    tissue = fct_relevel(tissue, names(abe_clrs)),
    sample = str_remove(sample, "_NHC$")
  ) %>%
  group_by(sample, tissue) %>%
  filter(n() > 20) %>%
  summarize(
    frac_high = sum(pred_grp == "Ag-competent") / n(),
    .groups = "drop"
  )

abe_bars_1 <- abe_bar_dat %>%
  ggplot(aes(sample, frac_high, fill = tissue)) +
  geom_col() +
  facet_grid(~ tissue, scales = "free_x", space = "free_x") +
  scale_fill_manual(values = abe_clrs) +
  scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(0, 1)) +
  labs(y = "fraction Ag-competent cLECs") +
  base_theme +
  theme(
    legend.position = "none",
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1)
  )

# Bargraphs for combined samples
p_dat <- abe_bar_dat %>%
  mutate(
    p = wilcox.test(
      frac_high[tissue == "FL"],
      frac_high[tissue == "MFLN"]
    )$p.value
  ) %>%
  group_by(tissue, p) %>%
  summarize(n = n(), .groups = "drop") %>%
  rowwise() %>%
  mutate(
    p_lab = str_c("italic(p) == ", .format_pvalue(p)),
    n_lab = str_c(tissue, "\nn = ", n)
  )

n_labs <- set_names(p_dat$n_lab, p_dat$tissue)

abe_bars_2 <- abe_bar_dat %>%
  group_by(tissue) %>%
  summarize(frac_high = median(frac_high), .groups = "drop") %>%
  ggplot(aes(tissue, frac_high, fill = tissue)) +
  geom_col() +
  geom_jitter(data = abe_bar_dat, size = 2, width = 0.03) +
  geom_text(
    aes(1.5, Inf),
    label = unique(p_dat$p_lab),
    parse = TRUE,
    size  = 12 / .pt,
    vjust = 1.3,
  ) +
  scale_fill_manual(values = abe_clrs) +
  scale_y_continuous(
    breaks = c(0, 0.5, 1),
    limits = c(0, 1)
  ) +
  scale_x_discrete(labels = n_labs) +
  labs(y = "fraction Ag-competent cLECs") +
  base_theme +
  theme(
    aspect.ratio    = 1.5,
    legend.position = "none",
    axis.title.x    = element_blank()
  )

# Create final figure
plot_grid(
  abe_bars_1, abe_bars_2,
  rel_widths = c(1, 0.5),
  labels     = c("A", "B"),
  label_size = ttl_pt2 * 1.5,
  align      = "h",
  axis       = "tb"
)
```

---

<br>

<br>

## Session info

```{r "session info"}
sessionInfo()
```



```{r "BETHS GRANT FIG", fig.width = 12, fig.height = 18, eval = FALSE}
# Features to plot
heat_feats <- rf_feats$cLEC$high
heat_feats <- heat_feats[heat_feats %in% rownames(chikv_so)]

# Format Ira's data
# object was saved with Seurat v5, need to convert to v4 object
ira_so <- read_rds(here(params$so_dir, "lec_complete_object.rds"))

cell_ids <- attr(ira_so[["RNA"]], "cells")@.Data %>%
  as.data.frame() %>%
  filter(counts)

genes <- attr(ira_so[["RNA"]], "features")@.Data %>%
  as.data.frame() %>%
  filter(counts)

counts <- attr(ira_so[["RNA"]], "layers")$counts

rownames(counts) <- rownames(genes)
colnames(counts) <- rownames(cell_ids)

# Check that sparse matrix has correct column/row names
count_sums <- colSums(counts)
feat_sums  <- colSums(counts > 0)

stopifnot(
  identical(colnames(counts), rownames(ira_so@meta.data)),
  all(count_sums == ira_so@meta.data$nCount_RNA),
  all(feat_sums  == ira_so@meta.data$nFeature_RNA)
)

# Create new object and format data
ira_dat <- counts %>%
  CreateSeuratObject(meta.data = ira_so@meta.data) %>%
  NormalizeData(normalization.method = "LogNormalize") %>%
  subset(orig.ident == "2d") %>%
  FetchData(c("orig.ident", "cell_type", heat_feats)) %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(
    !is.na(cell_type),
    cell_type == "ISG-high LEC"
  ) %>%
  mutate(
    rep = cell_type,
    treatment = "ISG\nhigh"
  ) %>%
  rename(lec_subtype = cell_type)

# Add Ira's data to CHIKV object
ira_dat <- chikv_so %>%
  FetchData(c("orig.ident", "treatment", "rep", "lec_subtype", heat_feats)) %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(lec_subtype %in% rf_cell_type) %>%
  
  bind_rows(ira_dat) %>%
  pivot_longer(all_of(heat_feats))

heat_dat <- ira_dat %>%
  group_by(orig.ident, lec_subtype, treatment, rep, name) %>%
  summarize(value = mean(value), .groups = "drop")
  
# Format heatmap data
gn_lvls <- heat_dat %>%
  filter(orig.ident != "2d") %>%
  select(treatment, rep, name, value) %>%
  pivot_wider(names_from = treatment, values_from = value) %>%
  mutate(value = CHIKV / mock, .keep = "unused") %>%
  arrange(desc(value)) %>%
  pull(name) %>%
  unique()

heat_dat <- heat_dat %>%
  group_by(name) %>%
  mutate(value = as.numeric(scale(value))) %>%
  ungroup() %>%
  mutate(
    treatment = fct_relevel(treatment, c("mock", "CHIKV")),
    name      = fct_relevel(name, rev(gn_lvls))
  )

# Ag heatmap
heat <- heat_dat %>%
  ggplot(aes(name, rep, fill = value)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("#56B4E9", "white", "#D7301F")) +
  facet_grid(treatment ~ ., scales = "free_y", space = "free_y", switch = "y") +
  guides(fill = guide_colorbar(
    barwidth = unit(150, "pt"),
    barheight = unit(6, "pt")
  )) +
  base_theme +
  theme(
    strip.placement = "outside",
    strip.clip      = "off",
    legend.position = "bottom",
    legend.title    = element_blank(),
    axis.title      = element_blank(),
    axis.text.x     = element_text(angle = 90, hjust = 1, vjust = 0.5),
    axis.text.y     = element_blank()
  )

# Create final figure
plot_grid(
  top, mid, heat,
  ncol       = 1,
  labels     = c("A", "", "F"),
  label_size = ttl_pt2 * 1.5,
  hjust      = c(0, 0, -0.5),
  vjust      = c(1.5, 1, 0),
  rel_heights = c(1, 2.5, 1)
)

ggsave(
  filename = here("results/figures/2024-04-02_fig5_for_grant.png"),
  device = "png",
  dpi    = 300,
  bg     = "white",
  width  = 12,
  height = 18,
)

# # Boxplots
# ira_dat %>%
#   mutate(
#     treatment = fct_relevel(treatment, c("mock", "CHIKV", "ISG\nhigh")),
#     name = fct_relevel(name, rev(gn_lvls))
#   ) %>%
#   ggplot(aes(orig.ident, value, fill = treatment)) +
#   geom_boxplot() +
#   facet_wrap(~ name) +
#   base_theme
```

```{r "TEMS GRANT FIG", eval = FALSE}
# Save figure for Tem's grant March 2024
c("C", "A") %>%
  iwalk(~ {
    chikv_u$cLEC[[.y]][[1]] <<- chikv_u$cLEC[[.y]][[1]] + labs(tag = .x)
  })

mid <- wrap_plots(
  chikv_u$cLEC[[2]],
  chikv_ag_bars + labs(tag = "B"),
  chikv_u$cLEC[[1]],
  chikv_ag_bxs + labs(tag = "D"),
  widths = c(1, 0.3)
) &
  theme(plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"))

new_fig5 <- plot_grid(
  mid, heat,
  ncol       = 1,
  labels     = c("", "E"),
  label_size = ttl_pt2 * 1.5,
  hjust      = c(0, -0.5),
  vjust      = c(1, 0),
  rel_heights = c(2.5, 1)
)

new_fig5 %>%
  ggsave(
    filename = here("results", str_c(Sys.Date(), "_fig5.png")),
    device = "png",
    width  = 12,
    height = 13.2,
    bg     = "white",
    dpi    = 600
  )
```



```{r "EXCHANGE SCATTER PLOTS EXTENDED", fig.width = 14, fig.height = 4, eval = FALSE}
# Plot data
dat <- lec_so@meta.data %>%
  as_tibble(rownames = "cell_id") %>%
  filter(
    exp == "exp-1",
    mouse == "6wk-3wk",
  ) %>%
  # group_by(subtype) %>%
  # filter(n() > 50) %>%
  # ungroup() %>%
  mutate(subtype = fct_relevel(subtype, ag_plt_typs$LEC))

# Create scatter plots
x_var <- "Ag_6wk_score"
y_var <- "Ag_3wk_score"

ag_labs <- c(
  Ag_3wk_score = "Ag-score (21 day)",
  Ag_6wk_score = "Ag-score (42 day)"
)

dat %>%
  ggplot(aes(!!sym(x_var), !!sym(y_var), color = Ag_class_dual)) +
  geom_abline(slope = 1, intercept = 0, linewidth = 0.5, color = "grey90") +
  geom_point(size = 1.5) +
  geom_smooth(
    method    = "lm",
    linetype  = 2,
    linewidth = 0.5,
    color     = "black",
    se        = FALSE,
    formula   = "y ~ x"
  ) +
  facet_wrap(~ subtype, nrow = 1) +
  coord_cartesian(
    xlim = c(0, max(c(dat[[x_var]], dat[[y_var]]))),
    ylim = c(0, max(c(dat[[x_var]], dat[[y_var]])))
  ) +
  labs(
    x = ag_labs[x_var],
    y = ag_labs[y_var]
  ) +
  base_theme +
  theme(
    aspect.ratio = 1,
    legend.position = "bottom",
    legend.title    = element_blank()
  )
```

```{r "AG-EX DOUBLE POSITIVE EXTENDED", fig.width = 14, fig.height = 3, eval = FALSE}
# Expression of Ag-high gene modules is shown for each LEC subset. Expression is
# similar for the day 42 cLEC Ag-low/high groups, likely due to the many Ag-low
# cells that are actually Ag-competent and should express the Ag-high gene
# module.
# 
# Format plot data
lvls <- c("Ag-low", "Ag-high", "single-high", "double-high")

plt_dat <- lec_so %>%
  FetchData(c("subtype", "mouse", "Ag_class_dual", ag_modules)) %>%
  filter(
    mouse %in% c("6wk", "3wk", "6wk-3wk"),
    subtype %in% rf_typs
  ) %>%
  pivot_longer(all_of(ag_modules)) %>%
  separate_wider_regex(
    name,
    c(module_type = ".*", "_", module_ag = ".*")
  ) %>%
  filter(
    subtype == module_type,
    module_ag == "high"
  ) %>%
  
  # Group all low cells together under single mouse, otherwise three boxplots
  # are shown
  mutate(
    subtype       = fct_relevel(subtype, rf_typs),
    mouse         = m_key[as.character(mouse)],
    Ag_class_dual = str_replace(Ag_class_dual, "^[0-9]+wk", "Ag"),
    Ag_class_dual = recode(Ag_class_dual, low = "Ag-low"),
    Ag_class_dual = fct_relevel(Ag_class_dual, lvls)
  ) %>%
  arrange(Ag_class_dual) %>%
  mutate(
    mouse_class = fct_inorder(str_c(mouse, "-", subtype, "-", Ag_class_dual))
  )

# Format n labels
n_dat <- plt_dat %>%
  group_by(Ag_class_dual, mouse_class, subtype, module_ag) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    n = str_c(Ag_class_dual, "\nn = ", scales::label_comma()(n))
  ) %>%
  filter(module_ag == "high")

n_lab <- set_names(n_dat$n, n_dat$mouse_class)

# Create boxplots
ex_mod_bxs <- rf_typs %>%
  map(~ {
    plt_dat %>%
      filter(subtype == .x) %>%
      ggplot(aes(mouse_class, value, fill = subtype)) +
      geom_boxplot(notch = TRUE, outlier.size = 0.1) +
      facet_grid(~ mouse, scales = "free_x", space = "free_x") +
      scale_fill_manual(values = lec_clrs) +
      scale_x_discrete(labels = n_lab) +
      labs(y = str_c(.x, " Ag-high\nmodule score")) +
      base_theme +
      theme(
        legend.position = "none",
        axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.x = element_blank()
      )
  })

ex_mod_bxs %>%
  plot_grid(
    plotlist = .,
    align    = "h",
    axis     = "tb",
    nrow     = 1
  )
```



```{r "OLD CHIKV Ag module score UMAPs", eval = FALSE}
# cLEC Ag-high module scores
ag_score_u <- chikv_dat %>%
  plot_scatter(
    "cLEC_high",
    x           = "hUMAP_1",
    y           = "hUMAP_2",
    group_col   = "treatment",
    plot_colors = c("#56B4E9", "white", "#D7301F"),
    outline     = TRUE,
    # size        = 0.5,
    size        = 0.75,
    stroke      = 0.7,
    n_label     = "none"
  ) +
  guides(fill = guide_colorbar(
    title = "cLEC\nAg-high module",
    title.hjust = 0.5,
    barwidth = unit(150, "pt"),
    barheight = unit(10, "pt")
  )) +
  chikv_u_theme +
  theme(
    legend.position = "bottom"
    # legend.title = element_blank()
  )
```

```{r "OLD EXTENDED 1-2", eval = FALSE}
lec_dat %>%
  group_by(subtype, tm, exp) %>%
  summarize(n = n(), .groups = "drop")
```

```{r "OLD EXTENDED 1-3", fig.width = 9, fig.height = 6, eval = FALSE}
# Ag-score is shown for all DC subsets for each experiment

# Data for boxplots
dc_dat <- dc_so@meta.data %>%
  filter(mouse != "6wk-3wk") %>%
  mutate(subtype = fct_reorder(subtype, Ag_score, mean, .desc = TRUE))

ln_dat <- dc_dat %>%
  group_by(subtype, exp, tm) %>%
  summarize(Ag_score = median(Ag_score), .groups = "drop")

# Create boxplots
exps <- c("exp-1", "exp-2")

exps %>%
  map(~ {
    dc_dat %>%
      filter(exp %in% c("exp-0", .x)) %>%
      
      ggplot(aes(tm, Ag_score, fill = subtype)) +
      
      geom_line(
        data = filter(ln_dat, exp %in% c("exp-0", .x)),
        color = "grey85", linewidth = 0.5, linetype = 2
      ) +
      geom_boxplot(aes(group = mouse), outlier.size = 0.2, width = 3, key_glyph = draw_key_point) +
      
      guides(
        fill = guide_legend(override.aes = list(size = 4, shape = 22), order = 1),
        linetype = guide_legend(override.aes = list(color = "black"), title = NULL)
      ) +
      facet_grid(~ subtype) +
      
      scale_fill_manual(values = dc_clrs) +
      labs(title = .x, x = "days post vaccination", y = "Ag-score") +
      base_theme +
      theme(
        aspect.ratio    = 0.9,
        legend.position = "none",
        plot.margin     = margin(0, 15, 0, 15)
      )
  }) %>%
  plot_grid(
    plotlist = .,
    ncol = 1
  )

```

```{r "OLD EXTENDED 1-4", eval = FALSE}
dc_dat %>%
  group_by(subtype, tm, exp) %>%
  summarize(n = n(), .groups = "drop")
```



```{r "EXTENDED 4-1", eval = FALSE}

dat %>%
  plot_scatter(
    # "Ag_6wk_score",
    "Ag_3wk_score",
    x           = "hUMAP_1",
    y           = "hUMAP_2",
    group_col   = "mouse",
    panel_nrow  = 1,
    # outline     = TRUE,
    size        = 0.5,
    stroke      = 0.7,
    # plot_colors = c("lightblue", "white", "#D7301F"),
    plot_colors = c("lightblue", "white", rep("#6A51A3", 3)),
    # max_q       = 99.9,
    label_params = list(size = ttl_pt1)
  ) +
  # scale_fill_gradientn(colours = c("lightblue", "white", ..2), limits = c(0, mx_sig)) +
  guides(fill = guide_colorbar(
    ticks = FALSE,
    # title = ag_labs_2[.x],
    barheight = unit(150, "pt"), barwidth = unit(10, "pt")
  )) +
  umap_theme +
  theme(
    legend.position = "right",
    panel.border    = element_rect(color = ln_col, linewidth = ln_pt)
  )

```

```{r "EXTENDED 4-2", eval = FALSE}

# Create UMAPs
plt_args <- list(
  ag_clmns = ag_clmns,
  clr      = tm_clrs[c("21", "42")]
)

mx_sig <- max(dat[plt_args$ag_clmns])

ex_u <- plt_args %>%
  pmap(~ {
    dat %>%
      plot_scatter(
        ..1,
        x           = "hUMAP_1",
        y           = "hUMAP_2",
        group_col   = "mouse",
        panel_nrow  = 1,
        outline     = TRUE,
        size        = 1,
        stroke      = 0.7,
        plot_colors = c("lightblue", "white", ..2),
        max_q       = 99,
        label_params = list(size = ttl_pt1)
      ) +
      # scale_fill_gradientn(colours = c("lightblue", "white", ..2), limits = c(0, mx_sig)) +
      guides(fill = guide_colorbar(
        ticks = FALSE,
        title = ag_labs_2[.x],
        barheight = unit(150, "pt"), barwidth = unit(10, "pt")
      )) +
      umap_theme +
      theme(
        legend.position = "right",
        panel.border    = element_rect(color = ln_col, linewidth = ln_pt)
      )
  })

```



```{r "Pdpn Ag correlation", eval = FALSE}
# Format data for plotting
clmns <- c(
  "Pdpn", "Ag_score",
  "subtype", "mouse",
  "pred_grp", "Ag_class", "training_data"
)

dat <- lec_so %>%
  FetchData(clmns) %>%
  mutate(
    Ag_class = fct_relevel(Ag_class, c("low", "high")),
    pred_grp = fct_relevel(pred_grp, names(pred_clrs)),
    subtype  = fct_relevel(subtype, rf_typs)
  ) %>%
  filter(!training_data)
  
# Plot Pdpn for Ag-class
dat %>%
  filter(subtype %in% rf_typs, mouse != "d2") %>%
  ggplot(aes(subtype, Pdpn, fill = pred_grp)) +
  geom_boxplot(outlier.size = 0.5) +
  scale_fill_manual(values = pred_clrs) +
  facet_wrap(~ mouse) +
  base_theme +
  theme(
    aspect.ratio = 0.6,
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

ggsave(
  here("results/Pdpn_Ag_class.png"),
  dpi = 300
)

# Plot Pdpn for LEC subsets
dat %>%
  plot_violin(
    "Pdpn",
    "subtype",
    "mouse",
    method = "boxplot",
    plot_colors = lec_clrs,
    color = "black",
    alpha = 1,
    outlier.size = 0.5
  ) +  
  base_theme +
  theme(
    aspect.ratio = 0.6,
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

ggsave(
  here("results/Pdpn_LEC_subsets.png"),
  dpi = 300
)

# Plot Ag-score for LEC subsets
dat %>%
  plot_violin(
    "Ag_score",
    "subtype",
    "mouse",
    method = "boxplot",
    plot_colors = lec_clrs,
    color = "black",
    alpha = 1,
    outlier.size = 0.5
  ) +  
  base_theme +
  theme(
    aspect.ratio = 0.6,
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

ggsave(
  here("results/Ag-score_LEC_subsets.png"),
  dpi = 300
)

# Scatter plot comparing Pdpn expression and Ag-score
dat %>%
  filter(Pdpn > 0, Ag_score > 0) %>%
  ggplot(aes(Ag_score, Pdpn, color = subtype)) +
  geom_point() +
  geom_smooth(method = "lm", color = "black", linetype = 2) +
  scale_color_manual(values = lec_clrs) +
  base_theme +
  theme(
    aspect.ratio = 1
  )

ggsave(
  here("results/Pdpn_Ag-score_scatter.png"),
  dpi = 300
)
```

```{r "RF TEST 1", eval = FALSE}
# GOALS
# 1. train model for LEC subsets (~80% accuracy for cLECs)
# 2. train general model for all LECs (~60-90% for all subsets)

# Identify Ag markers for all LECs
degs_1 <- lec_so %>%
  wilcoxauc(group_by = "Ag_class") %>%
  filter(
    padj < 0.05, logFC > 0.25,
    !grepl("^(mt-|R[pls])", feature)
  ) %>%
  arrange(padj)

# Identify Ag markers for subtype
degs_2 <- unique(lec_so$subtype) %>%
  map_dfr(~ {
    res <- lec_so %>%
      subset(subtype == .x) %>%
      wilcoxauc(group_by = "Ag_class")

    if (nrow(res) == 0) return(NULL)

    res %>%
      filter(
        padj < 0.05, logFC > 0.25,
        !grepl("^(mt-|R[pls])", feature)
      ) %>%
      mutate(subtype = .x) %>%
      arrange(padj)
  })

# Identify markers for subtype and mouse
degs_3 <- lec_so@meta.data %>%
  distinct(subtype, mouse) %>%
  pmap_dfr(~ {
    res <- lec_so %>%
      subset(subtype == ..1 & mouse == ..2)
    
    if (n_distinct(res$Ag_class) < 2) return(NULL)
    
    res <- res %>%
      wilcoxauc(group_by = "Ag_class")

    if (nrow(res) == 0) return(NULL)

    res %>%
      filter(
        padj < 0.05, logFC > 0.25,
        !grepl("^(mt-|R[pls])", feature)
      ) %>%
      mutate(subtype = ..1, mouse = ..2) %>%
      arrange(padj)
  })

degs <- bind_rows(degs_1, degs_2, degs_3) %>%
  distinct(feature, group)

# Format RF data
# cat_feats <- c("tm", "subtype", "Ag_class")
# dat_col   <- "Ag_score"
cat_feats <- c("tm", "mouse", "subtype")
feats     <- unique(degs$feature)
dat_col   <- "Ag_class"

rf_dat <- lec_so %>%
  FetchData(c(dat_col, cat_feats, feats)) %>%
  mutate(!!sym(dat_col) := fct_relevel(!!sym(dat_col), "high", "low")) %>%
  rename_with(~ str_replace_all(.x, "-", "_")) %>%
  rename_with(~ str_replace_all(.x, "\\(|\\)", "_")) %>%
  rename_with(~ str_c("x", .x), matches("^[0-9]"))

set.seed(42)

rf_split <- rf_dat %>%
  initial_split(prop = 0.1)

rf_test <- testing(rf_split)

rf_train <- training(rf_split) %>%
  dplyr::select(-all_of(cat_feats))

# Tune model
tune_rf <- function(dat_df, dat_col, param_lst, return_best = TRUE) {
  
  rf_params <- expand.grid(param_lst)
  
  res <- pmap_dfr(rf_params, ~ {
    res <- ranger(
      as.formula(str_c(dat_col, " ~ .")),
      data       = dat_df,
      importance = "impurity",
      seed       = 42,
      ...
    )
    
    err <- sqrt(res$prediction.error)
    
    tibble(..., oob_rmse = err)
  })
  
  if (return_best) {
    res <- res %>%
      dplyr::filter(oob_rmse == min(oob_rmse)) %>%
      dplyr::select(-oob_rmse) %>%
      head(1) %>%
      as.list()
  }
  
  res
}

param_lst <- list(
  mtry            = seq(1, ncol(rf_train), 50),
  min.node.size   = c(1:3, 5, 10)
)

rf_params <- rf_train %>%
  tune_rf(
    dat_col   = "Ag_class",
    param_lst = param_lst
  )

# Fit RF
rf_mod <- pmap(rf_params, ~ {
  ranger(
    Ag_class ~ .,
    data       = rf_train,
    importance = "impurity",
    ...
  )
}) %>%
  pluck(1)

# Get predictions
prds <- rf_mod %>%
  predict(rf_test) %>%
  predictions()

rf_res <- rf_test %>%
  as_tibble(rownames = ".cell_id") %>%
  dplyr::select(.cell_id, subtype, tm, mouse, Ag_class) %>%
  mutate(
    prediction = prds,
    correct    = Ag_class == prediction
  ) %>%
  group_by(subtype, mouse) %>%
  summarize(
    n       = n(),
    correct = sum(correct) / n,
    .groups = "drop"
  ) %>%
  arrange(desc(correct))

rf_res %>%
  mutate(lab = as.character(round(correct, 2))) %>%
  ggplot(aes(subtype, mouse, fill = correct)) +
  geom_tile() +
  geom_text(aes(label = lab)) +
  scale_fill_gradientn(colours = c("lightblue", "red")) +
  djvdj_theme()

# Get top regression variables
rf_marks <- tibble(
  gene   = names(rf_mod$variable.importance),
  impact = rf_mod$variable.importance
) %>%
  mutate(
    rank = row_number(desc(impact)),
    gene = ifelse(gene %in% cat_feats, gene, str_replace(gene, "_", "-"))
  ) %>%
  arrange(rank) %>%

  left_join(degs, by = c(gene = "feature"))

# # Create boxplots
# n_gns <- 50
# 
# plt_gns <- rf_marks %>%
#   filter(!gene %in% c("Ag-score", "Ag_class", "orig.ident", "tm")) %>%
#   pull(gene) %>%
#   head(n_gns) %>%
#   c("Ag_score", .)
# 
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns)) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
#   
# plt_dat %>%
#   ggplot(aes(tm, value, fill = name, color = name, alpha = Ag_class)) +
#   geom_boxplot() +
#   facet_wrap(~ name, scales = "free_y") +
#   scale_alpha_manual(values = c(low = 0.25, high = 0.75)) +
#   djvdj_theme()
# 
# # Create scatter plots
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns[-1])) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
# 
# plt_dat %>%
#   ggplot(aes(Ag_score, value, color = Ag_class)) +
#   geom_point(size = 0.1) +
#   facet_wrap(~ name, scales = "free") +
#   djvdj_theme() +
#   theme(aspect.ratio = 1)
```

```{r "RF TEST 2", eval = FALSE}
# GOALS
# 1. train model for LEC subsets for Ag groups (1-4)
# 2. identify cells in mock that are predicted to take up Ag

# Identify Ag markers for all LECs for each mouse
deg_so <- lec_so %>%
  SplitObject("mouse")

plan(multisession, workers = min(length(deg_so), 8))

degs_1 <- deg_so %>%
  future_imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = "Ag_group")

    if (nrow(res) == 0) return(NULL)

    res %>%
      filter(
        padj < 0.05, logFC > 0.25,
        !grepl("^(mt-|R[pls])", feature)
      ) %>%
      mutate(mouse = .y) %>%
      arrange(padj)
  })

# Identify Ag markers for subtype
deg_so <- lec_so %>%
  SplitObject("subtype")

plan(multisession, workers = min(length(deg_so), 8))

degs_2 <- deg_so %>%
  future_imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = "Ag_group")

    if (nrow(res) == 0) return(NULL)

    res %>%
      filter(
        padj < 0.05, logFC > 0.25,
        !grepl("^(mt-|R[pls])", feature)
      ) %>%
      mutate(subtype = .y) %>%
      arrange(padj)
  })

rm(deg_so)

# Identify markers for subtype and mouse
degs_3 <- lec_so@meta.data %>%
  distinct(subtype, mouse) %>%
  pmap_dfr(~ {
    res <- lec_so %>%
      subset(subtype == ..1 & mouse == ..2)
    
    if (n_distinct(res$Ag_class) < 2) return(NULL)
    
    res <- res %>%
      wilcoxauc(group_by = "Ag_group")

    if (nrow(res) == 0) return(NULL)

    res %>%
      filter(
        padj < 0.05, logFC > 0.25,
        !grepl("^(mt-|R[pls])", feature)
      ) %>%
      mutate(subtype = ..1, mouse = ..2) %>%
      arrange(padj)
  })

degs <- bind_rows(degs_1, degs_2, degs_3) %>%
  distinct(feature, group)

# Format RF data
dat_col   <- "Ag_group"
cat_feats <- c("tm", "mouse", "subtype")
feats     <- unique(degs$feature)

rf_dat <- lec_so %>%
  FetchData(c(dat_col, cat_feats, feats)) %>%
  rename_with(~ str_replace_all(.x, "-", "_")) %>%
  rename_with(~ str_replace_all(.x, "\\(|\\)", "_")) %>%
  rename_with(~ str_c("x", .x), matches("^[0-9]"))

set.seed(42)

rf_split <- rf_dat %>%
  initial_split(prop = 0.1)

rf_test <- testing(rf_split)

rf_train <- training(rf_split) %>%
  dplyr::select(-all_of(cat_feats))

# Tune model
tune_rf <- function(dat_df, dat_col, param_lst, return_best = TRUE) {
  
  rf_params <- expand.grid(param_lst)
  
  res <- pmap_dfr(rf_params, ~ {
    res <- ranger(
      as.formula(str_c(dat_col, " ~ .")),
      data       = dat_df,
      importance = "impurity",
      seed       = 42,
      ...
    )
    
    err <- sqrt(res$prediction.error)
    
    tibble(..., oob_rmse = err)
  })
  
  if (return_best) {
    res <- res %>%
      dplyr::filter(oob_rmse == min(oob_rmse)) %>%
      dplyr::select(-oob_rmse) %>%
      head(1) %>%
      as.list()
  }
  
  res
}

param_lst <- list(
  mtry            = seq(1, ncol(rf_train), 100),
  min.node.size   = c(1:3, 5, 10)
)

rf_params <- rf_train %>%
  tune_rf(
    dat_col   = "Ag_group",
    param_lst = param_lst
  )

# Fit model
rf_mod <- pmap(rf_params, ~ {
  ranger(
    Ag_class ~ .,
    data       = rf_train,
    importance = "impurity",
    ...
  )
}) %>%
  pluck(1)

# Get predictions
prds <- rf_mod %>%
  predict(rf_test) %>%
  predictions()

rf_res <- rf_test %>%
  as_tibble(rownames = ".cell_id") %>%
  dplyr::select(.cell_id, subtype, tm, mouse, Ag_class) %>%
  mutate(
    prediction = prds,
    correct    = Ag_class == prediction
  ) %>%
  group_by(subtype, mouse) %>%
  summarize(
    n       = n(),
    correct = sum(correct) / n,
    .groups = "drop"
  ) %>%
  arrange(desc(correct))

rf_res %>%
  mutate(lab = as.character(round(correct, 2))) %>%
  ggplot(aes(subtype, mouse, fill = correct)) +
  geom_tile() +
  geom_text(aes(label = lab)) +
  scale_fill_gradientn(colours = c("lightblue", "red")) +
  djvdj_theme()

# Get top regression variables
rf_marks <- tibble(
  gene   = names(rf_mod$variable.importance),
  impact = rf_mod$variable.importance
) %>%
  mutate(
    rank = row_number(desc(impact)),
    gene = ifelse(gene %in% cat_feats, gene, str_replace(gene, "_", "-"))
  ) %>%
  arrange(rank) %>%

  left_join(degs, by = c(gene = "feature"))



# # Create boxplots
# n_gns <- 50
# 
# plt_gns <- rf_marks %>%
#   filter(!gene %in% c("Ag-score", "Ag_class", "orig.ident", "tm")) %>%
#   pull(gene) %>%
#   head(n_gns) %>%
#   c("Ag_score", .)
# 
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns)) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
#   
# plt_dat %>%
#   ggplot(aes(tm, value, fill = name, color = name, alpha = Ag_class)) +
#   geom_boxplot() +
#   facet_wrap(~ name, scales = "free_y") +
#   scale_alpha_manual(values = c(low = 0.25, high = 0.75)) +
#   djvdj_theme()
# 
# # Create scatter plots
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns[-1])) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
# 
# plt_dat %>%
#   ggplot(aes(Ag_score, value, color = Ag_class)) +
#   geom_point(size = 0.1) +
#   facet_wrap(~ name, scales = "free") +
#   djvdj_theme() +
#   theme(aspect.ratio = 1)
```

```{r "RF TEST 3", eval = FALSE}
# GOALS
# 1. train model for LEC subsets for Ag groups (1-4)
# 2. identify cells in mock that are predicted to take up Ag

# Identify DEGs ----
# Identify Ag markers for all LECs for each mouse
grp <- "Ag_class"
fc  <- 0

deg_so <- lec_so %>%
  SplitObject("mouse")

degs_1 <- deg_so %>%
  imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = "all", mouse = .y) %>%
      arrange(padj)
  })

# Identify Ag markers for subtype
deg_so <- lec_so %>%
  SplitObject("subtype")

degs_2 <- deg_so %>%
  imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = .y, mouse = "all") %>%
      arrange(padj)
  })

rm(deg_so)

# Identify markers for subtype and mouse
degs_3 <- lec_so@meta.data %>%
  distinct(subtype, mouse) %>%
  pmap_dfr(~ {
    res <- lec_so %>%
      subset(subtype == ..1 & mouse == ..2)
    
    if (n_distinct(res$Ag_class) < 2) return(NULL)
    
    res <- res %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = ..1, mouse = ..2) %>%
      arrange(padj)
  })

degs <- bind_rows(degs_1, degs_2, degs_3) %>%
  filter(
    padj < 0.05, logFC > fc,
    !grepl("^(mt-|R[pls])", feature)
  )

# Format data ----
# dat_col    <- "Ag_group"
dat_col    <- "Ag_class_2"
feats      <- unique(degs$feature)
other_vars <- NULL

feats <- degs %>%
  filter(subtype == "cLEC", mouse == "d14") %>%
  pull(feature) %>%
  unique()

other_info <- c("tm", "subtype", "mouse")

rf_dat <- lec_so %>%
  subset(subtype == "cLEC" & mouse == "d14") %>%

  FetchData(unique(c(dat_col, other_info, other_vars, feats))) %>%
  
  # mutate(across(all_of(feats), ~ as.numeric(scale(.x)))) %>%
  # mutate(!!sym(dat_col) := factor(!!sym(dat_col))) %>%
  # mutate(
  #   across(all_of(c(dat_col, other_vars)), ~ as.numeric(factor(.x)))
  # ) %>%
  
  rename_with(~ str_replace_all(.x, "-", "_")) %>%
  rename_with(~ str_replace_all(.x, "\\(|\\)", "_")) %>%
  rename_with(~ str_c("x", .x), matches("^[0-9]"))

# Adjust groups so they are balanced
sam_sz <- rf_dat %>%
  group_by(!!sym(dat_col), mouse) %>%
  summarize(n = n(), .groups = "drop") %>%
  pull(n)

# sam_sz <- max(100, min(sam_sz))
sam_sz <- 135

# Split into train/test data
# rf_split <- initial_split(rf_dat, prop = 0.5)
# rf_train <- training(rf_split)
# rf_test  <- testing(rf_split)

rf_train <- rf_dat %>%
  rownames_to_column(".cell_id") %>%
  group_by(!!sym(dat_col), mouse) %>%
  filter(n() >= sam_sz) %>%
  dplyr::sample_n(sam_sz) %>%
  ungroup() %>%
  dplyr::select(-all_of(other_info[!other_info %in% other_vars])) %>%
  column_to_rownames(".cell_id")

rf_test <- rf_dat[!rownames(rf_dat) %in% rownames(rf_train), ]

# Tune RF model ----
tune_rf <- function(dat_df, dat_col, param_lst, return_best = TRUE) {
  
  rf_params <- expand.grid(param_lst)
  
  res <- pmap_dfr(rf_params, ~ {
    res <- ranger(
      as.formula(str_c("factor(", dat_col, ") ~ .")),
      data       = dat_df,
      importance = "impurity",
      seed       = 42,
      ...
    )
    
    err <- sqrt(res$prediction.error)
    
    tibble(..., oob_rmse = err)
  })
  
  if (return_best) {
    res <- res %>%
      dplyr::filter(oob_rmse == min(oob_rmse)) %>%
      dplyr::select(-oob_rmse) %>%
      head(1) %>%
      as.list()
  }
  
  res
}

param_lst <- list(
  # mtry          = c(10, 50, 200, 300),
  mtry          = seq(10, 60, 10),
  min.node.size = c(1, 2, 5),
  num.trees     = 1000
)

rf_params <- rf_train %>%
  tune_rf(
    dat_col   = dat_col,
    param_lst = param_lst
  )

# rf_params <- list(
#   mtry = 10,
#   min.node.size = 1,
#   num.trees = 1000
#   # max.depth = 5
# )

# Fit RF model
rf_mod <- pmap(rf_params, ~ {
  ranger(
    as.formula(str_c("factor(", dat_col, ") ~ .")),
    data       = rf_train,
    importance = "impurity",
    ...
  )
}) %>%
  pluck(1)

# Identify top features and re-fit
top_feats <- importance(rf_mod) %>%
  sort(decreasing = TRUE)

top_feats <- top_feats[top_feats > 0]

rf_params$mtry <- min(length(top_feats), rf_params$mtry)

rf_mod_top <- pmap(rf_params, ~ {
  ranger(
    as.formula(str_c("factor(", dat_col, ") ~ .")),
    data       = rf_train[, c(dat_col, names(top_feats))],
    importance = "impurity",
    ...
  )
}) %>%
  pluck(1)

# prds <- rf_mod_top %>%
prds <- rf_mod %>%
  predict(rf_test) %>%
  predictions()

# Try logit regression ----
logit_mod <- rf_train %>%
  mutate(
    Ag_class = ifelse(Ag_class == "high", 1, 0),
    Ag_class = factor(Ag_class)
  ) %>%
  glm(
    data = .,
    # as.factor(Ag_class) ~ Cav1 + Marcks | as.numeric(as.factor(mouse)),
    # as.factor(Ag_class) ~ Cav1 + Marcks,
    Ag_class ~ .,
    control = glm.control(maxit = 200),
    family = binomial
  )

prob <- logit_mod %>%
  predict(newdata = rf_test, type = "response")

prds <- vector("character", length(prob))
prds[prob > 0.7] <- "high"
prds[prob <= 0.7] <- "low"

# Try glmer ----
# library(lme4)
# 
# logit_mod <- rf_train %>%
#   mutate(
#     Ag_class = ifelse(Ag_class == "high", 1, 0),
#     Ag_class = factor(Ag_class)
#   ) %>%
#   glmer(
#     data = .,
#     # as.factor(Ag_class) ~ Cav1 + Marcks | as.numeric(as.factor(mouse)),
#     # as.factor(Ag_class) ~ Cav1 + Marcks,
#     # Ag_class ~ (. | factor(mouse)),
#     # control = glm.control(maxit = 200),
#     Ag_class ~ (. | factor(exp)),
#     family = binomial
#   )
# 
# prob <- logit_mod %>%
#   predict(newdata = rf_test, type = "response")
# 
# prds <- vector("character", length(prob))
# prds[prob > 0.5] <- "high"
# prds[prob <= 0.5] <- "low"


# Get predictions ----
rf_res <- rf_test %>%
  as_tibble(rownames = ".cell_id") %>%
  dplyr::select(.cell_id, subtype, all_of(c(dat_col, other_vars, other_info))) %>%
  mutate(
    prediction = prds,
    correct    = !!sym(dat_col) == prediction
  ) %>%
  # group_by(subtype, !!sym(other_vars)) %>%
  group_by(!!!syms(c(other_info, dat_col))) %>%
  summarize(
    n       = n(),
    correct = sum(correct) / n,
    .groups = "drop"
  ) %>%
  arrange(desc(correct))

# Plot results
rf_res %>%
  mutate(lab = as.character(round(correct, 2))) %>%
  ggplot(aes(subtype, mouse, fill = correct)) +
  geom_tile() +
  geom_text(aes(label = lab)) +
  facet_wrap(~ Ag_class_2) +
  scale_fill_gradientn(colours = c("lightblue", "red")) +
  djvdj_theme()

# Get top regression variables
rf_marks <- tibble(
  gene   = names(rf_mod$variable.importance),
  impact = rf_mod$variable.importance
) %>%
  mutate(
    rank = row_number(desc(impact)),
    gene = ifelse(gene %in% cat_feats, gene, str_replace(gene, "_", "-"))
  ) %>%
  arrange(rank) %>%

  left_join(degs, by = c(gene = "feature"))



# # Create boxplots
# n_gns <- 50
# 
# plt_gns <- rf_marks %>%
#   filter(!gene %in% c("Ag-score", "Ag_class", "orig.ident", "tm")) %>%
#   pull(gene) %>%
#   head(n_gns) %>%
#   c("Ag_score", .)
# 
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns)) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
#   
# plt_dat %>%
#   ggplot(aes(tm, value, fill = name, color = name, alpha = Ag_class)) +
#   geom_boxplot() +
#   facet_wrap(~ name, scales = "free_y") +
#   scale_alpha_manual(values = c(low = 0.25, high = 0.75)) +
#   djvdj_theme()
# 
# # Create scatter plots
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns[-1])) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
# 
# plt_dat %>%
#   ggplot(aes(Ag_score, value, color = Ag_class)) +
#   geom_point(size = 0.1) +
#   facet_wrap(~ name, scales = "free") +
#   djvdj_theme() +
#   theme(aspect.ratio = 1)
```

```{r "RF d14 TEST", eval = FALSE}
# GOALS
# 1. train model for LEC subsets for Ag groups (1-4)
# 2. identify cells in mock that are predicted to take up Ag

# Identify DEGs ----
# Identify Ag markers for all LECs for each mouse
grp <- "Ag_class_2"
fc  <- 0

deg_so <- lec_so %>%
  SplitObject("mouse")

degs_1 <- deg_so %>%
  imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = "all", mouse = .y) %>%
      arrange(padj)
  })

# Identify Ag markers for subtype
deg_so <- lec_so %>%
  SplitObject("subtype")

degs_2 <- deg_so %>%
  imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = .y, mouse = "all") %>%
      arrange(padj)
  })

rm(deg_so)

# Identify markers for subtype and mouse
degs_3 <- lec_so@meta.data %>%
  distinct(subtype, mouse) %>%
  pmap_dfr(~ {
    res <- lec_so %>%
      subset(subtype == ..1 & mouse == ..2)
    
    if (n_distinct(res[[grp]]) < 2) return(NULL)
    
    res <- res %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = ..1, mouse = ..2) %>%
      arrange(padj)
  })

degs <- bind_rows(degs_1, degs_2, degs_3) %>%
  filter(
    padj < 0.05, logFC > fc,
    !grepl("^(mt-|R[pls])", feature),
    feature != "Malat1"
  )

degs %>%
  write_tsv(here(params$so_dir, "ag-high_degs_Ag_class_2.tsv.gz"))

# Most common degs
top_degs <- degs %>%
  group_by(subtype, feature, group) %>%
  summarize(n = n_distinct(mouse), .groups = "drop") %>%
  arrange(desc(n)) %>%
  filter(n > 1) %>%
  filter(!subtype %in% c("all", "BEC", "unassigned")) %>%
  group_by(feature, group) %>%
  summarize(n_typ = n_distinct(subtype), .groups = "drop") %>%
  arrange(desc(n_typ)) %>%
  filter(n_typ > 1)

# Format data ----
dat_col    <- "Ag_class_2"
feats      <- unique(degs$feature)
other_vars <- NULL

# feats <- degs %>%
#   filter(subtype == "cLEC", mouse == "d14") %>%
#   pull(feature) %>%
#   unique()
feats <- unique(top_degs$feature)

other_info <- c("tm", "subtype", "mouse")

rf_dat <- rf_so %>%
  # subset(subtype != "BEC" & subtype == "cLEC") %>%
  subset(subtype == "cLEC" & mouse == "d14") %>%

  FetchData(unique(c(dat_col, other_info, other_vars, feats))) %>%
  
  rename_with(~ str_replace_all(.x, "-", "_")) %>%
  rename_with(~ str_replace_all(.x, "\\(|\\)", "_")) %>%
  rename_with(~ str_c("x", .x), matches("^[0-9]"))

# Adjust groups so they are balanced
# Split into train/test data
sam_sz <- 135

rf_train <- rf_dat %>%
  rownames_to_column(".cell_id") %>%
  group_by(!!sym(dat_col), subtype, mouse) %>%
  filter(n() >= sam_sz) %>%
  dplyr::sample_n(sam_sz) %>%
  ungroup() %>%
  dplyr::select(-all_of(other_info[!other_info %in% other_vars])) %>%
  column_to_rownames(".cell_id")

rf_test <- rf_dat[!rownames(rf_dat) %in% rownames(rf_train), ]

# Tune RF model ----
tune_rf <- function(dat_df, dat_col, param_lst, return_best = TRUE) {
  
  rf_params <- expand.grid(param_lst)
  
  res <- pmap_dfr(rf_params, ~ {
    res <- ranger(
      as.formula(str_c("factor(", dat_col, ") ~ .")),
      data       = dat_df,
      importance = "impurity",
      seed       = 42,
      ...
    )
    
    err <- sqrt(res$prediction.error)
    
    tibble(..., oob_rmse = err)
  })
  
  if (return_best) {
    res <- res %>%
      dplyr::filter(oob_rmse == min(oob_rmse)) %>%
      dplyr::select(-oob_rmse) %>%
      head(1) %>%
      as.list()
  }
  
  res
}

param_lst <- list(
  mtry          = seq(2, length(feats), 5),
  min.node.size = c(1, 2, 5),
  num.trees     = 1000
)

rf_params <- rf_train %>%
  tune_rf(
    dat_col   = dat_col,
    param_lst = param_lst
  )

# Fit RF model
rf_mod <- pmap(rf_params, ~ {
  ranger(
    as.formula(str_c("factor(", dat_col, ") ~ .")),
    data       = rf_train,
    importance = "impurity",
    ...
  )
}) %>%
  pluck(1)

# Identify top features and re-fit
top_feats <- importance(rf_mod) %>%
  sort(decreasing = TRUE)

top_feats <- top_feats[top_feats > 0]

rf_params$mtry <- min(length(top_feats), rf_params$mtry)

rf_mod_top <- pmap(rf_params, ~ {
  ranger(
    as.formula(str_c("factor(", dat_col, ") ~ .")),
    data       = rf_train[, c(dat_col, names(top_feats))],
    importance = "impurity",
    ...
  )
}) %>%
  pluck(1)

# prds <- rf_mod %>%
  # predict(rf_dat) %>%
prds <- rf_mod_top %>%
  predict(rf_test) %>%
  predictions()

# Get predictions ----
# rf_res <- rf_test %>%
rf_res <- rf_dat %>%
  as_tibble(rownames = ".cell_id") %>%
  dplyr::select(.cell_id, subtype, all_of(c(dat_col, other_vars, other_info))) %>%
  mutate(
    prediction = prds,
    correct    = !!sym(dat_col) == prediction
  ) %>%
  group_by(!!!syms(c(other_info, dat_col))) %>%
  summarize(
    n       = n(),
    correct = sum(correct) / n,
    .groups = "drop"
  ) %>%
  arrange(desc(correct))

# Plot results
rf_res %>%
  mutate(lab = as.character(round(correct, 2))) %>%
  ggplot(aes(subtype, mouse, fill = correct)) +
  geom_tile() +
  geom_text(aes(label = lab)) +
  facet_wrap(~ Ag_class_2) +
  scale_fill_gradientn(colours = c("lightblue", "red")) +
  djvdj_theme()

# Get top regression variables
rf_marks <- tibble(
  gene   = names(rf_mod$variable.importance),
  impact = rf_mod$variable.importance
) %>%
  mutate(
    rank = row_number(desc(impact)),
    gene = ifelse(gene %in% cat_feats, gene, str_replace(gene, "_", "-"))
  ) %>%
  arrange(rank) %>%

  left_join(degs, by = c(gene = "feature"))

# Create boxplots
n_gns <- 50

plt_gns <- rf_marks %>%
  filter(!gene %in% c("Ag_score", dat_col, "orig.ident", "tm")) %>%
  pull(gene) %>%
  unique() %>%
  head(n_gns) %>%
  c("Ag_score", .) %>%
  str_remove("^x")

plt_dat <- lec_so %>%
  subset(subtype == typ) %>%
  FetchData(c("tm", dat_col, plt_gns)) %>%
  pivot_longer(all_of(plt_gns)) %>%
  mutate(
    tm         = fct_relevel(as.character(tm), tms),
    Ag_class_2 = fct_relevel(Ag_class_2, c("low", "high")),
    name       = fct_relevel(name, plt_gns)
  )

plt_dat %>%
  ggplot(aes(tm, value, fill = name, color = name, alpha = Ag_class_2)) +
  geom_boxplot() +
  facet_wrap(~ name, scales = "free_y") +
  scale_alpha_manual(values = c(low = 0.25, high = 0.75)) +
  djvdj_theme()

# # Create scatter plots
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns[-1])) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
# 
# plt_dat %>%
#   ggplot(aes(Ag_score, value, color = Ag_class)) +
#   geom_point(size = 0.1) +
#   facet_wrap(~ name, scales = "free") +
#   djvdj_theme() +
#   theme(aspect.ratio = 1)
```



```{r "RF GO OLD", eval = FALSE}
# Load GO results
go <- qread(here(params$mod_dir, "ag-high_go.qs"))

endo_gns <- go@compareClusterResult %>%
  filter(
    Description == "endocytosis",
    p.adjust < 0.05
  ) %>%
  pull(geneID) %>%
  str_split("/") %>%
  unlist()

gns <- c("Ctsd", "Dab2", "Prcp", "Cltc", "Degs", "Clu", "Atp6v1a", endo_gns)

gns <- unique(gns[gns %in% rf_feats$cLEC$high])

lec_so %>%
  subset(subtype == "cLEC") %>%
  subset(mouse == "3wk") %>%
  mutate_meta(
    mutate,
    Ag_class_ml = ifelse(is.na(pred_grp), pred_grp, Ag_class_ml)
  ) %>%
  plot_violin(
    "Atp6v1a",
    "pred_grp",
    # "Ag_class_ml",
    "mouse",
    method = "boxplot"
  )

dat <- lec_so %>%
  FetchData(c("Atp6v1a", "pred_grp", "subtype", "mouse")) %>%
  filter(mouse == "3wk", subtype == "cLEC")

wilcox.test(
  dat %>% filter(pred_grp == "Ag-high") %>% pull(Atp6v1a),
  dat %>% filter(pred_grp == "Ag-competent") %>% pull(Atp6v1a)
)
```

```{r "RF DEGs", eval = FALSE}
# GOALS
# 1. train model for LEC subsets for Ag groups (1-4)
# 2. identify cells in mock that are predicted to take up Ag

# Identify Ag markers for all LECs for each mouse
grp <- "Ag_class_2"
fc  <- 0

deg_so <- lec_so %>%
  SplitObject("mouse")

degs_1 <- deg_so %>%
  imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = "all", mouse = .y) %>%
      arrange(padj)
  })

# Identify Ag markers for subtype
deg_so <- lec_so %>%
  SplitObject("subtype")

degs_2 <- deg_so %>%
  imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = .y, mouse = "all") %>%
      arrange(padj)
  })

rm(deg_so)

# Identify markers for subtype and mouse
degs_3 <- lec_so@meta.data %>%
  distinct(subtype, mouse) %>%
  pmap_dfr(~ {
    res <- lec_so %>%
      subset(subtype == ..1 & mouse == ..2)
    
    if (n_distinct(res[[grp]]) < 2) return(NULL)
    
    res <- res %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = ..1, mouse = ..2) %>%
      arrange(padj)
  })

degs <- bind_rows(degs_1, degs_2, degs_3) %>%
  filter(
    padj < 0.05, logFC > fc,
    !grepl("^(mt-|R[pls])", feature),
    feature != "Malat1"
  )

# degs %>%
#   write_tsv(here(params$so_dir, "ag-high_degs_Ag_class_2.tsv.gz"))

# # Most common degs
# top_degs <- degs %>%
#   group_by(subtype, feature, group) %>%
#   summarize(n = n_distinct(mouse), .groups = "drop") %>%
#   arrange(desc(n))
#   filter(n > 1)
```

```{r "RF FUNCTIONS", eval = FALSE}
# Function to predict
rf_predict <- function(mod, dat, data_clmn, return_stats = TRUE) {
  
  prds <- mod %>%
    predict(dat) %>%
    predictions()
  
  res <- dat %>%
    as_tibble(rownames = ".cell_id") %>%
    dplyr::select(.cell_id, all_of(data_clmn)) %>%
    mutate(
      pred = prds,
      correct = !!sym(data_clmn) == pred
    ) %>%
    group_by(!!sym(data_clmn)) %>%
    mutate(frac_correct = sum(correct) / n()) %>%
    ungroup()
  
  if (!return_stats) return(column_to_rownames(res, ".cell_id"))
  
  res <- res %>%
    distinct(!!sym(data_clmn), frac_correct) %>%
    pivot_wider(names_from = all_of(data_clmn), values_from = frac_correct)
  
  res
}

run_rf <- function(df_in, data_clmn, param_lst, sample_size = 135,
                   feat_p = 0.05, min_feats = 20, sample_info = NULL) {
  
  # Format train/test data
  rf_dat <- df_in %>%
    rename_with(~ str_replace_all(.x, "-", "_")) %>%
    rename_with(~ str_replace_all(.x, "\\(|\\)", "_")) %>%
    rename_with(~ str_c("x", .x), matches("^[0-9]"))
  
  # Adjust groups so they are balanced
  # Split into train/test data
  sam_sz <- min(sample_size, min(table(rf_dat[[data_clmn]])))
  min_sz <- 40
  
  if (sam_sz < min_sz) {
    cli::cli_abort("All groups must have at least {min_sz} cells")
    
  } else if (sam_sz < sample_size) {
    cli::cli_warn(
      "{sample_size} is too large for the number of cells,
       a sample size of {sam_sz} was used"
    )
  }
  
  rf_train <- rf_dat %>%
    rownames_to_column(".cell_id") %>%
    group_by(!!sym(data_clmn))
  
  set.seed(42)
  
  rf_train <- rf_train %>%
    dplyr::sample_n(sam_sz) %>%
    ungroup() %>%
    dplyr::select(-any_of(sample_info)) %>%
    column_to_rownames(".cell_id")
  
  rf_test <- rf_dat[!rownames(rf_dat) %in% rownames(rf_train), ]
  
  # Train models
  rf_params <- expand.grid(param_lst)
  
  res <- pmap_dfr(rf_params, ~ {
    args <- list(...)
    
    if (args$mtry > (ncol(rf_train) - 1)) return(NULL)
    
    args$formula    <- as.formula(str_c("factor(", data_clmn, ") ~ ."))
    args$data       <- rf_train
    args$importance <- "impurity_corrected"
    args$seed       <- 42
    
    set.seed(42)
    
    raw_mod <- lift_dl(ranger)(args)
    
    raw_err <- raw_mod$prediction.error

    # Feature selection
    key_feats <- raw_mod %>%
      importance_pvalues() %>%
      .[, "pvalue"] %>%
      sort()
    
    max_p <- key_feats %>%
      head(min_feats) %>%
      max()
    
    feat_p <- max(max_p, feat_p)
    
    key_feats <- key_feats[key_feats < feat_p] %>%
      names()
    
    n_feats <- length(key_feats)
    
    # Re-train model
    args$mtry       <- min(args$mtry, n_feats)
    args$data       <- rf_train[, c(data_clmn, key_feats)]
    args$importance <- "none"
    
    set.seed(42)
    
    mod <- lift_dl(ranger)(args)
    
    err <- mod$prediction.error
    
    # Test model    
    prds <- mod %>%
      rf_predict(
        dat = rf_test,
        data_clmn = data_clmn
      )
                                         # only used for continuous
    # err <- sqrt(res$prediction.error)  # response variables
    
    # Format output table
    tibble(
      raw_mod   = list(raw_mod),
      mod       = list(mod),
      feats     = list(key_feats),
      n_train   = nrow(rf_train),
      n_test    = nrow(rf_test),
      # train_bcs = list(rownames(rf_train)),
      # test_bcs  = list(rownames(rf_test)),
      ...,
      n_feats = length(key_feats),
      oob_raw = raw_err,
      oob     = err
    ) %>%
      bind_cols(prds)
  })
  
  res
}
```

```{r "RF MODELS", eval = FALSE}
# Train models
# use all cLEC features
dat_col <- "Ag_class_2"

feats <- degs %>%
  filter(subtype == "cLEC") %>%
  # filter(subtype == "cLEC", mouse == "d14") %>%
  pull(feature) %>%
  unique()

other_info <- c("subtype", "mouse")

rf_dat <- lec_so %>%
  subset(subtype == "cLEC" & mouse == "d14") %>%
  FetchData(unique(c(dat_col, other_info, feats)))

param_lst <- list(
  mtry          = seq(2, length(feats), 10),
  min.node.size = c(1:10),
  num.trees     = seq(50, 1000, 50)
)

rf_res <- rf_dat %>%
  run_rf(
    data_clmn   = dat_col,
    param_lst   = param_lst,
    sample_info = other_info,
    sample_size = 135
  ) %>%
  # arrange(desc(high), desc(low))
  arrange(desc(low), desc(high))

# Pull best model
best_mod <- rf_res %>%
  head(1) %>%
  pull(mod) %>%
  .[[1]]

best_feats <- rf_res %>%
  head(1) %>%
  pull(feats) %>%
  unlist()

# Predict for other timepoints
tm_dat <- lec_so %>%
  subset(subtype == "cLEC") %>%
  FetchData(unique(c(dat_col, other_info, feats))) %>%
  rename_with(~ str_replace_all(.x, "-", "_")) %>%
  rename_with(~ str_replace_all(.x, "\\(|\\)", "_")) %>%
  rename_with(~ str_c("x", .x), matches("^[0-9]")) %>%
  split(.$mouse)

preds <- tm_dat %>%
  imap_dfr(~ {
    best_mod %>%
      rf_predict(.x, data_clmn = dat_col, return_stats = FALSE) %>%
      mutate(mouse = .y)
  })

# Plot predicted classifications
pred_meta <- preds %>%
  dplyr::select(pred, correct)

lec_so <- lec_so %>%
  AddMetaData(pred_meta)

lec_so %>%
  plot_scatter(
    "pred", "hUMAP_1", "hUMAP_2",
    group_col = "mouse",
    size = 0.1,
    panel_nrow = 1
  ) +
  theme(aspect.ratio = 0.9)

# Fraction predicted in each mouse
plt_dat <- lec_so %>%
  subset(subtype == "cLEC") %>%
  AddModuleScore(features = best_feats, name = "Ag_module") %>%
  mutate_meta(
    mutate,
    pred_grp = case_when(
      Ag_class_2 == "high" ~ Ag_class_2,
      pred       == "high" ~ "high-predicted",
      TRUE                 ~ Ag_class_2
    ),
    mouse = fct_relevel(mouse, mice)
  )

plt_dat %>%
  plot_frequency("pred_grp", cluster_col = "mouse")

# Module scores for each prediction group
# expect that these genes will be highest expressed in high, high-predicted,
# lowest in low
# 3wk sample shows opposite
plt_dat@meta.data %>%
  # ggplot(aes(mouse, Ag_module1, fill = Ag_class_2)) +
  ggplot(aes(mouse, Ag_module1, fill = pred_grp)) +
  # geom_violin(scale = "width", draw_quantiles = c(0.25, 0.75, 0.5)) +
  geom_boxplot() +
  djvdj_theme()
  

```

```{r "COMPARE WITH MOCK DATA", eval = FALSE, fig.width = 18, fig.height = 15}

# Plot top features across data
plt_feats <- degs %>%
  group_by(subtype, feature, group) %>%
  summarize(n = n_distinct(mouse), .groups = "drop") %>%
  arrange(desc(n)) %>%
  filter(n > 1) %>%
  filter(group == "high") %>%
  split(.$subtype) %>%
  map(pull, feature)

# plt_feats <- top_degs %>%
#   split(.$group)

clmns <- str_c(names(plt_feats), seq_along(plt_feats))

plt_dat <- lec_so %>%
  Seurat::AddModuleScore(features = list(plt_feats$high$feature)) %>%
  # Seurat::AddModuleScore(features = plt_feats, name = names(plt_feats)) %>%
  FetchData(c("Ag_class_2", "Cluster1", "orig.ident", "mouse", "tm", "subtype"))

plt_dat %>%
  mutate(Ag_class_2 = fct_relevel(Ag_class_2, c("low", "high"))) %>%
  ggplot(aes(mouse, Cluster1, fill = mouse, alpha = Ag_class_2)) +
  geom_boxplot() +
  scale_alpha_manual(values = c(high = 0.75, low = 0.25)) +
  facet_wrap(~ subtype) +
  djvdj_theme()

# Load mock data
chikv_so <- qread("~/Dropbox/Ryan/Projects/morrison-lnsc/results/sobjs/so_merge.qs")

# Calculate module score
chikv_dat <- chikv_so %>%
  Seurat::AddModuleScore(features = plt_feats, name = names(plt_feats)) %>%
  FetchData(c("orig.ident", "lec_type", "tm", "cell_type", "rep", "cell_type_2", "treatment", clmns, "subtype"))

# Calculate mean signal
all_gns <- unique(unlist(plt_feats))

chikv_dat <- chikv_so %>%
  FetchData(c(
    "orig.ident", "lec_type", "tm", "cell_type",
    "rep", "cell_type_2", "treatment", "subtype",
    all_gns
  ))

# Plot example genes
chikv_so %>%
  FetchData(c("treatment", "tm", "lec_type", "rep", "Ctsd", "cell_type_2")) %>%
  filter(cell_type_2 == "LEC" | lec_type == "BEC") %>%
  mutate(
    treat_rep = str_c(treatment, "-", rep),
    treatment = fct_relevel(treatment, c("mock", "CHIKV"))
  ) %>%
  ggplot(aes(lec_type, Ctsd, fill = treatment, alpha = rep)) +
  geom_boxplot(outlier.size = 0.1) +
  scale_alpha_manual(values = c(0.2, 0.4, 0.6)) +
  facet_wrap(~ tm) +
  djvdj_theme() +
  theme(
    aspect.ratio = 0.4,
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1)
  )

# # Plot mean list expression
# plt_dat <- chikv_dat %>%
#   pivot_longer(any_of(all_gns)) %>%
#   filter(cell_type_2 == "LEC" | lec_type == "BEC")
# 
#   mutate(
#     treat_rep = str_c(treatment, "-", rep),
#     treatment = fct_relevel(treatment, c("mock", "CHIKV"))
#   ) %>%
#   ggplot(aes(lec_type, Ctsd, fill = treatment, alpha = rep)) +
#   geom_boxplot(outlier.size = 0.1) +
#   scale_alpha_manual(values = c(0.2, 0.4, 0.6)) +
#   facet_wrap(~ tm) +
#   djvdj_theme() +
#   theme(
#     aspect.ratio = 0.4,
#     axis.title.x = element_blank(),
#     axis.text.x  = element_text(angle = 45, hjust = 1)
#   )

# Plot module scores
plt <- clmns %>%
  map(~ {
    chikv_dat %>%
      filter(cell_type_2 == "LEC" | lec_type == "BEC") %>%
      mutate(
        treat_rep = str_c(treatment, "-", rep),
        treatment = fct_relevel(treatment, c("mock", "CHIKV"))
      ) %>%
      ggplot(aes(lec_type, !!sym(.x), fill = treatment, alpha = rep)) +
      geom_boxplot(outlier.size = 0.01, outlier.alpha = 1, size = 0.4) +
      scale_alpha_manual(values = c(0.2, 0.4, 0.6)) +
      facet_wrap(~ tm, nrow = 1) +
      guides(alpha = FALSE) +
      djvdj_theme() +
      theme(
        aspect.ratio = 0.4,
        axis.title.x = element_blank(),
        axis.text.x  = element_text(angle = 45, hjust = 1),
        legend.position = "top"
      )
  }) %>%
  plot_grid(plotlist = ., ncol = 2)

ggsave(
  here(params$so_dir, "ag_markers_chikv.png"),
  plot = plt,
  width = 18, height = 15
)
```



```{r "COMPARISON OF EMPTY DROPS", eval = FALSE}
# Paths to raw matrices to use for normalization factors
mat_dir <- names(params$res_dir)
mat_dir <- set_names(params[mat_dir], mat_dir)

# Calculate normalization factors
filt_counts <- qread(here("results/2022-10-28/sobjs/filt_norm_factors.qs"))
raw_counts  <- qread(here("results/2022-10-28/sobjs/norm_factors.qs"))

counts_df <- list(raw = raw_counts, filt = filt_counts) %>%
  imap_dfr(~ {
    reads <- .y
    
    .x %>%
      imap_dfr(~ {
        tibble(
          reads   = reads,
          run     = .y,
          library = names(.x),
          counts  = .x
        )
      })
  }) %>%
  pivot_wider(names_from = reads, values_from = counts) %>%
  mutate(frac_in_cells = filt / raw) %>%
  arrange(desc(frac_in_cells))

# Sample meta.data
lib_meta <- lec_so@meta.data %>%
  distinct(orig.ident, sample, tm, mouse, exp) %>%
  as_tibble()

counts_df %>%
  left_join(lib_meta, by = c(library = "orig.ident")) %>%
  filter(!is.na(sample)) %>%
  arrange(desc(frac_in_cells))
```

```{r "DSB TEST", eval = FALSE}
# Test the dsb package for correcting CITE-seq data based on empty droplets
# the results are not very convincing
# the signals do not decrease with time and the 3wk signals appear much higher
# than all the other timepoints

# data.frame with parameters
library(dsb)
library(furrr)

mat_dir <- names(params$res_dir)
mat_dir <- set_names(params[mat_dir], mat_dir)

mat_df <- mat_dir %>%
  imap_dfr(~ {
    rn <- .y
    
    .x %>%
      imap_dfr(~ {
        tibble(
          run    = rn,
          sample = .y,
          path   = here(params$res_dir[[rn]], .x, "outs"),
          rep    = str_extract(run, "^r[0-9]+(?=_)")
        )
      })
  })

# Normalize Ag matrices
plan(multisession, workers = 6)

norm_mats <- mat_df %>%
  future_pmap(~ {
    args <- list(...)
    filt_mat <- Read10X(here(args$path, "filtered_feature_bc_matrix"))$`Antibody Capture`
    raw_mat  <- Read10X(here(args$path, "raw_feature_bc_matrix"))$`Antibody Capture`
    raw_mat  <- raw_mat[, !colnames(raw_mat) %in% colnames(filt_mat)]
    
    DSBNormalizeProtein(
      cell_protein_matrix = filt_mat,
      empty_drop_matrix   = raw_mat,
      denoise.counts      = FALSE,
      pseudocount.use     = 1,
      use.isotype.control = FALSE,
      return.stats        = TRUE
    )
  })

names(norm_mats) <- mat_df$sample

# Create data.frame for corrected data
dsb_dat <- mat_df %>%
  pmap_dfr(~ {
    args <- list(...)
    
    dat <- t(norm_mats[[args$sample]]$dsb_normalized_matrix)
    
    if (!grepl("^CD45", args$sample)) {
      colnames(dat) <- params$ag_key[[args$rep]][colnames(dat)]
    }
    
    colnames(dat) <- colnames(dat) %>%
      str_c("dsb_", .)
    
    dat %>%
      as_tibble(rownames = "raw_cell_id") %>%
      mutate(run = args$fun, sample = args$sample)
  })

# Add corrected counts to object
lec_so <- qread(here(params$so_dir, "results/2022-10-28/sobjs/so_lec_new.qs"))

lec_so@meta.data <- lec_so@meta.data %>%
  mutate(raw_cell_id = str_remove(rownames(.), "_[_0-9]+$")) %>%
  as_tibble(rownames = ".cell_id") %>%
  left_join(dsb_dat, by = c("raw_cell_id", orig.ident = "sample")) %>%
  mutate(
    Ag_score = case_when(
      grepl("d[0-9]+$", mouse) ~ dsb_ovalbumin,
      grepl("^3wk$", mouse)    ~ dsb_3wk,
      grepl("^6wk$", mouse)    ~ dsb_6wk,
      
      Ag_3wk_score >= Ag_6wk_score ~ Ag_3wk_score,  # For 6wk-3wk use highest
      Ag_6wk_score >= Ag_3wk_score ~ Ag_6wk_score
    ),
    Ag_score_3 = ifelse(vaccination == "dual", Ag_3wk_score, Ag_score),
    Ag_score_6 = ifelse(vaccination == "dual", Ag_6wk_score, Ag_score),
    tm_3 = ifelse(vaccination == "dual", 21, tm),
    tm_6 = ifelse(vaccination == "dual", 42, tm)
  ) %>%
  column_to_rownames(".cell_id")

# Create plots
# timepoints
lec_so@meta.data %>%
  # filter(subtype == "fLEC") %>%
  # filter(tm != 21) %>%
  filter(vaccination == "single") %>%
  ggplot(aes(tm, Ag_score, group = tm, fill = subtype)) +
  geom_boxplot(width = 3) +
  facet_wrap(~ subtype, nrow = 1) +
  djvdj_theme()

# antigen-exchange
lec_so@meta.data %>%
  # filter(subtype == "fLEC") %>%
  mutate(vaccination = fct_relevel(vaccination, c("single", "dual"))) %>%
  ggplot(aes(tm_3, Ag_score_3, group = tm, fill = subtype, color = subtype, alpha = vaccination)) +
  geom_boxplot(width = 6, position = position_dodge2(preserve = "single")) +
  facet_wrap(~ subtype, nrow = 1) +
  scale_alpha_manual(values = c(0.25, 0.75)) +
  djvdj_theme()

lec_so@meta.data %>%
  # filter(tm_6 != 21) %>%
  # filter(subtype == "fLEC") %>%
  mutate(vaccination = fct_relevel(vaccination, c("single", "dual"))) %>%
  ggplot(aes(tm_6, Ag_score_6, group = tm, fill = subtype, color = subtype, alpha = vaccination)) +
  geom_boxplot(width = 6, position = position_dodge2(preserve = "single")) +
  facet_wrap(~ subtype, nrow = 1) +
  scale_alpha_manual(values = c(0.25, 0.75)) +
  djvdj_theme()
```

```{r "PLOT RAW COUNTS CORRELATION", eval = FALSE}
clmns <- c(
  "exp", "mouse", "vaccination", "orig.ident", "subtype",
  "adt_ABPS2", "adt_ABPS4",
  "Ag_3wk", "Ag_6wk"
)

dat <- lec_so %>%
  FetchData(clmns, slot = "counts") %>%
  filter(vaccination == "dual") %>%
  as_tibble(rownames = "cell_id")

xvar <- "adt_ABPS2"
yvar <- "adt_ABPS4"

# xvar <- "Ag_3wk"
# yvar <- "Ag_6wk"

dat %>%
  ggplot(aes(!!sym(xvar) + 1, !!sym(yvar) + 1, color = subtype)) +
  geom_point() +
  facet_grid(exp ~ subtype) +
  djvdj_theme() +
  theme(aspect.ratio = 1) +
  geom_abline(slope = 1, intercept = 0) +
  scale_x_log10() +
  scale_y_log10()
```