---
title: "Antigen Tracking Figures"
author: "Ryan Sheridan"
date: "`r Sys.Date()`"
output:
  html_document:
    toc:            true
    toc_float:      true
    toc_depth:      4
    df_print:       "paged"
    code_folding:   "hide"
    self_contained: true
    highlight:      "kate"
params:
  template_dir:    "src"                                                                  # Rmd templates
  so_dir:          "~/Dropbox/Ryan/Projects/antigen-exchange/results/sobjs/2024-07-01"    # Seurat objects
  mod_dir:         "~/Dropbox/Ryan/Projects/antigen-exchange/results/models/2025-03-20"   # ML models
  abe_dir:         "~/Dropbox/Ryan/Projects/antigen-exchange/results/abe"                 # Abe et al objects
  geomx_dir:       "results/geomx"                                                        # GeoMx objects
  chikv_dir:       "~/Dropbox/Ryan/Projects/morrison-lnsc/results/sobjs"                  # CHIKV objects
  chikv_type_clmn: "lec_subtype"
  xen_dir:         "~/Dropbox/Ryan/Projects/antigen-exchange/results/xenium/2024-03-22"   # Xenium objects
  ref_dir:         "ref"                                                                  # clustifyr references
  geo_dir:         "results/geo"                                                          # GEO submission files
  table_dir:       "results/tables/2025-03-20"                                            # tables
  sample_info:     "sample_info.xlsx"                                                     # samples names/info
  res_dir:
    value:
      exp-0: "~/Projects/antigen-tracking/results"
      exp-1: "~/Projects/tamburini-antigen-tracking/results/2022-03-11"
      exp-2: "~/Projects/tamburini-antigen-tracking/results/2022-10-28"
  xen_res_dir: "results/xenium/20240322__194506__032224_Tamburini_Run_1"                     # Directory with Xenium data
  xen_regex:   ["output-XETG00230__0022624__[A-Z]__", "output-XETG00230__0022841__[A-Z]__"]  # Regex to pull samples for Xenium slides
  xen_samples: 
    value:
      A: ["6wk", 1, 2]      # Sample identity and 6wk and 3wk barcode identity (in this order)
      B: ["6wk", 2, 1]      # e.g. `1` indicates GeoMx-Barcode01
      C: ["6wk-3wk", 1, 2]
      D: ["6wk-3wk", 2, 1]
      E: ["3wk", 2, 1]
      F: ["3wk", 1, 2]
  xen_ag_targets:
    value:
      Ag_tag_1: "GeoMx-Barcode01-ERCC00142"
      Ag_tag_2: "GeoMx-Barcode02-ERCC00144"
  xen_custom_targets: "ref/QX42XM_mMulti_98g_custom_sequences.csv" # Custom non-endogenous targets in panel
editor_options: 
  chunk_output_type: console
---

---

## Summary

A summary of the revised analysis needed to respond to reviewer comments is
included below.

<u>Primary</u>

- [ ] &nbsp;Train additional models to further corroborate predictions
      ([R3C7](#R3C7)) and to improve human predictions ([R1C5](#R1C5))
- [ ] &nbsp;Revise cell type annotations, check epithelial cells ([R3C3](#R3C3))
- [ ] &nbsp;Modify LEC subset labels used for CHIKV prediction plots, need these
      groups to be directly comparable with the groups assessed by flow
      ([R1C4](#R1C4))
- [x] &nbsp;Perform additional Xenium analysis to further assess spatial
      correlation between Ag+ DCs and LECs ([R3C8](#R3C8))
- [x] &nbsp;Compare expression of the Ag modules for LECs vs DC subsets
      ([R3C5](#R3C5))
- [x] &nbsp;Include a bargraph similar to [Figure 2]B that includes dual
      immunized mice, to show that predicted Ag-competent cells are taking up
      Ag after the 2nd immunization ([R1C1](#R1C1))

<u>Secondary</u>

- [ ] &nbsp;Check CCR7hi DC population for cDC1s ([R1C6](#R1C6))
- [ ] &nbsp;Report Pearson correlation for Fig S4E-F ([R3C6](#R3C6))
- [ ] &nbsp;Adjust [Figure 1] and [Figure S1] colors to make them more distinguishable
      ([R2-minor](#R2-minor))
- [ ] &nbsp;Generate source tables for all figures

```{r "setup", include = FALSE}
# Default chunk options
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo    = FALSE,
  dpi     = 300
)

options(java.parameters = "-Xmx12000m")

# Set environment for running UMAP
reticulate::use_condaenv("umap-2", conda = "~/.local/bin/micromamba")

# Main setup
knitr::knit(here::here(params$template_dir, "setup.Rmd"), "")

# Train and load models
knit(here(params$template_dir, "setup-ml.Rmd"), "")

# Plot parameters
min_cells <- 5
rf_typs   <- c("cLEC", "fLEC", "Collecting")
rf_tms    <- c(14, 21, 42)

ag_modules <- rf_typs %>%
  map(str_c, c("_high", "_low")) %>%
  unlist()

# Keys for naming labels
ag_clmns <- c("Ag_3wk_score", "Ag_6wk_score")

ag_labs_2 <- set_names(
  c("Ag-score\n21 day", "Ag-score\n42 day"),
  ag_clmns
)

ag_labs <- ag_labs_2 %>%
  str_replace_all("\n", " (") %>%
  str_replace_all("$", ")") %>%
  set_names(ag_clmns)

# Mouse key
m_key <- c(
  d2        = "2 day",
  d14       = "14 day",
  `3wk`     = "21 day",
  `6wk`     = "42 day",
  `6wk-3wk` = "dual"
)

# Experiment to show in main figures
ag_exps <- list(LEC = "exp-1", DC = "exp-2")

# LEC/DC subset colors
subtype_clrs <- list(LEC = lec_clrs, DC = dc_clrs)

# Column in CHIKV data containing LEC annotations
CHIKV_TYPE_CLMN <- params$chikv_type_clmn
```

```{r "load xenium data", include = FALSE}
# Xenium setup
xen_slides <- seq_along(params$xen_regex)

chunks <- xen_slides %>%
  map_chr(~ knit_expand(here(params$template_dir, "setup-xenium.Rmd")))

knit(text = chunks, output = "")

# Key Xenium marker genes
xen_gns <- c(
  "Ptprc",
  "Cd3d", "Cd19",
  "Cd8a",
  "Prox1", "Lyve1",
  "Pdpn", "Pecam1",
  "Itgam",  # cDC2
  "Itgax",  # DCs
  "Zbtb46", # DCs
  "Sirpa",
  "Xcr1",
  "Siglech",
  "Tbx21"
)

# Load Xenium data
xen <- xen_slides %>%
  map(~ {
    here(params$xen_dir, str_c("xen_s", .x, ".qs")) %>%
      qread() %>%
      FetchData(c(colnames(.@meta.data), xen_gns)) %>%
      as_tibble(rownames = "cell_id") %>%
      mutate(slide = .x)
  }) %>%
  bind_rows() %>%
  mutate(
    cell_type = if_else(grepl("^other", cell_type), "unassigned", cell_type),
    cell_type = if_else(
      cell_type %in% c("Macrophages", "Monocytes"),
      "Mon/Mac",
      cell_type
    )
  )

# Top cell types to plot
dcs <- xen$cell_type %>%
  unique() %>%
  grep("DC", ., value = TRUE) %>%
  sort()

ag_dcs <- xen %>%
  filter(cell_type %in% dcs) %>%
  group_by(cell_type) %>%
  filter(all(table(Ag_class) > 15)) %>%
  pull(cell_type) %>%
  unique()

top_types <- c("LEC", dcs)

# FOVs to plot
fovs    <- sort(unique(xen$fov))
top_fov <- "D"
bad_fov <- "C"  # section from very end of LN

# Format plotting data
xen_ag_classes <- c(
  `Ag-low`  = "Ag-",
  `Ag-high` = "Ag+"
)

xen <- xen %>%
  mutate(
    Ag_class = recode(Ag_class, !!!xen_ag_classes),
    Ag_class = fct_relevel(Ag_class, unname(xen_ag_classes)),
    slide    = str_c("section ", slide)
  ) %>%
  arrange(Ag_class)

# Xenium colors
xen_clrs <- typ_clrs

xen_clrs["Mon/Mac"] <- typ_clrs["Macrophages"]

xen_clrs <- xen_clrs[names(xen_clrs) %in% xen$cell_type]
```

```{r "load naive data", include = FALSE}
# Load 24 hpi mock-infected data
# * merge mock-infected with Ag LEC object
# * to compare timepoints, need all cells used for comparison to be present when
#   scores are calculated
chikv_naive <- load_obj(here(params$chikv_dir, "so_lec.qs"))

chikv_naive <- chikv_naive %>%
  subset(tm == "24hpi" & treatment == "mock") %>%
  mutate_meta(
    mutate,
    tm = 0,
    subtype = !!sym(CHIKV_TYPE_CLMN)  # Set correct subtype column used
  )                                   # for Lucas et al

naive_so <- lec_so %>%
  subset(mouse != "6wk-3wk") %>%
  merge(chikv_naive)

# Add module scores
# * need to recalculate scores since we are now including the mock-infected data
MODULE_FEATS[[1]] %>%
  iwalk(~ {
    clmn_nms <- str_c(.y, "_", names(.x))

    clmn_nms <- set_names(
      str_c(clmn_nms, seq_along(clmn_nms)),
      clmn_nms
    )

    naive_so <<- naive_so %>%
      mutate_meta(dplyr::select, -any_of(names(clmn_nms))) %>%
      AddModuleScore(
        features = .x,
        name     = names(clmn_nms)
      ) %>%
      mutate_meta(dplyr::rename, !!!syms(clmn_nms))
  })

naive_dat <- naive_so@meta.data %>%
  as_tibble(rownames = ".cell_id")

rm(chikv_naive, naive_so)
gc()
```

```{r "format lec data"}
# Objects to use for each cell type
ag_objs <- list(LEC = lec_so, DC = dc_so)

# Format prediction labels
# * make sure the correct predictions and module scores are being used
# * once the models are finalized, this information can be saved in the
#   meta.data
# * rf_pred will be NA if cell type was not used for training
# * pred_grp includes Ag-competent groups based on Ag predictions
pred_clmns <- str_c("ml_pred_", 1:2, "_grp")

ML_PREDS %>%
  iwalk(~ {
    lec_so <<- lec_so %>%
      AddMetaData(.x, col.name = str_c("ml_", colnames(.x), "_", .y))
  })

lec_so <- lec_so %>%
  mutate_meta(
    mutate,
    
    # Check for correct predictions
    across(
      matches("^ml_pred_[0-9]+$"),
      ~ as.character(.x) == as.character(!!sym(rf_dat_clmn)),
      .names = "{.col}_correct"
    ),
    
    # Format prediction labels
    across(
      matches("^ml_pred_[0-9]+$"),
      ~ {
        crct_clmn <- str_c(cur_column(), "_correct")
        
        ifelse(
          .x == "high" & !(cur_data()[[crct_clmn]]),
          "Ag-competent",
          str_c("Ag-", !!sym(rf_dat_clmn))
        )
      },
      .names = "{.col}_grp"
    ),
    
    across(
      matches("^ml_pred_[0-9]+$"),
      ~ str_c("Ag-", .x)
    ),
    
    # Label training data
    training_data = .cell_id %in% TRAINING_CELLS,
    mouse = fct_relevel(mouse, mice)
  )

# Add module scores
# * only add scores for the best model
MODULE_FEATS[[1]] %>%
  iwalk(~ {
    clmn_nms <- str_c(.y, "_", names(.x))

    clmn_nms <- set_names(
      str_c(clmn_nms, seq_along(clmn_nms)),
      clmn_nms
    )

    lec_so <<- lec_so %>%
      mutate_meta(dplyr::select, -any_of(names(clmn_nms))) %>%
      AddModuleScore(
        features = .x,
        name     = names(clmn_nms)
      ) %>%
      mutate_meta(dplyr::rename, !!!syms(clmn_nms))
  })
```

---

<br>

<br>

## Figure 1

*Antigen persists in discrete cell populations within the lymph node*

A.  Experimental design
B.  Mean Ag-score is shown for each cell type for 2, 14, 21, and 42 days post
    vaccination.
C.  UMAP projection shows LEC subsets.
D.  UMAP projections show Ag-scores for LEC subsets for each timepoint.
E.  Mean Ag-score is shown for LEC subsets for each timepoint.
F.  Ag-scores are shown for each timepoint for each LEC subset.
G.  UMAP projection shows DC subsets.
H.  UMAP projections show Ag-scores for DC subsets for each timepoint.
I.  Mean Ag-score is shown for DC subsets for each timepoint.
J.  Ag-scores are shown for each timepoint for each DC subset.

<u>Notes</u>

* Only experiment 1 is shown for the LEC 21 and 42 day timepoints (including
  for the UMAPs). Experiment 2 data include very few LECs for the 21 day
  timepoint.
* Experiment 2 is shown for the DC 21 and 42 day timepoints. Experiment 1 data
  are lower quality (sparse signal, mostly zeros).
* Subsets are only shown if there are >= `r min_cells` cells for each timepoint
* CD45+ and CD45- cell types are only shown for the CD45+ and CD45- datasets,
  respectively.

```{r "type Ag heat", fig.width = 16, fig.height = 5}
# Format Ag data for broad cell types
# exclude CD45+ cell types from CD45- datasets and vice versa
cd45neg_typs <- c("LEC", "BEC", "Fibroblasts", "Epithelial cells")

typ_ag_dat <- all_meta %>%
  filter(
    vaccination == "single",
    cell_type != "unassigned"
  ) %>%
  group_by(cell_sort) %>%
  mutate(
    tm        = fct_relevel(as.character(tm), tms),
    cell_type = fct_infreq(cell_type),
    cell_type = fct_relevel(cell_type, c("LEC", "Fibroblasts", "DC"))
  ) %>%
  ungroup() %>%
  filter(
    (cell_sort == "CD45neg" & cell_type %in% cd45neg_typs) |
      (cell_sort == "CD45pos" & !cell_type %in% cd45neg_typs)
  )

# Format data for heatmap
# exclude unassigned cells since these likely include multiple cell types
heat_dat <- typ_ag_dat %>%
  group_by(tm, cell_type) %>%
  summarize(
    Ag_score = mean(Ag_score), .groups = "drop",
    n        = n()
  ) %>%
  group_by(cell_type) %>%
  mutate(cell_type = str_c(cell_type, "\nn = ", label_comma()(sum(n)))) %>%
  ungroup() %>%
  mutate(
    tm      = fct_relevel(as.character(tm), rev(as.character(tms))),
    cell_type = fct_reorder(cell_type, Ag_score, mean, .desc = TRUE)
  )

# Create heatmap summarizing Ag scores
typ_heat <- heat_dat %>%
  ggplot(aes(cell_type, tm, fill = Ag_score)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("white", "#D55E00"), na.value = "black") +
  guides(fill = guide_colorbar(
    title.position = "top", ticks = FALSE, title = "Ag-score",
    barwidth = unit(150, "pt"), barheight = unit(10, "pt")
  )) +
  labs(y = "days post vaccination") +
  base_theme +
  theme(
    plot.margin  = margin(0, 15, 0, 15),
    legend.position = "top",
    legend.justification = 0,
    legend.text  = element_text(size = txt_pt1),
    aspect.ratio = n_distinct(heat_dat$tm) / n_distinct(heat_dat$cell_type) * 0.9,
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1),
  )
```

```{r "subtype Ag UMAPs"}
# Data for UMAPs
# show exp1 for LECs and exp2 for DCs
fig1_dat <- ag_objs %>%
  imap(~ {
    .x@meta.data %>%
      as_tibble(rownames = ".cell_id") %>%
      filter(
        exp %in% c("exp-0", ag_exps[[.y]]),
        vaccination == "single"
      ) %>%
      mutate(tm_lab = factor(as.character(tm), tms))    
  })

# Subtype UMAPs
fig1_typ_u <- fig1_dat %>%
  imap(~ {
    typ_lvls <- .x %>%
      pull(subtype) %>%
      unique() %>%
      as.character()
    
    res <- .x %>%
      plot_scatter(
        "subtype", "hUMAP_1", "hUMAP_2",
        size         = 0.1,
        plot_colors  = c(lec_clrs, dc_clrs),
        plot_lvls    = rev(typ_lvls),
        label_params = list(size = ttl_pt1),
        panel_nrow   = 2
      ) +
      guides(color = guide_legend(ncol = 1, override.aes = list(size = 4))) +
      umap_theme +
      theme(
        aspect.ratio = 0.9,
        legend.position = "right",
        legend.title = element_blank(),
        legend.text  = element_text(size = txt_pt2)
      )
    
    res
  })

# Ag-score UMAPs
u_clrs <- list(
  LEC = c("lightblue", "white", "#D7301F"),
  DC  = c("lightblue", "white", "#6A51A3", "#6A51A3")
  # DC  = c("white", "#6A51A3")
)

fig1_ag_u <- fig1_dat %>%
  imap(~ {
    .x %>%
      filter(tm_lab %in% as.character(tms)) %>%
      plot_scatter(
        "Ag_score", "hUMAP_1", "hUMAP_2",
        plot_colors  = u_clrs[[.y]],
        label_params = list(size = ttl_pt1),
        group_col    = "tm_lab",
        panel_nrow   = 1,
        outline      = TRUE,
        size         = 0.2,
        stroke       = 0.7
      ) +
      facet_wrap(
        ~ tm_lab, nrow = 1,
        labeller = as_labeller(function(x) str_c("day ", x))
      ) +
      guides(fill = guide_colorbar(
        title.position = "right", ticks = FALSE, title = "Ag-score",
        barheight = unit(120, "pt"), barwidth = unit(10, "pt")
      )) +
      umap_theme +
      theme(
        aspect.ratio         = 0.9,
        panel.border         = element_rect(color = ln_col, linewidth = ln_pt, fill = NA),
        legend.position      = "right",
        legend.justification = 0,
        legend.title         = element_text(angle = 270, hjust = 0.5),
        strip.text.y         = element_blank()
      )
  })
```

```{r "subtype Ag boxplots", fig.width = 8, fig.height = 2}
# Format data for boxplots
ag_bx_dat <- fig1_dat %>%
  map(~ {
    .x %>%
      filter(subtype != "unassigned") %>%
      mutate(mouse = as.character(mouse)) %>%
      group_by(subtype) %>%
      filter(all(table(mouse) >= min_cells)) %>%
      ungroup() %>%
      mutate(subtype = fct_reorder(subtype, Ag_score, mean, .desc = TRUE))
  })
  
# Create boxplots
ag_bxs <- ag_bx_dat %>%
  imap(~ create_ag_boxes(.x, clrs = subtype_clrs[[.y]]))

# Pull subsets used for main figures
ag_plt_typs <- ag_bx_dat %>%
  map(~ levels(.x$subtype))
```

```{r "subtype Ag heat"}
# Create heatmaps
heat_clrs <- c(LEC = "#D7301F", DC = "#6A51A3")

ag_heats <- ag_bx_dat %>%
  imap(~ {
    dat <- .x %>%
      group_by(mouse, tm, subtype) %>%
      summarize(Ag_score = mean(Ag_score), .groups = "drop") %>%
      mutate(
        subtype = fct_reorder(subtype, Ag_score, mean),
        tm      = fct_relevel(as.character(tm), as.character(tms)),
        ttl     = "Ag-score"
      )

    rat <- n_distinct(dat$subtype) / n_distinct(dat$tm)

    dat %>%
      ggplot(aes(tm, subtype, fill = Ag_score)) +
      geom_tile() +
      facet_wrap(~ ttl) +
      labs(x = "days post vaccination") +
      scale_fill_gradientn(colours = c("lightblue", "white", heat_clrs[[.y]])) +
      guides(fill = guide_colorbar(
        ticks = FALSE, title = "Ag-score",
        barheight = unit(105, "pt"), barwidth = unit(10, "pt")
      )) +
      base_theme +
      theme(
        legend.title = element_blank(),
        aspect.ratio = rat,
        axis.title.y = element_blank(),
        plot.margin  = margin(0, 15, 0, 15),
        plot.title   = element_text(hjust = 0.5, size = ttl_pt2)
      )
  })
```

```{r "Fig 1 diagram"}
img <- readPNG(here("results/figures/experiment_design/fig1_exp_design.png"))
img <- rasterGrob(img, interpolate = TRUE)

# Adjust plot limits if necessary
img_lims <- tibble(x = c(0, 1), y = c(0, 1))

# Create an empty plot for placing the image, adjusted to fill the area
fig1_dsgn <- ggplot(img_lims, aes(x, y)) + 
  annotation_custom(
    grob = img,
    xmin = -Inf, xmax = Inf,
    ymin = -Inf, ymax = Inf
  ) +
  xlim(c(-Inf, Inf)) +
  ylim(c(-Inf, Inf)) +
  theme_void()
```

```{r "Fig 1", fig.width = 16, fig.height = 15.5}
# Create final figure
top <- plot_grid(
  fig1_dsgn, typ_heat,
  labels     = c("A", "B"),
  label_size = ttl_pt2 * 1.5,
  nrow       = 1,
  hjust      = 1.1
)

bot <- list(
  "C" = fig1_typ_u$LEC,
  "D" = fig1_ag_u$LEC,
  "E" = ag_heats[[1]],
  "F" = ag_bxs$LEC,
  "G" = fig1_typ_u$DC,
  "H" = fig1_ag_u$DC,
  "I" = ag_heats[[2]],
  "J" = ag_bxs$DC
)

bot <- bot %>%
  imap(~ .x + labs(tag = .y))

fig1 <- top /
  (bot[[1]] + bot[[2]]) /
  (bot[[3]] + bot[[4]]) /
  (bot[[5]] + bot[[6]]) /
  (bot[[7]] + bot[[8]]) +
  plot_layout(heights = c(1.4, 1, 0.7, 1, 0.7)) &
  theme(
    plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"),
    plot.tag.position = c(0, 1)
  )

fig1
```

<br>

<br>

### Figure S2

A.  UMAP projections show cell types identified for CD45- and CD45+ samples.
B.  UMAP projections show LEC subsets identified for each timepoint.
C.  UMAP projections show DC subsets identified for each timepoint.
D.  The expression of key marker genes is shown for LEC subsets.
E.  The expression of key marker genes is shown for DC subsets.

```{r "type UMAPs"}
# Format data
typ_dat <- all_meta %>%
  mutate(
    vaccination = str_c(vaccination, "\nvaccination"),
    tm          = ifelse(is.na(tm), vaccination, str_c("day ", tm)),
    tm          = fct_relevel(tm, str_c("day ", tms)),
    cell_type   = fct_infreq(cell_type),
    cell_sort   = recode(cell_sort, CD45neg = "CD45-", CD45pos = "CD45+")
  )

n_dat <- typ_dat %>%
  group_by(tm, cell_sort) %>%
  summarize(n = str_c("n = ", label_comma()(n())), .groups = "drop")

# Create UMAP projections
typ_u <- typ_dat %>%
  plot_scatter(
    "cell_type",
    "hUMAP_1", "hUMAP_2",
    plot_colors  = typ_clrs,
    plot_lvls    = c("BEC", "LEC", "Epithelial cells", "Monocytes"),
    size         = 0.001,
    n_label      = "legend"
  ) +
  geom_text(
    aes(-Inf, Inf, label = n, color = NULL), data = n_dat,
    hjust = -0.1, vjust = 1.5, size = 10 / .pt,
    show.legend = FALSE
  ) +
  facet_grid(cell_sort ~ tm) +
  guides(color = guide_legend(nrow = 2, override.aes = list(size = 4))) +
  umap_theme_2 +
  theme(
    legend.position = "bottom",
    legend.title    = element_blank()
  )
```

```{r "subtype UMAPs"}
# Marker genes to plot
typ_gns <- list(
  LEC = c(
    "Lyve1", "Marco", "Madcam1",
    "Vcam1", "Ptx3",  "Ackr4"
  ),
  DC = c(
    "Xcr1", "Itgax", "Sirpa",
    "Lyz2", "Ccr7",  "Siglech"
  )
)

# Format data for subtype plots
subtype_dat <- ag_objs %>%
  imap(~ {
    clmns <- c("subtype", "tm", "vaccination", "hUMAP_1", "hUMAP_2")
    
    .x %>%
      FetchData(c(clmns, typ_gns[[.y]])) %>%
      as_tibble(rownames = ".cell_id") %>%
      mutate(
        vaccination = str_c(vaccination, "\nvaccination"),
        tm          = ifelse(is.na(tm), vaccination, str_c("day ", tm)),
        tm          = fct_relevel(tm, str_c("day ", tms)),
        subtype     = fct_relevel(subtype, levels(markers$annotation))
      )
  })

# Create UMAPs
subtype_u <- subtype_dat %>%
  imap(~ {
    .x %>%
      plot_scatter(
        "subtype", "hUMAP_1", "hUMAP_2",
        group_col    = "tm", panel_nrow = 2,
        n_label      = "corner",
        label_params = list(size = 14),
        size         = 0.1,
        plot_colors  = subtype_clrs[[.y]],
        plot_lvls    = levels(.x$subtype)
      ) +
      umap_theme_2 +
      theme(
        legend.position  = c(0.85, 0.25),
        legend.title     = element_blank()
      )
  })
```

```{r "subtype boxplots"}
# Create LEC/DC subtype boxplots
subtype_bxs <- subtype_dat %>%
  imap(~ {
    dat <- .x %>%
      pivot_longer(all_of(typ_gns[[.y]])) %>%
      mutate(
        subtype_tm = str_c(subtype, "_", tm),
        name       = fct_relevel(name, typ_gns[[.y]])
      ) %>%
      arrange(subtype) %>%
      mutate(subtype_tm = fct_inorder(subtype_tm))
    
    # Data for n labels
    n_labs <- dat %>%
      group_by(subtype, tm, subtype_tm) %>%
      summarize(
        n = label_comma()(n_distinct(.cell_id)),
        .groups = "drop"
      ) %>%
      mutate(
        n_lab = str_replace(subtype, "_", " "),
        n_lab = str_c(n_lab, " (", n, ")")
      )
    
    n_labs <- set_names(n_labs$n_lab, n_labs$subtype_tm)

    # Create boxplots
    n_typs <- n_distinct(dat$subtype)

    dat %>%
      ggplot(aes(subtype_tm, value, fill = subtype)) +
      geom_boxplot(outlier.size = 0.1, key_glyph = draw_key_point, alpha = 0.9) +
      facet_grid(name ~ tm, scales = "free") +
      guides(fill = guide_legend(override.aes = list(shape = 22, size = 4))) +
      scale_x_discrete(labels = n_labs) +
      scale_fill_manual(values = subtype_clrs[[.y]]) +
      base_theme +
      theme(
        aspect.ratio    = 4 / n_typs,
        legend.position = "none",
        strip.clip      = "off",
        axis.title      = element_blank(),
        axis.text.x     = element_text(angle = 90, hjust = 1, vjust = 0.5)
      )
  })
```

```{r "Fig S1", fig.width = 16, fig.height = 20}
# plts <- list(
#   "A" = typ_u,
#   "B" = 
# )

figS1 <- typ_u /
  (subtype_u$LEC + subtype_u$DC) /
  
  wrap_plots(subtype_bxs$LEC, subtype_bxs$DC, widths = c(1, 0.63)) +
  
  plot_layout(heights = c(0.7, 0.7, 1)) +
  plot_annotation(tag_levels = "A") &
  theme(plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"))

figS1
```

<br>

<br>

### Figure S3

A.  Mean Ag-score and mean MFI (fold change relative to naive)
    is shown for LECs from 2, 14, and 21 days post-immunization.
    Error bars show the 95% confidence interval.
A.  Ag-scores are shown for each timepoint for CD45- cell types.
    Two biological replicates are shown for the day 21 and day 42 timepoints
    (r1, r2),
    one biological replicate is shown for the day 2 and day 14 timepoints
    (this replicate is plotted for both r1 and r2).
    The number of cells identified for each cell type is shown above each
    timepoint.
B.  Ag-scores are shown for each timepoint for CD45+ cell types,
    as described in A.
C.  Ag-scores are shown for each timepoint for LEC subsets, as described in A.
D.  Ag-scores are shown for each timepoint for DC subsets, as described in A.

```{r "mfi vs Ag score"}
# Format MFI data
mfi_dat <- read.xlsx2(
  here("data/gMFI norm to naive 2d-14d-21d.xlsx"),
  sheetIndex = 1
)

mfi <- mfi_dat %>%
  as_tibble() %>%
  pivot_longer(everything()) %>%
  mutate(value = as.numeric(value)) %>%
  filter(!is.na(value)) %>%
  mutate(tm = str_extract(name, "(?<=X)[0-9]+"), .keep = "unused") %>%
  group_by(tm) %>%
  summarize(
    mfi         = mean(value),
    mfi_ci_low  = tidy(t.test(value))$conf.low,
    mfi_ci_high = tidy(t.test(value))$conf.high,
    .groups = "drop"
  )

# Format Ag data
ag_dat <- all_meta %>%
  filter(
    cell_type   == "LEC",
    cell_sort   == "CD45neg",
    vaccination == "single",
    tm %in% unique(mfi$tm)
  ) %>%
  group_by(tm) %>%
  summarize(
    Ag_ci_low  = tidy(t.test(Ag_score))$conf.low,
    Ag_ci_high = tidy(t.test(Ag_score))$conf.high,
    Ag_score   = mean(Ag_score),
    .groups    = "drop"
  ) %>%
  mutate(tm = as.character(tm))

dat <- ag_dat %>%
  left_join(mfi, by = "tm") %>%
  mutate(
    tm = str_c("day ", tm),
    tm = fct_relevel(tm, str_c("day ", tms))
  )

# Create scatter plot
mfi_plt <- dat %>%
  ggplot(aes(mfi, Ag_score, color = tm)) +
  geom_linerange(
    aes(ymin = Ag_ci_low, ymax = Ag_ci_high),
    color       = "grey90",
    linewidth   = 1,
    show.legend = FALSE
  ) +
  geom_linerange(
    aes(xmin = mfi_ci_low, xmax = mfi_ci_high),
    color       = "grey90",
    linewidth   = 1,
    show.legend = FALSE
  ) +
  geom_smooth(method = "lm", color = "black", linetype = 2, linewidth = 0.5, se = FALSE) +
  geom_point(size = 3) +
  geom_text_repel(
    aes(label = tm),
    size    = txt_pt2 / .pt,
    color   = "black",
    nudge_x = 2,
    nudge_y = 0.1,
    min.segment.length = Inf
  ) +
  scale_color_manual(values = c("black", "#E7301F", "#6A51A3")) +
  labs(y = "Ag-score", x = "MFI fold increase over naive") +
  
  base_theme +
  theme(
    aspect.ratio = 0.75,
    legend.position = "none"
  )
```

```{r "type Ag boxplots"}
# Format data for boxplots
dat <- c("exp-1", "exp-2") %>%
  map_dfr(~ {
    typ_ag_dat %>%
      filter(exp %in% c("exp-0", .x)) %>%
      mutate(
        exp       = str_replace(.x, "exp-", "r"),
        tm        = as.numeric(as.character(tm)),
        cell_sort = recode(cell_sort, CD45neg = "CD45-", CD45pos = "CD45+")
      )
  })

# Create boxplots
typ_ag_bxs <- c("CD45-", "CD45+") %>%
  set_names() %>%
  map(~ {
    dat <- dat %>%
      filter(cell_sort == .x)
      
    n_dat <- dat %>%
      group_by(cell_sort, exp, tm, cell_type) %>%
      summarize(n = label_comma()(n()), .groups = "drop")
    
    dat %>%
      create_ag_boxes(clrs = typ_clrs, facet_vars = c("exp", "cell_type")) +
      geom_text_repel(
        aes(y = Inf, label = n), data = n_dat,
        seed = 42, force = 0.005,
        size = 8 / .pt, vjust = 1.2
      ) +
      scale_y_continuous(expand = expansion(c(0.05, 0.15))) +
      theme(plot.margin = margin(15, 15, 30, 15))
  })
```

```{r "all subtype Ag boxplots"}
# Format data for boxplots
bx_dat <- ag_objs %>%
  map(~ {
    .x@meta.data %>%
      as_tibble(rownames = ".cell_id") %>%
      filter(vaccination == "single") %>%
      group_by(mouse, subtype, exp) %>%
      ungroup() %>%
      mutate(
        mouse   = as.character(mouse),
        tm_lab  = factor(as.character(tm), tms),
        subtype = fct_reorder(subtype, Ag_score, mean, .desc = TRUE)
      )
  })

# Create boxplots showing all LEC/DC subsets
all_ag_bxs <- bx_dat %>%
  imap(~ {
    dat <- .x
    
    dat <- c("exp-1", "exp-2") %>%
      map_dfr(~ {
        dat %>%
          filter(exp %in% c("exp-0", .x)) %>%
          mutate(exp = str_replace(.x, "exp-", "r"))
      })
    
    n_dat <- dat %>%
      group_by(tm, subtype, exp) %>%
      summarize(n = label_comma()(n()), .groups = "drop")
    
    dat %>%
      create_ag_boxes(clrs = subtype_clrs[[.y]], facet_vars = c("exp", "subtype")) +
      geom_text_repel(
        aes(y = Inf, label = n), data = n_dat,
        seed = 42, force = 0.005,
        size = 8 / .pt, vjust = 1.2
      ) +
      scale_y_continuous(expand = expansion(c(0.05, 0.15))) +
      theme(plot.margin = margin(15, 15, 30, 15))
  })
```

```{r "Fig S2", fig.width = 15, fig.height = 17.5}
top <- plot_grid(
  mfi_plt, typ_ag_bxs$`CD45-`,
  rel_widths = c(0.55, 1),
  labels = c("A", "B"),
  label_size = ttl_pt2 * 1.5,
  align = "h", axis = "tb",
  nrow = 1
)

# Create final figure
figS2 <- plot_grid(
  top, typ_ag_bxs$`CD45+`,
  all_ag_bxs$LEC, all_ag_bxs$DC,
  labels = c("", "C", "D", "E"),
  label_size = ttl_pt2 * 1.5,
  align = "v", axis = "rl",
  ncol = 1
)

figS2
```

---

<br>

<br>

## Figure 2

*Identification of an antigen archiving gene signature*

A.  Ag-score is shown for day 14 Ag-high and -low cells identified for LEC
    subsets with the highest Ag-score.
B.  The fraction of cells predicted to be Ag-competent is shown for each LEC
    subset 14, 21, and 42 days post vaccination.
C.  Ag-high module score is shown for Ag-low, Ag-high, and predicted
    Ag-competent LECs.
    The Spearman correlation between Ag classes and Ag-high module score is
    shown for each timepoint.
    One-sided p values were calculated and adjusted using Benjamini-Hochberg
    correction.
D.  UMAP projections show cLEC Ag-high module scores for each timepoint.
E.  The expression of select genes from the Ag-high (top four) and Ag-low
    (bottom four) gene modules is shown for cLECs. Genes identified as top
    predictors for multiple LEC subsets are shown. Triangles indicate the gene
    is differentially expressed when compared to Ag-low cells. P values were
    calculated using a one-sided Wilcoxon rank sum test with Benjamini-Hochberg
    correction.

<u>Notes</u>

* Ag-low and -high cells were identified by separately clustering each d14 LEC
  subset for each sample into two groups based on Ag-score.
* The cutoffs identified for the d14 dataset were then used to identify Ag-high
  cells for the other timepoints.
* With this approach there is a different cutoff for each LEC subset, but the
  cutoff is consistent across timepoints.

```{r "ML Ag signal", fig.width = 12, fig.height = 3.25}
# Format plotting data
plt_dat <- lec_so@meta.data %>%
  filter(
    subtype %in% rf_typs,
    mouse %in% rf_mouse
  ) %>%
  mutate(
    mouse = as.character(mouse),  # to remove factor
    !!sym(rf_dat_clmn) := fct_relevel(!!sym(rf_dat_clmn), c("low", "high"))
  ) %>%
  arrange(!!sym(rf_dat_clmn))

# Format plot labels
plt_labs <- plt_dat %>%
  group_by(subtype, mouse, tm, !!sym(rf_dat_clmn)) %>%
  summarize(
    n = label_comma()(n()),
    .groups = "drop"
  ) %>%
  mutate(n = str_c(!!sym(rf_dat_clmn), ": ", n, " cells")) %>%
  group_by(subtype, mouse, tm) %>%
  arrange(!!sym(rf_dat_clmn)) %>%
  mutate(n = str_c(n, collapse = "\n"))

plt_clrs <- lec_clrs

# Create bargraphs
brs <- rf_typs %>%
  imap(~ {
    dat <- plt_dat %>%
      filter(subtype == .x) %>%
      mutate(!!sym(rf_dat_clmn) := fct_rev(!!sym(rf_dat_clmn)))
    
    dat %>%
      ggplot(aes(y = mouse, fill = subtype, alpha = !!sym(rf_dat_clmn))) +
      geom_bar() +
      facet_wrap(~ subtype, nrow = 1, scales = "free") +
      scale_alpha_manual(values = c(low = 0.35, high = 1)) +
      scale_fill_manual(values = plt_clrs) +
      scale_y_discrete(expand = expansion(c(0, 0))) +
      scale_x_continuous(expand = expansion(c(0, 0))) +
      theme_void() +
      theme(
        legend.position = "none",
        plot.margin     = margin(t = 5),
        strip.text      = element_text(size = ttl_pt2)
      )
  })

hsts <- rf_typs %>%
  imap(~ {
    dat <- plt_dat %>%
      filter(subtype == .x)
    
    fst <- .y == 1
    lst <- .y == length(rf_typs)
    
    # Create histograms
    hst <- dat %>%
      ggplot(aes(Ag_score, fill = subtype, alpha = !!sym(rf_dat_clmn))) +
      geom_histogram(bins = 30, key_glyph = draw_key_point) +
      geom_text(
        aes(-Inf, Inf, label = n),
        data = filter(plt_labs, subtype == .x),
        check_overlap = TRUE,
        alpha = 1,
        hjust = -0.25,
        vjust = 1.1,
        size  = txt_pt2 / .pt
      ) +
      facet_wrap(~ subtype, scales = "free", nrow = 1) +
      guides(fill = "none") +
      scale_alpha_manual(
        values = c(low = 0.35, high = 1),
        labels = function(x) str_c("Ag-", x)
      ) +
      scale_fill_manual(values = plt_clrs) +
      scale_y_continuous(expand = expansion(c(0.05, 0.25))) +
      guides(alpha = guide_legend(override.aes = list(shape = 22, size = 5))) +
      labs(x = "Ag-score", y = "number of cells") +
      base_theme +
      theme(
        aspect.ratio    = 0.6,
        strip.text      = element_blank(),
        legend.position = "none",
        legend.title    = element_blank(),
        plot.margin     = margin(t = 0, r = 5, b = 0, l = 5)
      )
    
    if (!fst && !lst) {
      hst <- hst +
        theme(legend.position = "bottom")
    }
    
    if (!fst) {
      hst <- hst +
        theme(axis.title.y = element_blank())
    }
    
    if (fst || lst) {
      hst <- hst +
        theme(axis.title.x = element_blank())
    }
    
    hst
  })

# Create final figure
hsts <- wrap_plots(
  append(brs, hsts),
  ncol = length(rf_tms),
  heights = c(0.1, 1)
)
```

```{r "ML Ag-competent bars", fig.width = 8, fig.height = 5}
# Format data for bargraphs
# * exclude cells used to train models
rf_bar_dat <- lec_so@meta.data %>%
  .filter_rf_data() %>%
  mutate(
    subtype = fct_relevel(subtype, rf_typs),
    tm      = as.character(tm),
    
    across(
      matches("^ml_pred_[0-9]+_correct$"),
      ~ fct_relevel(as.character(.x), c("TRUE", "FALSE"))
    ),
    across(
      matches("^ml_pred_[0-9]+_grp$"),
      ~ fct_relevel(.x, rev(names(pred_clrs)))
    )
  )

bar_thm <- base_theme +
  theme(aspect.ratio = 1.4)

rf_bar_1 <- rf_bar_dat %>%
  ggplot(aes(tm, fill = ml_pred_1_grp)) +
  geom_bar(position = "fill", alpha = 0.85, key_glyph = draw_key_point) +
  facet_grid(~ subtype) +
  scale_fill_manual(values = pred_clrs) +
  scale_y_continuous(breaks = c(0, 0.5, 1)) +
  guides(fill = guide_legend(override.aes = list(shape = 22, size = 5))) +
  labs(y = "fraction of cells", x = "days post vaccination") +
  bar_thm +
  theme(
    legend.title = element_blank(),
    legend.position = "bottom"
  )
```

```{r "ML module boxplots", fig.width = 10, fig.height = 4}
# Format plot data
# * exclude dual vaccination data
clmns <- c(
  "tm", "subtype", pred_clmns,
  "Ag_score", "training_data",
  ag_modules
)

rf_bx_dat <- lec_so %>%
  FetchData(clmns) %>%
  as_tibble(rownames = ".cell_id") %>%
  .filter_rf_data()

rf_mod_args <- list(
  mod_typ = c("high", "low"),
  p_alt   = c("greater", "less"),
  p_hjust = c(0.7, 0.55),
  n_label = c(FALSE, TRUE)
)

mod_bxs <- rf_mod_args %>%
  pmap(~ {
    args <- list(...)
    
    module_clmns <- str_c(rf_typs, "_", args$mod_typ)
    module_title <- str_c("Ag-", args$mod_typ, " module score")
    
    rf_bx_dat %>%
      create_rf_module_boxes(
        module_clmns = module_clmns,
        module_title = module_title,
        p_alt        = args$p_alt,
        p_hjust      = c(args$p_hjust, 0.5),
        p_vjust      = c(1.2, 2),
        clrs         = pred_clrs,
        n_label      = args$n_label,
        linewidth    = 0.6
      )
  })
```

```{r "ML cLEC module UMAPs", fig.width = 8, fig.height = 6.5}
# Format plotting data
u_dat <- lec_so@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(tm %in% rf_tms) %>%
  mutate(Ag_class_ml = str_c("Ag-", Ag_class_ml))

# Create UMAPs
u_args <- list(
  typ = set_names(rf_typs),
  pt_size = list(1, c(0.5, 0.1), c(0.5, 0.1)),
  ag_clrs = list(
    c("lightblue", "lightblue", "lightblue", "white", "#D7301F"),
    c("lightblue", "white", "#D7301F"),
    c("lightblue", "white", "#D7301F")
  )
)

mod_u <- u_args %>%
  pmap(~ {
    args <- list(...)
    
    u_dat %>%
      create_rf_module_umaps(
        typ           = args$typ,
        ag_class_clmn = "Ag_class_ml",
        module_clmn   = str_c(args$typ, "_high"),
        module_title  = str_c(args$typ, " Ag-high module"),
        equal_scales  = FALSE,
        pt_size       = args$pt_size,
        ag_clrs       = args$ag_clrs
      )
  })
```

```{r "ML top features examples", fig.height = 8, fig.width = 8}
# Calculate p-values
# * use all features included in models, adjust based on this number of genes
# * adjust p-values separately for each tm and pred_grp
p_vals <- MODULE_FEATS[[1]][rf_typs] %>%
  imap_dfr(~ {
    p_alt <- c(high = "greater", low = "less")
    
    typ <- .y
    
    gns <- .x %>%
      imap(~ set_names(rep(.y, length(.x)), .x)) %>%
      reduce(c)
    
    ### FIX THIS, THERE SHOULD NOT BE DUPLICATES ###
    # gns <- gns[!duplicated(names(gns))]
    
    dat <- lec_so %>%
      .subset_rf_data(subtype == typ) %>%
      FetchData(c("ml_pred_1_grp", "tm", names(gns))) %>%
      as_tibble(rownames = ".cell_id") %>%      
      pivot_longer(all_of(names(gns)), names_to = "gene") %>%
      mutate(
        class = gns[gene],
        p_alt = p_alt[class]
      ) %>%
      split(.$gene)
      
    dat <- dat %>%
      map_dfr(~ {
        p_grps <- names(pred_clrs)[-1]
        p_con  <- names(pred_clrs)[1]
        
        dat <- .x
        
        p_grps %>%
          map_dfr(~ {
            dat %>%
              filter(ml_pred_1_grp %in% c(p_con, .x)) %>%
              group_by(tm) %>%
              summarize(
                p = wilcox.test(
                  value[ml_pred_1_grp == .x], value[ml_pred_1_grp == p_con],
                  alternative = unique(p_alt),
                  exact       = FALSE
                )$p.value,
                med_diff = abs(median(value[ml_pred_1_grp == .x]) - median(value[ml_pred_1_grp == p_con])),
                n        = length(value[ml_pred_1_grp == .x]),
                gene     = unique(gene),
                class    = unique(class),
                p_alt    = unique(p_alt),
                ml_pred_1_grp = .x,
                .groups  = "drop"
              )
          })
      }) %>%
      mutate(subtype = typ)
    
    dat
  })

# Adjust p-values for multiple testing
# * adjust each cell subtype separately
p_vals <- p_vals %>%
  group_by(subtype) %>%
  mutate(
    p_adj   = p.adjust(p, method = "BH"),
    n_genes = n()
  ) %>%
  ungroup()

# Identify genes shared between multiple cell types
shared_feats <- MODULE_FEATS[[1]] %>%
  imap_dfr(~ {
    typ <- .y
    
    .x %>%
      imap_dfr(~ {
        tibble(
          subtype = typ,
          class   = .y,
          gene    = .x,
        )
      })
  }) %>%
  group_by(class, gene) %>%
  filter(n_distinct(subtype) > 1) %>%
  ungroup()

p_dat <- p_vals %>%
  filter(
    subtype == rf_cell_type,
    p_adj < 0.05
  ) %>%
  group_by(gene) %>%
  filter(any(med_diff > 0)) %>%
  mutate(n_sig = n()) %>%
  ungroup() %>%
  arrange(desc(n_sig))

# Plot example genes
gene_examples <- lec_so %>%
  .subset_rf_data(subtype == rf_cell_type) %>%
  create_gn_plots(
    p_data = p_dat,
    top_gns = c(
      "Naaa",
      "Parm1",
      "Psap", "Ctsd",
      "Tgfa",
      # "Lgmn", "Grn",       # other endosome related genes
      "Col4a2", "Tmsb4x",  # cell migration, angiogenesis
      "Ctla2a", "Ltbp4"    # immune response
      
      # "Hmgb1", "Ccl21a"
      # "Col4a2", "Lamc1"
    ),
    n_gns     = 4,
    plt_clrs  = pred_clrs,
    x_lvls    = names(pred_clrs),
    draw_line = FALSE,
    pt_size   = 2.5
  )
```

```{r "Fig 2", fig.width = 15, fig.height = 16.1}
left <- plot_grid(
  plot_grid(hsts, nrow = 2, rel_heights = c(1, 0.1)),
  mod_bxs[[1]],
  mod_u$cLEC[[1]],
  mod_u$cLEC[[2]],
  ncol = 1,
  rel_heights = c(1, 1, 1.11, 1.1),
  labels = c("A", "C", "D"),
  label_size = ttl_pt2 * 1.5,
  label_y = c(1, 1.1, 1)
)

right <- plot_grid(
  plotlist = gene_examples,
  ncol     = 1,
  align    = "v",
  axis     = "rl",
  rel_heights = c(1, 1.225)
)

right <- plot_grid(
  rf_bar_1, right,
  ncol = 1,
  rel_heights = c(1, 3),
  labels = c("B", "E"),
  label_size = ttl_pt2 * 1.5
)

fig2 <- plot_grid(
  left, right,
  rel_widths = c(1, 0.5)
)

fig2
```

<br>

<br>

### Figure S4

A.  Venn diagram shows the number of genes shared between the cLEC, fLEC and
    collecting LEC gene modules.
A.  The top gene ontology terms (biological process, cellular component) are
    shown for the cLEC, fLEC, and collecting LEC Ag-high gene modules.
    The number of overlapping genes is shown to the right of each ontology term.
B.  Top gene ontology terms are shown for Ag-low gene modules as described in A.
C.  The expression of each Ag-high gene module is shown for naive LECs
    (0 days post vaccination) along with each time point.
D.  The expression of each Ag-low gene module is shown as described in D.
F.  The expression of the cLEC Ag-high gene module is shown for LEC and DC
    subsets for 14, 21, and 42 day post-immunization.
    Two-sided p values were calculated comparing cLECs (marked by arrow)
    with each cell subset using a Wilcoxon rank sum test and
    adjusted using Benjamini-Hochberg correction.
E.  The expression of the fLEC Ag-high gene module is shown as described in F.
G.  The expression of the collecting LEC Ag-high gene module is shown as
    described in F.

```{r "ML venn diagram"}
rf_venn <- MODULE_FEATS[[1]][rf_typs] %>%
  map(~ c(.x, recursive = TRUE, use.names = FALSE)) %>%
  ggVennDiagram(set_size = ttl_pt2 / .pt) +
  scale_fill_gradientn(colours = c("#56B4E9", "white", "#E7301F"), name = "number\nof genes") +
  theme(
    legend.key.height = unit(12, "pt"),
    legend.key.width  = unit(7, "pt"),
    legend.ticks      = element_blank()
  )
```

```{r "ML GO"}
# Term similarity for simplifying terms
dbs <- set_names(c("BP", "CC", "MF"))

go_sim_data <- dbs %>%
  map(~ {
    godata(
      OrgDb   = org.Mm.eg.db,
      keytype = "SYMBOL",
      ont     = .x
    )
  })

# Get GO terms
rf_grps <- MODULE_FEATS[[1]] %>%
  map(names) %>%
  reduce(unique)

go_res <- rf_grps %>%
  map_dfr(~ {
    grp <- .x
    
    go_feats <- MODULE_FEATS[[1]][rf_typs] %>%
      map(pluck, grp)
    
    go_sim_data %>%
      imap_dfr(~ {
        file <- str_c("ag-", grp, "_", .y, "_go")
        file <- here(params$table_dir, file)
        
        res <- go_feats %>%
          get_go(
            lec_so,
            type_clmn      = "subtype",
            max_term_size  = 1000,
            min_term_size  = 10,
            db             = .y,
            simplify_terms = TRUE,
            sim_data       = .x,
            file           = file
          )
        
        res@compareClusterResult %>%
          as_tibble() %>%
          mutate(db = .y, ml_pred_1_grp = grp)
      })
  })

# Filter for significant GO terms
go_sig <- go_res %>%
  filter(p.adjust < 0.05) %>%
  arrange(p.adjust)
```

```{r "ML GO plot"}
# Plot top CC GO terms
# * only plot top BP and CC terms
go_fig <- go_sig %>%
  # filter(db %in% c("CC", "BP")) %>%
  group_by(db, Cluster, ml_pred_1_grp) %>%
  slice(1:5) %>%
  ungroup() %>%
  arrange(p.adjust) %>%
  split(.$ml_pred_1_grp) %>%
  imap(~ {
    dat <- .x %>%
      mutate(
        y_var       = str_c(Cluster, "-", Description),
        y_var       = fct_reorder(y_var, p.adjust, min, .desc = TRUE),
        Description = str_trunc(Description, width = 60)
      )
      
    y_labs <- dat %>%
      distinct(Description, y_var)
    
    y_labs <- set_names(y_labs$Description, y_labs$y_var)
    
    dat %>%
      ggplot(aes(-log10(p.adjust), y_var, fill = Cluster)) +
      geom_col() +
      geom_vline(xintercept = -log10(0.05), color = "black", linetype = 2) +
      geom_text(
        aes(label = GeneRatio),
        hjust = -0.1
      ) +
      
      facet_grid(space = "free", scales = "free_y", rows = vars(Cluster)) +
      labs(title = str_c("Ag-", .y), x = "-log10(p-value)") +
      scale_x_continuous(expand = expansion(c(0, 0.35))) +
      scale_y_discrete(labels = y_labs) +
      scale_fill_manual(values = lec_clrs) +
      base_theme +
      theme(
        legend.position = "none",
        plot.title      = element_text(size = ttl_pt2, hjust = 0.25),
        strip.clip      = "off",
        axis.title.y    = element_blank(),
        axis.text.y     = element_text(size = txt_pt1)
      )
  }) %>%
  plot_grid(
    plotlist   = .,
    nrow       = 1,
    align      = "vh",
    axis       = "trbl",
    labels     = c("B", "C"),
    label_size = ttl_pt2 * 1.5
  )

# Create final GO figure
go_fig <- plot_grid(
  rf_venn, go_fig,
  nrow       = 1,
  rel_widths = c(0.5, 1),
  labels     = c("A", ""),
  label_size = ttl_pt2 * 1.5
)
```

```{r "ML naive expression", fig.width = 10, fig.height = 4}
naive_dat <- naive_dat %>%
  filter(subtype %in% rf_typs) %>%
  pivot_longer(all_of(ag_modules), names_to = "module_type") %>%
  separate(module_type, sep = "_", into = c("module_type", "module_ag")) %>%
  
  filter(subtype == module_type) %>%
  
  mutate(
    module_type = str_c(module_type, " Ag-", module_ag),
    tm = fct_relevel(as.character(tm), c("0", tms))
  )

naive_fig <- naive_dat %>%
  split(.$module_ag) %>%
  imap(~ {
    .x %>%
      ggplot(aes(tm, value, fill = subtype)) +
      geom_boxplot(linewidth = 0.7, outlier.size = 0.25, width = 0.5) +
      facet_grid(~ module_type) +
      scale_fill_manual(values = lec_clrs) +
      labs(x = "days post vaccination", y = "module score") +
      base_theme +
      theme(
        aspect.ratio = 1.1,
        legend.position = "none"
      )
  })

naive_fig <- plot_grid(
  plotlist = naive_fig,
  nrow  = 1,
  align = "h",
  axis  = "tb",
  rel_widths = c(1, 1),
  labels     = c("D", "E"),
  label_size = ttl_pt2 * 1.5
)

# naive_fig <- plot_grid(
#   plotlist = append(naive_fig, list(NULL)),
#   nrow  = 1,
#   align = "h",
#   axis  = "tb",
#   rel_widths = c(1, 1, 0.5),
#   labels     = c("D", "E", ""),
#   label_size = ttl_pt2 * 1.5
# )
```

```{r "ML module specificity data", include = FALSE}
mod_so <- dc_so %>%
  subset(tm %in% rf_tms)

mod_so <- lec_so %>%
  subset(tm %in% rf_tms) %>%
  merge(mod_so)

# Add module scores
# * need to recalculate scores since we are now including the mock-infected data
feats <- MODULE_FEATS$`1`

feats %>%
  iwalk(~ {
    clmn_nms <- str_c(.y, "_", names(.x))

    clmn_nms <- set_names(
      str_c(clmn_nms, seq_along(clmn_nms)),
      clmn_nms
    )
    
    mod_so <<- mod_so %>%
      mutate_meta(dplyr::select, -any_of(names(clmn_nms))) %>%
      AddModuleScore(
        features = .x,
        name     = names(clmn_nms)
      ) %>%
      mutate_meta(dplyr::rename, !!!syms(clmn_nms))
  })

mod_dat <- mod_so@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  mutate(tm = str_c("day ", tm))

rm(mod_so)
gc()
```

```{r "ML module specificity plots", fig.width = 10, fig.height = 6}
# Set module names
module_nms <- MODULE_FEATS$`1`[rf_typs] %>%
  imap(~ str_c(.y, "_", names(.x))) %>%
  flatten()

module_nms <- set_names(
  str_c(module_nms, seq_along(module_nms)),
  module_nms
)

# Format n labels
dat <- mod_dat %>%
  pivot_longer(all_of(names(module_nms)), names_to = "module") %>%
  mutate(tm_type = str_c(tm, "_", subtype))

n_dat <- dat %>%
  group_by(tm, subtype, tm_type) %>%
  summarize(n = n_distinct(.cell_id), .groups = "drop") %>%
  mutate(
    n_lab = ifelse(n > 1000, str_c(round(n / 1000, 1), "k"), as.character(n)),
    n_lab = str_c(subtype, " (", n_lab, ")")
  )

n_labs <- set_names(n_dat$n_lab, n_dat$tm_type)

# Calculate p values
p_dat <- dat %>%
  group_by(tm, module) %>%
  nest() %>%
  mutate(
    p = map(data, ~ {
      pairwise.wilcox.test(
        .x$value, .x$subtype,
        p.adjust.method = "none"
      )
    }),
    p = map2(p, module, ~ {
      typ <- str_extract(.y, "^[^_]+")
      
      .x %>%
        tidy() %>%
        filter(group1 == typ | group2 == typ) %>%
        mutate(module_type = typ)
    })
  ) %>%
  select(-data) %>%
  unnest(p) %>%
  ungroup()

# Adjust p values
p_dat <- p_dat %>%
  mutate(
    p_adj = p.adjust(
      p.value,
      method = "BH"
    )
  ) %>%
  rowwise() %>%
  mutate(
    p_lab = case_when(
      p_adj < 0.0001  ~ "***",
      p_adj < 0.001   ~ "**",
      p_adj < 0.05    ~ "*"
    ),
    subtype = c(group1, group2)[!c(group1, group2) %in% module_type]
  ) %>%
  ungroup() %>%
  filter(!is.na(p_lab)) %>%
  select(-group1, -group2) %>%
  mutate(tm_type = str_c(tm, "_", subtype))

# Plot module scores
# * only plot Ag-high modules
plts <- names(module_nms) %>%
  grep("_high$", ., value = TRUE) %>%
  set_names() %>%
  map(~ {
    ttl   <- str_replace(.x, "_(?=high|low)", " Ag-")
    ttl   <- str_c(ttl, " module expression")
    y_ttl <- str_remove(ttl, "^[^ ]+ ")
    typ   <- str_extract(.x, "^[^_]+")
    lvls  <- str_c("day ", rf_tms)
    
    dat <- mod_dat %>%
      mutate(
        tm_type = str_c(tm, "_", subtype),
        tm_type = fct_reorder(tm_type, !!sym(.x), .desc = TRUE),
        tm      = fct_relevel(tm, lvls)
      )
    
    mx_dat <- dat %>%
      group_by(tm, subtype, tm_type) %>%
      summarize(max = max(!!sym(.x)), .groups = "drop") %>%
      mutate(max = max(max)) %>%
      filter(subtype == typ)
    
    p <- p_dat %>%
      filter(module == .x) %>%
      mutate(
        max = max(mx_dat$max),
        tm  = fct_relevel(tm, lvls)
      )
    
    dat %>%
      ggplot(aes(x = tm_type, y = !!sym(.x), fill = subtype)) +
      geom_violin(color = "black", scale = "width", alpha = 0.75) +
      stat_summary(geom = "point", fun = median) +
      facet_wrap(~ tm, scales = "free_x") +
      scale_fill_manual(values = c(lec_clrs, dc_clrs)) +
      geom_text(
        aes(x = tm_type, y = max, label = p_lab),
        data  = p,
        color = "black",
        vjust = 0,
        size  = 12 / .pt
      ) +
      geom_point(
        aes(x = tm_type, y = max, fill = NULL),
        data     = mx_dat,
        position = position_nudge(y = 0.2),
        color    = "black",
        size     = 3,
        shape    = 25,
        stroke   = 1
      ) +
      scale_y_continuous(expand = expansion(c(0.05, 0.07))) +
      scale_x_discrete(labels = n_labs) +
      labs(title = ttl, y = y_ttl) +
      base_theme +
      theme(
        aspect.ratio    = 0.5,
        legend.position = "none",
        axis.title.x    = element_blank(),
        axis.text.x     = element_text(angle = 70, hjust = 1, color = "black")
      )
  })

mod_spec_plts <- plts %>%
  plot_grid(
    plotlist   = .,
    ncol       = 1,
    align      = "v",
    axis       = "rl",
    labels     = c("F", "G", "H"),
    label_size = ttl_pt2 * 1.5, label_y = 1.03
  )

# mod_spec_plts <- plot_grid(
#   mod_spec_plts, NULL,
#   nrow = 1,
#   rel_widths = c(2, 0.5)
# )
```

```{r "Fig S3", fig.width = 17, fig.height = 22}
figS3 <- plot_grid(
  go_fig, NULL, naive_fig, NULL, mod_spec_plts,
  ncol = 1,
  rel_heights = c(0.7, 0.1, 0.4, 0.1, 1.75)
)

figS3
```

<br>

<br>

### Figure S5

A.  The fraction of cells predicted to be Ag-low and Ag-high is shown for each
    LEC subset.
B.  Ag-low module score is shown for Ag-low, Ag-high, and predicted
    Ag-competent LECs.
    The Spearman correlation between Ag classes and Ag-high module score is
    shown for each timepoint.
    One-sided p values were calculated and adjusted using Benjamini-Hochberg
    correction.
    The number of cells plotted is shown below each boxplot.
C.  UMAP projections show fLEC Ag-high module scores for each timepoint.
D.  UMAP projections show collecting LEC Ag-high module scores for each
    timepoint.

```{r "ML predicted Ag class", fig.width = 8, fig.height = 5}
# Plot fraction predicted groups
rf_bar_2 <- rf_bar_dat %>%
  ggplot(aes(tm, fill = ml_pred_1, alpha = ml_pred_1_correct)) +
  geom_bar(position = "fill") +
  scale_alpha_manual(values = c(`TRUE` = 1, `FALSE` = 0.5)) +
  scale_fill_manual(values = pred_clrs) +
  scale_y_continuous(breaks = c(0, 0.5, 1)) +
  facet_grid(~ subtype) +
  labs(y = "fraction of cells", x = "days post vaccination") +
  guides(
    fill = guide_legend(order = 1, title = "prediction"),
    alpha = guide_legend(title.theme = element_blank())
  ) +
  bar_thm
```

```{r "Fig S4 NEW", fig.width = 20, fig.height = 13}
top <- plot_grid(
  rf_bar_2, mod_bxs[[2]],
  rel_widths = c(0.75, 1),
  labels = c("A", "B"),
  label_size = ttl_pt2 * 1.5,
  align = "h",
  axis  = "tb"
)

left <- plot_grid(
  plotlist = mod_u$fLEC,
  ncol  = 1,
  align = "v",
  axis  = "rl"
)

right <- plot_grid(
  plotlist = mod_u$Collecting,
  ncol  = 1,
  align = "v",
  axis  = "rl"
)

bot <- plot_grid(
  left, right,
  labels = c("C", "D"),
  label_size = ttl_pt2 * 1.5,
  nrow  = 1,
  align = "h",
  axis  = "tb"
)

figS4 <- plot_grid(
  top, bot,
  ncol = 1,
  rel_heights = c(0.5, 1)
)

figS4
```

```{r "GENE SIMILARITY", eval = FALSE, inclue = FALSE}
go_info <- go_sig %>%
  filter(Cluster == "cLEC", pred_grp == "high") %>%
  
  select(Cluster, ID, Description, geneID, p.adjust) %>%
  mutate(geneID = str_split(geneID, pattern = "/")) %>%
  unnest(geneID) %>%
  arrange(geneID, p.adjust)

# Simplified GO terms
gns <- MODULE_FEATS$cLEC$high

go <- gns %>%
  enrichGO(
    OrgDb   = org.Mm.eg.db,
    keyType = "SYMBOL",
    ont     = "BP"
  )



# go@result <- go[go$ID %in% go_info$ID]


sim <- pairwise_termsim(new)

x <- new %>%
  clusterProfiler::simplify(cutoff = 0.5)

# # GO slim terms
# go_slim_terms <- getBM(
#   attributes = c("go_id", "goslim_goa_accession", "goslim_goa_description"), 
#   filters = "go",
#   values = unique(go_sig$ID),
#   mart = m_mart
# )
# 
# # Prepare GO data to measure gene similarity
# go_sim_data <- godata(
#   OrgDb   = org.Mm.eg.db,
#   keytype = "SYMBOL",
#   ont     = "BP"
# )
# 

# # biomaRt GO terms
# biomrt <- "ensembl"
# m_mart <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")
# 
# gn_ids <- getBM(
#   attributes = c("external_gene_name", "entrezgene_id", "go_id"),
#   filters    = "external_gene_name",
#   values     = gns,
#   mart       = m_mart
# )
```

```{r "GENE SIMILARITY-2", eval = FALSE, include = FALSE}
# Calculate gene similarity
typ  <- "Collecting"
clss <- "high"
db   <- "CC"
gns  <- MODULE_FEATS[[typ]][[clss]]

gn_sim <- mgeneSim(gns, semData = go_sim_data[[db]])
gn_sim <- as.dist(1 - gn_sim)

# Cluster genes
clsts <- gn_sim %>%
  hclust() %>%
  cutree(k = 10)

clst_list <- split(names(clsts), as.factor(clsts))

# Identify GO terms for each gene cluster
res <- get_go(
  genes         = clst_list,
  so_in         = subset(lec_so, subtype == typ),
  max_term_size = 5000,
  min_term_size = 10,
  db            = db,
  sim_data      = go_sim_data[[db]]
)



# res@compareClusterResult %>%
#   select(-bgID) %>%
#   mutate(geneID = str_split(geneID, "/")) %>%
#   unnest(geneID) %>%
#   
#   group_by(ID) %>%
#   mutate(n_genes = n_distinct(geneID), term_size = as.numeric(str_extract(BgRatio, "^[0-9]+"))) %>%
#   filter(term_size < 2000) %>%
#   arrange(Cluster, desc(n_genes), p.adjust) %>%
#   
#   group_by(geneID) %>%
#   
#   ungroup() %>%
#   as.data.frame() %>%
#   arrange(Cluster, p.adjust) %>%
#   filter(p.adjust < 0.05)



# res@compareClusterResult %>%
#   select(-bgID) %>%
#   mutate(geneID = str_split(geneID, "/")) %>%
#   unnest(geneID) %>%
#   
#   group_by(ID) %>%
#   mutate(n_genes = n_distinct(geneID)) %>%
#   arrange(Cluster, desc(n_genes), p.adjust) %>%
#            
#   group_by(geneID) %>%
#   
#   slice(1) %>%
#   
#   ungroup()
  
  

# x <- go_res %>%
#   mutate(
#     geneID = str_split(geneID, "/")
#   ) %>%
#   unnest(geneID) %>%
#   filter(geneID %in% names(clsts)) %>%
#   mutate(go_clst = clsts[geneID]) %>%
#   group_by(go_clst, ID) %>%
#   mutate(n_genes = n_distinct(geneID)) %>%
#   ungroup() %>%
#   select(-bgID)
# 
# x %>%
#   filter(
#     Cluster == "Collecting",
#     pred_grp == "high",
#     db == "BP"
#   ) %>%
#   group_by(go_clst) %>%
#   slice_max(n_genes, n = 5) %>%
#   ungroup() %>%
#   arrange(go_clst, desc(n_genes)) %>%
# 
#   filter(go_clst == 1) %>%
#   group_by(geneID) %>%
#   slice_min(p.adjust, n = 1) %>%
#   slice_max(n_genes) %>%
#   ungroup() %>%
#   arrange(geneID, desc(n_genes), p.adjust)
# 
# go <- gost(
#   clsts,
#   organism = "mmusculus"
# )
# 
# # biomaRt GO terms
# biomrt <- "ensembl"
# m_mart <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")
# 
# attrs <- c(
#   "external_gene_name", "entrezgene_id",
#   "go_id", "goslim_goa_accession", "goslim_goa_description"
# )
# 
# gn_ids <- getBM(
#   attributes = attrs,
#   filters    = "external_gene_name",
#   values     = gns,
#   mart       = m_mart
# )
```

<br>

<br>

### Table S1

Stats are included for classification models.

```{r "Table S1"}
rf_clmns <- c(
  "F1", "Balanced Accuracy",
  "Specificity", "Sensitivity",
  "Precision"
)

tbl_s1 <- BEST_MLS %>%
  filter(subtype %in% rf_typs) %>%
  select(
    cell_type = subtype,
    algorithm,
    model_id,
    model_name = model_label,
    all_of(rf_clmns)
  ) %>%
  mutate(
    cell_type  = fct_relevel(cell_type, rf_typs),
    model_name = str_c("model ", model_name)
  ) %>%
  arrange(cell_type, model_name)

tbl_s1 %>%
  write_tsv(here(params$table_dir, "table_s1.tsv"))

tbl_s1
```

<br>

<br>

### Table S2

Variables used to train models are listed along with relative importance.
The top genes contributing to 99% of the cumulative importance were used for
the Ag-low and Ag-high gene modules.

```{r "Table S2"}
tbl_s2 <- BEST_MLS %>%
  filter(model_label == 1) %>%
  pmap_dfr(~ {
    args <- list(...)
    
    args$model@model$variable_importances %>%
      as_tibble() %>%
      mutate(
        subtype   = args$subtype,
        algorithm = args$algorithm,
        rank      = row_number(),
        .before   = 1
      )
  })

tbl_s2 %>%
  write_tsv(here(params$table_dir, "table_s2.tsv"))

tbl_s2
```

<br>

<br>

### Table S3

Gene ontology terms are shown for
Ag-low and Ag-high cells.

```{r "Table S3"}
ex_clmns <- c(
  "bgID", "n_ovlp", "tot_genes",
  "n_bg_ovlp", "tot_bg_genes",
  "Count", "enrichment"
)

tbl_s3 <- go_sig %>%
  rename(Ag_class = ml_pred_1_grp, cell_type = Cluster) %>%
  mutate(Ag_class = str_c("Ag-", Ag_class)) %>%
  arrange(cell_type, Ag_class, p.adjust) %>%
  select(-all_of(c(ex_clmns))) %>%
  relocate(Ag_class, .after = cell_type)
  
tbl_s3 %>%
  write_tsv(here(params$table_dir, "table_s3.tsv"))

tbl_s3
```

---

<br>

<br>

## Figure 4

*Antigen archiving is enhanced by sequential vaccinations*

A.  Experimental design
B.  Mean 21 day Ag-score is shown for LECs from mice that received a single
    vaccination (21 day or 42 day) or dual vaccination (21 day and 42 day).
C.  UMAP projections show 21 day Ag-scores for LEC subsets for single and dual
    vaccinations.
D.  Prior vaccination enhances antigen archiving.
    Ag-score is shown for single and dual vaccinations for the 21 day timepoint
    for each LEC subset. Other timepoints are shown in grey. P values
    were calculated using a one-sided Wilcoxon rank sum test with
    Benjamini-Hochberg correction.
E.  Mean 42 day Ag-score is shown as described in A.
F.  UMAP projections show 42 day Ag-scores for LEC subsets as described in B.
G.  Successive vaccinations enhances retention of previously archived antigen.
    Ag-score is shown for the 42 day timepoint as described in C.
H.  21 day and 42 day Ag-score is compared for LEC subsets.
I.  The fraction of cells that archived antigens from one vaccination (Ag-high)
    or both vaccinations (double-high) is shown for single (day 14, 21, 42)
    and dual-immunized mice (dual).
I.  The expression of Ag-high modules described in Figure 2 is shown for LECs
    that archived antigens from both vaccinations (double-high),
    from only one vaccination (single-high),
    or that have low levels of both antigens (Ag-low).
    Module scores are shown for the LEC subset matching the identified gene
    module.
    P values were calculated using a one-sided Wilcoxon rank sum test with
    Benjamini-Hochberg correction.
J.  Ag-low module scores are shown for LEC subset as described in H.

```{r "exchange UMAPs", fig.width = 16, fig.height = 4}
# Format Ag exchange plot data
# use experiment 1 since, experiment 2 is missing 3wk LECs
ex_dat <- lec_so@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(exp %in% c("exp-0", ag_exps$LEC)) %>%
  mutate(
    mouse = m_key[as.character(mouse)],
    mouse = fct_relevel(mouse, unname(m_key))
  )

# Create UMAPs
ex_heat_args <- list(
  ag_clmns = ag_clmns,
  clr      = tm_clrs[c("21", "42")]
)

ex_u <- ex_heat_args %>%
  pmap(~ {
    dat <- ex_dat %>%
      filter(exp == ag_exps$LEC)
    
    # mx_sig <- max(dat[ex_heat_args$ag_clmns])
    
    dat %>%
      plot_scatter(
        ..1,
        x           = "hUMAP_1",
        y           = "hUMAP_2",
        group_col   = "mouse",
        panel_nrow  = 1,
        outline     = TRUE,
        size        = 0.6,
        stroke      = 0.7,
        plot_colors = c("lightblue", "white", ..2),
        label_params = list(size = ttl_pt1)
      ) +
      # scale_fill_gradientn(colours = c("lightblue", "white", ..2), limits = c(0, mx_sig)) +
      guides(fill = guide_colorbar(
        ticks = FALSE,
        title = ag_labs[.x],
        barheight = unit(150, "pt"), barwidth = unit(10, "pt")
      )) +
      umap_theme +
      theme(
        legend.position = "right",
        panel.border    = element_rect(color = ln_col, linewidth = ln_pt)
      )
  })
```

```{r "exchange heatmaps"}
# Format data for heatmaps
heat_dat <- ex_dat %>%
  filter(
    exp == ag_exps$LEC,
    subtype %in% ag_plt_typs$LEC
  ) %>%
  group_by(mouse, subtype) %>%
  summarize(across(all_of(ag_clmns), mean), .groups = "drop")

# Create heatmaps
ex_heat <- ex_heat_args %>%
  pmap(~ {
    heat_dat %>%
      create_exchange_heats(
        ag_clmn = .x,
        ttl     = ag_labs_2[[.x]],
        clr     = .y
      )
  })
```

```{r "exchange boxes"}
# Effect of successive vaccinations on antigen uptake/retention
# use LEC subsets shown in Fig 1
ex_bx_args <- list(
  tm      = c("21", "42"),
  tm_clmn = c("tm_3", "tm_6"),
  ag_clmn = c("Ag_score_3", "Ag_score_6"),
  p_hjust = c(0.5, 0.8)
)

ex_bxs <- ex_bx_args %>%
  pmap(~ {
    args <- list(...)
    
    ex_dat %>%
      create_exchange_boxes(
        tm        = args$tm,
        tm_clmn   = args$tm_clmn,
        ag_clmn   = args$ag_clmn,
        typs      = ag_plt_typs$LEC,
        clrs      = tm_clrs,
        p_cutoff = Inf,
        p_hjust   = args$p_hjust,
        p_vjust   = 0.7
      )
  })
```

```{r "exchange correlation", fig.width = 10, fig.height = 4}
# Format scatter plot data
scat_dat <- lec_so@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(
    vaccination == "dual",
    subtype %in% ag_plt_typs$LEC,
  ) %>%
  mutate(subtype = fct_relevel(subtype, ag_plt_typs$LEC))

ex_scat <- scat_dat %>%
  split(.$exp) %>%
  map(create_exchange_scatter)
```

```{r "exchange Ag-high module scores", fig.width = 9, fig.height = 5.5}
# Format plot data
# module scores for biological replicates are similar
# pool all cells together to increase power
clmns <- c(
  "exp", "subtype", "mouse", "vaccination",
  "ml_pred_1", "Ag_class_dual", ag_modules
)

ex_mod_bx_dat <- lec_so %>%
  FetchData(clmns) %>%
  filter(vaccination == "dual") %>%
  mutate(Ag_class_dual = recode(Ag_class_dual, low = "Ag-low")) %>%
  
  mutate(Ag_class_dual = ifelse(
    Ag_class_dual == "Ag-low" & ml_pred_1 == "Ag-high",
    "Ag-competent",
    Ag_class_dual
  ))

# Create boxplots
bx_lvls <- c("Ag-low", "Ag-competent", "single-high", "double-high")

rf_mod_args$p_x <- c(1, 4)
rf_mod_args$p_hjust <- c(0.2, 0.8)

ex_mod_bxs <- rf_mod_args %>%
  pmap(~ {
    args <- list(...)

    module_clmns <- str_c(rf_typs, "_", args$mod_typ)
    module_title <- str_c("Ag-", args$mod_typ, " module score")

    ex_mod_bx_dat %>%
      create_module_boxes(
        module_clmns = module_clmns,
        module_title = module_title,
        lvls         = bx_lvls,
        p_alt        = args$p_alt,
        p_x          = args$p_x,
        p_hjust      = args$p_hjust
      ) +
      theme(plot.margin = margin(5, 5, 5, 15))
  })
```

```{r "exchange Ag-competent bars"}
# Format data
lvls <- c("Ag-low", "Ag-competent", "double-high", "Ag-high")

new_clrs <- pred_clrs
new_clrs["double-high"]  <- pred_clrs["Ag-competent"]
new_clrs["Ag-competent"] <- "grey75"

dat <- ex_mod_bx_dat %>%
  filter(subtype %in% rf_typs) %>%
  mutate(
    ml_pred_1_grp = Ag_class_dual,
    ml_pred_1_grp = recode(ml_pred_1_grp, `single-high` = "Ag-high"),
    tm = "dual"
  ) %>%
  bind_rows(rf_bar_dat) %>%
  mutate(
    ml_pred_1_grp = fct_relevel(ml_pred_1_grp, rev(lvls)),
    subtype       = fct_relevel(subtype, rf_typs)
  )

# Create bargraphs
dual_ag_class_brs <- dat %>%
  ggplot(aes(tm, fill = ml_pred_1_grp)) +
  geom_bar(position = "fill", key_glyph = draw_key_point) +
  facet_grid(~ subtype) +
  scale_fill_manual(values = new_clrs) +
  scale_y_continuous(breaks = c(0, 0.5, 1)) +
  guides(fill = guide_legend(override.aes = list(shape = 22, size = 5))) +
  labs(y = "fraction of cells", x = "days post vaccination") +
  bar_thm +
  theme(
    aspect.ratio = 1.2,
    legend.title = element_blank(),
    legend.position = "right"
  )
```

```{r "Fig 3 diagram", fig.width = 15.5, fig.height = 10}
img <- "results/figures/experiment_design/fig3_exp_design_landscape-2.png"
img <- readPNG(here(img))
img <- rasterGrob(img, interpolate = TRUE)

# Adjust plot limits if necessary
img_lims <- tibble(x = c(0, 1), y = c(0, 1))

# Create an empty plot for placing the image, adjusted to fill the area
fig3_dsgn <- ggplot(img_lims, aes(x, y)) + 
  annotation_custom(
    grob = img,
    xmin = -Inf, xmax = Inf,
    ymin = -Inf, ymax = Inf
  ) +
  xlim(c(-Inf, Inf)) +
  ylim(c(-Inf, Inf)) +
  theme_void() +
  theme(plot.margin = margin(0, 0, 30, 0))

fig3_dsgn <- plot_grid(
  fig3_dsgn, NULL,
  rel_widths = c(1, 0.4)
)
```

```{r "Fig 3", fig.width = 15.5, fig.height = 27}
fig3 <- fig3_dsgn /
  (ex_heat[[1]] + ex_u[[1]]) / ex_bxs[[1]] /
  (ex_heat[[2]] + ex_u[[2]]) / ex_bxs[[2]] /
  ex_scat[[ag_exps$LEC]] /
  (dual_ag_class_brs + plot_spacer() + plot_layout(widths = c(1, 1))) /
  (ex_mod_bxs[[1]] + ex_mod_bxs[[2]]) +
  # (dual_ag_class_brs + ex_mod_bxs[[1]]) / (plot_spacer() + ex_mod_bxs[[2]]) +
  plot_layout(heights = c(1.1, 1, 0.55, 1, 0.55, 0.8, 0.6, 0.8)) +
  plot_annotation(tag_levels = "A") &
  theme(
    plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"),
    plot.tag.position = c(0, 1)
  )

fig3
```

<br>

<br>

### Figure S6

A.  Ag-score is shown for single and dual vaccinations for the 21 day timepoint
    for DC subsets from a biological replicate.
    This analysis was not performed for a biological replicate for LEC subsets
    due to the low number of LECs recovered (<40 total cells).
    Other timepoints are shown in grey.
    P values were calculated using a one-sided Wilcoxon rank sum test with
    Benjamini-Hochberg correction.
B.  Ag-score is shown for the 42 day timepoint for LEC subsets from a
    biological replicate, as described in A.
C.  Ag-score is shown for the 42 day timepoint for DC subsets from a biological
    replicate, as described in A.
D.  21 day and 42 day Ag-score is compared for LEC subsets from a biological
    replicate.
D.  The correlation between Ag-high module scores and the 21 and 42 day Ag
    scores is shown for LEC subsets.
D.  The correlation between Ag-low module scores and Ag scores is shown as
    described in F.

```{r "rep exchange boxes", include = FALSE}
# Set input data
# * plot replicates not shown in main figure
exps        <- c("exp-1", "exp-2")
plt_lec_exp <- exps[exps != ag_exps$LEC]
plt_dc_exp  <- exps[exps != ag_exps$DC]

ex_bx_args$exps <-list(
  list(LEC = NULL,        DC = plt_dc_exp),
  list(LEC = plt_lec_exp, DC = plt_dc_exp)
)

# Create boxplots for biological replicates
rep_ex_bxs <- ex_bx_args %>%
  pmap(~ {
    args <- list(...)
    
    objs <- purrr::discard(args$exps, is.null)
    objs <- ag_objs[names(objs)]
    
    objs %>%
      imap(~ {
        dat <- .x@meta.data
        
        dat <- args$exps[[.y]] %>%
          map_dfr(~ {
            dat %>%
              filter(exp %in% c("exp-0", .x))
          })
        
        dat %>%
          create_exchange_boxes(
            tm         = args$tm,
            tm_clmn    = args$tm_clmn,
            ag_clmn    = args$ag_clmn,
            clrs       = tm_clrs,
            typs       = ag_plt_typs[[.y]],
            p_hjust    = args$p_hjust,
            facet_vars = "subtype",
            p_cutoff   = Inf
          ) +
          scale_y_continuous(expand = expansion(c(0.05, 0.15)))
      })
  }) %>%
  flatten()

# Remove unneeded objects
rm(ag_objs)
gc()
```

```{r "ML module exchange correlation", fig.height = 8}
# Set plot colors
ex_ag_clrs <- c(
  `double-high` = "#D7301F",
  `single-high` = "black",
  `Ag-low`      = "#56B4E9"
  # `single-high` = "#6A51A3",
)

# Format plotting data
scat_dat <- lec_so@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(vaccination == "dual") %>%
  mutate(Ag_class_dual = recode(Ag_class_dual, low = "Ag-low"))

# Create scatter plots
scat_args <- list(
  mod_typ     = c("high", "low"),
  label_x     = c(-Inf, Inf),
  label_hjust = c(-0.1, 1.1)
)

ex_mod_scat <- scat_args %>%
  pmap(~ {
    args <- list(...)
    
    module_clmns <- str_c(rf_typs, "_", args$mod_typ)
    module_title <- str_c("Ag-", args$mod_typ, " module score")
    
    scat_dat %>%
      create_module_scatter(
        ag_clmns     = ag_clmns,
        module_clmns = module_clmns,
        clrs         = ex_ag_clrs,
        module_title = module_title,
        label_x      = args$label_x,
        label_hjust  = args$label_hjust
      )
  })
```

```{r "Fig S4", fig.width = 13, fig.height = 16}
# Create final figure
top <- wrap_plots(rep_ex_bxs, ncol = 1) /
  ex_scat$`exp-2` /
  plot_layout(heights = c(1, 1, 1, 1.5)) +
  plot_annotation(tag_levels = "A") &
  theme(
    legend.margin     = margin(r = 5),
    plot.tag          = element_text(size = ttl_pt2 * 1.5, face = "bold"),
    plot.tag.position = c(0, 1)
  )

bot <- plot_grid(
  plotlist = ex_mod_scat[1:2],
  nrow = 1,
  labels = c("F", "G"),
  label_size = ttl_pt2 * 1.5 
)

# dual_ag_class_brs <- plot_grid(
#   dual_ag_class_brs, NULL,
#   nrow = 1,
#   rel_widths = c(1, 0.1),
#   labels = "E", "",
#   label_size = ttl_pt2 * 1.5
# )

figS4 <- plot_grid(
  top, bot,
  ncol = 1,
  rel_heights = c(4.5, 2.7)
  # rel_heights = c(4.5, 1.5, 2.7)
)

figS4
```

---

<br>

<br>

## Figure 5

*Antigen exchange with DCs correlates with levels of antigen archiving*

A.  Antigen counts are shown for adjacent LN tissue sections from
    a dual-immunized mouse analyzed with the Xenium platform.
B.  Localization of Ag+ and Ag- cells is shown for LECs for tissue sections
    from A.
C.  Localization of Ag+ and Ag- cells is shown for DCs for tissue sections from
    A.
D.  Cell-cell interaction enrichment results are shown for section 1 from A for
    Ag+/- LECs and DCs.
C.  The distance to the closest Ag+ LEC is shown for Ag+ and Ag- DCs for
    section 1 from A.
D.  Mean 21 day Ag-score is shown for DCs from mice that received a single
    vaccination (21 day or 42 day) or dual vaccination (21 day and 42 day).
E.  Ag-score is shown for single and dual vaccinations for the 21 day timepoint
    for DC subsets with highest Ag-score. Other timepoints are shown in grey.
    P values were calculated using a one-sided Wilcoxon rank sum test with
    Benjamini-Hochberg correction.
F.  Mean 42 day Ag-score is shown for DCs from mice that received single or
    dual vaccination as described in F.
G.  Ag-score is shown for single and dual vaccinations for the 42 day timepoint
    as described in G.

```{r "xenium Ag counts"}
# Format plot data
xen_dat <- xen %>%
  filter(fov %in% top_fov) %>%
  mutate(
    type_ag = if_else(
      cell_type %in% c("LEC", "DC", "B cells"),
      cell_type,
      "other"
    ),
    type_ag = fct_relevel(type_ag, c("LEC", "DC", "other")),
    
    Ag_class = fct_relevel(Ag_class, rev(unname(xen_ag_classes))),
    
    type_ag = if_else(
      type_ag %in% c("LEC", "DC"),
      str_c(type_ag, " (", Ag_class, ")"),
      type_ag
    ),
    
    lec_ag = ifelse(cell_type %in% c("LEC", "B cells"), type_ag, "other"),
    dc_ag  = ifelse(cell_type %in% c("DC", "B cells"),  type_ag, "other"),
  ) %>%
  arrange(type_ag, Ag_class) %>%
  mutate(
    across(c(type_ag, lec_ag, dc_ag), ~ fct_relevel(.x, unique(type_ag))),
    slide = fct_relevel(slide, str_c("section ", xen_slides))
  )

# Plot Ag+ cells
# * add pseudo count
ag_sig_fig <- xen_dat %>%
  mutate(Ag_counts = Ag_counts + 1) %>%
  plot_xenium(
    "Ag_counts",
    grp_clmn       = "slide",
    clrs           = c("white", xen_clrs[["LEC"]]),
    ttl            = "Ag counts + 1",
    trans          = "log10",
    trace_position = "bottom",
    panel_nrow     = 2
  )
```

```{r "xenium Ag class"}
# Adjust color palette
new_clrs <- c("white", xen_clrs[c("B cells", "DC", "LEC")]) %>%
  expand_colors(
    n             = c(1, 1, 2, 2),
    keep_original = TRUE,
    filter        = c("deutan", "protan", "tritan"),
    difference    = Inf,
    direction     = -1,
    range         = c(10, 90),
    maxit         = 2000
  ) %>%
  rev()

names(new_clrs) <- c(
  "LEC (Ag+)", "LEC (Ag-)",
  "DC (Ag+)", "DC (Ag-)",
  "B cells", "other"
)

# Plot Ag+ cells
ag_pos_fig <- c("lec_ag", "dc_ag") %>%
  map(~ {
    xen_dat %>%
      plot_xenium(
        .x,
        grp_clmn       = "slide",
        clrs           = new_clrs,
        pt_size        = 0.2,
        trace_position = "bottom",
        panel_nrow     = 2
      )
  })
```

```{r "xenium giotto proximity"}
# Format data
prox <- read_tsv("~/Projects/antigen-exchange-xenium/results/prox.tsv.gz")

dat <- prox %>%
  filter(type_int == "hetero") %>%
  arrange(slide, enrichm) %>%
  mutate(
    lec_dc = grepl("LEC", unified_int) & grepl("DC", unified_int),
    top    = int_ranking < min(int_ranking[lec_dc]),
    x      = str_c(slide, "-", unified_int),
    x      = fct_inorder(x)
  )

# Create bargraphs
prox_clrs <- c(
  `TRUE`  = typ_clrs[["LEC"]],
  `FALSE` = typ_clrs[["unassigned"]]
)

prox_brs <- dat %>%
  split(.$slide) %>%
  map(~ {
    .x %>%
      ggplot(aes(x, enrichm, fill = lec_dc)) +
      geom_col(width = 1) +
      geom_text_repel(
        aes(label = unified_int),
        data    = ~ filter(.x, lec_dc),
        nudge_y = 0.7,
        hjust   = 1
      ) +
      facet_wrap(~ slide, nrow = 1, scales = "free") +
      scale_fill_manual(values = prox_clrs) +
      labs(
        y = "cell-cell interaction\nenrichment",
        x = "cell-cell interactions"
      ) +
      base_theme +
      theme(
        legend.position = "none",
        axis.title.x = element_text(vjust = 5),
        axis.text.x  = element_blank(),
        axis.ticks.x = element_blank()
      )
  })
```

```{r "xenium LEC proximity", fig.width = 18, fig.height = 13}
# Format data for histograms
hist_dat <- xen_dat %>%
  filter(cell_type == "DC")
  
# Format data for median lines
ln_dat <- hist_dat %>%
  group_by(slide, fov, cell_type, Ag_class) %>%
  summarize(
    lec_dist = median(lec_dist),
    n        = n(),
    .groups  = "drop"
  )

# Format data for n labels
n_dat <- ln_dat %>%
  group_by(slide, fov, cell_type) %>%
  mutate(
    n_lab = str_c(Ag_class, " ", label_comma()(n), " cells")
  ) %>%
  summarize(
    n_lab = str_c(n_lab, collapse = "\n"),
    .groups = "drop"
  )

# Format data for p-values
p_dat <- hist_dat %>%
  group_by(slide, fov, cell_type) %>%
  summarize(
    p = wilcox.test(
      lec_dist[Ag_class == "Ag+"],
      lec_dist[Ag_class == "Ag-"],
      alternative = "less"
    )$p.value,
    .groups = "drop"
  ) %>%
  mutate(p_adj = p.adjust(p, method = "BH")) %>%
  rowwise() %>%
  mutate(
    p_lab = .format_pvalue(p_adj),
    p_lab = ifelse(
      p_adj != 0,
      str_c("italic(p) == ", p_lab),
      p_lab
    )
  ) %>%
  ungroup()

# Plot distance from nearest Ag+ LEC
prox_hists <- hist_dat %>%
  mutate(lec_dist = lec_dist + 1) %>%  # ADD 1
  split(.$slide) %>%
  imap(~ {
    dat <- list(
      n  = n_dat,
      p  = p_dat,
      ln = ln_dat
    ) %>%
      map(filter, slide == .y)
    
    .x %>%  
      ggplot(aes(lec_dist, fill = Ag_class, alpha = Ag_class)) +
      
      geom_density(aes(color = Ag_class)) +
      geom_vline(
        aes(xintercept = lec_dist, color = Ag_class, alpha = NULL),
        data      = dat$ln,
        linetype  = 2,
        linewidth = 0.75,
        show.legend = FALSE
      ) +
      
      geom_text(
        aes(x = 1, y = Inf, label = n_lab, fill = NULL, alpha = NULL),
        data = dat$n,
        hjust = 0,
        vjust = 1.2,
        size  = txt_pt2 / .pt,
        show.legend = FALSE
      ) +
      
      geom_text(
        aes(x = Inf, y = Inf, label = p_lab, fill = NULL, alpha = NULL),
        data  = dat$p,
        parse = TRUE,
        hjust = 1.1,
        vjust = 1.2,
        size  = txt_pt2 / .pt,
        show.legend = FALSE
      ) +
      
      facet_wrap(~ slide, nrow = 1) +
      labs(x = "distance from nearest Ag+ LEC (\U00B5m)", y = "density") +
      scale_fill_manual(values  = c(`Ag-` = other_clr, `Ag+` = xen_clrs[["LEC"]])) +
      scale_color_manual(values = c(`Ag-` = "black",  `Ag+` = xen_clrs[["LEC"]])) +
      scale_alpha_manual(values = c(`Ag-` = 0.5,      `Ag+` = 0.85)) +
      scale_x_log10() +
      scale_y_continuous(expand = expansion(c(0.05, 0.2))) +
      base_theme +
      theme(
        aspect.ratio    = 0.5,
        legend.position = "right",
        legend.title    = element_blank(),
        strip.text      = element_text(size = ttl_pt2),
        legend.text     = element_text(size = ttl_pt2),
        axis.title      = element_text(size = ttl_pt2)
      )
  })
```

```{r "xenium figure", fig.width = 14, fig.height = 8.5}
top <- append(list(ag_sig_fig), ag_pos_fig) %>%
  imap(~ .x + labs(tag = LETTERS[.y])) %>%
  wrap_plots(nrow = 1)

# bot <- wrap_plots(
#   plot_spacer(), hist + labs(tag = "D"), plot_spacer(),
#   nrow = 1,
#   widths = c(0.3, 1, 0.3)
# )

bot <- wrap_plots(
  prox_brs$`section 1`   + labs(tag = "D"),
  prox_hists$`section 1` + labs(tag = "E"),
  nrow = 1
  # widths = c(0.3, 1, 0.3)
)

xen_fig <- wrap_plots(
  top, bot,
  ncol = 1,
  heights = c(1, 0.35)
) &
  theme(
    plot.tag = element_text(
      size  = ttl_pt2 * 1.5,
      face  = "bold",
      hjust = 2,
      vjust = -0.2
    )
  )
```

```{r "DC exchange heatmaps", fig.width = 16, fig.height = 4}
# Format data for DC exchange plots
dc_ex_dat <- dc_so@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  filter(
    exp %in% c("exp-0", ag_exps$DC),
    subtype %in% ag_plt_typs$DC
  ) %>%
  mutate(
    mouse = m_key[as.character(mouse)],
    mouse = fct_relevel(mouse, unname(m_key))
  )

# Format data for heatmaps
dc_heat_dat <- dc_ex_dat %>%
  filter(exp == ag_exps$DC) %>%
  group_by(mouse, subtype) %>%
  summarize(across(all_of(ag_clmns), mean), .groups = "drop")

# Create heatmaps
dc_ex_heat <- ex_heat_args %>%
  pmap(~ {
    dc_heat_dat %>%
      create_exchange_heats(
        ag_clmn = .x,
        ttl     = ag_labs_2[[.x]],
        clr     = .y
      )
  })
```

```{r "DC exchange boxes", fig.width = 10, fig.height = 2}
# Effect of successive vaccinations on antigen uptake/retention
# * use DC subsets shown in figure 1
dc_ex_bxs <- ex_bx_args %>%
  pmap(~ {
    args <- list(...)
    
    dc_ex_dat %>%
      create_exchange_boxes(
        tm        = args$tm,
        tm_clmn   = args$tm_clmn,
        ag_clmn   = args$ag_clmn,
        typs      = ag_plt_typs$DC[1:3],
        clrs      = tm_clrs,
        p_size    = 11,
        p_hjust   = args$p_hjust,
        p_vjust   = 0.7
      ) +
      theme(legend.position = "bottom")
  })
```

```{r "Fig 4", fig.width = 14, fig.height = 21}
bot <- list(
  `F` = dc_ex_heat[[1]],
  G   = dc_ex_bxs[[1]],
  H   = dc_ex_heat[[2]],
  I   = dc_ex_bxs[[2]]
)

bot <- bot %>%
  imap(~ .x + labs(tag = .y))

bot <- wrap_plots(
  bot,
  nrow = 2,
  widths = c(0.25, 1)
) &
  theme(
    plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"),
    plot.tag.position = c(0, 1)
  )

fig4 <- plot_grid(
  xen_fig, NULL, bot,
  ncol = 1,
  rel_heights = c(1.7, 0.01, 1)
)

fig4
```

<br>

<br>

### Figure S7

A.  Localization of annotated cell types is shown for replicate tissue sections
    analyzed with the Xenium platform.
    Annotated cell types with >100 identified cells are shown.
    Color indicates cell density.
B.  Total Ag counts are shown for the 21 day and 42 day Ag tags, for mice
    immunized with the 21 day tag (left), 42 day tag (middle),
    or both 21 and 42 day tags (right).
    For each immunization group, tissue sections were analyzed for two LNs from
    separate mice (LN-1, LN-2), with the exception of the dual-immunization group.
    For each LN, two adjacent tissue sections were analyzed
    (section 1, section 2).
B.  The fraction of Ag+ cells is shown for each cell type for tissue sections
    shown in Figure 4A.
D.  Cell-cell interaction enrichment results are shown for section 2 from A for
    Ag+/- LECs and DCs.
C.  The distance to the closest Ag+ LEC is shown for Ag+ and Ag- DCs for
    section 2 from A.
A.  Sample identity is shown for all regions analyzed using the GeoMx DSP
    platform.
B.  Annotated regions are shown as described in F.
C.  Relative antigen tag signal is shown for each sample and each
    region segmented based on Lyve1 (LECs) and CD11C (DCs).
    Normalized 21 day antigen signal was scaled across all regions from 21 day
    mice and dual immunized mice,
    42 day antigen signal was scaled across all regions from 42 day mice and
    dual immunized mice.
    The number of plotted segments is shown above each boxplot.
    P values were calculated using a one-sided Wilcoxon rank sum test with
    Benjamini-Hochberg correction, p values <0.05 are shown.
D.  Normalized antigen tag signal is shown for regions that included both Lyve1+
    and CD11C+ segments.
    21 day antigen signal is shown for 21 day and dual immunized mice,
    42 day signal is shown for 42 day and dual immunized mice.
    The Spearman correlation coefficient is shown.    

```{r "xenium cell type sections", fig.width = 14, fig.height = 10}
# Format data for plots
xen_typs <- xen_dat$cell_type %>%
  table() %>%
  sort(decreasing = TRUE)

xen_typs <- xen_typs[xen_typs > 100]
xen_typs <- xen_typs[names(xen_typs) != "unassigned"]

# Function to set number of bins based on number of cells
fn <- approxfun(c(last(xen_typs), xen_typs[[1]]), c(45, 100))

# Create base plot to use for overlaying frequency data
base_xen <- xen_dat %>%
  plot_xenium(
    dat_clmn = NULL,
    grp_clmn = "slide",
    fill = "white"
  )

# Create plots
typ_plts <- names(xen_typs) %>%
  map(~ {
    typ <- .x
    
    typ_dat <- xen_dat %>%
      filter(cell_type == typ)
    
    ttl <- typ_dat %>%
      nrow() %>%
      label_comma()() %>%
      str_c(typ, " (n = ", ., ")")
    
    base_xen +
      geom_bin2d(
        aes(y_cell, x_cell),
        data = typ_dat,
        bins = floor(fn(xen_typs[[typ]]))
      ) +
      
      ggtitle(ttl) +
      guides(fill = guide_colorbar(ticks = FALSE)) +
      scale_fill_gradientn(
        # colours = c("white", xen_clrs[[typ]]),
        colours = c("white", xen_clrs[["LEC"]]),  # just use red for all slices
        name = "number of cells"
      ) +
      facet_wrap(~ slide, nrow = 1) +
      umap_theme_2 +
      theme(
        plot.title        = element_text(hjust = 0.5),
        legend.position   = "bottom",
        legend.key.height = unit(7, "pt"),
        legend.key.width  = unit(20, "pt"),
        legend.title      = element_text(size = 10, hjust = 0.5),
        legend.text       = element_text(size = 10),
        legend.title.position = "top"
      )
  })

typ_plts <- typ_plts %>%
  wrap_plots(ncol = 3)
```

```{r "xenium total Ag counts", fig.height = 6, fig.width = 20, include = FALSE}
# Format labels
xen_ag_clmns <- c(
  Ag_3wk_counts = "21 day",
  Ag_6wk_counts = "42 day"
)

xen_ag_clrs <- set_names(
  tm_clrs,
  str_c(names(tm_clrs), " day")
)

# Format bargraph data
xen_bar_dat <- xen %>%
  filter(!fov %in% bad_fov) %>%
  pivot_longer(all_of(names(xen_ag_clmns)), names_to = "Ag-tag") %>%
  group_by(slide, fov, mouse, `Ag-tag`) %>%
  mutate(
    tot_counts = sum(nCount_Xenium + nCount_Xenium_custom)
  ) %>%
  group_by(slide, fov, mouse, `Ag-tag`, tot_counts) %>%
  summarize(
    value   = sum(value),
    .groups = "drop"
  ) %>%
  group_by(slide, mouse, `Ag-tag`) %>%
  mutate(
    LN = str_c("LN-", seq_along(fov))
  ) %>%
  ungroup() %>%
  mutate(
    `Ag-tag` = xen_ag_clmns[`Ag-tag`],
    tm       = str_c(tm_key[mouse], " day"),
    tm       = replace_na(tm, "dual")
  )

# Create bargraph
xen_bars <- xen_bar_dat %>%
  ggplot(aes(LN, value, fill = `Ag-tag`)) +
  geom_col(position = position_dodge2()) +
  facet_grid(slide ~ tm, scales = "free_x", space = "free") +
  scale_fill_manual(values = xen_ag_clrs) +
  labs(y = "total Ag counts") +
  base_theme +
  theme(
    legend.position = "bottom",
    axis.title.x = element_blank()
  )
```

```{r "xenium Ag+ cell types", fig.width = 14, fig.height = 6}
bar_dat <- xen_dat %>%
  filter(cell_type %in% c(names(xen_typs), "unassigned")) %>%
  group_by(cell_type, fov, slide) %>%

  group_by(slide, fov, cell_type) %>%
  summarize(
    n = n_distinct(cell_id),
    n_ag = sum(Ag_class == "Ag+"),
    .groups = "drop"
  ) %>%
  filter(n_ag > 0) %>%
  mutate(
    pct_ag_pos = (n_ag / n),
    
    n_ag  = ifelse(n_ag > 1000, str_c(round(n_ag / 1000, 1), "k"), label_comma()(n_ag)),
    n     = ifelse(n > 1000, str_c(round(n / 1000, 1), "k"), label_comma()(n)),
    n_lab = str_c(n_ag, "/", n)
  )
    
# Create bargraphs
xen_typ_bars <- bar_dat %>%
  mutate(
    cell_type = fct_reorder(cell_type, pct_ag_pos),
    fov       = fct_reorder(fov, pct_ag_pos, .fun = max)
  ) %>%
  group_by(fov, slide) %>%
  
  ggplot(aes(pct_ag_pos, cell_type, fill = cell_type)) +
  geom_col() +
  geom_text(
    aes(label = n_lab),
    hjust = -0.25
  ) +
  facet_wrap(~ slide, nrow = 1) +
  scale_fill_manual(values = xen_clrs) +
  scale_x_continuous(
    labels = label_percent(),
    name   = "% Ag+ cells",
    expand = expansion(c(0.05, 0.3))
  ) +
  base_theme +
  theme(
    legend.position = "none",
    axis.title.y = element_blank()
  )
```

```{r "xenium supp figure", fig.width = 14, fig.height = 20}
bot <- wrap_plots(
  xen_bars + labs(tag = "B"),
  xen_typ_bars + labs(tag = "C"),
  nrow   = 1,
  widths = c(0.7, 1)
) &
  theme(plot.tag = element_text(face = "bold", size = ttl_pt2 * 1.5))

prox <- wrap_plots(
  prox_brs$`section 2`   + labs(tag = "D"),
  prox_hists$`section 2` + labs(tag = "E"),
  nrow = 1,
  widths = c(1, 1)
) &
  theme(plot.tag = element_text(face = "bold", size = ttl_pt2 * 1.5))

xen_supp_fig <- plot_grid(
  typ_plts, bot, prox,
  ncol = 1,
  rel_heights = c(1, 0.4, 0.4),
  labels = c("A", "", ""),
  label_size = ttl_pt2 * 1.5
)
```

```{r "geomx knit setup", include = FALSE}
knit(here(params$template_dir, "setup-geomx.Rmd"), "")
```

```{r "geomx uncropped images"}
# GeoMx arguments
img_args <- list(
  color = c("Sample", "Region", "Segment", "Ag-score"),
  clrs  = list(geo_m_clrs, reg_clrs, seg_clrs, c("white", "#D7301F")),
  labs  = list(geo_m_key, NULL, geo_seg_key, NULL),
  ttl   = list(NULL, NULL, "Segment type", "Ag signal")
)

# Plot uncropped images
all_img_fig <- img_args %>%
  pmap(~ {
    args <- list(...)
    
    res <- geomx_imgs %>%
      imap(~ {
        res <- .x %>%
          plot_geomx_image(
            color = args$color,
            clrs  = args$clrs,
            labs  = args$labs,
            ttl   = .y
          )
        
        # Only show legend for slide 1, since wrap_plots() does not correctly
        # 'collect' the legends
        if (!identical(.y, names(geomx_imgs[1]))) {
          res <- res +
            theme(legend.position = "none")
        }
        
        res
      }) %>%
      wrap_plots(
        guides = "collect",
        nrow = 1
      )
  })

# Create final figure
all_img_fig <- plot_grid(
  plotlist = all_img_fig[1:2],
  nrow     = 1,
  align    = "h",
  axis     = "tb",
  labels   = c("F", "G"),
  label_size = ttl_pt2 * 1.5
)
```

```{r "geomx Ag boxplots", fig.height = 4}
# Signals to plot
geo_clmns <- c(
  Ag_3wk_score = "Ag signal (21 day)",
  Ag_6wk_score = "Ag signal (42 day)"
)

# Function for strip labels
geo_lab_fn <- function(x) {
  keys <- list(geo_m_key, geo_seg_key)
  
  keys %>%
    walk(~ {
      idx    <- x %in% names(.x)
      x[idx] <- .x[x[idx]]
      x      <<- x
    })
  
  x
}

# Calculate relative Ag signal
# want to plot both Ag tags on same plot
# this is not useful without scaling values since Ag-3wk signal is much higher
# than Ag-6wk signal
bx_dat <- geo_clmns %>%
  imap_dfr(~ {
    sam <- str_extract(.y, "(?<=Ag_)[0-9]+wk")
    
    geomx_dat %>%
      filter(
        sample %in% c(sam, "6wk-3wk"),
        Segment %in% c("Lyve1", "CD11C")
      ) %>%
      mutate(
        key = .y,
        Ag_rel = as.numeric(scale(!!sym(.y))),
        Segment = fct_relevel(Segment, names(seg_clrs))
      )
  })

# Calculate p-values
# * one-sided wilcox test with BH correction
p_dat <- bx_dat %>%
  split(.$Segment) %>%
  imap_dfr(~ {
    mx <- max(.x$Ag_rel)
    mn <- min(.x$Ag_rel)
    
    pairwise.wilcox.test(
      .x$Ag_rel, .x$region_3,
      p.adjust.method = "none",
      alternative = "greater"
    ) %>%
      tidy() %>%
      mutate(
        Segment = .y,
        y_max   = mx,
        y_dif   = diff(c(mn, mx))
      )
  }) %>%
  mutate(
    p_adj = p.adjust(p.value, method = "BH"),
    y_max = max(y_max),
    y_dif = max(y_dif)
  ) %>%
  rowwise() %>%
  mutate(p_lab = .format_pvalue(p_adj))

p_dat <- p_dat %>%
  filter(p_adj < 0.05) %>%
  group_by(Segment) %>%
  
  arrange(
    desc(match(group1, names(reg_clrs))),
    desc(match(group2, names(reg_clrs)))
  ) %>%
  
  mutate(
    y = row_number(),
    y = y_max + (y_dif * (0.15 * y))
  ) %>%
  ungroup() %>%
  rowwise() %>%
  mutate(
    x1   = match(group1, names(reg_clrs)),
    x2   = match(group2, names(reg_clrs)),
    x    = min(c(x1, x2)),
    xend = max(c(x1, x2)),
    p_x  = x + ((xend - x) / 2)
  ) %>%
  ungroup() %>%
  mutate(Segment = fct_relevel(Segment, names(seg_clrs)))

# Calculate n values
n_dat <- bx_dat %>%
  group_by(region_3, Segment) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(n = str_c("n = ", n))

# Create boxplots
geo_bxs <- bx_dat %>%  
  ggplot(aes(region_3, Ag_rel, fill = region_3)) +
  geom_boxplot(
    linewidth = ln_pt * 1.5, alpha = 1, outlier.color = NA, width = 0.65,
    key_glyph = draw_key_point
  ) +
  geom_jitter(size = 2, width = 0.1) +
  geom_text(
    aes(region_3, Inf, label = n, fill = NULL),
    data  = n_dat,
    vjust = 1.4,
    size  = txt_pt2 / .pt
  ) +
  
  # Add p-values
  geom_segment(
    aes(x = x, xend = xend, y = y, fill = NULL),
    data = p_dat
  ) +
  geom_text(
    aes(x = p_x, y = y, fill = NULL, label = p_lab),
    data = p_dat,
    vjust = -0.2,
    hjust = 0.4,
    parse = TRUE
  ) +
  
  facet_grid(~ Segment, scales = "free_y", labeller = as_labeller(geo_lab_fn)) +
  scale_fill_manual(values = reg_clrs) +
  scale_y_continuous(expand = expansion(c(0.05, 0.2))) +
  labs(y = "Relative Ag signal") +
  base_theme +
  theme(
    aspect.ratio    = 1.4,
    legend.position = "none",
    axis.text.x     = element_text(size = ttl_pt2, angle = 45, hjust = 1),
    axis.title.y    = element_text(size = ttl_pt2),
    axis.title.x    = element_blank()
  )
```

```{r "geomx Ag scatter"}
# Format data
# * want to plot both Ag tags on same plot
sct_dat <- geo_clmns %>%
  imap_dfr(~ {
    sam <- str_extract(.y, "(?<=Ag_)[0-9]+wk")
    
    geomx_dat %>%
      filter(
        sample %in% c(sam, "6wk-3wk"),
        Segment %in% c("Lyve1", "CD11C")
      ) %>%
      mutate(
        key      = .y,
        Ag_score = !!sym(.y)
      )
  })

sct_dat <- sct_dat %>%
  dplyr::select(
    `ROI Coordinate X`, `ROI Coordinate Y`,
    region_3, Segment, sample, key,
    Ag_score
  ) %>%
  group_by(key, `ROI Coordinate X`, `ROI Coordinate Y`) %>%
  filter(all(c("Lyve1", "CD11C") %in% Segment)) %>%
  ungroup() %>%
  mutate(Segment = str_c(Segment, "+ Ag signal")) %>%
  pivot_wider(names_from = Segment, values_from = Ag_score) %>%
  mutate(
    tm = str_extract(key, "[0-9+]wk"),
    tm = as.character(tm_key[tm])
  )

# Legend labels
sct_labs <- sct_dat %>%
  group_by(key, tm) %>%
  summarize(n = str_c("n = ", n()), .groups = "drop") %>%
  mutate(n = str_c(geo_clmns[key], "\n", n))

sct_labs <- set_names(sct_labs$n, sct_labs$tm)

# Calculate correlation
cor <- cor(
  sct_dat$`CD11C+ Ag signal`,
  sct_dat$`Lyve1+ Ag signal`,
  method = "spearman"
)
cor <- round(cor, digits = 2)
cor = str_c("italic(r[s]) == ", cor)

geo_sct <- sct_dat %>%
  ggplot(aes(`CD11C+ Ag signal`, `Lyve1+ Ag signal`, color = tm)) +
  geom_smooth(method = "lm", color = "black", alpha = 0.1, linewidth = 0.5, linetype = 2) +
  geom_point(size = 3) +
  annotate(
    "text",
    x = min(sct_dat$`CD11C+ Ag signal`),
    y = max(sct_dat$`Lyve1+ Ag signal`),
    label = cor,
    parse = TRUE,
    hjust = -0.3,
    vjust = 0.5,
    color = "black",
    size  = ttl_pt1 / .pt,
  ) +
  scale_color_manual(values = tm_clrs, labels = sct_labs) +
  scale_x_continuous(transform = "log10") +
  scale_y_continuous(transform = "log10") +
  guides(color = guide_legend(override.aes = list(size = 4))) +
  base_theme +
  theme(
    aspect.ratio = 1,
    legend.title = element_blank(),
    legend.text  = element_text(size = ttl_pt2),
    axis.title   = element_text(size = ttl_pt2)
  )
```

```{r "Fig S7", fig.width = 14, fig.height = 27}
bot <- plot_grid(
  geo_bxs, geo_sct,
  nrow   = 1,
  align  = "h",
  axis   = "tb",
  labels = c("H", "I"),
  label_size = ttl_pt2 * 1.5
)

geo_fig <- plot_grid(
  all_img_fig, bot,
  ncol = 1,
  rel_heights = c(1, 0.7)
)

# Create final figure
figS7 <- plot_grid(
  xen_supp_fig, geo_fig,
  ncol = 1,
  rel_heights = c(1, 0.5)
)

figS7
```

<br>

<br>

## Figure 6

*Antigen archiving is impaired during CHIKV infection*

A.  UMAP projections show LEC subsets for mock and CHIKV-infected mice.
B.  UMAP projections show predicted Ag-competent cLECs for mock and
    CHIKV-infected mice.
C.  The fraction of predicted Ag-competent cLECs is shown for mock and
    CHIKV-infected mice for each biological replicate. P values were calculated
    using Fisher's exact test with Benjamini-Hochberg correction.
D.  UMAP projections show cLEC Ag-high module scores for mock and CHIKV-infected
    mice.
E.  cLEC Ag-high module scores are shown for mock and CHIKV-infected mice for
    each biological replicate.
    P values were calculated using a two-sided Wilcoxon rank sum test with
    Benjamini-Hochberg correction.
F.  Expression of the top 70 genes from the cLEC Ag-high gene module is shown
    for mock and CHIKV-infected mice for cLECs from three biological replicates.

```{r "chikv knit setup", include = FALSE}
knit(here(params$template_dir, "setup-chikv.Rmd"), "")
```

```{r "chikv subset UMAPs"}
# LEC subsets
top <- CHIKV_DAT %>%
  plot_scatter(
    CHIKV_TYPE_CLMN,
    x            = "hUMAP_1",
    y            = "hUMAP_2",
    group_col    = "treatment",
    plot_colors  = lec_clrs,
    size         = 0.75,
    label_params = list(size = ttl_pt2)
  ) +
  chikv_u_theme +
  theme(
    legend.title = element_blank(),
    legend.text  = element_text(size = 11.5),
    panel.border = element_blank()
  )
```

```{r "chikv ag class UMAPs"}
# Create UMAPs
u_args <- list(
  typ = set_names(rf_typs),
  pt_size = list(1, c(0.5, 0.1), c(0.5, 0.1))
)

chikv_u <- u_args %>%
  pmap(~ {
    CHIKV_DAT %>%
      create_rf_module_umaps(
        typ           = .x,
        ag_class_clmn = "ml_pred_1_grp",
        module_clmn   = str_c(.x, "_high"),
        module_title  = str_c(.x, " Ag-high module"),
        facet_var     = "treatment",
        lab_fn        = "label_value",
        class_clrs    = chikv_clrs,
        ag_clrs       = c("lightblue", "white", "#D7301F"),
        pt_size       = 0.75,
        circle_cells  = FALSE,
        show_scale_title = TRUE
      )
  })
```

```{r "chikv ag class bargraphs"}
# Format data for bargraphs
chikv_bar_dat <- CHIKV_DAT %>%
  filter(subtype %in% rf_typs) %>%
  mutate(subtype = fct_relevel(subtype, rf_typs))

# Create Ag class bargraphs
chikv_ag_bars <- pred_clmns %>%
  map(~ {
    chikv_bar_dat %>%
      filter(subtype == rf_cell_type) %>%
      create_chikv_class_bars(
        x         = "treatment",
        n_p       = n_distinct(chikv_bar_dat$subtype),
        ttl       = str_c("fraction Ag-competent ", rf_cell_type, "s"),
        pred_clmn = .x
      ) +
      theme(strip.text = element_blank())
  })
```

```{r "chikv ag module score boxplots"}
# Format data for plotting
chikv_bx_dat <- CHIKV_DAT %>%
  filter(subtype %in% rf_typs) %>%
  mutate(subtype = fct_relevel(subtype, rf_typs))

# Plot Ag module scores
chikv_ag_bxs <- chikv_bx_dat %>%
  filter(subtype == rf_cell_type) %>%
  create_chikv_module_boxes(
    module_clmns = "cLEC_high",
    ttl          = "cLEC Ag-high module",
    n_p          = length(ag_modules)
  ) +
  theme(strip.text = element_blank())
```

```{r "chikv ag-high heatmap"}
# Ag-high heatmap data
# * two genes are not in the CHIKV object
heat_dat <- CHIKV_DAT %>%
  filter(!!sym(CHIKV_TYPE_CLMN) %in% rf_cell_type) %>%
  pivot_longer(all_of(CHIKV_FEATS)) %>%
  group_by(!!sym(CHIKV_TYPE_CLMN), treatment, rep, name) %>%
  summarize(value = mean(value), .groups = "drop")

gn_lvls <- heat_dat %>%
  pivot_wider(names_from = treatment, values_from = value) %>%
  mutate(value = CHIKV / mock, .keep = "unused") %>%
  arrange(desc(value)) %>%
  pull(name) %>%
  unique()
  
heat_dat <- heat_dat %>%
  group_by(name) %>%
  mutate(value = as.numeric(scale(value))) %>%
  ungroup() %>%
  mutate(
    treatment = fct_relevel(treatment, c("mock", "CHIKV")),
    name      = fct_relevel(name, rev(gn_lvls))
  )

# CHIKV heatmap
heat <- heat_dat %>%
  ggplot(aes(name, rep, fill = value)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("#56B4E9", "white", "#D7301F")) +
  facet_wrap(~ treatment, ncol = 1, strip.position = "left") +
  guides(fill = guide_colorbar(
    barwidth = unit(150, "pt"),
    barheight = unit(6, "pt")
  )) +
  base_theme +
  theme(
    aspect.ratio    = length(rf_typs) / length(CHIKV_FEATS),
    strip.placement = "outside",
    strip.clip      = "off",
    legend.position = "bottom",
    legend.title    = element_blank(),
    axis.title      = element_blank(),
    axis.text.x     = element_text(angle = 90, hjust = 1, vjust = 0.5)
  )
```

```{r "cormac figure", eval = FALSE}
# DROPPED FROM PAPER

img_g <- readPNG(here("results/figures/morrison_chikv/fig5g.png"))
img_h <- readPNG(here("results/figures/morrison_chikv/fig5h.png"))

fig5gh <- list(img_g, img_h) %>%
  map(~ {
    img <- rasterGrob(.x, interpolate = TRUE)
    
    # Adjust plot limits if necessary
    img_lims <- tibble(x = c(0, 1), y = c(0, 1))
    
    # Create an empty plot for placing the image, adjusted to fill the area
    ggplot(img_lims, aes(x, y)) + 
      annotation_custom(
        grob = img,
        xmin = -Inf, xmax = Inf,
        ymin = -Inf, ymax = Inf
      ) +
      xlim(c(-Inf, Inf)) +
      ylim(c(-Inf, Inf)) +
      theme_void()
  }) %>%
  plot_grid(
    plotlist = .,
    nrow     = 1,
    align    = "h",
    axis     = "tb",
    labels   = c("G", "H"),
    label_size = ttl_pt2 * 1.5,
    rel_widths = c(1, 0.7)
  )
```

```{r "cormac data", eval = FALSE}
# Format data for plot
tbl_nm <- "ova+ LECs (%)"

pzfx_dat <- read_pzfx(
  here("results/figures/morrison_chikv/TEM2306 results.pzfx"),
  table = tbl_nm
)

pzfx_lvls <- colnames(pzfx_dat)

pzfx_clrs <- set_names(
  c("#56B4E9", "#E69F00", "#D7301F"),
  pzfx_lvls
)

pzfx_dat <- pzfx_dat %>%
  as_tibble() %>%
  filter(if_any(everything(), ~ !is.na(.x))) %>%
  mutate(mouse = row_number()) %>%
  pivot_longer(-mouse) %>%
  filter(!is.na(value)) %>%
  mutate(name = fct_relevel(name, pzfx_lvls))

# Add significance labels based on Cormac's original plot
pzfx_sig <- list(
  `CHIKV 181/25` = "*",
  `polyI:C/ova`  = "**"
)

p_dat <- pzfx_dat %>%
  pull(name) %>%
  unique() %>%
  expand.grid(., .) %>%
  filter(
    Var1 != Var2,
    Var1 == last(pzfx_lvls),
  ) %>%
  mutate(Var2 = fct_relevel(Var2, pzfx_lvls)) %>%
  arrange(desc(Var2)) %>%
  mutate(
    sig  = pzfx_sig[as.character(Var2)],
    y    = 110 + (10 * (row_number() - 1)),
    x    = match(Var2, pzfx_lvls),
    x    = x + ((length(pzfx_lvls) - x) / 2)
  )

# Format data for bargraphs
bar_dat <- pzfx_dat %>%
  group_by(name) %>%
  summarize(value = median(value), .groups = "drop")

# Create bargraphs
pzfx_plt <- pzfx_dat %>%
  ggplot(aes(name, value, fill = name)) +
  geom_col(data = bar_dat, width = 0.75) +
  geom_point(size = 3, show.legend = FALSE) +
  
  geom_segment(
    aes(x = Var1, xend = Var2, y = y, yend = y, fill = NULL),
    data = p_dat
  ) +
  geom_text(
    aes(x = x, y = y, label = sig, fill = NULL),
    data = p_dat,
    size = ttl_pt2 * 1.5 / .pt
  ) +
  
  scale_fill_manual(values = pzfx_clrs) +
  scale_y_continuous(expand = expansion(c(0.01, 0.05))) +
  labs(y = tbl_nm) +
  base_theme +
  theme(
    plot.margin     = margin(20, 0, 0, 0),
    aspect.ratio    = 1.3,
    legend.position = "none",
    axis.title.x    = element_blank(),
    axis.text.x     = element_text(size = ttl_pt2, angle = 30, hjust = 1)
  )
```

```{r "Fig 5", fig.width = 12, fig.height = 18}
# Combine and annotate center panels
c("D", "B") %>%
  iwalk(~ {
    chikv_u$cLEC[[.y]][[1]] <<- chikv_u$cLEC[[.y]][[1]] + labs(tag = .x)
  })

chikv_u$cLEC[[1]] <- chikv_u$cLEC[[1]] &
  theme(
    legend.position   = "left",
    legend.title      = element_text(size = ttl_pt2, hjust = 0.5, angle = 90),
    legend.key.height = unit(30, "pt"),
    legend.key.width  = unit(6, "pt")
  )

mid <- wrap_plots(
  chikv_u$cLEC[[2]],
  chikv_ag_bars[[1]] + labs(tag = "C"),
  chikv_u$cLEC[[1]],
  chikv_ag_bxs + labs(tag = "E"),
  widths = c(1, 0.3)
) &
  theme(plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"))

# Final figure
fig5 <- plot_grid(
  top, mid, heat,
  ncol       = 1,
  labels     = c("A", "", "F"),
  label_size = ttl_pt2 * 1.5,
  hjust      = c(0, 0, -0.5),
  vjust      = c(1.5, 1, 0),
  rel_heights = c(1, 2.5, 1)
)

fig5
```

<br>

<br>

### Figure S9

A.  Ag-high and Ag-low module scores are shown for the predicted Ag classes.
    P values were calculated using a two-sided Wilcoxon rank sum test with
    Benjamini-Hochberg correction.
B.  Ag-high and Ag-low module scores are shown for mock and CHIKV-infected mice
    for each biological replicate. P values were calculated using a two-sided
    Wilcoxon rank sum test with Benjamini-Hochberg correction.
C.  The total fraction of cLECs, fLECs, and collecting LECs predicted to be
    Ag-competent is shown for mock and CHIKV-infected mice for two sets of
    predictions made using independent models.
    P values were calculated using Fisher's exact test with Benjamini-Hochberg
    correction.
D.  The fraction of cLECs, fLECs, and collecting LECs predicted to be
    Ag-competent is shown as described in C.

```{r "CHIKV NEW BARGRAPHS", eval = FALSE}
# Format data for bargraphs
icam_dat <- CHIKV_DAT %>%
  filter(subtype %in% rf_typs) %>%
  mutate(subtype = icam_grp)

# Create bargraphs
.adjust_axis_theme <- function(plt, idx) {
  if (idx == 1) {
    plt <- plt +
      theme(
        axis.text.x  = element_blank(),
        axis.ticks.x = element_blank()
      )
  } else {
    plt <- plt +
      theme(strip.text = element_blank())
  }
  
  plt
}

icam_bars <- pred_clmns %>%
  imap(~ {
    plt <- icam_dat %>%
      create_chikv_class_bars(
        x         = "treatment",
        n_p       = n_distinct(chikv_bar_dat$subtype),
        ttl       = str_c("model ", .y, "\nfraction Ag-competent"),
        pred_clmn = .x
      ) %>%
      .adjust_axis_theme(.y)
  }) %>%
  wrap_plots(ncol = 1)
```

```{r "Fig S8", fig.width = 13.5, fig.height = 15}
# Plot Ag module scores for predicted classes
all_chikv_bxs_1 <- chikv_bx_dat %>%
  create_chikv_module_boxes(
    module_clmns = ag_modules,
    x            = "ml_pred_1_grp",
    clrs         = pred_clrs,
    ttl          = "module score"
  ) +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))

# Plot Ag module scores for treatment groups for all cell types
all_chikv_bxs_2 <- chikv_bx_dat %>%
  create_chikv_module_boxes(
    module_clmns = ag_modules,
    ttl = "module score"
  )

# Bargraphs for combined LEC subsets
.adjust_axis_theme <- function(plt, idx) {
  if (idx == 1) {
    plt <- plt +
      theme(
        axis.text.x  = element_blank(),
        axis.ticks.x = element_blank()
      )
  } else {
    plt <- plt +
      theme(strip.text = element_blank())
  }
  
  plt
}

all_chikv_bars <- pred_clmns %>%
  imap(~ {
    chikv_bar_dat %>%
      mutate(subtype = "cLEC, fLEC, Collecting") %>%
      create_chikv_class_bars(
        x         = "treatment",
        n_p       = n_distinct(chikv_bar_dat$subtype),
        ttl       = str_c("model ", .y, "\nfraction Ag-competent"),
        pred_clmn = .x
      ) %>%
      .adjust_axis_theme(.y)
  }) %>%
  wrap_plots(ncol = 1)

# Bargraphs for other LEC subtypes
other_chikv_bars <- pred_clmns %>%
  imap(~ {
    plt <- chikv_bar_dat %>%
      create_chikv_class_bars(
        x         = "treatment",
        n_p       = n_distinct(chikv_bar_dat$subtype),
        ttl       = str_c("model ", .y, "\nfraction Ag-competent"),
        pred_clmn = .x
      ) %>%
      .adjust_axis_theme(.y)
  }) %>%
  wrap_plots(ncol = 1)

# Create final figure
figS8 <- plot_grid(
  all_chikv_bxs_1, all_chikv_bxs_2,
  all_chikv_bars,  other_chikv_bars,
  labels = c("A", "B", "C", "D"),
  label_size = ttl_pt2 * 1.5,
  nrow  = 2,
  align = "vh",
  axis  = "trbl",
  rel_heights = c(1, 1)
)

figS8
```

<br>

<br>

### Figure S10

A.  Mean expression is shown for LEC marker genes for
    metastasis-free control lymph node (MFLN) and follicular lymphoma (FL)
    samples from Abe et al.
B.  The fraction of cLECs predicted to be Ag-competent is shown for each 
    MFLN and FL sample with at least 20 identified cLECs (left).
    P values were calculated using a two-sided Wilcoxon rank sum test with
    Benjamini-Hochberg correction (middle).
    The fraction of predicted Ag-competent cells is shown for a second
    independent model (right).
C.  The fraction of fLECs predicted to be Ag-competent is shown as in A.
D.  The fraction of collecting LECs predicated to be Ag-compentent is shown as
    in A.

```{r "abe knit setup", include = FALSE}
knit(here(params$template_dir, "setup-abe.Rmd"), "")
```

```{r "abe marker expression"}
# Format marker data
abe_gns <- c(
  "ACKR4",
  "MFAP4",
  "SERPINE1",
  "BMP2",
  "CXCL5",
  "MARCO",
  "PTX3",
  "CLDN11",
  "GJA4"
)

tissue_nms <- c(
  "MFLN" = "Metastasis-free LN",
  "FL"   = "Follicular Lymphoma"
)

abe_gn_dat <- abe_lec_so %>%
  FetchData(c("tissue", "orig.ident", "subtype", abe_gns)) %>%
  as_tibble(rownames = "cell_id") %>%
  filter(subtype != "unassigned") %>%
  pivot_longer(all_of(abe_gns)) %>%
  mutate(
    name   = fct_relevel(name, abe_gns),
    tissue = fct_relevel(tissue, names(tissue_nms))
  )

# Create heatmap
abe_heat_dat <- abe_gn_dat %>%
  group_by(tissue, subtype, name) %>%
  summarize(
    n       = n_distinct(cell_id),
    value   = mean(value),
    .groups = "drop"
  ) %>%
  group_by(tissue, name) %>%
  mutate(value = as.numeric(scale(value))) %>%
  mutate(
    name  = fct_relevel(name, rev(abe_gns)),
    n_lab = str_c(subtype, "\nn = ", label_comma()(n))
  )

abe_gn_heat <- abe_heat_dat %>%
  ggplot(aes(n_lab, name, fill = value)) +
  geom_tile() +
  facet_wrap(~ tissue, nrow = 1, scales = "free_x", labeller = as_labeller(tissue_nms)) +
  scale_fill_gradientn(colours = c("white", abe_clrs[["FL"]])) +
  guides(fill = guide_colorbar(ticks = FALSE, title = "z-score")) +
  base_theme +
  theme(
    aspect.ratio = length(abe_gns) / n_distinct(abe_gn_dat$subtype) * 0.65,
    legend.key.width = unit(7, "pt"),
    axis.title   = element_blank(),
    axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5)
  )
```

```{r "abe frac Ag-competent", include = FALSE}
# Format Abe data
# * cLECs are the LEC1 population
# * only include samples with at least 20 cLECs
abe_bar_dat <- pred_clmns %>%
  set_names() %>%
  map(~ {
    abe_lec_so@meta.data %>%
      as_tibble(rownames = "cell_id") %>%
      filter(subtype %in% rf_typs) %>%
      mutate(
        tissue = fct_relevel(tissue, names(abe_clrs)),
        sample = str_remove(sample, "_NHC$")
      ) %>%
      group_by(sample, tissue, subtype) %>%
      mutate(n = n_distinct(cell_id)) %>%
      filter(n > 20) %>%
      group_by(sample, tissue, subtype, n) %>%
      summarize(
        n_high = sum(!!sym(.x) == "Ag-competent"),
        .groups = "drop"
      ) %>%
      mutate(
        frac_high = n_high / n,
        n_lab     = str_c(sample, "\nn = ", label_comma()(n))
      )
  })

# Create full bargraphs
abe_bars_1 <- abe_bar_dat %>%
  unname() %>%
  imap(~ {
    dat       <- .x
    mod_idx   <- .y
    pred_clmn <- pred_clmns[.y]
    
    brs <- rf_typs %>%
      imap(~ {
        abe_type <- .x
        
        y_ttl <- str_c(abe_type, " - model ", mod_idx, "\nfraction Ag-competent")
        
        dat <- dat %>%
          filter(subtype == abe_type)
        
        # Create bar graph for all samples
        dat %>%
          ggplot(aes(n_lab, frac_high, fill = tissue)) +
          geom_col() +
          facet_grid(
            ~ tissue,
            scales = "free_x",
            space = "free_x",
            labeller = as_labeller(tissue_nms)
          ) +
          scale_fill_manual(values = abe_clrs) +
          scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(0, 1)) +
          labs(y = y_ttl, tag = LETTERS[-1][.y]) +
          base_theme +
          theme(
            legend.position = "none",
            plot.tag        = element_text(size = ttl_pt2 * 1.5, face = "bold"),
            strip.clip      = "off",
            axis.title.x    = element_blank(),
            axis.text.x     = element_text(angle = 90, hjust = 1, vjust = 0.5)
          )
      }) %>%
      wrap_plots(ncol = 1)
    
    brs
  })
    
# Create bargraphs for combined samples
# * use wilcox test instead of fisher's exact test since there is a large
#   difference in the number of cells for each sample
# * this will cause one sample to disproportionately influence the results
abe_bars_2 <- abe_bar_dat %>%
  unname() %>%
  imap(~ {
    dat     <- .x
    mod_idx <- .y
    
    brs <- set_names(rf_typs) %>%
      map(~ {
        abe_type <- .x
        
        y_ttl <- str_c(abe_type, " - model ", mod_idx, "\nfraction Ag-competent")
        
        dat <- dat %>%
          filter(subtype == abe_type)
        
        p_dat <- dat %>%
          mutate(
            p = wilcox.test(
              frac_high[tissue == "FL"],
              frac_high[tissue == "MFLN"]
            )$p.value
          ) %>%
          group_by(tissue, p) %>%
          summarize(n = n(), .groups = "drop") %>%
          mutate(p_adj = p.adjust(p, method = "BH", n = length(rf_typs))) %>%
          rowwise() %>%
          mutate(
            p_lab = str_c("italic(p) == ", .format_pvalue(p_adj)),
            n_lab = str_c(tissue, "\nn = ", n)
          )
        
        n_labs <- set_names(p_dat$n_lab, p_dat$tissue)
        
        set.seed(1)
        
        p <- dat %>%
          group_by(tissue) %>%
          summarize(frac_high = median(frac_high), .groups = "drop") %>%
          ggplot(aes(tissue, frac_high, fill = tissue)) +
          geom_col() +
          geom_jitter(data = dat, size = 1.5, width = 0.03) +
          geom_text(
            aes(1.5, Inf),
            label = unique(p_dat$p_lab),
            parse = TRUE,
            size  = 12 / .pt,
            vjust = 1.25,
          ) +
          scale_fill_manual(values = abe_clrs) +
          scale_y_continuous(
            breaks = c(0, 0.5, 1),
            limits = c(0, 1),
            expand = expansion(c(0.05, 0.2))
          ) +
          scale_x_discrete(labels = n_labs) +
          labs(y = y_ttl) +
          base_theme +
          theme(
            aspect.ratio    = 1.75,
            legend.position = "none",
            axis.title.x    = element_blank()
          )
        
        p
      }) %>%
      wrap_plots(ncol = 1)
    
    brs
  })

# Create final figure
abe_bars <- wrap_plots(
  append(list(abe_bars_1[[1]]), abe_bars_2),
  widths = c(1, 0.3, 0.3)
)

rm(abe_lec_so)
gc()
```

```{r "Fig S9", fig.width = 13, fig.height = 12}
plot_grid(
  abe_gn_heat,
  abe_bars,
  ncol        = 1,
  rel_heights = c(0.4, 1),
  labels      = c("A", ""),
  label_size  = ttl_pt2 * 1.5
)
```

---

<br>

<br>

`r knit_child("reviews.Rmd")`

```{r "cleanup"}
# `r knit_child("cell-type-annotations.Rmd")`
h2o.shutdown(prompt = FALSE)
```

## Session info

```{r "session info"}
sessionInfo()
```

```{r "matrices for geo", eval = FALSE}
# Metadata columns to include
clmns_1 <- c(
  "cell_id",
  "exp",
  "sample",
  "mouse",
  "tm",
  "vaccination",
  "RNA_snn_res.5",
  "RNA_snn_res.10",
  "cell_type",
  "subtype",
  "Ag_score",
  "Ag_score_3",
  "Ag_score_6",
  "Ag_class_ml",
  "Ag_class_dual",
  "hUMAP_1",
  "hUMAP_2"
)

clmns_2 <- clmns_1[clmns_1 != "subtype"]

# Write files for GEO
geo_objs <- list(
  obj   = c("cd45neg", "cd45pos", "lec", "dc"),
  clmns = list(clmns_2, clmns_2, clmns_1, clmns_1)
)

geo_objs %>%
  pwalk(~ {
    prfx      <- str_c(.x, "_")
    geo_files <- str_c(prfx, c("count_matrix.h5", "metadata.tsv.gz"))
    geo_files <- here(params$geo_dir, geo_files)
    
    # Write matrices and meta.data
    if (all(file.exists(geo_files))) return(NULL)
    
    obj <- qread(here(params$so_dir, str_c(prfx, "so.qs")))
    
    obj %>%
      export_matrices(
        assays      = c("RNA", "ADT"),
        feat_type   = c("Gene Expression", "Antibody Capture"),
        columns     = .y,
        out_dir     = params$geo_dir,
        file_prefix = prfx
      )
  })
```

```{r "xenium metadata for geo", eval = FALSE}
clmns <- c(
  "cell_id",
  "slide",
  "fov",
  "mouse",
  "tm",
  "nCount_Xenium",
  "nFeature_Xenium",
  "UMAP_1", "UMAP_2",
  "cell_type",
  "Ag_class",
  "x_cell", "y_cell",
  "lec_dist",
  "Ag_3wk_counts",
  "Ag_6wk_counts"
)

metadat <- xen %>%
  mutate(
    tm = str_c(tm_key[mouse], " day"),
    tm = replace_na(tm, "dual")
  ) %>%
  select(all_of(clmns))

metadat %>%
  write_tsv(here(params$geo_dir, "xenium_metadata.tsv.gz"))
```

```{r "geomx metadata for geo", eval = FALSE}
clmns <- c(
  "dcc",
  "Slide Name",
  "slide",
  "sample",
  "LN",
  "rep",
  "Aoi",
  "Segment",
  "Area",
  "Nuclei",
  "region_3",
  "UMAP_1", "UMAP_2",
  "Ag_3wk_score",
  "Ag_6wk_score"
)

metadat <- geomx_dat %>%
  select(all_of(clmns))

metadat %>%
  write_tsv(here(params$geo_dir, "geomx_metadata.tsv.gz"))
```
