---
title: "Antigen Tracking Figures"
author: "Ryan Sheridan"
date: "`r Sys.Date()`"
output:
  html_document:
    toc:            true
    toc_float:      true
    toc_depth:      4
    df_print:       "paged"
    code_folding:   "hide"
    self_contained: true
    highlight:      "kate"
params:
  template_dir:  "src"                                                     # template Rmarkdowns
  so_dir:        "~/Dropbox/Ryan/Projects/antigen-exchange/results/sobjs"  # Seurat objects
  mod_dir:       "~/Dropbox/Ryan/Projects/antigen-exchange/results/models/2023-11-01" # ML models
  chikv_dir:     "~/Dropbox/Ryan/Projects/morrison-lnsc/results/sobjs"     # CHIKV objects
  ref_dir:       "ref"                                                     # clustifyr references
  sample_info:   "sample_info.xlsx"
  res_dir:                                                
    value:
      exp-0: "~/Projects/antigen-tracking/results"
      exp-1: "~/Projects/tamburini-antigen-tracking/results/2022-03-11"
      exp-2: "~/Projects/tamburini-antigen-tracking/results/2022-10-28"
  chikv_so: "~/Dropbox/Ryan/Projects/morrison-lnsc/results/sobjs/so_merge.qs"
editor_options: 
  chunk_output_type: inline
---

```{r "setup", include = FALSE}
# Default chunk options
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo    = FALSE,
  dpi     = 300
)

knitr::knit(here::here(params$template_dir, "setup.Rmd"), "")
knitr::knit(here::here(params$template_dir, "setup-ml.Rmd"), "")

# Plot parameters
min_cells <- 5
rf_typs   <- c("cLEC", "fLEC", "Collecting")
rf_tms    <- c(14, 21, 42)

ag_modules <- rf_typs %>%
  map(str_c, c("_high", "_low")) %>%
  unlist()

# Keys for naming labels
ag_clmns <- c("Ag_3wk_score", "Ag_6wk_score")

ag_labs <- set_names(
  c("Ag-score\n21 day", "Ag-score\n42 day"),
  ag_clmns
)

# Mouse key
m_key <- c(
  d2        = "2 day",
  d14       = "14 day",
  `3wk`     = "21 day",
  `6wk`     = "42 day",
  `6wk-3wk` = "dual"
)

# Experiment to show in main figures
ag_exps <- list(LEC = "exp-1", DC = "exp-2")

# Objects to use for each cell type
ag_objs <- list(LEC = lec_so, DC = dc_so)
```

---

<br>

```{r "CD45- heatmaps", fig.width = 30, fig.height = 20, eval = FALSE}
# Function to create marker heatmaps
create_marker_heatmap <- function(so_in, markers) {
  
  gns <- markers %>%
    pull(gene) %>%
    as.character()
  
  clmns <- c("sample", "mouse", "subtype", "cell_type", gns)
  
  dat <- so_in %>%
    FetchData(clmns) %>%
    as_tibble(rownames = ".cell_id") %>%
    pivot_longer(any_of(gns), names_to = "gene") %>%
    left_join(markers, by = "gene") %>%
    
    group_by(subtype, mouse, gene, label) %>%
    summarize(
      pct_expr = sum(value > 0) / n(),
      pct_expr = as.character(round(pct_expr, 2)),
      value    = mean(value),
      .groups  = "drop"
    ) %>%
    group_by(gene, mouse) %>%
    mutate(value = as.numeric(scale(value))) %>%
    ungroup() %>%
    mutate(
      subtype = fct_relevel(subtype, levels(markers$annotation)),
      gene    = fct_relevel(gene, rev(levels(markers$gene)))
    )
  
  res <- dat %>%
    ggplot(aes(subtype, gene, fill = value)) +
    geom_tile() +
    geom_text(aes(label = pct_expr), size = 6 / .pt) +
    facet_grid(label ~ mouse, scales = "free_y", space = "free_y") +
    scale_fill_gradientn(colours = c("lightblue", "white", "red")) +
    base_theme +
    theme(
      strip.clip   = "off",
      strip.text.y = element_text(angle = 0, hjust = 0),
      axis.title   = element_blank(),
      axis.text.x  = element_text(angle = 45, hjust = 1)
    )
  
  res
}

# Load data
rm(lec_so, dc_so)

cd45neg_so <- qread(here(params$so_dir, "cd45neg_so.qs"))
cd45pos_so <- qread(here(params$so_dir, "cd45pos_so.qs"))

# Load cell type markers
xlsx_path <- here(params$ref_dir, "cell-type-markers.xlsx")
markers   <- xlsx::read.xlsx(xlsx_path, sheetName = "markers")

markers <- markers %>%
  mutate(
    cell_type = fct_inorder(cell_type),
    subtype   = fct_inorder(subtype)
  ) %>%
  group_by(gene) %>%
  summarize(
    label      = str_c(cell_type, collapse = "/"),
    cell_type  = dplyr::first(cell_type),
    annotation = dplyr::first(subtype),
    .groups    = "drop"
  ) %>%
  arrange(cell_type, annotation) %>%
  mutate(
    gene  = fct_inorder(gene),
    label = fct_inorder(label)
  )

# CD45- heatmaps
create_marker_heatmap(cd45neg_so, markers)
```

```{r "CD45+ heatmaps", fig.width = 25, fig.height = 20, eval = FALSE}
create_marker_heatmap(cd45pos_so, markers)

# Load data for other figures
lec_so <- qread(here(params$so_dir, "lec_so.qs"))
dc_so  <- qread(here(params$so_dir, "dc_so.qs"))

rm(cd45neg_so, cd45pos_so)
```

```{r "MARKER PLOTS", eval = FALSE}
cd45neg_so %>%
  plot_scatter(
    data_col = "cell_type",
    x = "hUMAP_1",
    y = "hUMAP_2",
    group_col = "mouse",
    size = 0.1
  )

cd45pos_so %>%
  plot_violin(
    "Fcer1g",
    "cell_type",
    "mouse",
    method = "boxplot"
  )

cd45pos_so %>%
  plot_scatter(
    data_col = "Ceacam8",
    x = "hUMAP_1",
    y = "hUMAP_2",
    group_col = "mouse",
    size = 0.1,
    plot_colors = c("lightblue", "white", "red")
  )

cd45pos_so %>%
  plot_scatter(
    data_col = "Zbtb46",
    x = "hUMAP_1",
    y = "hUMAP_2",
    group_col = "mouse",
    size = 0.1,
    plot_colors = c("lightblue", "white", "red")
  )
```

## Figure 1

Antigen persists in discrete cell populations within the lymph node

A.  Experimental design
B.  Mean Ag-score is shown for each cell type for 2, 14, 21, and 42 days post
    vaccination.
A.  UMAP projection shows LEC subsets.
B.  UMAP projections show Ag-scores for LEC subsets for each timepoint.
C.  Mean Ag-score is shown for LEC subsets for each timepoint.
D.  Ag-scores are shown for each timepoint for each LEC subset.
E.  UMAP projection shows DC subsets.
F.  UMAP projections show Ag-scores for DC subsets for each timepoint.
G.  Mean Ag-score is shown for DC subsets for each timepoint.
H.  Ag-scores are shown for each timepoint for each DC subset.

<u>Notes</u>

* Experiment 1 is shown for the LEC 21 and 42 day timepoints. Experiment 2 data
  include very few LECs for the 21 day timepoint.
* Experiment 2 is shown for the DC 21 and 42 day timepoints. Experiment 1 data
  are lower quality (sparse signal, mostly zeros).
* Subsets are only shown if there are >= `r min_cells` cells for each timepoint
* CD45+ and CD45- cell types are only shown for the CD45+ and CD45- datasets,
  respectively.

```{r "load data"}
# Format prediction labels
# make sure the correct predictions and module scores are being used
# once the models are finalized, this information can be saved in the meta.data
# training data was removed before predicting
lec_so <- lec_so %>%
  AddMetaData(rf_preds, col.name = str_c("rf_", colnames(rf_preds)))

lec_so <- lec_so %>%
  mutate_meta(
    mutate,
    pred_grp = ifelse(
      rf_pred == "high" & !rf_correct,
      "Ag-competent", str_c("Ag-", !!sym(rf_dat_clmn))
    ),
    rf_pred       = str_c("Ag-", rf_pred),
    mouse         = fct_relevel(mouse, mice),
    training_data = subtype %in% rf_mods$subtype & is.na(rf_pred)
  )

# Add module scores
rf_feats %>%
  iwalk(~ {
    clmn_nms <- str_c(.y, "_", names(.x))

    clmn_nms <- set_names(
      str_c(clmn_nms, seq_along(clmn_nms)),
      clmn_nms
    )

    lec_so <<- lec_so %>%
      mutate_meta(dplyr::select, -any_of(names(clmn_nms))) %>%
      AddModuleScore(
        features = .x,
        name     = names(clmn_nms)
      ) %>%
      mutate_meta(dplyr::rename, !!!syms(clmn_nms))
  })
```

```{r "type frequency"}
# Format data for plotting
typ_dat <- all_meta %>%
  filter(
    vaccination == "single",
    cell_type != "unassigned"
  ) %>%
  group_by(cell_sort) %>%
  mutate(
    tm        = fct_relevel(as.character(tm), tms),
    cell_type = fct_infreq(cell_type),
    cell_type = fct_relevel(cell_type, c("LEC", "Fibroblasts", "DC"))
  ) %>%
  ungroup() %>%
  arrange(tm, cell_type)

freq_dat <- typ_dat %>%
  mutate(
    tm_group = str_c(tm, "_", cell_sort),
    tm_group = fct_inorder(tm_group)
  )

# Create bargraphs
# these results are not very convincing since CD45 sorting varied significantly
# between our two experiments
group_lab <- c(CD45pos = "CD45+", CD45neg = "CD45-")

x_labs <- get_nlab_fun(freq_dat, "tm_group", l = "\n", r = "")
x_labs <- x_labs() %>%
  map_chr(~ str_remove_all(.x, "_.+"))

typ_freq <- freq_dat %>%
  ggplot(aes(tm_group, fill = cell_type, color = cell_type)) +
  geom_bar(position = "fill", key_glyph = draw_key_point) +
  facet_wrap(~ cell_sort, nrow = 1, scales = "free_x", labeller = as_labeller(group_lab)) +
  scale_fill_manual(values = typ_clrs) +
  scale_color_manual(values = typ_clrs) +
  scale_x_discrete(labels = x_labs) +
  scale_y_continuous(expand = expansion(c(0.01, 0.05))) +
  guides(fill = guide_legend(ncol = 2, override.aes = list(shape = 19, size = 4))) +
  labs(x = "days post vaccination") +
  base_theme +
  theme(
    aspect.ratio = 1.5,
    legend.title = element_blank(),
    panel.border = element_blank(),
    axis.title.y = element_blank(),
    axis.text.x  = element_text(size = txt_pt2, angle = 45, hjust = 1, vjust = 1),
    axis.text.y  = element_blank(),
    axis.ticks   = element_blank()
  )
```

```{r "type Ag heat", fig.width = 16, fig.height = 5}
# Create heatmap summarize Ag scores
# exclude CD45+ cell types from CD45- datasets and vice versa
cd45neg_typs <- c("LEC", "BEC", "Fibroblasts", "Epithelial cells")

heat_dat <- typ_dat %>%
  filter(
    (cell_sort == "CD45neg" & cell_type %in% cd45neg_typs) |
      (cell_sort == "CD45pos" & !cell_type %in% cd45neg_typs)
  ) %>%
  
  group_by(tm, cell_type) %>%
  summarize(
    Ag_score = mean(Ag_score), .groups = "drop",
    n        = n()
  ) %>%
  group_by(cell_type) %>%
  mutate(cell_type = str_c(cell_type, "\nn = ", label_comma()(sum(n)))) %>%
  ungroup() %>%
  mutate(
    tm = fct_relevel(as.character(tm), rev(as.character(tms))),
    cell_type = fct_reorder(cell_type, Ag_score, mean, .desc = TRUE)
  )

typ_heat <- heat_dat %>%
  ggplot(aes(cell_type, tm, fill = Ag_score)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("white", "#D55E00"), na.value = "black") +
  guides(fill = guide_colorbar(
    title.position = "top", ticks = FALSE, title = "Ag-score",
    barwidth = unit(150, "pt"), barheight = unit(10, "pt")
  )) +
  labs(y = "days post vaccination") +
  base_theme +
  theme(
    plot.margin  = margin(0, 15, 0, 15),
    legend.position = "top",
    legend.justification = 0,
    legend.text  = element_text(size = txt_pt1),
    aspect.ratio = n_distinct(heat_dat$tm) / n_distinct(heat_dat$cell_type) * 0.9,
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1),
  )
```

```{r "subtype Ag UMAPs"}
# Data for UMAPs
# show exp1 for LECs and exp2 for DCs
fig1_dat <- ag_objs %>%
  imap(~ {
    .x@meta.data %>%
      as_tibble(rownames = ".cell_id") %>%
      filter(
        exp %in% c("exp-0", ag_exps[[.y]]),
        mouse != "6wk-3wk"
      ) %>%
      mutate(tm_lab = factor(as.character(tm), tms))    
  })

# Subtype UMAPs
fig1_typ_u <- fig1_dat %>%
  imap(~ {
    typ_lvls <- .x %>%
      pull(subtype) %>%
      unique() %>%
      as.character()
    
    res <- .x %>%
      plot_scatter(
        "subtype", "hUMAP_1", "hUMAP_2",
        size         = 0.1,
        plot_colors  = c(lec_clrs, dc_clrs),
        plot_lvls    = rev(typ_lvls),
        label_params = list(size = ttl_pt1),
        panel_nrow   = 2
      ) +
      guides(color = guide_legend(ncol = 1, override.aes = list(size = 4))) +
      umap_theme +
      theme(
        aspect.ratio = 0.9,
        legend.position = "right",
        legend.title = element_blank(),
        legend.text  = element_text(size = txt_pt2)
      )
    
    res
  })

# Ag-score UMAPs
u_clrs <- list(
  LEC = c("lightblue", "white", "#D7301F"),
  DC  = c("lightblue", "white", "#6A51A3", "#6A51A3")
  # DC  = c("white", "#6A51A3")
)

fig1_ag_u <- fig1_dat %>%
  imap(~ {
    .x %>%
      filter(tm_lab %in% as.character(tms)) %>%
      plot_scatter(
        "Ag_score", "hUMAP_1", "hUMAP_2",
        plot_colors  = u_clrs[[.y]],
        label_params = list(size = ttl_pt1),
        group_col    = "tm_lab",
        panel_nrow   = 1,
        outline      = TRUE,
        size         = 0.7,
        stroke       = 0.7
      ) +
      facet_wrap(
        ~ tm_lab, nrow = 1,
        labeller = as_labeller(function(x) str_c("day ", x))
      ) +
      guides(fill = guide_colorbar(
        title.position = "right", ticks = FALSE, title = "Ag-score",
        barheight = unit(120, "pt"), barwidth = unit(10, "pt")
      )) +
      umap_theme +
      theme(
        aspect.ratio         = 0.9,
        panel.border         = element_rect(color = ln_col, linewidth = ln_pt, fill = NA),
        legend.position      = "right",
        legend.justification = 0,
        legend.title         = element_text(angle = 270, hjust = 0.5),
        strip.text.y         = element_blank()
      )
  })
```

```{r "subtype Ag boxplots", fig.width = 8, fig.height = 2}
# Format data for boxplots
ag_bx_dat <- fig1_dat %>%
  map(~ {
    .x %>%
      filter(subtype != "unassigned") %>%
      mutate(mouse = as.character(mouse)) %>%
      group_by(subtype) %>%
      filter(all(table(mouse) >= min_cells)) %>%
      ungroup() %>%
      mutate(subtype = fct_reorder(subtype, Ag_score, mean, .desc = TRUE))
  })
  
# Create boxplots
bx_clrs <- list(LEC = lec_clrs, DC = dc_clrs)

ag_bxs <- ag_bx_dat %>%
  imap(~ create_ag_boxes(.x, clrs = bx_clrs[[.y]]))

# Pull subsets used for main figures
ag_plt_typs <- ag_bx_dat %>%
  map(~ levels(.x$subtype))
```

```{r "subtype Ag heat"}
# Create heatmaps
heat_clrs <- c(LEC = "#D7301F", DC = "#6A51A3")

ag_heats <- ag_bx_dat %>%
  imap(~ {
    dat <- .x %>%
      group_by(mouse, tm, subtype) %>%
      summarize(Ag_score = mean(Ag_score), .groups = "drop") %>%
      mutate(
        subtype = fct_reorder(subtype, Ag_score, mean),
        tm      = fct_relevel(as.character(tm), as.character(tms)),
        ttl     = "Ag-score"
      )

    rat <- n_distinct(dat$subtype) / n_distinct(dat$tm)

    dat %>%
      ggplot(aes(tm, subtype, fill = Ag_score)) +
      geom_tile() +
      facet_wrap(~ ttl) +
      labs(x = "days post vaccination") +
      scale_fill_gradientn(colours = c("lightblue", "white", heat_clrs[[.y]])) +
      guides(fill = guide_colorbar(
        ticks = FALSE, title = "Ag-score",
        barheight = unit(105, "pt"), barwidth = unit(10, "pt")
      )) +
      base_theme +
      theme(
        legend.title = element_blank(),
        aspect.ratio = rat,
        axis.title.y = element_blank(),
        plot.margin  = margin(0, 15, 0, 15),
        plot.title   = element_text(hjust = 0.5, size = ttl_pt2)
      )
  })
```

```{r "Fig 1", fig.width = 15.5, fig.height = 17}
# Create mock panel A
mock_a <- ggplot() +
  geom_blank() +
  theme(panel.background = element_blank())

# Create final figure
fig1 <- (mock_a + typ_heat) /
  (fig1_typ_u$LEC + fig1_ag_u$LEC) /
  (ag_heats[[1]] + ag_bxs$LEC) /
  (fig1_typ_u$DC + fig1_ag_u$DC) /
  (ag_heats[[2]] + ag_bxs$DC) +
  plot_layout(heights = c(0.7, 1, 0.7, 1, 0.7), tag_level = "keep") +
  plot_annotation(tag_levels = "A") &
  theme(plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"))

fig1
```

<br>

<br>

### Figure S1

A.  UMAP projections show cell types identified for CD45- and CD45+ samples.
B.  Ag-scores are shown for each timepoint for each LEC subset. Two biological
    replicates are shown for the day 21 and day 42 timepoints. The number of
    cells identified for each LEC subset is shown above each timepoint.
C.  Ag-scores are shown for each timepoint fo each DC subset, as described in B.

```{r "type UMAPs", fig.width = 14, fig.height = 5}
# Format data
u_dat <- all_meta %>%
  mutate(
    vaccination = str_c(vaccination, "\nvaccination"),
    tm          = ifelse(is.na(tm), vaccination, str_c("day ", tm)),
    tm          = fct_relevel(tm, str_c("day ", tms)),
    cell_type   = fct_infreq(cell_type),
    cell_sort   = recode(cell_sort, CD45neg = "CD45-", CD45pos = "CD45+")
  )

n_dat <- u_dat %>%
  group_by(tm, cell_sort) %>%
  summarize(n = str_c("n = ", label_comma()(n())), .groups = "drop")

# Create UMAP projections
typ_u <- u_dat %>%
  plot_scatter(
    "cell_type",
    "hUMAP_1", "hUMAP_2",
    plot_colors  = typ_clrs,
    plot_lvls    = c("BEC", "LEC", "Epithelial cells", "Monocytes"),
    size         = 0.001,
    n_label      = "legend"
  ) +
  geom_text(
    aes(-Inf, Inf, label = n, color = NULL), data = n_dat,
    hjust = -0.1, vjust = 1.5, size = 10 / .pt
  ) +
  facet_grid(cell_sort ~ tm) +
  guides(color = guide_legend(nrow = 2, override.aes = list(size = 4))) +
  umap_theme_2 +
  theme(
    legend.position = "bottom",
    legend.title = element_blank()
  )
```

```{r "all subtype Ag boxplots"}
# Format data for boxplots
bx_dat <- ag_objs %>%
  map(~ {
    .x@meta.data %>%
      as_tibble(rownames = ".cell_id") %>%
      filter(mouse != "6wk-3wk") %>%
      group_by(mouse, subtype, exp) %>%
      ungroup() %>%
      mutate(
        mouse   = as.character(mouse),
        tm_lab  = factor(as.character(tm), tms),
        subtype = fct_reorder(subtype, Ag_score, mean, .desc = TRUE)
      )
  })

# Create boxplots showing all LEC/DC subsets
all_ag_bxs <- bx_dat %>%
  imap(~ {
    dat <- .x
    
    dat <- c("exp-1", "exp-2") %>%
      map_dfr(~ {
        dat %>%
          filter(exp %in% c("exp-0", .x)) %>%
          mutate(exp = str_replace(.x, "exp-", "r"))
      })
    
    n_dat <- dat %>%
      group_by(tm, subtype, exp) %>%
      summarize(n = label_comma()(n()), .groups = "drop")
    
    dat %>%
      create_ag_boxes(clrs = bx_clrs[[.y]], facet_vars = c("exp", "subtype")) +
      geom_text_repel(
        aes(y = Inf, label = n), data = n_dat,
        seed = 42, force = 0.005,
        size = 8 / .pt, vjust = 1.2
      ) +
      scale_y_continuous(expand = expansion(c(0.05, 0.15))) +
      theme(plot.margin = margin(15, 15, 30, 15))
  })
```

```{r "Fig S1", fig.width = 14, fig.height = 15}
# Create final figure
figS1 <- plot_grid(
  typ_u, all_ag_bxs$LEC, all_ag_bxs$DC,
  labels = c("A", "B", "C"),
  label_size = ttl_pt2 * 1.5,
  rel_heights = c(1.6, 1, 1.1),
  align = "v", axis = "rl",
  ncol = 1
)

figS1
```



```{r "OLD EXTENDED 1-2", eval = FALSE}
lec_dat %>%
  group_by(subtype, tm, exp) %>%
  summarize(n = n(), .groups = "drop")
```

```{r "OLD EXTENDED 1-3", fig.width = 9, fig.height = 6, eval = FALSE}
# Ag-score is shown for all DC subsets for each experiment

# Data for boxplots
dc_dat <- dc_so@meta.data %>%
  filter(mouse != "6wk-3wk") %>%
  mutate(subtype = fct_reorder(subtype, Ag_score, mean, .desc = TRUE))

ln_dat <- dc_dat %>%
  group_by(subtype, exp, tm) %>%
  summarize(Ag_score = median(Ag_score), .groups = "drop")

# Create boxplots
exps <- c("exp-1", "exp-2")

exps %>%
  map(~ {
    dc_dat %>%
      filter(exp %in% c("exp-0", .x)) %>%
      
      ggplot(aes(tm, Ag_score, fill = subtype)) +
      
      geom_line(
        data = filter(ln_dat, exp %in% c("exp-0", .x)),
        color = "grey85", linewidth = 0.5, linetype = 2
      ) +
      geom_boxplot(aes(group = mouse), outlier.size = 0.2, width = 3, key_glyph = draw_key_point) +
      
      guides(
        fill = guide_legend(override.aes = list(size = 4, shape = 22), order = 1),
        linetype = guide_legend(override.aes = list(color = "black"), title = NULL)
      ) +
      facet_grid(~ subtype) +
      
      scale_fill_manual(values = dc_clrs) +
      labs(title = .x, x = "days post vaccination", y = "Ag-score") +
      base_theme +
      theme(
        aspect.ratio    = 0.9,
        legend.position = "none",
        plot.margin     = margin(0, 15, 0, 15)
      )
  }) %>%
  plot_grid(
    plotlist = .,
    ncol = 1
  )

```

```{r "OLD EXTENDED 1-4", eval = FALSE}
dc_dat %>%
  group_by(subtype, tm, exp) %>%
  summarize(n = n(), .groups = "drop")
```

---

<br>

<br>

## Figure 2

Identification of an antigen archiving gene signature

A.  Ag-score is shown for day 14 Ag-high and -low cells identified for LEC
    subsets with the highest Ag-score.
B.  The fraction of cells predicted to be Ag-competent is shown for each LEC
    subset 14, 21, and 42 days post vaccination.
C.  Ag-high module score is shown for Ag-high and predicted Ag-competent LECs
    for each timepoint. P values were calculated by pooling cells from all three
    timepoints for each Ag class.
    P values were calculated using a one-sided t test with Bonferroni correction
    to compare Ag-high vs Ag-competent and Ag-competent vs Ag-low cells for each
    LEC subset.
D.  UMAP projections show cLEC Ag-high module scores for each timepoint.
E.  The expression of select genes from the Ag-high (top four) and Ag-low
    (bottom four) gene modules is shown for cLECs. Genes identified as top
    predictors for all three LEC subsets are shown.

<u>Notes</u>

* Ag-low and -high cells were identified by separately clustering each d14 LEC
  subset for each sample into two groups based on Ag-score.
* The cutoffs identified for the d14 dataset were then used to identify Ag-high
  cells for the other timepoints.
* With this approach there is a different cutoff for each LEC subset, but the
  cutoff is consistent across timepoints.

```{r "RF Ag signal", fig.width = 12, fig.height = 3.25}
# Format plotting data
plt_dat <- lec_so@meta.data %>%
  filter(
    subtype %in% rf_typs,
    mouse %in% rf_mouse
  ) %>%
  mutate(
    mouse = as.character(mouse),  # to remove factor
    !!sym(rf_dat_clmn) := fct_relevel(!!sym(rf_dat_clmn), c("low", "high"))
  ) %>%
  arrange(!!sym(rf_dat_clmn))

# Format plot labels
plt_labs <- plt_dat %>%
  group_by(subtype, mouse, tm, !!sym(rf_dat_clmn)) %>%
  summarize(
    n = label_comma()(n()),
    .groups = "drop"
  ) %>%
  mutate(n = str_c(!!sym(rf_dat_clmn), ": ", n, " cells")) %>%
  group_by(subtype, mouse, tm) %>%
  arrange(!!sym(rf_dat_clmn)) %>%
  mutate(n = str_c(n, collapse = "\n"))

plt_clrs <- lec_clrs

# Create bargraphs
brs <- rf_typs %>%
  imap(~ {
    dat <- plt_dat %>%
      filter(subtype == .x) %>%
      mutate(!!sym(rf_dat_clmn) := fct_rev(!!sym(rf_dat_clmn)))
    
    dat %>%
      ggplot(aes(y = mouse, fill = subtype, alpha = !!sym(rf_dat_clmn))) +
      geom_bar() +
      facet_wrap(~ subtype, nrow = 1, scales = "free") +
      scale_alpha_manual(values = c(low = 0.35, high = 1)) +
      scale_fill_manual(values = plt_clrs) +
      scale_y_discrete(expand = expansion(c(0, 0))) +
      scale_x_continuous(expand = expansion(c(0, 0))) +
      theme_void() +
      theme(
        legend.position = "none",
        plot.margin     = margin(t = 5),
        strip.text      = element_text(size = ttl_pt2)
      )
  })

hsts <- rf_typs %>%
  imap(~ {
    dat <- plt_dat %>%
      filter(subtype == .x)
    
    fst <- .y == 1
    lst <- .y == length(rf_typs)
    
    # Create histograms
    hst <- dat %>%
      ggplot(aes(Ag_score, fill = subtype, alpha = !!sym(rf_dat_clmn))) +
      geom_histogram(bins = 30, key_glyph = draw_key_point) +
      geom_text(
        aes(-Inf, Inf, label = n),
        data = filter(plt_labs, subtype == .x),
        check_overlap = TRUE,
        alpha = 1,
        hjust = -0.25,
        vjust = 1.1,
        size  = txt_pt2 / .pt
      ) +
      facet_wrap(~ subtype, scales = "free", nrow = 1) +
      guides(fill = "none") +
      scale_alpha_manual(
        values = c(low = 0.35, high = 1),
        labels = function(x) str_c("Ag-", x)
      ) +
      scale_fill_manual(values = plt_clrs) +
      scale_y_continuous(expand = expansion(c(0.05, 0.25))) +
      guides(alpha = guide_legend(override.aes = list(shape = 22, size = 5))) +
      labs(x = "Ag-score", y = "number of cells") +
      base_theme +
      theme(
        aspect.ratio    = 0.6,
        strip.text      = element_blank(),
        legend.position = "none",
        legend.title    = element_blank(),
        plot.margin     = margin(t = 0, r = 5, b = 0, l = 5)
      )
    
    if (!fst && !lst) {
      hst <- hst +
        theme(legend.position = "bottom")
    }
    
    if (!fst) {
      hst <- hst +
        theme(axis.title.y = element_blank())
    }
    
    if (fst || lst) {
      hst <- hst +
        theme(axis.title.x = element_blank())
    }
    
    hst
  })

# Create final figure
hsts <- wrap_plots(
  append(brs, hsts),
  ncol = length(rf_tms),
  heights = c(0.1, 1)
)
```

```{r "RF Ag-competent bars", fig.width = 8, fig.height = 5}
# Format data
bar_dat <- lec_so@meta.data %>%
  filter(
    tm %in% rf_tms,
    subtype %in% rf_typs,
    !training_data
  ) %>%
  mutate(
    subtype    = fct_relevel(subtype, rf_typs),
    rf_correct = fct_relevel(as.character(rf_correct), c("TRUE", "FALSE")),
    pred_grp   = fct_relevel(pred_grp, rev(names(pred_clrs))),
    tm         = as.character(tm)
  )

bar_thm <- base_theme +
  theme(aspect.ratio = 1.4)

# # Plot fraction predicted groups
# bar_1 <- bar_dat %>%
#   ggplot(aes(tm, fill = rf_pred, alpha = rf_correct)) +
#   geom_bar(position = "fill") +
#   scale_alpha_manual(values = c(`TRUE` = 1, `FALSE` = 0.5)) +
#   scale_fill_manual(values = pred_clrs) +
#   scale_y_continuous(breaks = c(0, 0.5, 1)) +
#   facet_wrap(~ subtype) +
#   labs(y = "fraction of cells", x = "days post vaccination") +
#   bar_thm

bar_2 <- bar_dat %>%
  ggplot(aes(tm, fill = pred_grp)) +
  geom_bar(position = "fill", alpha = 0.85, key_glyph = draw_key_point) +
  facet_grid(~ subtype) +
  scale_fill_manual(values = pred_clrs) +
  scale_y_continuous(breaks = c(0, 0.5, 1)) +
  guides(fill = guide_legend(override.aes = list(shape = 22, size = 5))) +
  labs(y = "fraction of cells", x = "days post vaccination") +
  bar_thm +
  theme(
    legend.title = element_blank(),
    legend.position = "bottom"
  )

# plot_grid(
#   bar_1, bar_2, 
#   ncol = 1,
#   align = "v",
#   axis  = "rl"
# )
```

```{r "RF module boxplots", fig.width = 10, fig.height = 4}
# Format plot data
clmns <- c("mouse", "tm", "subtype", "pred_grp", "training_data", ag_modules)

plt_dat <- lec_so %>%
  FetchData(clmns) %>%
  filter(
    tm %in% rf_tms,
    subtype %in% rf_typs,
    !training_data
  ) %>%
  pivot_longer(all_of(ag_modules)) %>%
  separate_wider_regex(
    name,
    c(module_type = ".*", "_", module_ag = ".*")
  ) %>%
  filter(subtype == module_type) %>%
  mutate(
    subtype  = fct_relevel(subtype, rf_typs),
    pred_grp = fct_relevel(pred_grp, names(pred_clrs))
  )

# Calculate p-values
p_dat <- plt_dat %>%
  split(.$module_ag) %>%
  imap_dfr(~ {
    .x %>%
      group_by(subtype, module_ag) %>%
      summarize(
        p_high = t.test(
          value[pred_grp == "Ag-high"],
          value[pred_grp == "Ag-competent"],
          alternative = switch(.y, high = "greater", low = "less")
        )$p.value,
        p_comp = t.test(
          value[pred_grp == "Ag-competent"],
          value[pred_grp == "Ag-low"],
          alternative = switch(.y, high = "greater", low = "less")
        )$p.value,
        .groups = "drop"
      )
  }) %>%
  mutate(
    across(
      starts_with("p_"),
      ~ p.adjust(.x, method = "bonferroni"),
      .names = "{.col}_adj"
    )
  ) %>%
  rowwise() %>%
  mutate(
    across(ends_with("_adj"), ~ str_c(": ", .format_pvalue(.x))),
    
    # plab = .format_pvalue(padj),
    # plab = str_c('paste(', subtype, ', ":", ~ italic(p) == ', plab, ')')
    # plab = str_c('paste(', subtype, ', ":", ~ italic(p), " == ', plab, '")')
    # across(ends_with("_adj"), ~ {
    #   if (!grepl("^", .x, fixed = TRUE)) .x <- str_c(.x, "^''")
    #   .x
    # }),
    p_high_adj = str_c("high~v~comp", p_high_adj),
    p_comp_adj = str_c("comp~v~low", p_comp_adj)
  ) %>%
  ungroup()

# Create boxplots
mod_bxs <- plt_dat %>%
  split(.$module_ag) %>%
  imap(~ {
    p_dat <- p_dat %>%
      filter(module_ag == .y)
    
    # labs <- set_names(p_dat$plab, p_dat$subtype)
      
    .x %>%
      # mutate(subtype = labs[subtype]) %>%
      ggplot(aes(as.character(tm), value, fill = pred_grp)) +
      geom_boxplot(outlier.size = 0.1, key_glyph = draw_key_point) +
      
      geom_text(
        aes(1.5, Inf, fill = NULL, label = p_high_adj),
        size = 11 / .pt,
        data = p_dat,
        hjust = 0,
        vjust = 1.2,
        parse = TRUE
      ) +
      geom_text(
        aes(1.5, Inf, fill = NULL, label = p_comp_adj),
        size = 11 / .pt,
        data = p_dat,
        hjust = 0,
        vjust = 2.4,
        parse = TRUE
      ) +
      
      scale_fill_manual(values = pred_clrs) +
      scale_y_continuous(expand = expansion(c(0.05, 0.2))) +
      facet_wrap(~ subtype, nrow = 1, scales = "free_y") +
      # facet_wrap(~ subtype, nrow = 1, scales = "free_y", labeller = label_parsed) +
      labs(x = "days post vaccination", y = str_c("Ag-", .y, " module score")) +
      guides(fill = guide_legend(override.aes = list(shape = 22, size = 5))) +
      base_theme +
      theme(
        aspect.ratio    = 0.75,
        plot.title      = element_text(hjust = 0.5),
        legend.position = "bottom",
        legend.title    = element_blank(),
        plot.margin     = margin(t = 0, r = 5, b = 5, l = 5)
      )
  })
```

```{r "RF cLEC module UMAPs", fig.width = 8, fig.height = 6.5}
# Format plotting data
plt_typ     <- "cLEC"
module_clmn <- str_c(plt_typ, "_high")

u_so <- lec_so %>%
  subset(tm %in% rf_tms) %>%
  mutate_meta(
    mutate,
    pred_grp = ifelse(subtype == plt_typ, pred_grp, "other"),
    pred_grp = recode(pred_grp, `Ag-competent` = "Ag-low"),
    pred_grp = fct_relevel(pred_grp, c("other", names(pred_clrs))),
    tm       = as.character(tm)
  )

# Set coordinates for drawing circle
circ_dat <- lec_so@meta.data %>%
  filter(subtype == rf_cell_type) %>%
  group_by(tm) %>%
  summarize(across(c(hUMAP_1, hUMAP_2), mean))

# Plot module scores
ag_mod_u <- rf_tms %>%
  imap(~ {
    u_so %>%
      subset(tm == .x) %>%
      create_sig_umap(
        dat_col = module_clmn,
        grp_col = "tm",
        clrs    = c("lightblue", "lightblue", "white", pred_clrs[["Ag-high"]]),
        pt_size    = 1,
        pt_stroke  = 0.7
      ) +
      geom_circle(
        aes(x0 = hUMAP_1, y0 = hUMAP_2, fill = NULL, color = NULL, r = 2.5),
        data = filter(circ_dat, tm == .x),
        linetype = 2
      ) +
      facet_wrap(~ tm, labeller = as_labeller(function(x) str_c("day ", x)))
      # guides(fill = guide_colorbar(
      #   title = str_c(plt_typ, " Ag-high module"),
      #   ticks = FALSE, title.position = "top",
      #   barheight = unit(6, "pt"), barwidth = unit(120, "pt"), 
      # )) +
      # theme(legend.title = element_text())
  }) %>%
  wrap_plots(nrow = 1) +
  plot_annotation(
    title = str_c(plt_typ, " Ag-high module"),
    theme = theme(
      plot.title = element_text(hjust = 0.5, size = 18)
    )
  )

# Plot Ag-high cells
clrs <- rev(pred_clrs)
clrs["Ag-low"] <- "#56B4E9"
clrs["other"]  <- "white"

ag_low_u <- rf_tms %>%
  imap(~ {
    u_so %>%
      subset(tm == .x) %>%
      create_grp_umap(
        dat_col = "pred_grp",
        grp_col = "tm",
        clrs    = clrs,
        lvls    = c("Ag-high", "Ag-low", "other"),
        pt_size = 1,
        pt_stroke = 0.8
      ) +
      facet_wrap(~ tm, labeller = as_labeller(function(x) str_c("day ", x)))
      # theme(strip.text = element_text(size = 14))
  }) %>%
  wrap_plots(nrow = 1)
```

```{r "RF top features examples", fig.height = 8, fig.width = 8}
# Identify genes shared between multiple cell types
typ <- "cLEC"

shared_feats <- c("high", "low") %>%
  set_names() %>%
  map(~ {
    ag <- .x
    
    fts <- rf_feats %>%
      map(pluck, ag)
    
    fts <- fts %>%
      map(~ {
        .x[.x %in% rf_feats$cLEC[[ag]]]
      })
    
    fts <- fts[!names(fts) %in% c(typ, "BEC")]
    
    fts %>%
      purrr::reduce(c) %>%
      unique() %>%
      sort()
  })

# Plot example genes
gene_examples <- lec_so %>%
  subset(subtype == rf_cell_type) %>%
  subset(tm %in% rf_tms) %>%
  create_gn_plots(
    ag_high_gns = shared_feats$high,
    ag_low_gns  = shared_feats$low,
    plt_clrs    = tm_clrs,
    x_lvls      = names(pred_clrs),
    n_gns       = 4,
    
    # top_gns   = c("Gdf10", "Tgfa", "Ctsd", "Grn", "Col4a1", "Col4a2", "Fau", "Lama4"),
    
    p_test    = wilcox.test,
    draw_line = FALSE,
    pt_size   = 2.5
  ) +
  ggtitle(str_c(rf_cell_type, " expression")) +
  theme(plot.title = element_text(size = 18, hjust = 0.5))
```

```{r "RF GO", eval = FALSE}
# # Load GO results
# go <- qread(here(params$mod_dir, "ag-high_go.qs"))
# 
# endo_gns <- go@compareClusterResult %>%
#   filter(
#     Description == "endocytosis",
#     p.adjust < 0.05
#   ) %>%
#   pull(geneID) %>%
#   str_split("/") %>%
#   unlist()
# 
# gns <- c("Ctsd", "Dab2", "Prcp", "Cltc", "Degs", "Clu", "Atp6v1a", endo_gns)
# 
# gns <- unique(gns[gns %in% rf_feats$cLEC$high])
# 
# lec_so %>%
#   subset(subtype == "cLEC") %>%
#   subset(mouse == "3wk") %>%
#   mutate_meta(
#     mutate,
#     Ag_class_ml = ifelse(is.na(pred_grp), pred_grp, Ag_class_ml)
#   ) %>%
#   plot_violin(
#     "Atp6v1a",
#     "pred_grp",
#     # "Ag_class_ml",
#     "mouse",
#     method = "boxplot"
#   )
# 
# dat <- lec_so %>%
#   FetchData(c("Atp6v1a", "pred_grp", "subtype", "mouse")) %>%
#   filter(mouse == "3wk", subtype == "cLEC")
# 
# wilcox.test(
#   dat %>% filter(pred_grp == "Ag-high") %>% pull(Atp6v1a),
#   dat %>% filter(pred_grp == "Ag-competent") %>% pull(Atp6v1a)
# )
```

```{r "Fig 2", fig.width = 15, fig.height = 16.1}
left <- plot_grid(
  plot_grid(hsts, nrow = 2, rel_heights = c(1, 0.1)),
  mod_bxs$high,
  ag_mod_u,
  ag_low_u,
  ncol = 1,
  rel_heights = c(1, 1, 1.11, 1.1),
  labels = c("A", "C", "D"),
  label_size = ttl_pt2 * 1.5,
  label_y = c(1, 1.1, 1)
)

right <- plot_grid(
  bar_2, gene_examples,
  ncol = 1,
  rel_heights = c(1, 3),
  labels = c("B", "E"),
  label_size = ttl_pt2 * 1.5
)

plot_grid(
  left, right,
  rel_widths = c(1, 0.5)
)
```

---

<br>

<br>

## Figure 3

Antigen archiving is enhanced by successive vaccinations

A.  Mean 21 day Ag-score is shown for LECs from mice that received a single
    vaccination (21 day or 42 day) or dual vaccination (21 day and 42 day).
B.  UMAP projections show 21 day Ag-scores for LEC subsets for single and dual
    vaccinations.
C.  Prior vaccination enhances antigen archiving.
    Ag-score is shown for single and dual vaccinations for the 21 day timepoint
    for each LEC subset. Other timepoints are shown in grey. P values
    were calculated using a one-sided t test with Bonferroni correction.
D.  Mean 42 day Ag-score is shown as described in A.
E.  UMAP projections show 42 day Ag-scores for LEC subsets as described in B.
F.  Successive vaccinations enhances retention of previously archived antigen.
    Ag-score is shown for the 42 day timepoint as described in C.
G.  21 day and 42 day Ag-score is compared for each LEC subset.
H.  Ag-high module scores described in Figure 2 are shown for LECs that archived
    antigens from both vaccinations (double-high), from only one vaccination
    (single-high), or have low levels of both antigens (Ag-low).

<u>Notes</u>

* H shows that the best archivers (the ones that archive Ag from both
  vaccinations) show strongest expression of the Ag-high gene modules.
  Need to add p values.

```{r "exchange UMAPs", fig.width = 16, fig.height = 4}
# Format plot data
# Using experiment 1 since, experiment is missing 3wk LECs


# DON'T FILTER CELL TYPES HERE, SHOW ALL FOR UMAP
dat <- lec_so@meta.data %>%
  group_by(subtype) %>%
  filter(all(table(mouse) >= min_cells)) %>%
  ungroup() %>%
  mutate(mouse = fct_relevel(mouse, names(m_key))) %>%
  filter(exp == "exp-1") %>%
  arrange(mouse) %>%
  mutate(
    mouse = m_key[as.character(mouse)],
    mouse = fct_inorder(mouse)
  )

# Create UMAPs
plt_args <- list(
  ag_clmns = ag_clmns,
  clr      = tm_clrs[c("21", "42")]
)

mx_sig <- max(dat[plt_args$ag_clmns])

ex_u <- plt_args %>%
  pmap(~ {
    dat %>%
      plot_scatter(
        ..1,
        x           = "hUMAP_1",
        y           = "hUMAP_2",
        group_col   = "mouse",
        panel_nrow  = 1,
        outline     = TRUE,
        size        = 1,
        stroke      = 0.7,
        plot_colors = c("lightblue", "white", ..2),
        label_params = list(size = ttl_pt1)
      ) +
      # scale_fill_gradientn(colours = c("lightblue", "white", ..2), limits = c(0, mx_sig)) +
      guides(fill = guide_colorbar(
        ticks = FALSE,
        title = ag_labs[.x],
        barheight = unit(150, "pt"), barwidth = unit(10, "pt")
      )) +
      umap_theme +
      theme(
        legend.position = "right",
        panel.border    = element_rect(color = ln_col, linewidth = ln_pt)
      )
  })

# Create heatmaps
heat_dat <- dat %>%
  group_by(mouse, subtype) %>%
  summarize(across(all_of(ag_clmns), mean), .groups = "drop")

ex_heat <- plt_args %>%
  pmap(~ {
    heat_dat %>%
      mutate(subtype = fct_reorder(subtype, !!sym(..1), mean)) %>%
      ggplot(aes(mouse, subtype, fill = !!sym(..1))) +
      geom_tile() +
      labs(x = "vaccination group") +
      scale_fill_gradientn(colours = c("lightblue", "white", ..2)) +
      scale_x_discrete(labels = ~ str_replace(.x, " day", "d")) +
      guides(fill = guide_colorbar(
        ticks = FALSE,
        title = ag_labs[.x],
        barheight = unit(150, "pt"), barwidth = unit(10, "pt")
      )) +
      base_theme +
      theme(
        aspect.ratio = n_distinct(heat_dat$subtype) / n_distinct(heat_dat$mouse),
        plot.margin  = margin(0, 15, 0, 15),
        axis.title.y = element_blank(),
        axis.text.y  = element_text(size = ttl_pt2)
      )
  })

# Create final panel
ex_panels <- map2(ex_heat, ex_u, ~ {
  plot_grid(
    .x, .y,
    nrow = 1,
    align = "h",
    axis  = "tb",
    rel_widths = c(0.37, 1)
  )
})
```

```{r "exchange 3wk boxes", fig.width = 10, fig.height = 2}
# Effect of previous vaccination on uptake of new antigen
# Comparison using exp-2 isn't very useful since we captured very few LECs
# for the 3wk timepoint
# Use LEC subsets shown in Fig 1
ex_3_bxs <- lec_so@meta.data %>%
  filter(
    !subtype %in% c("unassigned"),
    exp %in% c("exp-0", "exp-1")
  ) %>%
  create_exchange_boxes(
    tm        = "21",
    tm_clmn   = "tm_3",
    ag_clmn   = "Ag_score_3",
    typs      = ag_plt_typs$LEC,
    clrs      = tm_clrs
  )
```

```{r "exchange 6wk boxes", fig.width = 10, fig.height = 2.15}
# Effect of previous vaccination on uptake of new antigen
# Comparison using exp-2 isn't very useful since we captured very few LECs
# for the 3wk timepoint
# Use LEC subsets shown in Fig 1
ex_6_bxs <- lec_so@meta.data %>%
  filter(
    !subtype %in% c("unassigned"),
    exp %in% c("exp-0", "exp-1")
  ) %>%
  create_exchange_boxes(
    tm        = "42",
    tm_clmn   = "tm_6",
    ag_clmn   = "Ag_score_6",
    typs      = ag_plt_typs$LEC,
    clrs      = tm_clrs,
    p_hjust   = 0.75
  )
```

```{r "exchange correlation", fig.width = 10, fig.height = 4}
# Plot data
dat <- lec_so@meta.data %>%
  as_tibble(rownames = "cell_id") %>%
  filter(
    exp == "exp-1",
    mouse == "6wk-3wk",
    # cell_type == "LEC",
    subtype != "unassigned"
  ) %>%
  group_by(subtype) %>%
  filter(n() > 50) %>%
  ungroup() %>%
  mutate(subtype = fct_relevel(subtype, ag_plt_typs$LEC))

# Calculate correlation
lab_dat <- dat %>%
  group_by(subtype) %>%
  summarize(
    cor   = cor.test(Ag_3wk_score, Ag_6wk_score, method = "pearson")$estimate[[1]],
    p_val = cor.test(Ag_3wk_score, Ag_6wk_score, method = "pearson")$p.value,
    n     = n_distinct(cell_id),
    .groups = "drop"
  ) %>%
  mutate(p_adj = p.adjust(p_val, method = "bonferroni")) %>%
  rowwise() %>%
  mutate(
    p_lab = .format_pvalue(p_adj),
    r_lab = str_c("r = ", round(cor, digits = 2), "\nn = ", n)
    # r_lab = str_c("atop(italic(r) == ", round(cor, digits = 2), ", n == ", n, ")")
    # r_lab = str_c("italic(r) == ", round(cor, digits = 2), " ~~ n == ", n)
    # r_lab = str_c("italic(r) == ", round(cor, digits = 2))
    # r_lab = str_c("italic(r) == ", round(cor, digits = 2), " ~~ italic(p) == ", p_lab)
  )

# Create scatter plots
x_var <- "Ag_6wk_score"
y_var <- "Ag_3wk_score"

ag_labs <- c(
  Ag_3wk_score = "Ag-score (21 day)",
  Ag_6wk_score = "Ag-score (42 day)"
)

ex_scat <- dat %>%
  ggplot(aes(!!sym(x_var), !!sym(y_var), color = subtype)) +
  geom_abline(slope = 1, intercept = 0, linewidth = 0.5, color = "grey90") +
  geom_point(size = 1.5) +
  geom_smooth(
    method    = "lm",
    linetype  = 2,
    linewidth = 0.5,
    color     = "black",
    se        = FALSE,
    formula   = "y ~ x"
  ) +
  geom_text(
    aes(x = -Inf, y = Inf, label = r_lab),
    data  = lab_dat,
    hjust = -0.1,
    vjust = 1.1,
    color = "black",
    size  = ttl_pt1 / .pt
  ) +
  facet_wrap(~ subtype, nrow = 1) +
  
  scale_color_manual(values = lec_clrs) +
  coord_cartesian(
    xlim = c(0, max(c(dat[[x_var]], dat[[y_var]]))),
    ylim = c(0, max(c(dat[[x_var]], dat[[y_var]])))
  ) +
  labs(
    x = ag_labs[x_var],
    y = ag_labs[y_var]
  ) +
  base_theme +
  theme(
    aspect.ratio = 1,
    legend.position = "none"
  )
```

```{r "exchange Ag-high module scores", fig.width = 9, fig.height = 5.5}
## CHECK FOR ANY DIFFERENCE BETWEEN THE BIOLOGICAL REPLICATES ##
# replicates are similar
# pool all cells together to increase power

# Format plot data
lvls <- c("Ag-low", "Ag-high", "single-high", "double-high")

plt_dat <- lec_so %>%
  FetchData(c("subtype", "mouse", "Ag_class_dual", "exp", ag_modules)) %>%
  filter(
    mouse %in% c("6wk", "3wk", "6wk-3wk"),
    subtype %in% rf_typs
  ) %>%
  pivot_longer(all_of(ag_modules)) %>%
  separate_wider_regex(
    name,
    c(module_type = ".*", "_", module_ag = ".*")
  ) %>%
  filter(
    subtype == module_type,
    module_ag == "high"
  ) %>%
  
  # Group all low cells together under single mouse, otherwise three boxplots
  # are shown
  mutate(
    subtype       = fct_relevel(subtype, rf_typs),
    mouse         = m_key[as.character(mouse)],
    Ag_class_dual = str_replace(Ag_class_dual, "^[0-9]+wk", "Ag"),
    Ag_class_dual = recode(Ag_class_dual, low = "Ag-low"),
    Ag_class_dual = fct_relevel(Ag_class_dual, lvls)
  ) %>%
  arrange(Ag_class_dual) %>%
  mutate(
    mouse_class = fct_inorder(str_c(mouse, "-", subtype, "-", Ag_class_dual))
  )

# Calculate p-values
p_dat <- plt_dat %>%
  filter(mouse == "dual") %>%
  split(.$subtype) %>%
  map(~ {
      pairwise.t.test(
        .x$value, .x$Ag_class_dual,
        alternative = "greater",
        p.adjust.method = "bonferroni"
      ) %>%
      broom::tidy()
    }) %>%
  bind_rows(.id = "subtype") %>%
  filter(group1 == "double-high") %>%
  mutate(
    ln_x = match(group2, lvls[lvls %in% group2]),
    x    = (3 + ln_x) / 2
  ) %>%
  rowwise() %>%
  mutate(padj = .format_pvalue(p.value))

# Format n labels
n_dat <- plt_dat %>%
  filter(mouse == "dual") %>%
  group_by(Ag_class_dual, mouse_class, subtype, module_ag) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    n = str_c(Ag_class_dual, "\nn = ", scales::label_comma()(n))
  ) %>%
  filter(module_ag == "high")

n_lab <- set_names(n_dat$n, n_dat$mouse_class)

# Create boxplots
ex_mod_bxs <- plt_dat %>%
  filter(mouse == "dual") %>%
  ggplot(aes(mouse_class, value, fill = subtype)) +
  geom_boxplot(notch = TRUE, outlier.size = 0.1) +
  
  facet_wrap(~ subtype, scales = "free") +
  scale_fill_manual(values = lec_clrs) +
  scale_x_discrete(labels = n_lab) +
  labs(y = "Ag-high module score") +
  base_theme +
  theme(
    aspect.ratio    = 1.3,
    legend.position = "none",
    axis.text.x     = element_text(angle = 45, hjust = 1),
    axis.title.x    = element_blank()
  )
```

```{r "Fig 3", fig.width = 14, fig.height = 20}
(ex_heat[[1]] + ex_u[[1]]) / ex_3_bxs /
  (ex_heat[[2]] + ex_u[[2]]) / ex_6_bxs /
  ex_scat /
  wrap_plots(ex_mod_bxs, plot_spacer()) +
  # wrap_plots(ex_mod_bxs, nrow = 1) +
  # wrap_plots(append(list(plot_spacer()), ex_mod_bxs), nrow = 1, widths = c(0, 1, 1, 1)) +
  # (ex_mod_bxs[[1]] + ex_mod_bxs[[2]] + ex_mod_bxs[[3]]) +
  # plot_grid(plotlist = ex_mod_bxs, nrow = 1) +
  plot_layout(heights = c(1, 0.55, 1, 0.55, 0.8, 0.8)) +
  plot_annotation(tag_levels = "A") &
  theme(plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"))
```

<br>

<br>

### Figure S3

A.  Ag-score is shown for single and dual vaccinations for the 21 day timepoint
    for each LEC subset. Other timepoints are shown in grey. P values
    were calculated using a one-sided t test with Bonferroni correction.
B.  Ag-score is shown for the 21 day timepoint for DC subsets,
    as described in A.
C.  Ag-score is shown for the 42 day timepoint for LEC subsets,
    as described in A.
D.  Ag-score is shown for the 42 day timepoint for DC subsets,
    as described in C. 

```{r "Fig S3", fig.width = 18, fig.height = 8}
# Set input data
exps <- c("exp-1", "exp-2")

bx_args <- list(
  tm      = c("21", "42"),
  tm_clmn = c("tm_3", "tm_6"),
  ag_clmn = c("Ag_score_3", "Ag_score_6"),
  p_hjust = c(0.5, 0.9),
  exps    = list(
    list(LEC = "exp-1", DC = exps),
    list(LEC = exps, DC = exps)
  )
)

# Create boxplots
all_ex_bxs <- bx_args %>%
  pmap(~ {
    bx_args <- list(...)
    
    ag_objs %>%
      imap(~ {
        dat <- .x@meta.data
        
        dat <- bx_args$exps[[.y]] %>%
          map_dfr(~ {
            dat %>%
              filter(exp %in% c("exp-0", .x)) %>%
              mutate(exp = str_replace(.x, "exp-", "r"))
          })
        
        dat %>%
          create_exchange_boxes(
            tm         = bx_args$tm,
            tm_clmn    = bx_args$tm_clmn,
            ag_clmn    = bx_args$ag_clmn,
            clrs       = tm_clrs,
            typs       = ag_plt_typs[[.y]],
            p_hjust    = bx_args$p_hjust,
            facet_vars = c("exp", "subtype"),
            p_vars     = "exp",
            p_cutoff   = Inf
          ) +
          scale_y_continuous(expand = expansion(c(0.05, 0.15))) +
          theme(legend.position = "bottom")
      })
  }) %>%
  flatten()

# Create final figure
plot_grid(
  plotlist = all_ex_bxs,
  labels = c("A", "B", "C", "D"),
  label_size = ttl_pt2 * 1.5,
  # rel_heights = c(0.7, 0.7, 1, 0.7),
  ncol = 2, align = "v", axis = "rl"
)
```



```{r "extended scatter plots", fig.width = 14, fig.height = 4, eval = FALSE}
# Plot data
dat <- lec_so@meta.data %>%
  as_tibble(rownames = "cell_id") %>%
  filter(
    exp == "exp-1",
    mouse == "6wk-3wk",
  ) %>%
  # group_by(subtype) %>%
  # filter(n() > 50) %>%
  # ungroup() %>%
  mutate(subtype = fct_relevel(subtype, ag_plt_typs$LEC))

# Create scatter plots
x_var <- "Ag_6wk_score"
y_var <- "Ag_3wk_score"

ag_labs <- c(
  Ag_3wk_score = "Ag-score (21 day)",
  Ag_6wk_score = "Ag-score (42 day)"
)

dat %>%
  ggplot(aes(!!sym(x_var), !!sym(y_var), color = Ag_class_dual)) +
  geom_abline(slope = 1, intercept = 0, linewidth = 0.5, color = "grey90") +
  geom_point(size = 1.5) +
  geom_smooth(
    method    = "lm",
    linetype  = 2,
    linewidth = 0.5,
    color     = "black",
    se        = FALSE,
    formula   = "y ~ x"
  ) +
  facet_wrap(~ subtype, nrow = 1) +
  coord_cartesian(
    xlim = c(0, max(c(dat[[x_var]], dat[[y_var]]))),
    ylim = c(0, max(c(dat[[x_var]], dat[[y_var]])))
  ) +
  labs(
    x = ag_labs[x_var],
    y = ag_labs[y_var]
  ) +
  base_theme +
  theme(
    aspect.ratio = 1,
    legend.position = "bottom",
    legend.title    = element_blank()
  )
```

```{r "AG-EX DOUBLE POSITIVE EXTENDED", fig.width = 14, fig.height = 3, eval = FALSE}
# Expression of Ag-high gene modules is shown for each LEC subset. Expression is
# similar for the day 42 cLEC Ag-low/high groups, likely due to the many Ag-low
# cells that are actually Ag-competent and should express the Ag-high gene
# module.
# 
# Format plot data
lvls <- c("Ag-low", "Ag-high", "single-high", "double-high")

plt_dat <- lec_so %>%
  FetchData(c("subtype", "mouse", "Ag_class_dual", ag_modules)) %>%
  filter(
    mouse %in% c("6wk", "3wk", "6wk-3wk"),
    subtype %in% rf_typs
  ) %>%
  pivot_longer(all_of(ag_modules)) %>%
  separate_wider_regex(
    name,
    c(module_type = ".*", "_", module_ag = ".*")
  ) %>%
  filter(
    subtype == module_type,
    module_ag == "high"
  ) %>%
  
  # Group all low cells together under single mouse, otherwise three boxplots
  # are shown
  mutate(
    subtype       = fct_relevel(subtype, rf_typs),
    mouse         = m_key[as.character(mouse)],
    Ag_class_dual = str_replace(Ag_class_dual, "^[0-9]+wk", "Ag"),
    Ag_class_dual = recode(Ag_class_dual, low = "Ag-low"),
    Ag_class_dual = fct_relevel(Ag_class_dual, lvls)
  ) %>%
  arrange(Ag_class_dual) %>%
  mutate(
    mouse_class = fct_inorder(str_c(mouse, "-", subtype, "-", Ag_class_dual))
  )

# Format n labels
n_dat <- plt_dat %>%
  group_by(Ag_class_dual, mouse_class, subtype, module_ag) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    n = str_c(Ag_class_dual, "\nn = ", scales::label_comma()(n))
  ) %>%
  filter(module_ag == "high")

n_lab <- set_names(n_dat$n, n_dat$mouse_class)

# Create boxplots
ex_mod_bxs <- rf_typs %>%
  map(~ {
    plt_dat %>%
      filter(subtype == .x) %>%
      ggplot(aes(mouse_class, value, fill = subtype)) +
      geom_boxplot(notch = TRUE, outlier.size = 0.1) +
      facet_grid(~ mouse, scales = "free_x", space = "free_x") +
      scale_fill_manual(values = lec_clrs) +
      scale_x_discrete(labels = n_lab) +
      labs(y = str_c(.x, " Ag-high\nmodule score")) +
      base_theme +
      theme(
        legend.position = "none",
        axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.x = element_blank()
      )
  })

ex_mod_bxs %>%
  plot_grid(
    plotlist = .,
    align    = "h",
    axis     = "tb",
    nrow     = 1
  )
```

---

<br>

<br>

## Figure 4

Antigen uptake by DCs is enhanced by successive vaccinations

A.  Mean 21 day Ag-score is shown for DCs from mice that received a single
    vaccination (21 day or 42 day) or dual vaccination (21 day and 42 day).
B.  Ag-score is shown for single and dual vaccinations for the 21 day timepoint
    for DC subsets with highest Ag-score. Other timepoints are shown in grey.
    P values were calculated using a one-sided t test with Bonferroni
    correction.
C.  Mean 42 day Ag-score is shown for DCs from mice that received single or
    dual vaccination as described in A.
D.  Ag-score is shown for single and dual vaccinations for the 42 day timepoint
    as described in B.
    
<u>Notes</u>

* Need to add GeoMx figures showing Ag signal is highest in medula and sinus
  regions for both LECs and DCs.

```{r "DC exchange heatmaps", fig.width = 16, fig.height = 4}
# Format plot data
# EXPERIMENT 2
plt_args <- list(
  ag_clmns = ag_clmns,
  clr      = tm_clrs[c("21", "42")]
)

dat <- dc_so@meta.data %>%
  group_by(subtype) %>%
  filter(all(table(mouse) >= min_cells)) %>%
  ungroup() %>%
  mutate(mouse = fct_relevel(mouse, names(m_key))) %>%
  filter(exp == "exp-2") %>%
  arrange(mouse) %>%
  mutate(
    mouse = m_key[as.character(mouse)],
    mouse = fct_inorder(mouse)
  )

# Create heatmaps
heat_dat <- dat %>%
  group_by(mouse, subtype) %>%
  summarize(across(all_of(ag_clmns), mean), .groups = "drop")

ex_heat <- plt_args %>%
  pmap(~ {
    heat_dat %>%
      mutate(subtype = fct_reorder(subtype, !!sym(..1), mean)) %>%
      ggplot(aes(mouse, subtype, fill = !!sym(..1))) +
      geom_tile() +
      labs(x = "vaccination group") +
      scale_fill_gradientn(colours = c("lightblue", "white", ..2)) +
      scale_x_discrete(labels = ~ str_replace(.x, " day", "d")) +
      guides(fill = guide_colorbar(
        ticks = FALSE,
        title = ag_labs[.x],
        barheight = unit(150, "pt"), barwidth = unit(10, "pt")
      )) +
      base_theme +
      theme(
        aspect.ratio = n_distinct(heat_dat$subtype) / n_distinct(heat_dat$mouse),
        plot.margin  = margin(0, 15, 0, 15),
        axis.title.y = element_blank(),
        axis.text.y  = element_text(size = ttl_pt2)
      )
  })
```

```{r "DC exchange 3wk boxes", fig.width = 10, fig.height = 2}
# Effect of previous vaccination on uptake of new antigen
ex_3_bxs <- dc_so@meta.data %>%
  filter(
    !subtype %in% c("unassigned"),
    exp %in% c("exp-0", "exp-2")
  ) %>%
  create_exchange_boxes(
    tm        = "21",
    tm_clmn   = "tm_3",
    ag_clmn   = "Ag_score_3",
    typs      = ag_plt_typs$DC[1:3],
    clrs      = tm_clrs,
    p_size    = 10
  ) +
  theme(legend.position = "bottom")
```

```{r "DC exchange 6wk boxes", fig.width = 10, fig.height = 2.15}
# Effect of previous vaccination on uptake of new antigen
ex_6_bxs <- dc_so@meta.data %>%
  filter(
    !subtype %in% c("unassigned"),
    exp %in% c("exp-0", "exp-2")
  ) %>%
  create_exchange_boxes(
    tm        = "42",
    tm_clmn   = "tm_6",
    ag_clmn   = "Ag_score_6",
    typs      = ag_plt_typs$DC[1:3],
    clrs      = tm_clrs,
    p_size    = 10,
    p_hjust   = 0.75
  ) +
  theme(legend.position = "bottom")
```

```{r "Fig 4", fig.width = 15, fig.height = 9}
wrap_plots(
  ex_heat[[1]], ex_3_bxs,
  ex_heat[[2]], ex_6_bxs,
  nrow = 2,
  widths = c(0.25, 1)
) +
  plot_annotation(tag_levels = "A") &
  theme(plot.tag = element_text(size = ttl_pt2 * 1.5, face = "bold"))
```



```{r "EXTENDED 4-1", eval = FALSE}

dat %>%
  plot_scatter(
    # "Ag_6wk_score",
    "Ag_3wk_score",
    x           = "hUMAP_1",
    y           = "hUMAP_2",
    group_col   = "mouse",
    panel_nrow  = 1,
    # outline     = TRUE,
    size        = 0.5,
    stroke      = 0.7,
    # plot_colors = c("lightblue", "white", "#D7301F"),
    plot_colors = c("lightblue", "white", rep("#6A51A3", 3)),
    # max_q       = 99.9,
    label_params = list(size = ttl_pt1)
  ) +
  # scale_fill_gradientn(colours = c("lightblue", "white", ..2), limits = c(0, mx_sig)) +
  guides(fill = guide_colorbar(
    ticks = FALSE,
    # title = ag_labs[.x],
    barheight = unit(150, "pt"), barwidth = unit(10, "pt")
  )) +
  umap_theme +
  theme(
    legend.position = "right",
    panel.border    = element_rect(color = ln_col, linewidth = ln_pt)
  )

```

```{r "EXTENDED 4-2", eval = FALSE}

# Create UMAPs
plt_args <- list(
  ag_clmns = ag_clmns,
  clr      = tm_clrs[c("21", "42")]
)

mx_sig <- max(dat[plt_args$ag_clmns])

ex_u <- plt_args %>%
  pmap(~ {
    dat %>%
      plot_scatter(
        ..1,
        x           = "hUMAP_1",
        y           = "hUMAP_2",
        group_col   = "mouse",
        panel_nrow  = 1,
        outline     = TRUE,
        size        = 1,
        stroke      = 0.7,
        plot_colors = c("lightblue", "white", ..2),
        max_q       = 99,
        label_params = list(size = ttl_pt1)
      ) +
      # scale_fill_gradientn(colours = c("lightblue", "white", ..2), limits = c(0, mx_sig)) +
      guides(fill = guide_colorbar(
        ticks = FALSE,
        title = ag_labs[.x],
        barheight = unit(150, "pt"), barwidth = unit(10, "pt")
      )) +
      umap_theme +
      theme(
        legend.position = "right",
        panel.border    = element_rect(color = ln_col, linewidth = ln_pt)
      )
  })

```

```{r "GEOMX PROCESSING", eval = FALSE}
library(NanoStringNCTools)
library(GeomxTools)
library(GeoMxWorkflows)
library(umap)

proj      <- "tamburini-antigen-tracking"
geomx_dat <- file.path("~/Dropbox/Ryan/Projects", proj)

geomx_dir  <- file.path("~/Projects", proj)
geomx_res  <- "results/2023-06-23"
counts_dir <- "results/2023-06-23/counts"
worksheet  <- "geomx_worksheets/Tamburini_061423_20230619T1930_annotations.xlsx"
obj_dir    <- "sobjs"                            # Directory to load/save Seurat objects
table_dir  <- "tables"
pkcs       <- c("Mm_R_NGS_WTA_v1.0.pkc", "Hesselberth_02-SO-90138_v1.0.pkc")

# Load data
dcc <- dir(file.path(geomx_dat, counts_dir), ".dcc$", full.names = TRUE)
pkc <- file.path(geomx_dir, "ref", pkcs)
ann <- file.path(geomx_dir, geomx_res, worksheet)

dat <- readNanoStringGeoMxSet(
  dccFiles               = dcc,
  pkcFiles               = pkc,
  phenoDataFile          = ann,
  phenoDataSheet         = "LabWorksheet",
  phenoDataDccColName    = "Sample_ID",
  protocolDataColNames   = c("Aoi", "Roi"),
  experimentDataColNames = "Panel"
)

pkcs    <- annotation(dat)
modules <- gsub(".pkc", "", pkcs)

# Add 1 to counts
dat <- dat %>%
  shiftCountsOne(useDALogic = TRUE)

# Default QC cutoffs are shown in parenthesis
# study-specific values were selected after visualizing the QC results in more
# detail below
QC_params <- list(
  minSegmentReads   = 1000,  # Minimum number of reads (1000)
  percentTrimmed    = 80,    # Minimum % of reads trimmed (80%)
  percentStitched   = 80,    # Minimum % of reads stitched (80%)
  percentAligned    = 75,    # Minimum % of reads aligned (80%)
  percentSaturation = 50,    # Minimum sequencing saturation (50%)
  minNegativeCount  = 1,     # Minimum negative control counts (10)
  maxNTCCount       = 1000,  # Maximum counts observed in NTC well (1000)
  minNuclei         = 100,   # Minimum # of nuclei estimated (100)
  minArea           = 5000   # Minimum segment area in um2 (5000)
)

dat <- dat %>%
  setSegmentQCFlags(qcCutoffs = QC_params)

# Collate QC Results
QCResults    <- protocolData(dat)[["QCFlags"]]
flag_columns <- colnames(QCResults)

QC_Summary <- data.frame(
  Pass    = colSums(!QCResults[, flag_columns]),
  Warning = colSums(QCResults[, flag_columns])
)

QCResults$QCStatus <- apply(QCResults, 1L, function(x) {
    ifelse(sum(x) == 0L, "PASS", "WARNING")
})

QC_Summary["TOTAL FLAGS", ] <- c(
  sum(QCResults[, "QCStatus"] == "PASS"),
  sum(QCResults[, "QCStatus"] == "WARNING")
)

# Calculate geometric mean for negative control probes for each pkc module 
neg_geo_mns <- esBy(
  negativeControlSubset(dat), 
  GROUP = "Module", 
  FUN = function(x) assayDataApply(x, MARGIN = 2, FUN = ngeoMean, elt = "exprs")
)

protocolData(dat)[["NegGeoMean"]] <- neg_geo_mns

# explicitly copy the Negative geoMeans from sData to pData
negCols <- paste0("NegGeoMean_", modules)

pData(dat)[, negCols] <- sData(dat)[["NegGeoMean"]][, modules]

# Filter data
dat <- dat[, QCResults$QCStatus == "PASS"]

## Probe QC
# A probe is removed globally from the dataset if either of the following is
# true, Nanostring does not recommend adjusting these parameters.
# A probe is removed locally (from a given segment) if the probe is an outlier
# according to the Grubb’s test in that segment.
# 
# * The geometric mean of that probe’s counts from all segments divided by the
#   geometric mean of all probe counts representing the target from all segments
#   is less than 0.1
# * The probe is an outlier according to the Grubb’s test in at least 20% of the
#   segments
# 
# Generally keep the qcCutoffs parameters unchanged. Set removeLocalOutliers to 
# FALSE if you do not want to remove local outliers
dat <- dat %>%
  setBioProbeQCFlags(
    qcCutoffs = list(minProbeRatio = 0.1, percentFailGrubbs = 20),
    removeLocalOutliers = TRUE
  )

ProbeQCResults <- fData(dat)[["QCFlags"]]

# Probe QC summary
qc_df <- data.frame(
  Passed = sum(rowSums(ProbeQCResults[, -1]) == 0),
  Global = sum(ProbeQCResults$GlobalGrubbsOutlier),
  Local  = sum(rowSums(ProbeQCResults[, -2:-1]) > 0 & !ProbeQCResults$GlobalGrubbsOutlier)
)

# Collapse probe counts for each gene
# This is important since multiple probes are included for some genes
# Remove neg_geomean columns from pData since aggregateCounts()
# will add these columns
pData(dat) <- pData(dat)[, !colnames(pData(dat)) %in% negCols]
gene_dat   <- aggregateCounts(dat)

# In addition to Segment and Probe QC, also need to determine the limit of
# quantification (LOQ) per segment. The LOQ is calculated based on the
# distribution of negative control probes and is intended to approximate the
# quantifiable limit of gene expression per segment. It should be noted that this
# process is more stable for larger segments. Likewise, the LOQ may not be as
# accurately reflective of true signal detection rates in segments with low
# negative probe counts (e.g. <2).
# 
# Use 2 geometric standard deviations (n = 2) above the geometric mean as the LOQ, which is reasonable for most studies.
# Nanostring also recommends that a minimum LOQ of 2 be used if the LOQ calculated in a segment is below this threshold.

# Define LOQ SD threshold and minimum value
cutoff <- 2
minLOQ <- 2

# Calculate LOQ per module tested
LOQ <- data.frame(row.names = colnames(gene_dat))

for (module in modules) {
    vars <- paste0(c("NegGeoMean_", "NegGeoSD_"), module)
    
    if(all(vars[1:2] %in% colnames(pData(gene_dat)))) {
        LOQ[, module] <- pmax(minLOQ, pData(gene_dat)[, vars[1]] * pData(gene_dat)[, vars[2]] ^ cutoff)
    }
}

# Add LOQ to gene-level data
pData(gene_dat)$LOQ <- LOQ

# Calculate detected genes in each segment
LOQ_Mat <- c()

for(module in modules) {
  ind     <- fData(gene_dat)$Module == module
  Mat_i   <- t(esApply(gene_dat[ind, ], MARGIN = 1, FUN = function(x) x > LOQ[, module]))
  LOQ_Mat <- rbind(LOQ_Mat, Mat_i)
}

# Ensure ordering since this is stored outside of the geomxSet
LOQ_Mat <- LOQ_Mat[fData(gene_dat)$TargetName, ]

# Save detection rate information to pheno data
pData(gene_dat)$GenesDetected     <- colSums(LOQ_Mat, na.rm = TRUE)
pData(gene_dat)$GeneDetectionRate <- pData(gene_dat)$GenesDetected / nrow(gene_dat)

# Determine detection thresholds
# Bin segments based on detection rate
pData(gene_dat)$DetectionThreshold <- cut(
  pData(gene_dat)$GeneDetectionRate,
  breaks = c(0, 0.01, 0.05, 0.1, 0.15, 1),
  labels = c("<1%", "1-5%", "5-10%", "10-15%", ">15%")
)

# Filter segments
rate_cutoff <- 0.01

gene_dat <- gene_dat[, pData(gene_dat)$GeneDetectionRate >= rate_cutoff]

# Calculate detection rate
LOQ_Mat <- LOQ_Mat[, colnames(gene_dat)]

fData(gene_dat)$DetectedSegments <- rowSums(LOQ_Mat, na.rm = TRUE)
fData(gene_dat)$DetectionRate    <- fData(gene_dat)$DetectedSegments / nrow(pData(gene_dat))

# Plot detection rate:
plot_detect <- data.frame(Freq = c(1, 5, 10, 20, 30, 50))

plot_detect$Number <- unlist(lapply(
  c(0.01, 0.05, 0.1, 0.2, 0.3, 0.5),
  function(x) sum(fData(gene_dat)$DetectionRate >= x)
))

plot_detect$Rate <- plot_detect$Number / nrow(fData(gene_dat))

rownames(plot_detect) <- plot_detect$Freq

# USING CUTOFF OF 1%
# Subset to target genes detected in at least 1% of the samples.
# Also manually include the negative control probe, for downstream use
rate_cutoff <- 0.01

negativeProbefData <- subset(fData(gene_dat), CodeClass == "Negative")
neg_probes         <- unique(negativeProbefData$TargetName)

gene_dat <- gene_dat[fData(gene_dat)$DetectionRate >= rate_cutoff | fData(gene_dat)$TargetName %in% neg_probes, ]

# Background normalization for WTA/CTA without custom spike-in
gene_dat <- gene_dat %>%
  normalize(norm_method = "neg", fromElt = "exprs", toElt = "neg_norm")

# Q3 norm (75th percentile) for WTA/CTA  with or without custom spike-ins
gene_dat <- gene_dat %>%
  normalize(norm_method = "quant", desiredQuantile = 0.75, toElt = "q_norm")

# update defaults for umap to contain a stable random_state (seed)
custom_umap <- umap::umap.defaults

custom_umap$random_state <- 42

# run UMAP
# exclude data for Ag tags and control tags
exclude <- rownames(gene_dat)
exclude <- grepl("^(Custom|Barcode)|_CTL$", exclude)
u_dat   <- gene_dat[!exclude, ]

umap_out <- umap(
  t(log2(assayDataElement(u_dat, elt = "q_norm"))),  
  config = custom_umap
)

pData(gene_dat)[, c("UMAP1", "UMAP2")] <- umap_out$layout[, c(1,2)]
```

```{r "GEOMX PLOT DATA", eval = FALSE}
# THIS SYNTAX DOES NOT WORK, WHY?
# pData(plt_dat)[, "Barcode01"] <- gn_dat$Barcode01
rf_gns <- c(
  # rf_feats$cLEC$low
  "Barcode01", "Barcode02",
  "ova-3wk", "ova-6wk",
  rf_feats$cLEC$high
)

rf_gns <- rf_gns[rf_gns %in% rownames(gene_dat)]

gn_dat <- gene_dat[rf_gns, ] %>%
  assayDataElement(elt = "q_norm") %>%
  # assayDataElement(elt = "neg_norm") %>%
  # assayDataElement(elt = "exprs") %>%
  t() %>%
  as_tibble(rownames = "dcc")

plt_dat <- pData(gene_dat) %>%
  as_tibble(rownames = "dcc") %>%
  left_join(gn_dat, by = "dcc")

# Format segment names
plt_dat <- plt_dat %>%
  mutate(
    LN = str_remove(LN, "(?<=LN) "),
    sample = str_remove(LN, " LN[0-9]+$"),
    sample = str_remove(sample, "^ova-GeoMX\\+VV "),
    sample = str_replace_all(sample, "week", "wk"),
    sample = str_replace_all(sample, " and ", "-"),
    rep    = str_extract(LN, "LN[ 0-9]+"),
    rep    = str_remove_all(rep, " "),
    slide  = as.numeric(str_extract(`Slide Name`, "[0-9]+$"))
  )

# Set sample colors
sam_cols <- c("naive", "3wk", "6wk", "6wk-3wk")

sam_cols <- set_names(
  ito_cols[seq_along(sam_cols)],
  sam_cols
)

ag_cols <- c(`ova-3wk` = "#56B4E9", `ova-6wk` = "#6A51A3")

# Assign barcode signals
bc_key <- plt_dat %>%
  filter(sample %in% c("3wk", "6wk")) %>%
  group_by(slide, sample) %>%
  summarize(across(starts_with("Barcode0"), mean), .groups = "drop") %>%
  pivot_longer(starts_with("Barcode0")) %>%
  group_by(slide, sample) %>%
  filter(value == max(value)) %>%
  split(.$slide) %>%
  map(~ set_names(.x$name, .x$sample))

plt_dat <- plt_dat %>%
  group_by(slide) %>%
  rowwise() %>%
  mutate(
    `ova-3wk` = bc_key[[as.character(slide)]][["3wk"]],
    `ova-6wk` = bc_key[[as.character(slide)]][["6wk"]]
  ) %>%
  ungroup() %>%
  mutate(
    across(matches("ova-[0-9]+wk"), ~ {
      case_when(
        .x == "Barcode01" ~ Barcode01,
        .x == "Barcode02" ~ Barcode02
      )
    }),
    # Segment = fct_relevel(Segment, names(seg_cols)),
    sample  = fct_relevel(sample, names(sam_cols))
  )

plt_dat <- plt_dat %>%
  rowwise() %>%
  mutate(top_ova = max(c(`ova-3wk`, `ova-6wk`))) %>%
  ungroup()
```

```{R "GEOMX PLOTS", eval = FALSE}
# Boxplots
# uses all segments regardless of Lyve1/CD11c/CD4
bx_gns <- rf_gns

bx <- plt_dat %>%
  pivot_longer(all_of(bx_gns)) %>%
  mutate(name = fct_relevel(name, bx_gns)) %>%
  ggplot(aes(region_2, value, fill = region_2, color = region_2)) +
  geom_boxplot(alpha = 0.5, key_glyph = draw_key_point) +
  facet_wrap(~ name, scales = "free_y") +
  # scale_color_manual(values = seg_cols) +
  # scale_fill_manual(values = seg_cols) +
  # labs(x = "Segment type", y = "Normalized expression") +
  base_theme +
  theme(
    aspect.ratio    = 1.2,
    legend.position = "none",
    axis.text.x     = element_text(angle = 45, hjust = 1)
  )

# Heatmap
# ht_dat %>%
#   ggplot(aes(region_2, name, fill = value)) +
#   geom_tile() +
#   scale_fill_gradientn(colours = c("lightblue", "white", "red")) +
#   base_theme
ht_dat <- plt_dat %>%
  pivot_longer(all_of(rf_gns)) %>%
  filter(Segment == "Lyve1", region_2 != "lymphatics") %>%
  group_by(name, region_2) %>%
  summarize(value = mean(value), .groups = "drop") %>%
  group_by(name) %>%
  mutate(value = as.numeric(scale(value))) %>%
  ungroup()

# Complex Heatmap
ht <- ht_dat %>%
  pivot_wider(names_from = name, values_from = value) %>%
  column_to_rownames("region_2") %>%
  as.matrix() %>%
  t() %>%
  Heatmap()

# Scatter plots
sc_gns <- rf_gns[!grepl("^ova-|Barcode", rf_gns)]

sc <- plt_dat %>%
  pivot_longer(all_of(sc_gns)) %>%
  filter(Segment == "Lyve1") %>%
  
  ggplot(aes(`ova-6wk`, value, color = name)) +
  geom_point() +
  facet_wrap(~ name, scales = "free") +
  scale_x_log10() +
  base_theme +
  theme(
    aspect.ratio = 1,
    axis.text.x  = element_text(angle = 45, hjust = 1)
  )

# UMAP projections
plt_dat %>%
  plot_scatter(
    "Segment",
    "UMAP1", "UMAP2",
    size = 3
  ) +
  theme(aspect.ratio = 1)
```

```{r "GEOMX HEATMAP", fig.width = 5, fig.height = 5, eval = FALSE}
 ht
```

```{r "GEOMX SLIDES", eval = FALSE}
library(SpatialOmicsOverlay)

readSpatialOverlay(
  ometiff = "~/Dropbox/Ryan/Projects/tamburini-antigen-tracking/data/230623_A00405_0707_AHCK7MDSX7/images/Slide 1.png",
  annots = "~/Projects/tamburini-antigen-tracking/results/2023-06-23/geomx_worksheets/Tamburini_061423_20230619T1930_annotations_merged.xlsx"
)


```

---

<br>

<br>

## Figure 5

Antigen archiving is impaired during CHIKV infection

A.  UMAP projections show LEC subsets for mock and CHIKV-infected mice.
B.  UMAP projections show predicted Ag-competent cLECs for mock and
    CHIKV-infected mice.
C.  The fraction of predicted Ag-competent cLECs is shown for mock and
    CHIKV-infected mice for each biological replicate.
D.  UMAP projections show cLEC Ag-high module scores for mock and CHIKV-infected
    mice.
E.  cLEC Ag-high module scores are shown for mock and CHIKV-infected mice for
    each biological replicate.
F.  Expression of cLEC Ag-high gene module is shown for mock and CHIKV-infected
    mice for top LEC subsets.
    
<u>Notes</u>

* Need to add p values for C and E.
* Predicted Ag-competent fLECs and collecting LECs can be shown as a
  supplemental figure (no difference between mock and CHIKV).

```{r "CHIKV data"}
# Load CHIKV data
rm(lec_so, dc_so)

chikv_so <- qread(here(params$chikv_dir, "so_lec.qs"))
  
chikv_so <- chikv_so %>%
  subset(tm == "24hpi") %>%
  AddMetaData(chikv_preds, col.name = str_c("rf_", colnames(chikv_preds)))

chikv_so <- chikv_so %>%
  mutate_meta(
    mutate,
    pred_grp = dplyr::recode(rf_pred, high = "Ag-competent", low = "Ag-low"),
    treatment = fct_relevel(treatment, c("mock", "CHIKV"))
  )

# Add module scores
rf_feats %>%
  iwalk(~ {
    clmn_nms <- str_c(.y, "_", names(.x))

    clmn_nms <- set_names(
      str_c(clmn_nms, seq_along(clmn_nms)),
      clmn_nms
    )

    chikv_so <<- chikv_so %>%
      AddModuleScore(
        features = .x,
        name     = names(clmn_nms)
      ) %>%
      mutate_meta(dplyr::rename, !!!syms(clmn_nms))
  })
```

```{r "CHIKV figure", fig.width = 15, fig.height = 19}
# CHIKV color palette
chikv_clrs <- c(`Ag-low` = "#56B4E9", `Ag-competent` = "#D7301F")
treat_clrs <- set_names(chikv_clrs, c("mock", "CHIKV"))

chikv_u_theme <- umap_theme +
  theme(aspect.ratio = 0.7)

# LEC subsets
top <- chikv_so %>%
  plot_scatter(
    "lec_subtype",
    x            = "hUMAP_1",
    y            = "hUMAP_2",
    group_col    = "treatment",
    plot_colors  = lec_clrs,
    size         = 0.75,
    label_params = list(size = ttl_pt2)
  ) +
  chikv_u_theme +
  theme(
    legend.title = element_blank()
  )

# Ag class UMAPs
clmns <- c(
  "treatment", "rep", "lec_subtype", "pred_grp",
  "hUMAP_1", "hUMAP_2",
  "cLEC_high", "cLEC_low"
)

plt_dat <- chikv_so %>%
  FetchData(clmns) %>%
  mutate(
    treatment = fct_relevel(treatment, c("mock", "CHIKV")),
    pred_grp  = fct_relevel(pred_grp, rev(names(chikv_clrs)))
  )

ag_class_u <- plt_dat %>%
  ggplot(aes(hUMAP_1, hUMAP_2, fill = pred_grp)) +
  geom_point_trace(
    trace_position = lec_subtype == "cLEC",
    background_params = list(fill = "grey85", color = NA),
    color = NA,
    size = 0.75
  ) +
  facet_wrap(~ treatment) +
  scale_fill_manual(values = chikv_clrs) +
  guides(fill = guide_legend(override.aes = list(size = 4))) +
  chikv_u_theme +
  theme(
    legend.position = "bottom",
    legend.title = element_blank()
  )

# Ag class bargraphs
bar_dat <- plt_dat %>%
  filter(lec_subtype %in% rf_typs) %>%
  group_by(treatment, rep, lec_subtype) %>%
  summarize(
    frac_high = sum(pred_grp == "Ag-competent") / n(),
    .groups = "drop"
  )

ag_class_bars <- bar_dat %>%
  filter(lec_subtype == "cLEC") %>%
  ggplot(aes(treatment, frac_high, fill = treatment, alpha = rep)) +
  geom_col(position = position_dodge2(padding = 0.15)) +
  # facet_wrap(~ lec_subtype) +
  scale_fill_manual(values = treat_clrs) +
  scale_alpha_manual(values = rep(1, n_distinct(plt_dat$rep))) +
  scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(0, 1)) +
  guides(alpha = "none") +
  labs(y = "cLEC fraction Ag-competent") +
  base_theme +
  theme(
    # aspect.ratio    = 1.1,
    legend.position = "none",
    axis.title.x    = element_blank(),
    axis.text.x     = element_text(size = ttl_pt2)
  )

mid <- plot_grid(
  ag_class_u, ag_class_bars,
  align      = "h",
  axis       = "tb",
  rel_widths = c(1, 0.3), 
  labels     = c("B", "C"),
  label_size = ttl_pt2 * 1.5
)

# cLEC Ag-high module scores
ag_score_u <- plt_dat %>%
  plot_scatter(
    "cLEC_high",
    x           = "hUMAP_1",
    y           = "hUMAP_2",
    group_col   = "treatment",
    plot_colors = c("#56B4E9", "white", "#D7301F"),
    outline     = TRUE,
    # size        = 0.5,
    size        = 0.75,
    stroke      = 0.7,
    n_label     = "none"
  ) +
  guides(fill = guide_colorbar(
    title = "cLEC\nAg-high module",
    title.hjust = 0.5,
    barwidth = unit(150, "pt"),
    barheight = unit(10, "pt")
  )) +
  chikv_u_theme +
  theme(
    legend.position = "bottom"
    # legend.title = element_blank()
  )

ag_score_bx <- plt_dat %>%
  filter(lec_subtype == "cLEC") %>%
  ggplot(aes(treatment, cLEC_high, fill = treatment, alpha = rep)) +
  geom_boxplot(outlier.size = 0.5) +
  scale_fill_manual(values = treat_clrs) +
  scale_alpha_manual(values = rep(1, n_distinct(plt_dat$rep))) +
  scale_y_continuous(expand = expansion(c(0.05, 0.2))) +
  guides(alpha = "none") +
  labs(y = "cLEC Ag-high module") +
  base_theme +
  theme(
    # aspect.ratio    = 1.1,
    legend.position = "none",
    axis.title.x    = element_blank(),
    axis.text.x     = element_text(size = ttl_pt2)
  )

bot <- plot_grid(
  ag_score_u, ag_score_bx,
  align = "h",
  axis  = "tb",
  rel_widths = c(1, 0.3),
  labels     = c("D", "E"),
  label_size = ttl_pt2 * 1.5
)

# Ag-high heatmap data
heat_feats <- rf_feats$cLEC$high
heat_feats <- heat_feats[heat_feats %in% rownames(chikv_so)]

heat_dat <- chikv_so %>%
  FetchData(c("treatment", "lec_subtype", heat_feats)) %>%
  filter(lec_subtype %in% rf_typs) %>%
  # filter(!lec_subtype %in% c("BEC", "unassigned")) %>%
  pivot_longer(all_of(heat_feats)) %>%
  group_by(lec_subtype, treatment, name) %>%
  summarize(value = mean(value), .groups = "drop")

gn_lvls <- heat_dat %>%
  pivot_wider(names_from = treatment, values_from = value) %>%
  mutate(value = CHIKV / mock, .keep = "unused") %>%
  pivot_wider(names_from = lec_subtype, values_from = value) %>%
  arrange(desc(!!sym(rf_typs[1])), desc(!!sym(rf_typs[2])), desc(!!sym(rf_typs[3]))) %>%
  pull(name)
  
heat_dat <- heat_dat %>%
  group_by(name) %>%
  mutate(value = as.numeric(scale(value))) %>%
  ungroup() %>%
  mutate(
    treatment   = fct_relevel(treatment, c("mock", "CHIKV")),
    name        = fct_relevel(name, rev(gn_lvls)),
    lec_subtype = fct_relevel(lec_subtype, rev(rf_typs))
  )

# Ag heatmap
heat <- heat_dat %>%
  ggplot(aes(name, lec_subtype, fill = value)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("#56B4E9", "white", "#D7301F")) +
  facet_wrap(~ treatment, ncol = 1, strip.position = "left") +
  guides(fill = guide_colorbar(
    barwidth = unit(150, "pt"),
    barheight = unit(10, "pt")
  )) +
  base_theme +
  theme(
    aspect.ratio      = length(rf_typs) / length(heat_feats),
    strip.placement   = "outside",
    strip.clip        = "off",
    legend.position   = "bottom",
    legend.title      = element_blank(),
    axis.title        = element_blank(),
    axis.text.x       = element_text(angle = 90, hjust = 1, vjust = 0.5)
  )

# Final figure
plot_grid(
  top, mid, bot, heat,
  ncol       = 1,
  labels     = c("A", "", "", "F"),
  label_size = ttl_pt2 * 1.5
)
```

---

<br>

<br>

## Session info

```{r "session info"}
sessionInfo()
```



```{r "RF TEST 1", eval = FALSE}
# GOALS
# 1. train model for LEC subsets (~80% accuracy for cLECs)
# 2. train general model for all LECs (~60-90% for all subsets)

# Identify Ag markers for all LECs
degs_1 <- lec_so %>%
  wilcoxauc(group_by = "Ag_class") %>%
  filter(
    padj < 0.05, logFC > 0.25,
    !grepl("^(mt-|R[pls])", feature)
  ) %>%
  arrange(padj)

# Identify Ag markers for subtype
degs_2 <- unique(lec_so$subtype) %>%
  map_dfr(~ {
    res <- lec_so %>%
      subset(subtype == .x) %>%
      wilcoxauc(group_by = "Ag_class")

    if (nrow(res) == 0) return(NULL)

    res %>%
      filter(
        padj < 0.05, logFC > 0.25,
        !grepl("^(mt-|R[pls])", feature)
      ) %>%
      mutate(subtype = .x) %>%
      arrange(padj)
  })

# Identify markers for subtype and mouse
degs_3 <- lec_so@meta.data %>%
  distinct(subtype, mouse) %>%
  pmap_dfr(~ {
    res <- lec_so %>%
      subset(subtype == ..1 & mouse == ..2)
    
    if (n_distinct(res$Ag_class) < 2) return(NULL)
    
    res <- res %>%
      wilcoxauc(group_by = "Ag_class")

    if (nrow(res) == 0) return(NULL)

    res %>%
      filter(
        padj < 0.05, logFC > 0.25,
        !grepl("^(mt-|R[pls])", feature)
      ) %>%
      mutate(subtype = ..1, mouse = ..2) %>%
      arrange(padj)
  })

degs <- bind_rows(degs_1, degs_2, degs_3) %>%
  distinct(feature, group)

# Format RF data
# cat_feats <- c("tm", "subtype", "Ag_class")
# dat_col   <- "Ag_score"
cat_feats <- c("tm", "mouse", "subtype")
feats     <- unique(degs$feature)
dat_col   <- "Ag_class"

rf_dat <- lec_so %>%
  FetchData(c(dat_col, cat_feats, feats)) %>%
  mutate(!!sym(dat_col) := fct_relevel(!!sym(dat_col), "high", "low")) %>%
  rename_with(~ str_replace_all(.x, "-", "_")) %>%
  rename_with(~ str_replace_all(.x, "\\(|\\)", "_")) %>%
  rename_with(~ str_c("x", .x), matches("^[0-9]"))

set.seed(42)

rf_split <- rf_dat %>%
  initial_split(prop = 0.1)

rf_test <- testing(rf_split)

rf_train <- training(rf_split) %>%
  dplyr::select(-all_of(cat_feats))

# Tune model
tune_rf <- function(dat_df, dat_col, param_lst, return_best = TRUE) {
  
  rf_params <- expand.grid(param_lst)
  
  res <- pmap_dfr(rf_params, ~ {
    res <- ranger(
      as.formula(str_c(dat_col, " ~ .")),
      data       = dat_df,
      importance = "impurity",
      seed       = 42,
      ...
    )
    
    err <- sqrt(res$prediction.error)
    
    tibble(..., oob_rmse = err)
  })
  
  if (return_best) {
    res <- res %>%
      dplyr::filter(oob_rmse == min(oob_rmse)) %>%
      dplyr::select(-oob_rmse) %>%
      head(1) %>%
      as.list()
  }
  
  res
}

param_lst <- list(
  mtry            = seq(1, ncol(rf_train), 50),
  min.node.size   = c(1:3, 5, 10)
)

rf_params <- rf_train %>%
  tune_rf(
    dat_col   = "Ag_class",
    param_lst = param_lst
  )

# Fit RF
rf_mod <- pmap(rf_params, ~ {
  ranger(
    Ag_class ~ .,
    data       = rf_train,
    importance = "impurity",
    ...
  )
}) %>%
  pluck(1)

# Get predictions
prds <- rf_mod %>%
  predict(rf_test) %>%
  predictions()

rf_res <- rf_test %>%
  as_tibble(rownames = ".cell_id") %>%
  dplyr::select(.cell_id, subtype, tm, mouse, Ag_class) %>%
  mutate(
    prediction = prds,
    correct    = Ag_class == prediction
  ) %>%
  group_by(subtype, mouse) %>%
  summarize(
    n       = n(),
    correct = sum(correct) / n,
    .groups = "drop"
  ) %>%
  arrange(desc(correct))

rf_res %>%
  mutate(lab = as.character(round(correct, 2))) %>%
  ggplot(aes(subtype, mouse, fill = correct)) +
  geom_tile() +
  geom_text(aes(label = lab)) +
  scale_fill_gradientn(colours = c("lightblue", "red")) +
  djvdj_theme()

# Get top regression variables
rf_marks <- tibble(
  gene   = names(rf_mod$variable.importance),
  impact = rf_mod$variable.importance
) %>%
  mutate(
    rank = row_number(desc(impact)),
    gene = ifelse(gene %in% cat_feats, gene, str_replace(gene, "_", "-"))
  ) %>%
  arrange(rank) %>%

  left_join(degs, by = c(gene = "feature"))

# # Create boxplots
# n_gns <- 50
# 
# plt_gns <- rf_marks %>%
#   filter(!gene %in% c("Ag-score", "Ag_class", "orig.ident", "tm")) %>%
#   pull(gene) %>%
#   head(n_gns) %>%
#   c("Ag_score", .)
# 
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns)) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
#   
# plt_dat %>%
#   ggplot(aes(tm, value, fill = name, color = name, alpha = Ag_class)) +
#   geom_boxplot() +
#   facet_wrap(~ name, scales = "free_y") +
#   scale_alpha_manual(values = c(low = 0.25, high = 0.75)) +
#   djvdj_theme()
# 
# # Create scatter plots
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns[-1])) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
# 
# plt_dat %>%
#   ggplot(aes(Ag_score, value, color = Ag_class)) +
#   geom_point(size = 0.1) +
#   facet_wrap(~ name, scales = "free") +
#   djvdj_theme() +
#   theme(aspect.ratio = 1)
```

```{r "RF TEST 2", eval = FALSE}
# GOALS
# 1. train model for LEC subsets for Ag groups (1-4)
# 2. identify cells in mock that are predicted to take up Ag

# Identify Ag markers for all LECs for each mouse
deg_so <- lec_so %>%
  SplitObject("mouse")

plan(multisession, workers = min(length(deg_so), 8))

degs_1 <- deg_so %>%
  future_imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = "Ag_group")

    if (nrow(res) == 0) return(NULL)

    res %>%
      filter(
        padj < 0.05, logFC > 0.25,
        !grepl("^(mt-|R[pls])", feature)
      ) %>%
      mutate(mouse = .y) %>%
      arrange(padj)
  })

# Identify Ag markers for subtype
deg_so <- lec_so %>%
  SplitObject("subtype")

plan(multisession, workers = min(length(deg_so), 8))

degs_2 <- deg_so %>%
  future_imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = "Ag_group")

    if (nrow(res) == 0) return(NULL)

    res %>%
      filter(
        padj < 0.05, logFC > 0.25,
        !grepl("^(mt-|R[pls])", feature)
      ) %>%
      mutate(subtype = .y) %>%
      arrange(padj)
  })

rm(deg_so)

# Identify markers for subtype and mouse
degs_3 <- lec_so@meta.data %>%
  distinct(subtype, mouse) %>%
  pmap_dfr(~ {
    res <- lec_so %>%
      subset(subtype == ..1 & mouse == ..2)
    
    if (n_distinct(res$Ag_class) < 2) return(NULL)
    
    res <- res %>%
      wilcoxauc(group_by = "Ag_group")

    if (nrow(res) == 0) return(NULL)

    res %>%
      filter(
        padj < 0.05, logFC > 0.25,
        !grepl("^(mt-|R[pls])", feature)
      ) %>%
      mutate(subtype = ..1, mouse = ..2) %>%
      arrange(padj)
  })

degs <- bind_rows(degs_1, degs_2, degs_3) %>%
  distinct(feature, group)

# Format RF data
dat_col   <- "Ag_group"
cat_feats <- c("tm", "mouse", "subtype")
feats     <- unique(degs$feature)

rf_dat <- lec_so %>%
  FetchData(c(dat_col, cat_feats, feats)) %>%
  rename_with(~ str_replace_all(.x, "-", "_")) %>%
  rename_with(~ str_replace_all(.x, "\\(|\\)", "_")) %>%
  rename_with(~ str_c("x", .x), matches("^[0-9]"))

set.seed(42)

rf_split <- rf_dat %>%
  initial_split(prop = 0.1)

rf_test <- testing(rf_split)

rf_train <- training(rf_split) %>%
  dplyr::select(-all_of(cat_feats))

# Tune model
tune_rf <- function(dat_df, dat_col, param_lst, return_best = TRUE) {
  
  rf_params <- expand.grid(param_lst)
  
  res <- pmap_dfr(rf_params, ~ {
    res <- ranger(
      as.formula(str_c(dat_col, " ~ .")),
      data       = dat_df,
      importance = "impurity",
      seed       = 42,
      ...
    )
    
    err <- sqrt(res$prediction.error)
    
    tibble(..., oob_rmse = err)
  })
  
  if (return_best) {
    res <- res %>%
      dplyr::filter(oob_rmse == min(oob_rmse)) %>%
      dplyr::select(-oob_rmse) %>%
      head(1) %>%
      as.list()
  }
  
  res
}

param_lst <- list(
  mtry            = seq(1, ncol(rf_train), 100),
  min.node.size   = c(1:3, 5, 10)
)

rf_params <- rf_train %>%
  tune_rf(
    dat_col   = "Ag_group",
    param_lst = param_lst
  )

# Fit model
rf_mod <- pmap(rf_params, ~ {
  ranger(
    Ag_class ~ .,
    data       = rf_train,
    importance = "impurity",
    ...
  )
}) %>%
  pluck(1)

# Get predictions
prds <- rf_mod %>%
  predict(rf_test) %>%
  predictions()

rf_res <- rf_test %>%
  as_tibble(rownames = ".cell_id") %>%
  dplyr::select(.cell_id, subtype, tm, mouse, Ag_class) %>%
  mutate(
    prediction = prds,
    correct    = Ag_class == prediction
  ) %>%
  group_by(subtype, mouse) %>%
  summarize(
    n       = n(),
    correct = sum(correct) / n,
    .groups = "drop"
  ) %>%
  arrange(desc(correct))

rf_res %>%
  mutate(lab = as.character(round(correct, 2))) %>%
  ggplot(aes(subtype, mouse, fill = correct)) +
  geom_tile() +
  geom_text(aes(label = lab)) +
  scale_fill_gradientn(colours = c("lightblue", "red")) +
  djvdj_theme()

# Get top regression variables
rf_marks <- tibble(
  gene   = names(rf_mod$variable.importance),
  impact = rf_mod$variable.importance
) %>%
  mutate(
    rank = row_number(desc(impact)),
    gene = ifelse(gene %in% cat_feats, gene, str_replace(gene, "_", "-"))
  ) %>%
  arrange(rank) %>%

  left_join(degs, by = c(gene = "feature"))



# # Create boxplots
# n_gns <- 50
# 
# plt_gns <- rf_marks %>%
#   filter(!gene %in% c("Ag-score", "Ag_class", "orig.ident", "tm")) %>%
#   pull(gene) %>%
#   head(n_gns) %>%
#   c("Ag_score", .)
# 
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns)) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
#   
# plt_dat %>%
#   ggplot(aes(tm, value, fill = name, color = name, alpha = Ag_class)) +
#   geom_boxplot() +
#   facet_wrap(~ name, scales = "free_y") +
#   scale_alpha_manual(values = c(low = 0.25, high = 0.75)) +
#   djvdj_theme()
# 
# # Create scatter plots
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns[-1])) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
# 
# plt_dat %>%
#   ggplot(aes(Ag_score, value, color = Ag_class)) +
#   geom_point(size = 0.1) +
#   facet_wrap(~ name, scales = "free") +
#   djvdj_theme() +
#   theme(aspect.ratio = 1)
```

```{r "RF TEST 3", eval = FALSE}
# GOALS
# 1. train model for LEC subsets for Ag groups (1-4)
# 2. identify cells in mock that are predicted to take up Ag

# Identify DEGs ----
# Identify Ag markers for all LECs for each mouse
grp <- "Ag_class"
fc  <- 0

deg_so <- lec_so %>%
  SplitObject("mouse")

degs_1 <- deg_so %>%
  imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = "all", mouse = .y) %>%
      arrange(padj)
  })

# Identify Ag markers for subtype
deg_so <- lec_so %>%
  SplitObject("subtype")

degs_2 <- deg_so %>%
  imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = .y, mouse = "all") %>%
      arrange(padj)
  })

rm(deg_so)

# Identify markers for subtype and mouse
degs_3 <- lec_so@meta.data %>%
  distinct(subtype, mouse) %>%
  pmap_dfr(~ {
    res <- lec_so %>%
      subset(subtype == ..1 & mouse == ..2)
    
    if (n_distinct(res$Ag_class) < 2) return(NULL)
    
    res <- res %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = ..1, mouse = ..2) %>%
      arrange(padj)
  })

degs <- bind_rows(degs_1, degs_2, degs_3) %>%
  filter(
    padj < 0.05, logFC > fc,
    !grepl("^(mt-|R[pls])", feature)
  )

# Format data ----
# dat_col    <- "Ag_group"
dat_col    <- "Ag_class_2"
feats      <- unique(degs$feature)
other_vars <- NULL

feats <- degs %>%
  filter(subtype == "cLEC", mouse == "d14") %>%
  pull(feature) %>%
  unique()

other_info <- c("tm", "subtype", "mouse")

rf_dat <- lec_so %>%
  subset(subtype == "cLEC" & mouse == "d14") %>%

  FetchData(unique(c(dat_col, other_info, other_vars, feats))) %>%
  
  # mutate(across(all_of(feats), ~ as.numeric(scale(.x)))) %>%
  # mutate(!!sym(dat_col) := factor(!!sym(dat_col))) %>%
  # mutate(
  #   across(all_of(c(dat_col, other_vars)), ~ as.numeric(factor(.x)))
  # ) %>%
  
  rename_with(~ str_replace_all(.x, "-", "_")) %>%
  rename_with(~ str_replace_all(.x, "\\(|\\)", "_")) %>%
  rename_with(~ str_c("x", .x), matches("^[0-9]"))

# Adjust groups so they are balanced
sam_sz <- rf_dat %>%
  group_by(!!sym(dat_col), mouse) %>%
  summarize(n = n(), .groups = "drop") %>%
  pull(n)

# sam_sz <- max(100, min(sam_sz))
sam_sz <- 135

# Split into train/test data
# rf_split <- initial_split(rf_dat, prop = 0.5)
# rf_train <- training(rf_split)
# rf_test  <- testing(rf_split)

rf_train <- rf_dat %>%
  rownames_to_column(".cell_id") %>%
  group_by(!!sym(dat_col), mouse) %>%
  filter(n() >= sam_sz) %>%
  dplyr::sample_n(sam_sz) %>%
  ungroup() %>%
  dplyr::select(-all_of(other_info[!other_info %in% other_vars])) %>%
  column_to_rownames(".cell_id")

rf_test <- rf_dat[!rownames(rf_dat) %in% rownames(rf_train), ]

# Tune RF model ----
tune_rf <- function(dat_df, dat_col, param_lst, return_best = TRUE) {
  
  rf_params <- expand.grid(param_lst)
  
  res <- pmap_dfr(rf_params, ~ {
    res <- ranger(
      as.formula(str_c("factor(", dat_col, ") ~ .")),
      data       = dat_df,
      importance = "impurity",
      seed       = 42,
      ...
    )
    
    err <- sqrt(res$prediction.error)
    
    tibble(..., oob_rmse = err)
  })
  
  if (return_best) {
    res <- res %>%
      dplyr::filter(oob_rmse == min(oob_rmse)) %>%
      dplyr::select(-oob_rmse) %>%
      head(1) %>%
      as.list()
  }
  
  res
}

param_lst <- list(
  # mtry          = c(10, 50, 200, 300),
  mtry          = seq(10, 60, 10),
  min.node.size = c(1, 2, 5),
  num.trees     = 1000
)

rf_params <- rf_train %>%
  tune_rf(
    dat_col   = dat_col,
    param_lst = param_lst
  )

# rf_params <- list(
#   mtry = 10,
#   min.node.size = 1,
#   num.trees = 1000
#   # max.depth = 5
# )

# Fit RF model
rf_mod <- pmap(rf_params, ~ {
  ranger(
    as.formula(str_c("factor(", dat_col, ") ~ .")),
    data       = rf_train,
    importance = "impurity",
    ...
  )
}) %>%
  pluck(1)

# Identify top features and re-fit
top_feats <- importance(rf_mod) %>%
  sort(decreasing = TRUE)

top_feats <- top_feats[top_feats > 0]

rf_params$mtry <- min(length(top_feats), rf_params$mtry)

rf_mod_top <- pmap(rf_params, ~ {
  ranger(
    as.formula(str_c("factor(", dat_col, ") ~ .")),
    data       = rf_train[, c(dat_col, names(top_feats))],
    importance = "impurity",
    ...
  )
}) %>%
  pluck(1)

# prds <- rf_mod_top %>%
prds <- rf_mod %>%
  predict(rf_test) %>%
  predictions()

# Try logit regression ----
logit_mod <- rf_train %>%
  mutate(
    Ag_class = ifelse(Ag_class == "high", 1, 0),
    Ag_class = factor(Ag_class)
  ) %>%
  glm(
    data = .,
    # as.factor(Ag_class) ~ Cav1 + Marcks | as.numeric(as.factor(mouse)),
    # as.factor(Ag_class) ~ Cav1 + Marcks,
    Ag_class ~ .,
    control = glm.control(maxit = 200),
    family = binomial
  )

prob <- logit_mod %>%
  predict(newdata = rf_test, type = "response")

prds <- vector("character", length(prob))
prds[prob > 0.7] <- "high"
prds[prob <= 0.7] <- "low"

# Try glmer ----
# library(lme4)
# 
# logit_mod <- rf_train %>%
#   mutate(
#     Ag_class = ifelse(Ag_class == "high", 1, 0),
#     Ag_class = factor(Ag_class)
#   ) %>%
#   glmer(
#     data = .,
#     # as.factor(Ag_class) ~ Cav1 + Marcks | as.numeric(as.factor(mouse)),
#     # as.factor(Ag_class) ~ Cav1 + Marcks,
#     # Ag_class ~ (. | factor(mouse)),
#     # control = glm.control(maxit = 200),
#     Ag_class ~ (. | factor(exp)),
#     family = binomial
#   )
# 
# prob <- logit_mod %>%
#   predict(newdata = rf_test, type = "response")
# 
# prds <- vector("character", length(prob))
# prds[prob > 0.5] <- "high"
# prds[prob <= 0.5] <- "low"


# Get predictions ----
rf_res <- rf_test %>%
  as_tibble(rownames = ".cell_id") %>%
  dplyr::select(.cell_id, subtype, all_of(c(dat_col, other_vars, other_info))) %>%
  mutate(
    prediction = prds,
    correct    = !!sym(dat_col) == prediction
  ) %>%
  # group_by(subtype, !!sym(other_vars)) %>%
  group_by(!!!syms(c(other_info, dat_col))) %>%
  summarize(
    n       = n(),
    correct = sum(correct) / n,
    .groups = "drop"
  ) %>%
  arrange(desc(correct))

# Plot results
rf_res %>%
  mutate(lab = as.character(round(correct, 2))) %>%
  ggplot(aes(subtype, mouse, fill = correct)) +
  geom_tile() +
  geom_text(aes(label = lab)) +
  facet_wrap(~ Ag_class_2) +
  scale_fill_gradientn(colours = c("lightblue", "red")) +
  djvdj_theme()

# Get top regression variables
rf_marks <- tibble(
  gene   = names(rf_mod$variable.importance),
  impact = rf_mod$variable.importance
) %>%
  mutate(
    rank = row_number(desc(impact)),
    gene = ifelse(gene %in% cat_feats, gene, str_replace(gene, "_", "-"))
  ) %>%
  arrange(rank) %>%

  left_join(degs, by = c(gene = "feature"))



# # Create boxplots
# n_gns <- 50
# 
# plt_gns <- rf_marks %>%
#   filter(!gene %in% c("Ag-score", "Ag_class", "orig.ident", "tm")) %>%
#   pull(gene) %>%
#   head(n_gns) %>%
#   c("Ag_score", .)
# 
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns)) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
#   
# plt_dat %>%
#   ggplot(aes(tm, value, fill = name, color = name, alpha = Ag_class)) +
#   geom_boxplot() +
#   facet_wrap(~ name, scales = "free_y") +
#   scale_alpha_manual(values = c(low = 0.25, high = 0.75)) +
#   djvdj_theme()
# 
# # Create scatter plots
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns[-1])) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
# 
# plt_dat %>%
#   ggplot(aes(Ag_score, value, color = Ag_class)) +
#   geom_point(size = 0.1) +
#   facet_wrap(~ name, scales = "free") +
#   djvdj_theme() +
#   theme(aspect.ratio = 1)
```

```{r "RF d14 TEST", eval = FALSE}
# GOALS
# 1. train model for LEC subsets for Ag groups (1-4)
# 2. identify cells in mock that are predicted to take up Ag

# Identify DEGs ----
# Identify Ag markers for all LECs for each mouse
grp <- "Ag_class_2"
fc  <- 0

deg_so <- lec_so %>%
  SplitObject("mouse")

degs_1 <- deg_so %>%
  imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = "all", mouse = .y) %>%
      arrange(padj)
  })

# Identify Ag markers for subtype
deg_so <- lec_so %>%
  SplitObject("subtype")

degs_2 <- deg_so %>%
  imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = .y, mouse = "all") %>%
      arrange(padj)
  })

rm(deg_so)

# Identify markers for subtype and mouse
degs_3 <- lec_so@meta.data %>%
  distinct(subtype, mouse) %>%
  pmap_dfr(~ {
    res <- lec_so %>%
      subset(subtype == ..1 & mouse == ..2)
    
    if (n_distinct(res[[grp]]) < 2) return(NULL)
    
    res <- res %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = ..1, mouse = ..2) %>%
      arrange(padj)
  })

degs <- bind_rows(degs_1, degs_2, degs_3) %>%
  filter(
    padj < 0.05, logFC > fc,
    !grepl("^(mt-|R[pls])", feature),
    feature != "Malat1"
  )

degs %>%
  write_tsv(here(params$so_dir, "ag-high_degs_Ag_class_2.tsv.gz"))

# Most common degs
top_degs <- degs %>%
  group_by(subtype, feature, group) %>%
  summarize(n = n_distinct(mouse), .groups = "drop") %>%
  arrange(desc(n)) %>%
  filter(n > 1) %>%
  filter(!subtype %in% c("all", "BEC", "unassigned")) %>%
  group_by(feature, group) %>%
  summarize(n_typ = n_distinct(subtype), .groups = "drop") %>%
  arrange(desc(n_typ)) %>%
  filter(n_typ > 1)

# Format data ----
dat_col    <- "Ag_class_2"
feats      <- unique(degs$feature)
other_vars <- NULL

# feats <- degs %>%
#   filter(subtype == "cLEC", mouse == "d14") %>%
#   pull(feature) %>%
#   unique()
feats <- unique(top_degs$feature)

other_info <- c("tm", "subtype", "mouse")

rf_dat <- rf_so %>%
  # subset(subtype != "BEC" & subtype == "cLEC") %>%
  subset(subtype == "cLEC" & mouse == "d14") %>%

  FetchData(unique(c(dat_col, other_info, other_vars, feats))) %>%
  
  rename_with(~ str_replace_all(.x, "-", "_")) %>%
  rename_with(~ str_replace_all(.x, "\\(|\\)", "_")) %>%
  rename_with(~ str_c("x", .x), matches("^[0-9]"))

# Adjust groups so they are balanced
# Split into train/test data
sam_sz <- 135

rf_train <- rf_dat %>%
  rownames_to_column(".cell_id") %>%
  group_by(!!sym(dat_col), subtype, mouse) %>%
  filter(n() >= sam_sz) %>%
  dplyr::sample_n(sam_sz) %>%
  ungroup() %>%
  dplyr::select(-all_of(other_info[!other_info %in% other_vars])) %>%
  column_to_rownames(".cell_id")

rf_test <- rf_dat[!rownames(rf_dat) %in% rownames(rf_train), ]

# Tune RF model ----
tune_rf <- function(dat_df, dat_col, param_lst, return_best = TRUE) {
  
  rf_params <- expand.grid(param_lst)
  
  res <- pmap_dfr(rf_params, ~ {
    res <- ranger(
      as.formula(str_c("factor(", dat_col, ") ~ .")),
      data       = dat_df,
      importance = "impurity",
      seed       = 42,
      ...
    )
    
    err <- sqrt(res$prediction.error)
    
    tibble(..., oob_rmse = err)
  })
  
  if (return_best) {
    res <- res %>%
      dplyr::filter(oob_rmse == min(oob_rmse)) %>%
      dplyr::select(-oob_rmse) %>%
      head(1) %>%
      as.list()
  }
  
  res
}

param_lst <- list(
  mtry          = seq(2, length(feats), 5),
  min.node.size = c(1, 2, 5),
  num.trees     = 1000
)

rf_params <- rf_train %>%
  tune_rf(
    dat_col   = dat_col,
    param_lst = param_lst
  )

# Fit RF model
rf_mod <- pmap(rf_params, ~ {
  ranger(
    as.formula(str_c("factor(", dat_col, ") ~ .")),
    data       = rf_train,
    importance = "impurity",
    ...
  )
}) %>%
  pluck(1)

# Identify top features and re-fit
top_feats <- importance(rf_mod) %>%
  sort(decreasing = TRUE)

top_feats <- top_feats[top_feats > 0]

rf_params$mtry <- min(length(top_feats), rf_params$mtry)

rf_mod_top <- pmap(rf_params, ~ {
  ranger(
    as.formula(str_c("factor(", dat_col, ") ~ .")),
    data       = rf_train[, c(dat_col, names(top_feats))],
    importance = "impurity",
    ...
  )
}) %>%
  pluck(1)

# prds <- rf_mod %>%
  # predict(rf_dat) %>%
prds <- rf_mod_top %>%
  predict(rf_test) %>%
  predictions()

# Get predictions ----
# rf_res <- rf_test %>%
rf_res <- rf_dat %>%
  as_tibble(rownames = ".cell_id") %>%
  dplyr::select(.cell_id, subtype, all_of(c(dat_col, other_vars, other_info))) %>%
  mutate(
    prediction = prds,
    correct    = !!sym(dat_col) == prediction
  ) %>%
  group_by(!!!syms(c(other_info, dat_col))) %>%
  summarize(
    n       = n(),
    correct = sum(correct) / n,
    .groups = "drop"
  ) %>%
  arrange(desc(correct))

# Plot results
rf_res %>%
  mutate(lab = as.character(round(correct, 2))) %>%
  ggplot(aes(subtype, mouse, fill = correct)) +
  geom_tile() +
  geom_text(aes(label = lab)) +
  facet_wrap(~ Ag_class_2) +
  scale_fill_gradientn(colours = c("lightblue", "red")) +
  djvdj_theme()

# Get top regression variables
rf_marks <- tibble(
  gene   = names(rf_mod$variable.importance),
  impact = rf_mod$variable.importance
) %>%
  mutate(
    rank = row_number(desc(impact)),
    gene = ifelse(gene %in% cat_feats, gene, str_replace(gene, "_", "-"))
  ) %>%
  arrange(rank) %>%

  left_join(degs, by = c(gene = "feature"))

# Create boxplots
n_gns <- 50

plt_gns <- rf_marks %>%
  filter(!gene %in% c("Ag_score", dat_col, "orig.ident", "tm")) %>%
  pull(gene) %>%
  unique() %>%
  head(n_gns) %>%
  c("Ag_score", .) %>%
  str_remove("^x")

plt_dat <- lec_so %>%
  subset(subtype == typ) %>%
  FetchData(c("tm", dat_col, plt_gns)) %>%
  pivot_longer(all_of(plt_gns)) %>%
  mutate(
    tm         = fct_relevel(as.character(tm), tms),
    Ag_class_2 = fct_relevel(Ag_class_2, c("low", "high")),
    name       = fct_relevel(name, plt_gns)
  )

plt_dat %>%
  ggplot(aes(tm, value, fill = name, color = name, alpha = Ag_class_2)) +
  geom_boxplot() +
  facet_wrap(~ name, scales = "free_y") +
  scale_alpha_manual(values = c(low = 0.25, high = 0.75)) +
  djvdj_theme()

# # Create scatter plots
# plt_dat <- rf_so %>%
#   subset(subtype == typ) %>%
#   FetchData(c("tm", "Ag_class", plt_gns)) %>%
#   pivot_longer(all_of(plt_gns[-1])) %>%
#   mutate(
#     tm       = fct_relevel(as.character(tm), tms),
#     Ag_class = fct_relevel(Ag_class, c("low", "high")),
#     name     = fct_relevel(name, plt_gns)
#   )
# 
# plt_dat %>%
#   ggplot(aes(Ag_score, value, color = Ag_class)) +
#   geom_point(size = 0.1) +
#   facet_wrap(~ name, scales = "free") +
#   djvdj_theme() +
#   theme(aspect.ratio = 1)
```



```{r "RF DEGs", eval = FALSE}
# GOALS
# 1. train model for LEC subsets for Ag groups (1-4)
# 2. identify cells in mock that are predicted to take up Ag

# Identify Ag markers for all LECs for each mouse
grp <- "Ag_class_2"
fc  <- 0

deg_so <- lec_so %>%
  SplitObject("mouse")

degs_1 <- deg_so %>%
  imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = "all", mouse = .y) %>%
      arrange(padj)
  })

# Identify Ag markers for subtype
deg_so <- lec_so %>%
  SplitObject("subtype")

degs_2 <- deg_so %>%
  imap_dfr(~ {
    res <- .x %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = .y, mouse = "all") %>%
      arrange(padj)
  })

rm(deg_so)

# Identify markers for subtype and mouse
degs_3 <- lec_so@meta.data %>%
  distinct(subtype, mouse) %>%
  pmap_dfr(~ {
    res <- lec_so %>%
      subset(subtype == ..1 & mouse == ..2)
    
    if (n_distinct(res[[grp]]) < 2) return(NULL)
    
    res <- res %>%
      wilcoxauc(group_by = grp)

    if (nrow(res) == 0) return(NULL)

    res %>%
      mutate(subtype = ..1, mouse = ..2) %>%
      arrange(padj)
  })

degs <- bind_rows(degs_1, degs_2, degs_3) %>%
  filter(
    padj < 0.05, logFC > fc,
    !grepl("^(mt-|R[pls])", feature),
    feature != "Malat1"
  )

# degs %>%
#   write_tsv(here(params$so_dir, "ag-high_degs_Ag_class_2.tsv.gz"))

# # Most common degs
# top_degs <- degs %>%
#   group_by(subtype, feature, group) %>%
#   summarize(n = n_distinct(mouse), .groups = "drop") %>%
#   arrange(desc(n))
#   filter(n > 1)
```

```{r "RF FUNCTIONS", eval = FALSE}
# Function to predict
rf_predict <- function(mod, dat, data_clmn, return_stats = TRUE) {
  
  prds <- mod %>%
    predict(dat) %>%
    predictions()
  
  res <- dat %>%
    as_tibble(rownames = ".cell_id") %>%
    dplyr::select(.cell_id, all_of(data_clmn)) %>%
    mutate(
      pred = prds,
      correct = !!sym(data_clmn) == pred
    ) %>%
    group_by(!!sym(data_clmn)) %>%
    mutate(frac_correct = sum(correct) / n()) %>%
    ungroup()
  
  if (!return_stats) return(column_to_rownames(res, ".cell_id"))
  
  res <- res %>%
    distinct(!!sym(data_clmn), frac_correct) %>%
    pivot_wider(names_from = all_of(data_clmn), values_from = frac_correct)
  
  res
}

run_rf <- function(df_in, data_clmn, param_lst, sample_size = 135,
                   feat_p = 0.05, min_feats = 20, sample_info = NULL) {
  
  # Format train/test data
  rf_dat <- df_in %>%
    rename_with(~ str_replace_all(.x, "-", "_")) %>%
    rename_with(~ str_replace_all(.x, "\\(|\\)", "_")) %>%
    rename_with(~ str_c("x", .x), matches("^[0-9]"))
  
  # Adjust groups so they are balanced
  # Split into train/test data
  sam_sz <- min(sample_size, min(table(rf_dat[[data_clmn]])))
  min_sz <- 40
  
  if (sam_sz < min_sz) {
    cli::cli_abort("All groups must have at least {min_sz} cells")
    
  } else if (sam_sz < sample_size) {
    cli::cli_warn(
      "{sample_size} is too large for the number of cells,
       a sample size of {sam_sz} was used"
    )
  }
  
  rf_train <- rf_dat %>%
    rownames_to_column(".cell_id") %>%
    group_by(!!sym(data_clmn))
  
  set.seed(42)
  
  rf_train <- rf_train %>%
    dplyr::sample_n(sam_sz) %>%
    ungroup() %>%
    dplyr::select(-any_of(sample_info)) %>%
    column_to_rownames(".cell_id")
  
  rf_test <- rf_dat[!rownames(rf_dat) %in% rownames(rf_train), ]
  
  # Train models
  rf_params <- expand.grid(param_lst)
  
  res <- pmap_dfr(rf_params, ~ {
    args <- list(...)
    
    if (args$mtry > (ncol(rf_train) - 1)) return(NULL)
    
    args$formula    <- as.formula(str_c("factor(", data_clmn, ") ~ ."))
    args$data       <- rf_train
    args$importance <- "impurity_corrected"
    args$seed       <- 42
    
    set.seed(42)
    
    raw_mod <- lift_dl(ranger)(args)
    
    raw_err <- raw_mod$prediction.error

    # Feature selection
    key_feats <- raw_mod %>%
      importance_pvalues() %>%
      .[, "pvalue"] %>%
      sort()
    
    max_p <- key_feats %>%
      head(min_feats) %>%
      max()
    
    feat_p <- max(max_p, feat_p)
    
    key_feats <- key_feats[key_feats < feat_p] %>%
      names()
    
    n_feats <- length(key_feats)
    
    # Re-train model
    args$mtry       <- min(args$mtry, n_feats)
    args$data       <- rf_train[, c(data_clmn, key_feats)]
    args$importance <- "none"
    
    set.seed(42)
    
    mod <- lift_dl(ranger)(args)
    
    err <- mod$prediction.error
    
    # Test model    
    prds <- mod %>%
      rf_predict(
        dat = rf_test,
        data_clmn = data_clmn
      )
                                         # only used for continuous
    # err <- sqrt(res$prediction.error)  # response variables
    
    # Format output table
    tibble(
      raw_mod   = list(raw_mod),
      mod       = list(mod),
      feats     = list(key_feats),
      n_train   = nrow(rf_train),
      n_test    = nrow(rf_test),
      # train_bcs = list(rownames(rf_train)),
      # test_bcs  = list(rownames(rf_test)),
      ...,
      n_feats = length(key_feats),
      oob_raw = raw_err,
      oob     = err
    ) %>%
      bind_cols(prds)
  })
  
  res
}
```

```{r "RF MODELS", eval = FALSE}
# Train models
# use all cLEC features
dat_col <- "Ag_class_2"

feats <- degs %>%
  filter(subtype == "cLEC") %>%
  # filter(subtype == "cLEC", mouse == "d14") %>%
  pull(feature) %>%
  unique()

other_info <- c("subtype", "mouse")

rf_dat <- lec_so %>%
  subset(subtype == "cLEC" & mouse == "d14") %>%
  FetchData(unique(c(dat_col, other_info, feats)))

param_lst <- list(
  mtry          = seq(2, length(feats), 10),
  min.node.size = c(1:10),
  num.trees     = seq(50, 1000, 50)
)

rf_res <- rf_dat %>%
  run_rf(
    data_clmn   = dat_col,
    param_lst   = param_lst,
    sample_info = other_info,
    sample_size = 135
  ) %>%
  # arrange(desc(high), desc(low))
  arrange(desc(low), desc(high))

# Pull best model
best_mod <- rf_res %>%
  head(1) %>%
  pull(mod) %>%
  .[[1]]

best_feats <- rf_res %>%
  head(1) %>%
  pull(feats) %>%
  unlist()

# Predict for other timepoints
tm_dat <- lec_so %>%
  subset(subtype == "cLEC") %>%
  FetchData(unique(c(dat_col, other_info, feats))) %>%
  rename_with(~ str_replace_all(.x, "-", "_")) %>%
  rename_with(~ str_replace_all(.x, "\\(|\\)", "_")) %>%
  rename_with(~ str_c("x", .x), matches("^[0-9]")) %>%
  split(.$mouse)

preds <- tm_dat %>%
  imap_dfr(~ {
    best_mod %>%
      rf_predict(.x, data_clmn = dat_col, return_stats = FALSE) %>%
      mutate(mouse = .y)
  })

# Plot predicted classifications
pred_meta <- preds %>%
  dplyr::select(pred, correct)

lec_so <- lec_so %>%
  AddMetaData(pred_meta)

lec_so %>%
  plot_scatter(
    "pred", "hUMAP_1", "hUMAP_2",
    group_col = "mouse",
    size = 0.1,
    panel_nrow = 1
  ) +
  theme(aspect.ratio = 0.9)

# Fraction predicted in each mouse
plt_dat <- lec_so %>%
  subset(subtype == "cLEC") %>%
  AddModuleScore(features = best_feats, name = "Ag_module") %>%
  mutate_meta(
    mutate,
    pred_grp = case_when(
      Ag_class_2 == "high" ~ Ag_class_2,
      pred       == "high" ~ "high-predicted",
      TRUE                 ~ Ag_class_2
    ),
    mouse = fct_relevel(mouse, mice)
  )

plt_dat %>%
  plot_frequency("pred_grp", cluster_col = "mouse")

# Module scores for each prediction group
# expect that these genes will be highest expressed in high, high-predicted,
# lowest in low
# 3wk sample shows opposite
plt_dat@meta.data %>%
  # ggplot(aes(mouse, Ag_module1, fill = Ag_class_2)) +
  ggplot(aes(mouse, Ag_module1, fill = pred_grp)) +
  # geom_violin(scale = "width", draw_quantiles = c(0.25, 0.75, 0.5)) +
  geom_boxplot() +
  djvdj_theme()
  

```

```{r "COMPARE WITH MOCK DATA", eval = FALSE, fig.width = 18, fig.height = 15}

# Plot top features across data
plt_feats <- degs %>%
  group_by(subtype, feature, group) %>%
  summarize(n = n_distinct(mouse), .groups = "drop") %>%
  arrange(desc(n)) %>%
  filter(n > 1) %>%
  filter(group == "high") %>%
  split(.$subtype) %>%
  map(pull, feature)

# plt_feats <- top_degs %>%
#   split(.$group)

clmns <- str_c(names(plt_feats), seq_along(plt_feats))

plt_dat <- lec_so %>%
  Seurat::AddModuleScore(features = list(plt_feats$high$feature)) %>%
  # Seurat::AddModuleScore(features = plt_feats, name = names(plt_feats)) %>%
  FetchData(c("Ag_class_2", "Cluster1", "orig.ident", "mouse", "tm", "subtype"))

plt_dat %>%
  mutate(Ag_class_2 = fct_relevel(Ag_class_2, c("low", "high"))) %>%
  ggplot(aes(mouse, Cluster1, fill = mouse, alpha = Ag_class_2)) +
  geom_boxplot() +
  scale_alpha_manual(values = c(high = 0.75, low = 0.25)) +
  facet_wrap(~ subtype) +
  djvdj_theme()

# Load mock data
chikv_so <- qread("~/Dropbox/Ryan/Projects/morrison-lnsc/results/sobjs/so_merge.qs")

# Calculate module score
chikv_dat <- chikv_so %>%
  Seurat::AddModuleScore(features = plt_feats, name = names(plt_feats)) %>%
  FetchData(c("orig.ident", "lec_type", "tm", "cell_type", "rep", "cell_type_2", "treatment", clmns, "subtype"))

# Calculate mean signal
all_gns <- unique(unlist(plt_feats))

chikv_dat <- chikv_so %>%
  FetchData(c(
    "orig.ident", "lec_type", "tm", "cell_type",
    "rep", "cell_type_2", "treatment", "subtype",
    all_gns
  ))

# Plot example genes
chikv_so %>%
  FetchData(c("treatment", "tm", "lec_type", "rep", "Ctsd", "cell_type_2")) %>%
  filter(cell_type_2 == "LEC" | lec_type == "BEC") %>%
  mutate(
    treat_rep = str_c(treatment, "-", rep),
    treatment = fct_relevel(treatment, c("mock", "CHIKV"))
  ) %>%
  ggplot(aes(lec_type, Ctsd, fill = treatment, alpha = rep)) +
  geom_boxplot(outlier.size = 0.1) +
  scale_alpha_manual(values = c(0.2, 0.4, 0.6)) +
  facet_wrap(~ tm) +
  djvdj_theme() +
  theme(
    aspect.ratio = 0.4,
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1)
  )

# # Plot mean list expression
# plt_dat <- chikv_dat %>%
#   pivot_longer(any_of(all_gns)) %>%
#   filter(cell_type_2 == "LEC" | lec_type == "BEC")
# 
#   mutate(
#     treat_rep = str_c(treatment, "-", rep),
#     treatment = fct_relevel(treatment, c("mock", "CHIKV"))
#   ) %>%
#   ggplot(aes(lec_type, Ctsd, fill = treatment, alpha = rep)) +
#   geom_boxplot(outlier.size = 0.1) +
#   scale_alpha_manual(values = c(0.2, 0.4, 0.6)) +
#   facet_wrap(~ tm) +
#   djvdj_theme() +
#   theme(
#     aspect.ratio = 0.4,
#     axis.title.x = element_blank(),
#     axis.text.x  = element_text(angle = 45, hjust = 1)
#   )

# Plot module scores
plt <- clmns %>%
  map(~ {
    chikv_dat %>%
      filter(cell_type_2 == "LEC" | lec_type == "BEC") %>%
      mutate(
        treat_rep = str_c(treatment, "-", rep),
        treatment = fct_relevel(treatment, c("mock", "CHIKV"))
      ) %>%
      ggplot(aes(lec_type, !!sym(.x), fill = treatment, alpha = rep)) +
      geom_boxplot(outlier.size = 0.01, outlier.alpha = 1, size = 0.4) +
      scale_alpha_manual(values = c(0.2, 0.4, 0.6)) +
      facet_wrap(~ tm, nrow = 1) +
      guides(alpha = FALSE) +
      djvdj_theme() +
      theme(
        aspect.ratio = 0.4,
        axis.title.x = element_blank(),
        axis.text.x  = element_text(angle = 45, hjust = 1),
        legend.position = "top"
      )
  }) %>%
  plot_grid(plotlist = ., ncol = 2)

ggsave(
  here(params$so_dir, "ag_markers_chikv.png"),
  plot = plt,
  width = 18, height = 15
)
```



```{r "COMPARISON OF EMPTY DROPS", eval = FALSE}
# Paths to raw matrices to use for normalization factors
mat_dir <- names(params$res_dir)
mat_dir <- set_names(params[mat_dir], mat_dir)

# Calculate normalization factors
filt_counts <- qread(here("results/2022-10-28/sobjs/filt_norm_factors.qs"))
raw_counts  <- qread(here("results/2022-10-28/sobjs/norm_factors.qs"))

counts_df <- list(raw = raw_counts, filt = filt_counts) %>%
  imap_dfr(~ {
    reads <- .y
    
    .x %>%
      imap_dfr(~ {
        tibble(
          reads   = reads,
          run     = .y,
          library = names(.x),
          counts  = .x
        )
      })
  }) %>%
  pivot_wider(names_from = reads, values_from = counts) %>%
  mutate(frac_in_cells = filt / raw) %>%
  arrange(desc(frac_in_cells))

# Sample meta.data
lib_meta <- lec_so@meta.data %>%
  distinct(orig.ident, sample, tm, mouse, exp) %>%
  as_tibble()

counts_df %>%
  left_join(lib_meta, by = c(library = "orig.ident")) %>%
  filter(!is.na(sample)) %>%
  arrange(desc(frac_in_cells))
```

```{r "DSB TEST", eval = FALSE}
# Test the dsb package for correcting CITE-seq data based on empty droplets
# the results are not very convincing
# the signals do not decrease with time and the 3wk signals appear much higher
# than all the other timepoints

# data.frame with parameters
library(dsb)
library(furrr)

mat_dir <- names(params$res_dir)
mat_dir <- set_names(params[mat_dir], mat_dir)

mat_df <- mat_dir %>%
  imap_dfr(~ {
    rn <- .y
    
    .x %>%
      imap_dfr(~ {
        tibble(
          run    = rn,
          sample = .y,
          path   = here(params$res_dir[[rn]], .x, "outs"),
          rep    = str_extract(run, "^r[0-9]+(?=_)")
        )
      })
  })

# Normalize Ag matrices
plan(multisession, workers = 6)

norm_mats <- mat_df %>%
  future_pmap(~ {
    args <- list(...)
    filt_mat <- Read10X(here(args$path, "filtered_feature_bc_matrix"))$`Antibody Capture`
    raw_mat  <- Read10X(here(args$path, "raw_feature_bc_matrix"))$`Antibody Capture`
    raw_mat  <- raw_mat[, !colnames(raw_mat) %in% colnames(filt_mat)]
    
    DSBNormalizeProtein(
      cell_protein_matrix = filt_mat,
      empty_drop_matrix   = raw_mat,
      denoise.counts      = FALSE,
      pseudocount.use     = 1,
      use.isotype.control = FALSE,
      return.stats        = TRUE
    )
  })

names(norm_mats) <- mat_df$sample

# Create data.frame for corrected data
dsb_dat <- mat_df %>%
  pmap_dfr(~ {
    args <- list(...)
    
    dat <- t(norm_mats[[args$sample]]$dsb_normalized_matrix)
    
    if (!grepl("^CD45", args$sample)) {
      colnames(dat) <- params$ag_key[[args$rep]][colnames(dat)]
    }
    
    colnames(dat) <- colnames(dat) %>%
      str_c("dsb_", .)
    
    dat %>%
      as_tibble(rownames = "raw_cell_id") %>%
      mutate(run = args$fun, sample = args$sample)
  })

# Add corrected counts to object
lec_so <- qread(here(params$so_dir, "results/2022-10-28/sobjs/so_lec_new.qs"))

lec_so@meta.data <- lec_so@meta.data %>%
  mutate(raw_cell_id = str_remove(rownames(.), "_[_0-9]+$")) %>%
  as_tibble(rownames = ".cell_id") %>%
  left_join(dsb_dat, by = c("raw_cell_id", orig.ident = "sample")) %>%
  mutate(
    Ag_score = case_when(
      grepl("d[0-9]+$", mouse) ~ dsb_ovalbumin,
      grepl("^3wk$", mouse)    ~ dsb_3wk,
      grepl("^6wk$", mouse)    ~ dsb_6wk,
      
      Ag_3wk_score >= Ag_6wk_score ~ Ag_3wk_score,  # For 6wk-3wk use highest
      Ag_6wk_score >= Ag_3wk_score ~ Ag_6wk_score
    ),
    Ag_score_3 = ifelse(vaccination == "dual", Ag_3wk_score, Ag_score),
    Ag_score_6 = ifelse(vaccination == "dual", Ag_6wk_score, Ag_score),
    tm_3 = ifelse(vaccination == "dual", 21, tm),
    tm_6 = ifelse(vaccination == "dual", 42, tm)
  ) %>%
  column_to_rownames(".cell_id")

# Create plots
# timepoints
lec_so@meta.data %>%
  # filter(subtype == "fLEC") %>%
  # filter(tm != 21) %>%
  filter(vaccination == "single") %>%
  ggplot(aes(tm, Ag_score, group = tm, fill = subtype)) +
  geom_boxplot(width = 3) +
  facet_wrap(~ subtype, nrow = 1) +
  djvdj_theme()

# antigen-exchange
lec_so@meta.data %>%
  # filter(subtype == "fLEC") %>%
  mutate(vaccination = fct_relevel(vaccination, c("single", "dual"))) %>%
  ggplot(aes(tm_3, Ag_score_3, group = tm, fill = subtype, color = subtype, alpha = vaccination)) +
  geom_boxplot(width = 6, position = position_dodge2(preserve = "single")) +
  facet_wrap(~ subtype, nrow = 1) +
  scale_alpha_manual(values = c(0.25, 0.75)) +
  djvdj_theme()

lec_so@meta.data %>%
  # filter(tm_6 != 21) %>%
  # filter(subtype == "fLEC") %>%
  mutate(vaccination = fct_relevel(vaccination, c("single", "dual"))) %>%
  ggplot(aes(tm_6, Ag_score_6, group = tm, fill = subtype, color = subtype, alpha = vaccination)) +
  geom_boxplot(width = 6, position = position_dodge2(preserve = "single")) +
  facet_wrap(~ subtype, nrow = 1) +
  scale_alpha_manual(values = c(0.25, 0.75)) +
  djvdj_theme()
```

```{r "PLOT RAW COUNTS CORRELATION", eval = FALSE}
clmns <- c(
  "exp", "mouse", "vaccination", "orig.ident", "subtype",
  "adt_ABPS2", "adt_ABPS4",
  "Ag_3wk", "Ag_6wk"
)

dat <- lec_so %>%
  FetchData(clmns, slot = "counts") %>%
  filter(vaccination == "dual") %>%
  as_tibble(rownames = "cell_id")

xvar <- "adt_ABPS2"
yvar <- "adt_ABPS4"

# xvar <- "Ag_3wk"
# yvar <- "Ag_6wk"

dat %>%
  ggplot(aes(!!sym(xvar) + 1, !!sym(yvar) + 1, color = subtype)) +
  geom_point() +
  facet_grid(exp ~ subtype) +
  djvdj_theme() +
  theme(aspect.ratio = 1) +
  geom_abline(slope = 1, intercept = 0) +
  scale_x_log10() +
  scale_y_log10()
```