
```{r "main setup", include = FALSE, warning = FALSE, message = FALSE}
# Default chunk options
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo    = FALSE,
  dpi     = 300
)

# scRNA-seq
library(Seurat)
library(clustifyr)
library(clustifyrdata)
library(presto)
library(harmony)
library(djvdj)
library(M3Drop)

# Gene ontology
library(gprofiler2)
library(biomaRt)
library(clusterProfiler)
library(enrichplot)
library(msigdbr)
library(org.Mm.eg.db)
library(GOSemSim)

# Models
library(mixtools)
library(boot)
library(tools)
library(gtools)
library(rsample)
library(ranger)
library(furrr)
library(caret)

# Plotting
library(ggspruce, lib.loc = "/home/rmsheridan/.cache/R/renv/library/ggspruce-c2135e66/R-4.3/x86_64-pc-linux-gnu")
library(scales)
library(ggforce)
library(ggbeeswarm)
library(cowplot)
library(ggrepel)
library(RColorBrewer)
library(colorblindr)
library(patchwork)
library(ggtext)
library(ggtrace)
library(MetBrewer)
library(ggtree)
library(qs)
library(openxlsx)
library(xlsx)
library(ComplexHeatmap)
library(png)
library(pzfx)

# Tidyverse
library(tidyverse)
library(knitr)
library(broom)
library(here)
library(devtools)
library(cli)

# GeoMx
library(NanoStringNCTools)
library(GeomxTools)
library(GeoMxWorkflows)
library(umap)
library(SpatialOmicsOverlay)

# Xenium
library(sf)
library(sp)

source(here(params$template_dir, "funs.R"))

# Load sample info
xlsx_path <- here(params$ref_dir, params$sample_info)
sam_info  <- xlsx::read.xlsx(xlsx_path, sheetName = "samples")
ag_key    <- xlsx::read.xlsx(xlsx_path, sheetName = "tags")

ags <- unique(ag_key$tag)

ag_key <- ag_key %>%
  split(.$experiment) %>%
  map(~ set_names(.x$tag, .x$identity))

# Check for saved objects
create_sobjs <- !file.exists(here(params$so_dir, "lec_so.qs"))
```

```{r "functions"}
#' Integrate Seurat objects using harmony
#' 
#' @param obj Input object
#' @param group_vars Variable(s) to using for grouping cells
#' @noRd
integrate_sobjs <- function(so_in, group_vars = "mouse", dims = 1:40,
                            resolution = c(1, 3, 5, 10)) {
  rm_clmns <- c(str_c("UMAP_", 1:2), str_c("hUMAP_", 1:2))
  
  res <- so_in %>%
    mutate_meta(dplyr::select, -any_of(rm_clmns)) %>%
    FindVariableFeatures(nfeatures = 2000) %>%
    ScaleData() %>%
    RunPCA() %>%
    RunUMAP(dims = dims) %>%
    AddMetaData(FetchData(., c("UMAP_1", "UMAP_2")))
  
  res <- res %>%
    RunHarmony(
      group.by.vars = group_vars,
      dims = dims,
      plot_convergence = TRUE
    ) %>%
    RunUMAP(
      dims = dims,
      reduction      = "harmony",
      reduction.name = "humap",
      reduction.key  = "hUMAP_"
    ) %>%
    FindNeighbors(reduction = "harmony", dims = dims) %>%
    FindClusters(resolution = resolution) %>%
    AddMetaData(FetchData(., c("hUMAP_1", "hUMAP_2")))
  
  res
}

#' Calculate antigen score by subtracting B/T cell signal
#' 
#' @param obj Input object
#' @param clmns Columns containing antigen counts
#' @param control_types Cell types to use for calculating control signal
#' @param type_clmn Column containing cell types, this should contain the types
#' provided to control_types
#' @param norm_factors Named vector where names are sample names found in
#' `sample_clmn` and values are normalization factors for each sample,
#' e.g. total counts expressed in millions. If NULL, signal is not transformed
#' prior to correcting for background.
#' @param sample_clmn Column containing sample names that match those provided
#' by `norm_factors`
#' @param q Quantile to use as cutoff for control signal
#' @param operation Function to use for normalizing by control signal, this 
#' should take two arguments, the first is the signal to normalize,
#' the second is the control signal.
#' @param slot Slot in object to fetch data, default is counts.
#' @param final_trans Final transformation to perform after normalizing and
#' correcting background
#' @param suffix Suffix to add to new columns
#' @noRd
calc_ag_score <- function(obj, clmns, control_types = c("B cells", "T cells"),
                          type_clmn = "cell_type", sample_clmn = "directory",
                          norm_factors = NULL, q = 0.75,
                          operation = `-`, slot = "counts",
                          final_trans = log1p, suffix = "_score"
                          ) {
  
  # Fetch clmns
  assay             <- DefaultAssay(obj)
  DefaultAssay(obj) <- "ADT"
  
  obj <- obj %>%
    AddMetaData(FetchData(., clmns, slot = slot))
  
  DefaultAssay(obj) <- assay
  
  # Normalize using norm_factors
  # if norm_factors not provided, values are not transformed
  norm_clmns <- clmns
    
  if (!is.null(norm_factors)) {
    norm_clmns <- str_c(clmns, "_norm")
    
    obj <- obj %>%  
      mutate_meta(~ {
        .x %>%
          group_by(!!sym(sample_clmn)) %>%
          mutate(
            across(all_of(unname(clmns)), ~ {
              fctr <- norm_factors[!!sym(sample_clmn)]
              .x / fctr
            },
            .names = "{.col}_norm"
            )
          ) %>%
          ungroup()
      })
  }
  
  # Calculate background signal
  if (!is.null(control_types)) {
    obj <- obj %>%
      mutate_meta(~ {
        .x %>%
          group_by(!!sym(sample_clmn)) %>%
          mutate(across(
            all_of(norm_clmns),
            ~ {
              missing <- control_types[!control_types %in% !!sym(type_clmn)]
              
              if (!is_empty(missing)) {
                cli_warn("{missing} w{?as/ere} not found in {type_clmn}.")
              }
              
              idx <- !!sym(type_clmn) %in% control_types  # rows to use
              res <- quantile(.x[idx], q)                 # set cutoff
              
              res
            },
            .names = str_c("{.col}_threshold")
          )) %>%
          ungroup()
      })
    
    # Subtract background signal
    obj <- obj %>%
      mutate_meta(~ {
        .x %>%
          group_by(!!sym(sample_clmn)) %>%
          mutate(across(
            all_of(norm_clmns),
            ~ {
              clmn   <- str_c(cur_column(), "_threshold")
              thresh <- obj@meta.data[cur_group_rows(), clmn]
              thresh <- unique(thresh)
              
              if (length(thresh) > 1) {
                cli_abort("Threshold must be a single value for each group")
              }
              
              res <- .x - thresh

              res[res < 0] <- 0

              final_trans(res)
            },
            .names = str_c("{.col}", suffix)
          )) %>%
          ungroup()
      })
    
    norm_clmns <- str_c(norm_clmns, suffix)
  }
  
  # Rename final columns
  final_clmns <- set_names(
    norm_clmns,
    names(clmns) %||% str_c(clmns, suffix)
  )
  
  res <- obj %>%
    mutate_meta(~ dplyr::rename(.x, !!!final_clmns))
  
  res
}

#' Format and normalize Ag meta.data columns
#' 
#' @param so_in Input object
#' @param norm_factors Named vector containing normalization factors for
#' `calc_ag_score()`
#' @param ag_key List of named vectors containing labels for each Ag tag for
#' each experiment. List names should be experiment name, vector names should
#' be new label for the Ag tag.
#' @noRd
format_ag_data <- function(so_in, norm_factors, ag_key, ...) {
  
  # Assign labels for Ag tags (ABPS2, ABPS4)
  # * the identify of each tag is differs depending on the experiment
  res <- so_in %>%
    mutate_meta(~ {
      dat <- .x
      
      c("3wk", "6wk") %>%
        walk(~ {
          clmn <- str_c("Ag_", .x)
          
          dat <<- dat %>%
            group_by(exp) %>%
            mutate(
              !!sym(clmn) := ag_key[[as.character(cur_group())]][.x],
              
              !!sym(clmn) := case_when(
                !!sym(clmn) == "ABPS2" ~ ABPS2,
                !!sym(clmn) == "ABPS4" ~ ABPS4
              )
            ) %>%
            ungroup()
        })
      
      dat
    })
  
  # Calculate Ag scores
  ag_clmns <- c("ovalbumin", "Ag_3wk", "Ag_6wk")
  
  res <- res %>%
    mutate_meta(mutate, across(all_of(ag_clmns), ~ replace_na(.x, 0))) %>%
    calc_ag_score(
      clmns        = ag_clmns,
      type_clmn    = "cell_type",
      sample_clmn  = "directory",
      norm_factors = norm_factors,
      ...
    )
  
  # Format Ag score columns
  res <- res %>%
    mutate_meta(
      mutate,
      Ag_score = case_when(
        exp   == "exp-0" ~ ovalbumin_score,
        mouse == "3wk" ~ Ag_3wk_score,
        mouse == "6wk" ~ Ag_6wk_score,
        
        # for 6wk-3wk use highest score
        mouse == "6wk-3wk" & Ag_3wk_score >= Ag_6wk_score ~ Ag_3wk_score,
        mouse == "6wk-3wk" & Ag_6wk_score >= Ag_3wk_score ~ Ag_6wk_score
      ),
      
      # these columns are used for antigen-exchange plots
      Ag_score_3 = ifelse(vaccination == "dual", Ag_3wk_score, Ag_score),
      Ag_score_6 = ifelse(vaccination == "dual", Ag_6wk_score, Ag_score),
      
      tm_3 = ifelse(vaccination == "dual", 21, tm),
      tm_6 = ifelse(vaccination == "dual", 42, tm)
    )
}

#' Create boxplots comparing RF module scores for predicted Ag classes
#' 
#' Used for figure 2
create_rf_module_boxes <- function(df_in, module_clmns, module_title, clrs,
                                   p_method = wilcox.test, p_alt = "two.sided",
                                   p_hjust = c(0.7, 0.5), p_vjust = c(1.2, 2.4),
                                   n_label = FALSE
                                   ) {

  # Format data for boxplots  
  typs <- str_remove(module_clmns, "_[^_]+$")
  
  dat <- df_in %>%
    pivot_longer(all_of(module_clmns)) %>%
    mutate(module_type = str_remove(name, "_[^_]+$")) %>%
    filter(subtype == module_type) %>%
    mutate(
      subtype  = fct_relevel(subtype, typs),
      pred_grp = fct_relevel(pred_grp, names(clrs))  # must set as factor to
    )                                                # calculate correlation
  
  # Calculate Spearman correlation
  p_dat <- dat %>%
    group_by(subtype, tm) %>%
    summarize(corr = list(tidy(cor.test(
      as.numeric(pred_grp), value,
      method      = "spearman",
      continuity  = TRUE,
      exact       = FALSE,
      alternative = p_alt
    ))), .groups = "drop") %>%
    unnest(corr) %>%
    mutate(p_adj = p.adjust(p.value, method = "BH")) %>%
    rowwise() %>%
    mutate(
      p_adj = .format_pvalue(p_adj, show_decimal = Inf),
      p_adj = str_c("italic(p) == ", p_adj),
      r_lab = str_c("italic(r[s]) == ", round(estimate, 2))
    ) %>%
    ungroup()
  
  # Format n labels
  # the order of n labels in n_dat determines order on plot
  # need to make sure n_dat is ordered based on factor level which is set
  # upstream
  n_dat <- dat %>%
    group_by(tm, subtype, pred_grp) %>%
    summarize(n = n_distinct(.cell_id), .groups = "drop") %>%
    mutate(n_lab = label_comma()(n)) %>%
    arrange(pred_grp)
  
  # Create boxplots
  y_exp <- c(0.05, 0.25)
  
  if (n_label) y_exp[1] <- 0.1
  
  res <- dat %>%
    ggplot(aes(as.character(tm), value, fill = pred_grp)) +
    geom_boxplot(
      outlier.size = 0.1,
      position     = position_dodge2(preserve = "single"),
      key_glyph    = draw_key_point
    ) +
    
    # r label
    geom_text(
      aes(y = Inf, fill = NULL, label = r_lab),
      data = p_dat,
      size = 10 / .pt,
      hjust = p_hjust[1],
      vjust = p_vjust[1],
      parse = TRUE
    ) +
    
    # p label
    geom_text(
      aes(y = Inf, fill = NULL, label = p_adj),
      data  = p_dat,
      size  = 10 / .pt,
      hjust = p_hjust[2],
      vjust = p_vjust[2],
      parse = TRUE
    ) +
    
    scale_fill_manual(values = clrs) +
    scale_y_continuous(expand = expansion(y_exp)) +
    facet_wrap(~ subtype, nrow = 1, scales = "free_y") +
    labs(x = "days post vaccination", y = module_title) +
    guides(fill = guide_legend(override.aes = list(shape = 22, size = 5))) +
    base_theme +
    theme(
      aspect.ratio    = 0.85,
      plot.title      = element_text(hjust = 0.5),
      legend.position = "bottom",
      legend.title    = element_blank(),
      plot.margin     = margin(t = 0, r = 5, b = 5, l = 5)
    )
  
  # n label
  if (n_label) {
    res <- res +
      geom_text(
        aes(y = -Inf, fill = NULL, label = n_lab),
        data     = n_dat,
        position = position_dodge2(preserve = "single", width = 1),
        size     = 10 / .pt,
        vjust    = -0.25
      )
  }
  
  res
}

#' Create UMAPs showing module scores for each timepoint
#' 
#' Used for figures 2 and 5
create_rf_module_umaps <- function(df_in, typ, module_clmn, module_title,
                                  ag_class_clmn, pt_size = 1, facet_var = "tm",
                                  equal_scales = TRUE, outline = FALSE,
                                  circle_cells = TRUE, show_scale_title = FALSE,
                                  ag_clrs = c(
                                    "lightblue", "lightblue",
                                    "white", "#D7301F"
                                  ),
                                  class_clrs = c(
                                    `Ag-high` = "#D7301F",
                                    `Ag-low`  = "#56B4E9",
                                    other     = "white"
                                  ),
                                  lab_fn = as_labeller(function(x) str_c("day ", x))) {
  
  # Format plot data
  if (length(pt_size) == 1) pt_size <- rep(pt_size, 2)
  
  dat <- df_in %>%
    mutate(
      !!sym(ag_class_clmn) := ifelse(
        subtype == typ,
        as.character(!!sym(ag_class_clmn)),
        as.character(NA)
      ),
      !!sym(ag_class_clmn) := fct_relevel(!!sym(ag_class_clmn), names(class_clrs)),
      !!sym(facet_var) := as.character(!!sym(facet_var))
    )
  
  # Set coordinates for drawing circle
  circ_dat <- dat %>%
    filter(subtype == typ) %>%
    group_by(!!sym(facet_var)) %>%
    summarize(across(c(hUMAP_1, hUMAP_2), mean), .groups = "drop")
  
  # Plot module scores
  grps         <- sort(unique(df_in[[facet_var]]))
  facet_frm    <- as.formula(str_c("~ ", facet_var))
  scale_limits <- guides <- NULL
  
  if (equal_scales) {
    scale_limits <- c(min(dat[[module_clmn]]), max(dat[[module_clmn]]))
    guides       <- "collect"
  }
  
  ag_mod_u <- grps %>%
    imap(~ {
      u <- dat %>%
        filter(!!sym(facet_var) == .x) %>%
        create_sig_umap(
          dat_col      = module_clmn,
          grp_col      = facet_var,
          clrs         = ag_clrs,
          pt_size      = pt_size[1],
          pt_stroke    = 0.7,
          scale_limits = scale_limits
        ) +
        facet_wrap(facet_frm, labeller = lab_fn) +
        theme(legend.text = element_text(size = 10))
      
      if (show_scale_title) {
        u <- u +
          guides(fill = guide_colorbar(
            ticks = FALSE, title.position = "left", title = module_title
          )) +
          theme(
            legend.title = element_text(size = ttl_pt2, hjust = 0.5),
            legend.key.height = unit(6, "pt"),
            legend.key.width  = unit(30, "pt")
          )
      }
      
      if (circle_cells) {
        u <- u +
          geom_circle(
            aes(x0 = hUMAP_1, y0 = hUMAP_2, fill = NULL, color = NULL, r = 2.5),
            data = filter(circ_dat, !!sym(facet_var) == .x),
            linetype = 2
          )
      }
      
      u
    }) %>%
    wrap_plots(nrow = 1, guides = guides) &
    theme(legend.position = "bottom")
  
  if (!show_scale_title) {
    ag_mod_u <- ag_mod_u +
      plot_annotation(
        title = module_title,
        theme = theme(
          plot.title = element_text(hjust = 0.5, size = 18)
        )
      )
  }
  
  # Plot Ag-high cells
  ag_class_u <- grps %>%
    imap(~ {
      dat %>%
        filter(!!sym(facet_var) == .x) %>%
        create_grp_umap(
          dat_col  = ag_class_clmn,
          grp_col  = facet_var,
          clrs     = class_clrs,
          lvls     = names(class_clrs),
          size     = pt_size[2],
          na_color = "grey85"
        ) +
        facet_wrap(facet_frm, labeller = lab_fn) +
        theme(
          legend.position = c(0.5, -0.1),
          legend.direction = "horizontal"
        )
    }) %>%
    wrap_plots(nrow = 1)
  
  # Return list of UMAPs
  res <- list(ag_mod_u, ag_class_u)
}

#' Create boxplots comparing single and dual vaccinated mice
#' 
#' Used for figures 3 and 4
create_exchange_boxes <- function(df_in, tm = NULL, tm_clmn, ag_clmn,
                                  typs = NULL, clrs, facet_vars = "subtype",
                                  p_method = wilcox.test, p_vars = NULL,
                                  p_hjust = 0.5, p_vjust = 0.5, p_size = 10,
                                  p_cutoff = Inf) {
  
  # Set plot colors
  ln_clrs   <- clrs
  ln_clrs[] <- "black"
  lgnd_clr  <- "black"
  
  if (!is.null(tm)) {
    ln_clrs[names(ln_clrs) != tm] <- "grey65"
    clrs[names(clrs) != tm]       <- "white"
    lgnd_clr                      <- clrs[[tm]]
  }
  
  # Format plot data
  x_facet <- dplyr::last(facet_vars)
  
  dat <- df_in %>%
    mutate(
      !!sym(x_facet) := fct_reorder(
        !!sym(x_facet), !!sym(ag_clmn), mean,
        .desc = TRUE
      ),
      vaccination = fct_relevel(vaccination, vacs)
    )
  
  if (!is.null(typs)) {
    dat <- dat %>%
      filter(!!sym(x_facet) %in% typs)
  }
  
  # Data to draw line connecting timepoints
  ln_dat <- dat %>%
    filter(vaccination == "single") %>%
    group_by(!!!syms(c(tm_clmn, facet_vars))) %>%
    summarize(!!sym(ag_clmn) := median(!!sym(ag_clmn)), .groups = "drop")
  
  # p-values for vaccination
  y_max <- max(pull(dat, ag_clmn))
  
  p_dat <- dat %>%
    group_by(!!!syms(c(tm_clmn, facet_vars))) %>%
    filter(all(vacs %in% vaccination)) %>%
    summarize(
      pval = p_method(
        (!!sym(ag_clmn))[vaccination == vacs[1]],
        (!!sym(ag_clmn))[vaccination == vacs[2]],
        alternative = "less"
      )$p.value,
      y = max(!!sym(ag_clmn)),
      y = y + (y_max * 0.15),
      .groups = "drop"
    )
  
  # Adjust p-values separately for each replicate
  if (!is.null(p_vars)) {
    p_dat <- p_dat %>%
      group_by(!!!syms(p_vars))
  }
  
  p_dat <- p_dat %>%
    mutate(padj = p.adjust(pval, method = "BH")) %>%
    rowwise() %>%
    mutate(
      plab = .format_pvalue(padj),
      plab = str_c("italic(p) == ", plab)
    ) %>%
    ungroup() %>%
    filter(padj < p_cutoff)
  
  # Set facet variables
  if (length(facet_vars) == 1) facet_vars <- c("", facet_vars)
  
  facet_vars <- str_c(facet_vars, collapse = "~")
  
  # Create boxplots
  res <- dat %>%
    ggplot(aes(!!sym(tm_clmn), !!sym(ag_clmn))) +
    
    geom_line(
      data = ln_dat,
      color = "grey85", linewidth = 0.5, linetype = 2
    ) +
    geom_boxplot(
      aes(
        fill = as.character(!!sym(tm_clmn)),
        color = as.character(!!sym(tm_clmn)),
        alpha = vaccination
      ),
      outlier.size = 0.2, width = 5, key_glyph = draw_key_point,
      position = position_dodge2(preserve = "single")
    ) +
    geom_text(
      aes(y = y, label = plab, alpha = NULL),
      show.legend = FALSE,
      data  = p_dat,
      parse = TRUE,
      size  = p_size / .pt,
      hjust = p_hjust,
      vjust = p_vjust
    ) +
    facet_grid(as.formula(facet_vars)) +
    scale_alpha_manual(values = c(single = 0.35, dual = 1)) +
    scale_fill_manual(values = clrs) +
    scale_color_manual(values = ln_clrs) +
    guides(
      fill = "none", color = "none",
      alpha = guide_legend(override.aes = list(
        shape = 15, size = 4, color = lgnd_clr
      ))
    ) +
    labs(x = "days post vaccination", y = "Ag-score") +
    base_theme +
    theme(aspect.ratio = 0.7)
  
  res
}

#' Create heatmaps comparing single and dual vaccinated mice
#' 
#' Used for figures 3 and 4
create_exchange_heats <- function(df_in, ag_clmn, ttl, clr) {

  # Format data for heatmaps  
  dat <- df_in %>%
    mutate(subtype = fct_reorder(subtype, !!sym(ag_clmn), mean))
  
  # Create heatmaps
  res <- dat %>%
    ggplot(aes(mouse, subtype, fill = !!sym(ag_clmn))) +
    geom_tile() +
    labs(x = "vaccination group") +
    scale_fill_gradientn(colours = c("lightblue", "white", clr)) +
    scale_x_discrete(labels = ~ str_replace(.x, " day", "d")) +
    guides(fill = guide_colorbar(
      ticks = FALSE,
      title = ttl,
      barheight = unit(150, "pt"), barwidth = unit(10, "pt")
    )) +
    base_theme +
    theme(
      aspect.ratio = n_distinct(dat$subtype) / n_distinct(dat$mouse),
      plot.margin  = margin(0, 15, 0, 15),
      axis.title.y = element_blank(),
      axis.text.y  = element_text(size = ttl_pt2)
    )
  
  res
}

#' Create scatter plots comparing Ag-scores for dual vaccinated mice
#' 
#' Used for figure 3 
create_exchange_scatter <- function(df_in,
                                    x = c(Ag_3wk_score = "Ag-score (21 day)"),
                                    y = c(Ag_6wk_score = "Ag-score (42 day)"),
                                    clrs = lec_clrs) {
  
  # Set x y variables
  x_var <- names(x)
  y_var <- names(y)
  
  # Calculate correlation
  lab_dat <- df_in %>%
    group_by(subtype, exp) %>%
    summarize(
      cor_res = list(cor.test(!!sym(x_var), !!sym(y_var), method = "spearman")),
      cor     = map_dbl(cor_res, ~ .x$estimate[[1]]),
      p_val   = map_dbl(cor_res, ~ .x$p.value),
      n       = n_distinct(.cell_id),
      n       = label_comma()(n),
      .groups = "drop"
    ) %>%
    mutate(p_adj = p.adjust(p_val, method = "BH")) %>%
    rowwise() %>%
    mutate(
      p_lab = .format_pvalue(p_adj),
      r_lab = str_c("italic(r[s]) == ", round(cor, digits = 2)),
      n_lab = str_c("n = ", n)
    )
  
  # Create scatter plot
  res <- df_in %>%
    ggplot(aes(!!sym(x_var), !!sym(y_var), color = subtype)) +
    geom_point(size = 1.5) +
    geom_smooth(
      method    = "lm",
      linetype  = 2,
      linewidth = 0.5,
      color     = "black",
      se        = FALSE,
      formula   = "y ~ x"
    ) +
    geom_text(
      aes(x = -Inf, y = Inf, label = r_lab),
      data  = lab_dat,
      hjust = -0.1,
      vjust = 1.2,
      color = "black",
      size  = txt_pt2 / .pt,
      parse = TRUE
    ) +
    geom_text(
      aes(x = -Inf, y = Inf, label = n_lab),
      data  = lab_dat,
      hjust = -0.1,
      vjust = 2.7,
      color = "black",
      size  = txt_pt2 / .pt
    ) +
    facet_wrap(~ subtype, nrow = 1) +
    
    scale_color_manual(values = clrs) +
    labs(x = unname(x), y = unname(y)) +
    base_theme +
    theme(
      aspect.ratio = 1,
      legend.position = "none"
    )
  
  res
}

#' Create boxplots comparing RF module scores and Ag-classes 
#' 
#' Used for figure 3
create_module_boxes <- function(df_in, ag_class_clmn = "Ag_class_dual",
                                module_clmns, module_title,
                                clrs = lec_clrs, lvls, p_alt = "two.sided",
                                p_method = "correlation",
                                p_x = 1,
                                p_hjust = c(0, 0),
                                p_vjust = c(1.2, 1.9),
                                ...) {
  
  # Format plotting data
  # cell types filtered based on module_clmns names
  # module score is only plotted for the corresponding cell type
  # group all low cells together under single mouse, otherwise three boxplots
  # are shown
  typs <- str_remove(module_clmns, "_[^_]+$")
  
  dat <- df_in %>%
    pivot_longer(all_of(module_clmns)) %>%
    mutate(module_type = str_remove(name, "_[^_]+$")) %>%
    filter(subtype == module_type) %>%
    mutate(
      subtype = fct_relevel(subtype, typs),
      !!sym(ag_class_clmn) := fct_relevel(!!sym(ag_class_clmn), lvls)
    ) %>%
    arrange(!!sym(ag_class_clmn)) %>%
    mutate(
      mouse_class = fct_inorder(str_c(mouse, "-", subtype, "-", !!sym(ag_class_clmn)))
    )
  
  n_grps <- n_distinct(dat[[ag_class_clmn]])
             
  # Calculate p-values
  if (!identical(p_method, "correlation")) {
    p_dat <- dat %>%
      split(.$subtype) %>%
      map(~ {
        pairwise.wilcox.test(
          .x$value, .x[[ag_class_clmn]],
          alternative     = p_alt,
          p.adjust.method = "none"  # adjust below using all p-values
        ) %>%
          broom::tidy() %>%
          mutate(
            ymax = max(.x$value),
            ymin = min(.x$value)
          )
      }) %>%
      bind_rows(.id = "subtype") %>%
      filter(group1 == "double-high") %>%
      mutate(
        p_adj   = p.adjust(p.value, method = "BH"),
        ln_x    = match(group2, lvls),
        ln_xend = match(group1, lvls),
        y_fctr  = ((ymax - ymin) * 0.1),
        ln_y    = ymax + (y_fctr * ln_x),
        y       = ymax + (y_fctr * (ln_x + 1)),
        x       = ln_x + 0.5,
        ln_x    = ln_x + 0.1,
        ln_xend = ln_xend - 0.1
      ) %>%
      rowwise() %>%
      mutate(p_adj = .format_pvalue(p_adj)) %>%
      ungroup()
    
  # Calculate Spearman correlation
  } else {
    p_dat <- dat %>%
      group_by(subtype) %>%
      summarize(
        corr = list(tidy(cor.test(
          as.numeric(!!sym(ag_class_clmn)), value,
          method      = "spearman",
          continuity  = TRUE,
          exact       = FALSE,
          alternative = p_alt
        ))),
        .groups = "drop"
      ) %>%
      unnest(corr) %>%
      mutate(p_adj = p.adjust(p.value, method = "BH")) %>%
      rowwise() %>%
      mutate(
        p_adj = .format_pvalue(p_adj, show_decimal = Inf),
        p_adj = str_c("italic(p) == ", p_adj),
        r_lab = str_c("italic(r[s]) == ", round(estimate, 2))
      ) %>%
      ungroup()
  }
  
  # Format n labels for x axis
  n_dat <- dat %>%
    group_by(!!sym(ag_class_clmn), mouse_class, subtype) %>%
    summarize(n = n(), .groups = "drop") %>%
    mutate(
      n = str_c(!!sym(ag_class_clmn), "\nn = ", scales::label_comma()(n))
    )
  
  n_lab <- set_names(n_dat$n, n_dat$mouse_class)
  
  # Set p label position
  # by default center on plot
  p_x <- p_x %||% ((n_grps / 2) + 0.5)
  
  if (length(p_x) == 1)     p_x     <- rep(p_x, 2)
  if (length(p_hjust) == 1) p_hjust <- rep(p_hjust, 2)
  if (length(p_vjust) == 1) p_vjust <- rep(p_vjust, 2)
  
  # Create boxplots
  x_idx   <- 1:n_grps
  x_hjust <- (x_idx - min(x_idx)) / (max(x_idx) - min(x_idx))
  # x_hjust <- rev(x_hjust)
  
  res <- dat %>%
    ggplot(aes(mouse_class, value, fill = subtype)) +
    geom_boxplot(outlier.size = 0.1, ...) +
    
    facet_wrap(~ subtype, scales = "free") +
    scale_fill_manual(values = clrs) +
    scale_x_discrete(labels = n_lab) +
    scale_y_continuous(expand = expansion(c(0.05, 0.07))) +
    labs(y = module_title) +
    base_theme +
    theme(
      aspect.ratio    = 1.5,
      legend.position = "none",
      axis.text.x     = element_text(angle = 90, hjust = 1, vjust = x_hjust),
      # axis.text.x     = element_text(angle = 90, hjust = 1),
      axis.title.x    = element_blank()
    )
  
  if (!identical(p_method, "correlation")) {
    res <- res +
      geom_segment(
        aes(x = ln_x, xend = ln_xend, y = ln_y, yend = ln_y),
        data = p_dat,
        color = ln_col
      ) +
      
      geom_text(
        aes(x = x, y = y, label = p_adj),
        # direction = "x", seed = 42, force = 0.1, force_pull = 0, box.padding = 0.5,
        data = p_dat,
        parse = TRUE,
        size = 9 / .pt
      )
    
  } else {
    res <- res +
      geom_text(
        aes(x = !!(p_x[1]), y = Inf, fill = NULL, label = r_lab),
        size = 10 / .pt,
        data = p_dat,
        hjust = p_hjust[1],
        vjust = p_vjust[1],
        parse = TRUE
      ) +
      geom_text(
        aes(x = !!(p_x[2]), y = Inf, fill = NULL, label = p_adj),
        size = 10 / .pt,
        data = p_dat,
        hjust = p_hjust[2],
        vjust = p_vjust[2],
        parse = TRUE
      )
  }
  
  res
}

#' Create scatter plots comparing RF module scores and Ag-scores
#'
#' Used for figure 3
create_module_scatter <- function(df_in, ag_clmns, module_clmns, module_title,
                                  ag_class_clmn = "Ag_class_dual", clrs,
                                  label_x = -Inf, label_hjust = -0.1,
                                  label_start = -0.5, label_pad = 0.15) {
  
  # Format plotting data
  # cell types are filtered based on the module_clmns names
  typs <- str_remove(module_clmns, "_[^_]+$")
  
  dat <- df_in %>%
    pivot_longer(
      all_of(ag_clmns),
      names_to = "ag_clmn", values_to = "Ag-score"
    ) %>%
    pivot_longer(
      all_of(module_clmns),
      names_to = "module_name", values_to = module_title
    ) %>%
    mutate(module_type = str_remove(module_name, "_[^_]+$")) %>%
    filter(subtype == module_type) %>%
    mutate(
      subtype       = fct_relevel(subtype, typs),
      !!sym(ag_class_clmn) := fct_relevel(!!sym(ag_class_clmn), rev(names(clrs)))
    )
  
  # Format r label data
  lab_dat <- dat %>%
    group_by(ag_clmn) %>%
    mutate(
      min_score = min(`Ag-score`),
      max_score = max(`Ag-score`)
    ) %>%
    group_by(subtype, ag_clmn, min_score, max_score) %>%
    summarize(
      cor_res = list(cor.test(!!sym(module_title), `Ag-score`, method = "spearman")),
      cor     = map_dbl(cor_res, ~ .x$estimate[[1]]),
      p_val   = map_dbl(cor_res, ~ .x$p.value),
      n       = n_distinct(.cell_id),  # tricky to include commas with plotmath
      # n       = label_comma()(n_distinct(.cell_id)),
      .groups = "drop"
    ) %>%
    mutate(p_adj = p.adjust(p_val, method = "BH")) %>%
    rowwise() %>%
    mutate(
      r_lab = str_c("italic(r[s]) == ", round(cor, digits = 2)),
      p_lab = str_c("italic(p) == ", .format_pvalue(p_adj)),
      n_lab = str_c("n == ", n)
    )
  
  # Set label coordinates
  lab_dat <- lab_dat %>%
    mutate(
      y_fctr = (max_score - min_score) * label_pad,
      r_y    = max_score + (y_fctr * (label_start + 2)),
      p_y    = max_score + (y_fctr * (label_start + 1)),
      n_y    = max_score + (y_fctr * label_start)
    )
  
  # Labelling function
  lab_fn <- function(x) {
    modify <- x %in% names(ag_labs_2)
    
    x[modify] <- ag_labs_2[x[modify]]
    
    x
  }
  
  # Create scatter plots
  res <- dat %>%
    ggplot(aes(!!sym(module_title), `Ag-score`, color = !!sym(ag_class_clmn))) +
    geom_point(size = 0.7) +
    geom_smooth(
      method = "lm", formula = y ~ x,
      color = "black", linetype = 2, linewidth = 1, alpha = 0.1
    ) +
    facet_grid(
      ag_clmn ~ subtype,
      scales = "free", switch = "y",
      labeller = as_labeller(lab_fn)
    ) +
    scale_color_manual(values = clrs) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    base_theme +
    theme(
      aspect.ratio    = 1,
      strip.placement = "outside",
      legend.position = "bottom",
      legend.title    = element_blank(),
      axis.title.y    = element_blank()
    )
  
  # Add labels
  lab_args <- list(
    y = str_c(c("r", "p", "n"), "_y"),
    label = str_c(c("r", "p", "n"), "_lab")
  )
  
  lab_args %>%
    pwalk(~ {
      res <<- res +
        geom_text(
          aes(!!label_x, !!sym(.x), color = NULL, label = !!sym(.y)),
          data  = lab_dat,
          hjust = label_hjust,
          parse = TRUE,
          show.legend = FALSE
        )
    })
  
  res
}

#' Create boxplots comparing Ag-score for each LEC and DC subset
#' 
#' Used for figure 1
create_ag_boxes <- function(df_in, clrs, facet_vars = "subtype") {
  
  # Line data
  # make sure tm is numeric
  grp_vars <- unique(c(facet_vars, "tm"))
  
  df_in <- df_in %>%
    mutate(tm = as.numeric(as.character(tm)))
  
  ln_dat <- df_in %>%
    group_by(!!!syms(grp_vars)) %>%
    summarize(Ag_score = median(Ag_score), .groups = "drop")
  
  # Create boxplots
  x_facet <- dplyr::last(facet_vars)
  
  if (length(facet_vars) == 1) facet_vars <- c("", facet_vars)
  
  facet_vars <- str_c(facet_vars, collapse = "~")
  
  res <- df_in %>%
    ggplot(aes(tm, Ag_score, fill = !!sym(x_facet))) +
    geom_line(
      data = ln_dat,
      color = "grey85", linewidth = 0.5, linetype = 2
    ) +
    geom_boxplot(
      aes(group = mouse),
      outlier.size = 0.2,
      width = 3,
      key_glyph = draw_key_point
    ) +
    
    guides(
      fill     = guide_legend(override.aes = list(size = 4, shape = 22), order = 1),
      linetype = guide_legend(override.aes = list(color = "black"), title = NULL)
    ) +
    facet_grid(as.formula(facet_vars)) +
    
    scale_fill_manual(values = clrs) +
    labs(x = "days post vaccination", y = "Ag-score") +
    base_theme +
    theme(
      aspect.ratio    = 0.9,
      legend.position = "none",
      plot.margin     = margin(0, 15, 0, 15)
    )
  
  res
}

#' Create bargraphs summarizing the fraction of Ag-competent cells
#' 
#' Used for figure 5
create_chikv_class_bars <- function(df_in, x = "treatment", clrs = treat_clrs,
                                    rep_clmn = "rep", ttl, n_p = NULL) {
  
  # Set colors
  grps <- unique(df_in[[x]])
  clrs <- clrs[names(clrs) %in% grps]
  
  # Format data for bargraphs
  df_in <- df_in %>%
    mutate(
      !!sym(x) := fct_relevel(!!sym(x), names(clrs)),
      x_clmn    = str_c(subtype, "-", !!sym(x))
    ) %>%
    arrange(!!sym(x)) %>%
    mutate(x_clmn = fct_inorder(x_clmn))
  
  dat <- df_in %>%
    group_by(!!!syms(c("x_clmn", x, "subtype", rep_clmn))) %>%
    summarize(
      frac_comp = sum(pred_grp == "Ag-competent") / n(),
      .groups = "drop"
    )
  
  # Perform fisher exact test
  p_dat <- df_in %>%
    group_by(x_clmn, subtype, !!sym(x)) %>%
    mutate(n = n()) %>%
    group_by(x_clmn, subtype, !!sym(x), n) %>%
    summarize(
      n_comp = sum(pred_grp == "Ag-competent"),
      .groups = "drop"
    ) %>%
    group_by(subtype) %>%
    mutate(
      frac_comp = n_comp / n,
      tot_comp  = sum(n_comp),
      tot       = sum(n)
    ) %>%
    ungroup()
  
  p_vals <- p_dat %>%
    rowwise() %>%
    mutate(p = .calc_fisher(
      x = n_comp,
      k = n,
      m = tot_comp,
      n = tot - tot_comp
    )) %>%
    ungroup() %>%
    distinct(subtype, p) %>%    # there should be a single p-value per cell type
    mutate(
      p_adj = p.adjust(
        p,
        method = "BH",
        n = n_p %||% length(p)  # specifying n_p is useful when only plotting a 
      )                         # single panel and showing the other panels in
    ) %>%                       # other figures
    rowwise() %>%
    mutate(p_adj = str_c("italic(p) == ", .format_pvalue(p_adj))) %>%
    ungroup()
  
  if (nrow(p_vals) > n_distinct(p_vals$subtype)) {
    cli_abort("Malformed p-value data, check calculations")
  }
  
  # Format x-axis labels
  x_labs <- p_dat %>%
    mutate(
      n = str_c("n = ", label_comma()(n)),
      n = str_c(!!sym(x), "\n", n)
    )
  
  x_labs <- set_names(x_labs$n, x_labs$x_clmn)
  
  # Create bargraphs
  plt_aes <- aes(
    x    = x_clmn,
    y    = frac_comp,
    fill = !!sym(x)
  )
  
  if (!is.null(rep_clmn)) plt_aes$alpha <- sym(rep_clmn)
  
  res <- dat %>%
    ggplot(plt_aes) +
    geom_col(position = position_dodge2(padding = 0.15)) +
    
    geom_text(
      aes(1.5, Inf, fill = NULL, alpha = NULL, label = p_adj),
      data  = p_vals,
      parse = TRUE,
      vjust = 1.3,
      size  = 12 / .pt
    ) +
    
    facet_wrap(~ subtype, scales = "free_x") +
    scale_fill_manual(values = clrs) +
    scale_y_continuous(breaks = c(0, 0.5, 1), limits = c(0, 1)) +
    scale_x_discrete(labels = x_labs) +
    guides(alpha = "none") +
    labs(y = ttl) +
    base_theme +
    theme(
      aspect.ratio    = 1.5,
      legend.position = "none",
      axis.title.x    = element_blank(),
      axis.text.x     = element_text(size = ttl_pt2)
    )
  
  if (!is.null(rep_clmn)) {
    al_vals <- rep(1, n_distinct(df_in[[rep_clmn]]))
    
    res <- res +
      scale_alpha_manual(values = al_vals)
  }
  
  res
}

#' Create boxplots comparing Ag module scores for mock vs CHIKV
#' 
#' Used for figure 5
create_chikv_module_boxes <- function(df_in, x = "treatment", module_clmns,
                                      clrs = treat_clrs, ttl, n_p = NULL, ...) {
  
  # Format colors
  grps <- unique(df_in[[x]])
  clrs <- clrs[names(clrs) %in% grps]
  
  # Format data for plotting
  if (length(grps) > 2) {
    cli_warn("More than two groups present in `x`, using the first two")
  }
  
  dat <- df_in %>%
    pivot_longer(all_of(module_clmns)) %>%
    mutate(
      module_subtype = str_remove(name, "_[^_]+$"),
      module_type    = str_remove(name, "^.+_"),
      module_type    = str_c("Ag-", module_type),
      !!sym(x)      := fct_relevel(!!sym(x), names(clrs)),
      x_clmn         = str_c(subtype, "-", !!sym(x))
    ) %>%
    filter(subtype == module_subtype) %>%
    arrange(!!sym(x)) %>%
    mutate(x_clmn = fct_inorder(x_clmn))
  
  # Calculate p-values
  p_dat <- dat %>%
    group_by(subtype, module_type) %>%
    summarize(
      p = wilcox.test(
        value[!!sym(x) == grps[1]],
        value[!!sym(x) == grps[2]]
      )$p.value,
      .groups = "drop"
    ) %>%
    mutate(
      p_adj = p.adjust(
        p,
        method = "BH",
        n = n_p %||% length(p)  # specifying n_p is useful when only plotting a 
      )                         # single panel and showing the other panels in
    ) %>%                       # other figures
    rowwise() %>%
    mutate(p_adj = str_c("italic(p) == ", .format_pvalue(p_adj))) %>%
    ungroup()
    
  # Format x-axis labels
  x_labs <- dat %>%
    group_by(!!sym(x), x_clmn) %>%
    summarize(n = n_distinct(.cell_id), .groups = "drop") %>%
    mutate(
      n = str_c("n = ", label_comma()(n)),
      n = str_c(!!sym(x), "\n", n)
    )
  
  x_labs <- set_names(x_labs$n, x_labs$x_clmn)
  
  # Create boxplots
  res <- dat %>%
    ggplot(aes(x_clmn, value, fill = !!sym(x), alpha = rep)) +
    geom_boxplot(outlier.size = 0.5, ...) +
    
    geom_text(
      aes(1.5, Inf, fill = NULL, alpha = NULL, label = p_adj),
      data  = p_dat,
      parse = TRUE,
      vjust = 1.3,
      size  = 12 / .pt
    ) +
    
    facet_grid(module_type ~ subtype, scales = "free") +
    scale_fill_manual(values = clrs) +
    scale_alpha_manual(values = rep(1, n_distinct(df_in$rep))) +
    scale_y_continuous(expand = expansion(c(0.05, 0.1))) +
    scale_x_discrete(labels = x_labs) +
    guides(alpha = "none") +
    labs(y = ttl) +
    base_theme +
    theme(
      aspect.ratio    = 1.5,
      legend.position = "none",
      axis.title.x    = element_blank(),
      axis.text.x     = element_text(size = ttl_pt2)
    )
  
  res
}
```

```{r "references"}
# LEC reference
ref_lec <- ref_immgen[, grepl("^Endothelial cells", colnames(ref_immgen))]
ref_lec <- ref_lec[rownames(ref_lec) %in% rownames(ref_LEC_xiang), ]

colnames(ref_lec) <- colnames(ref_lec) %>%
  str_replace("Endothelial cells \\(BEC\\)", "BEC")

ref_LEC_xiang <- ref_LEC_xiang[rownames(ref_lec), ]

stopifnot(identical(rownames(ref_LEC_xiang), rownames(ref_lec)))

ref_lec <- cbind(ref_LEC_xiang, ref_lec)
ref_lec <- ref_lec[, !grepl("^Endothelial cells", colnames(ref_lec))]

# Fibroblast/stromal cell reference
ref_frc <- ref_immgen[, grepl("^Fibroblast", colnames(ref_immgen))]
ref_frc <- ref_frc[rownames(ref_frc) %in% rownames(ref_lymphnodestromal), ]

ref_lymphnodestromal <- ref_lymphnodestromal[rownames(ref_frc), ]

stopifnot(identical(rownames(ref_lymphnodestromal), rownames(ref_frc)))

ref_frc <- cbind(ref_lymphnodestromal, ref_frc)

# Combined Immgen/DC reference
# * DC reference from Brown et al. Cell 2019
rownames(ref_mousespleenDC) <- str_to_title(rownames(ref_mousespleenDC))

shared_gns    <- intersect(rownames(ref_immgen), rownames(ref_mousespleenDC))
ref_immgen_dc <- cbind(ref_immgen[shared_gns, ], ref_mousespleenDC[shared_gns, ])
```

```{r "Ag normalization factors", eval = create_sobjs}
# Calculate total UMI counts for each library using raw matrix
# do not use filtered matrix since a poor quality library would have a high
# fraction of counts in empty drops that would not be taken into account when
# normalizing data
norm_file <- here("results/norm_factors.tsv")

.calc_lib_size <- function(mat_path, mat_file, factor = 1e6) {
  mat <- here(mat_path, "outs", mat_file)
  res <- Read10X(mat)
  res <- sum(res$`Antibody Capture`) / factor
  
  res
}

if (!file.exists(norm_file)) {
  plan(strategy = multisession, workers = 6)
  
  norm_info <- sam_info %>%
    future_pmap_dfr(~ {
      args    <- list(...)
      dat_dir <- here(params$res_dir[[args$experiment]], args$directory)
      raw     <- .calc_lib_size(dat_dir, "raw_feature_bc_matrix")
      filt    <- .calc_lib_size(dat_dir, "filtered_feature_bc_matrix")
      
      tibble(
        !!!args,
        Ag_counts_raw      = raw,
        Ag_counts_filtered = filt
      )
    })
  
  plan(strategy = sequential)
  
  norm_info %>%
    write_tsv(norm_file)
}

norm_info <- read_tsv(norm_file)

norm_factors <- set_names(
  norm_info$Ag_counts_raw,
  norm_info$directory
)
```

```{r "create seurat objects", eval = create_sobjs}
# Create objects
sobjs <- sam_info %>%
  pmap(~ {
    args <- list(...)
    
    mat <- args$directory %>%
      here(params$res_dir[args$experiment], .) %>%
      here("outs/filtered_feature_bc_matrix")
    
    mat %>%
      create_sobj(
        proj_name = args$sample,
        gene_min  = 250,
        gene_max  = 5000,
        mito_max  = 20
      ) %>%
      mutate_meta(
        mutate,
        exp         = args$experiment,
        directory   = args$directory,
        sample      = args$sample,
        tm          = args$tm,
        mouse       = args$mouse,
        vaccination = args$vaccination,
        cell_sort   = args$cell_sort
      ) %>%
      norm_sobj()
  })

names(sobjs) <- sam_info$directory

sobjs %>%
  qsave(here(params$so_dir, "raw_sobjs.qs"))

# Save sample stats
stat_clmns <- c(
  "nCount_RNA", "nFeature_RNA",
  "nCount_ADT", "nFeature_ADT",
  "pct_mito"
)

sobjs %>%
  map_dfr(~ {
    .x@meta.data %>%
      as_tibble(rownames = "cell_id") %>%
      group_by(directory, qc_class) %>%
      mutate(across(all_of(stat_clmns), median, .names = "med_{.col}")) %>%
      group_by(qc_class, !!!syms(str_c("med_", stat_clmns)), .add = TRUE) %>%
      summarize(n = n(), .groups = "drop")
  }) %>%
  left_join(norm_info, by = "directory") %>%
  rename()
  mutate(Ag_counts_frac = Ag_counts_filtered / Ag_counts_raw) %>%
  write_tsv(here("results/sample_stats.tsv"))

# Save filtered objects
sobjs <- sobjs %>%
  map(subset, qc_class == "pass")

sobjs %>%
  qsave(here(params$so_dir, "sobjs.qs"))
```

```{r "CD45- samples", eval = create_sobjs}
# Merge objects
neg_idx    <- sam_info$cell_sort == "CD45neg"
sobjs      <- sobjs[neg_idx]
cd45neg_so <- merge(sobjs[[1]], sobjs[-1])

rm(sobjs)
gc()

cd45neg_so <- cd45neg_so %>%
  integrate_sobjs(group_vars = "mouse")

# Broad cell type annotations
clst_clmn <- "RNA_snn_res.5"

cd45neg_so <- cd45neg_so %>%
  mutate_meta(dplyr::select, -any_of(c("UMAP_1", "UMAP_2", "type", "r"))) %>%
  clustify(
    ref_mat     = ref_immgen,
    cluster_col = clst_clmn,
    n_genes     = 2000,
    threshold   = 0.55
  ) %>%
  mutate_meta(mutate, rough_type = type)

# Adjust B and T cells
# * adjust stem cells based on Cd79a, these appear to be B cells
cd45neg_so <- cd45neg_so %>%
  classify_markers(
    feats    = c("Ptprc", "Cd79a", "Cd19"),
    filt     = Ptprc > 1 & Cd79a > 1.5 & Cd19 > 0.5,
    type     = "B cells",
    clst_col = clst_clmn,
    type_col = "rough_type"
  ) %>%
  classify_markers(
    feats    = c("Ptprc", "Cd3e"),
    filt     = Ptprc > 1.25 & Cd3e > 0.75,
    type     = "T cells",
    clst_col = clst_clmn,
    type_col = "rough_type"
  )

# Adjust BECs
cd45neg_so <- cd45neg_so %>%
  classify_markers(
    feats    = c("Ptprc", "Pdpn", "Pecam1"),
    filt     = Ptprc < 0.5 & Pdpn < 0.5 & Pecam1 > 2,
    type     = "Endothelial cells",
    clst_col = clst_clmn,
    type_col = "rough_type"
  )

# Adjust epithelial cells based on E-cadherin and cytokeratin-8
# * some unassigned cells appear to be epithelial cells
cd45neg_so <- cd45neg_so %>%
  classify_markers(
    feats    = c("Cdh1", "Krt8"),
    filt     = Cdh1 > 0.75 & Krt8 > 0.75,
    type     = "Epithelial cells",
    clst_col = clst_clmn,
    type_col = "rough_type"
  )

# Adjust cell type labels
# * fibroblasts also include DN stromal cells
cd45neg_so <- cd45neg_so %>%
  mutate_meta(
    mutate,
    cell_type = str_remove(rough_type, " \\(.+$"),
    cell_type = recode(cell_type, `Stromal cells` = "Fibroblasts"),
    subtype   = cell_type
  )
```

```{r "CD45+ samples", eval = create_sobjs}
# NEED TO:
# 1. check neutrophils, they show expression of Cd14, which is not expected

# Load objects
sobjs <- qread(here(params$so_dir, "sobjs.qs"))

# Merge objects
pos_idx    <- sam_info$cell_sort == "CD45pos"
sobjs      <- sobjs[pos_idx]
cd45pos_so <- merge(sobjs[[1]], sobjs[-1])

rm(sobjs)
gc()

cd45pos_so <- cd45pos_so %>%
  integrate_sobjs(group_vars = "mouse")

# Cell type annotations
# assign broad cell types and DC subset using a combined reference
clst_clmn <- "RNA_snn_res.5"

cd45pos_so <- cd45pos_so %>%
  mutate_meta(dplyr::select, -any_of(c("UMAP_1", "UMAP_2", "type", "r"))) %>%
  clustify(
    ref_mat     = ref_immgen_dc,
    cluster_col = clst_clmn,
    n_genes     = 2000
  ) %>%
  mutate_meta(mutate, rough_type = type)

# NK cells
# ILCs show high expression of Nkg7, label these are NK cells
cd45pos_so <- cd45pos_so %>%
  classify_markers(
    feats    = c("Ptprc", "Nkg7"),
    filt     = Ptprc > 1.25 & Nkg7 > 2,
    type     = "NK cells",
    clst_col = clst_clmn,
    type_col = "rough_type"
  )

# T cells
cd45pos_so <- cd45pos_so %>%
  classify_markers(
    feats    = c("Ptprc", "Cd3e", "Cd19"),
    filt     = Ptprc > 1.25 & Cd3e > 1 & Cd19 < 0.3,
    type     = "T cells",
    clst_col = clst_clmn,
    type_col = "rough_type"
  )

# CD45- cells
# a cluster of B cells show poor expression of Ptprc, label as unassigned
cd45pos_so <- cd45pos_so %>%
  classify_markers(
    feats    = "Ptprc",
    filt     = Ptprc < 0.75,
    type     = "unassigned",
    clst_col = clst_clmn,
    type_col = "rough_type"
  )

# cDC2 Tbet+
# evidence is lacking that these are actually Tbet+ cells, label as cDC2s
cd45pos_so <- cd45pos_so %>%
  mutate_meta(
    mutate,
    rough_type = ifelse(
      rough_type %in% c("cDC2 Mixed", "cDC2 Tbet+"),
      "cDC2", rough_type
    ),
    cell_type = str_remove(rough_type, " \\(.+$"),
    subtype   = cell_type,
    cell_type = ifelse(grepl("DC", cell_type), "DC", cell_type)
  )
```

```{r "LEC object", eval = create_sobjs}
# Re-cluster and integrate endothelial cells
lec_so <- cd45neg_so %>%
  subset(cell_type == "Endothelial cells") %>%
  integrate_sobjs(group_vars = "mouse")

# Annotate LEC subsets
.annotate_lecs <- function(so_in, clst_clmn) {
  res <- so_in %>%
    mutate_meta(dplyr::select, -any_of(c("UMAP_1", "UMAP_2", "type", "r"))) %>%
    clustify(
      ref_mat     = ref_lec,
      cluster_col = clst_clmn,
      n_genes     = 2000,
      threshold   = 0.53
    ) %>%
    mutate_meta(
      mutate,
      subtype   = type,
      cell_type = ifelse(subtype == "BEC", subtype, "LEC")
    )
  
  # Use higher correlation cutoff for LECs
  res <- res %>%
    mutate_meta(
      mutate,
      subtype = ifelse(
        cell_type == "LEC" & r < 0.75,
        "unassigned",
        subtype
      )
    )
  
  res
}

clst_clmn <- "RNA_snn_res.10"

lec_so <- .annotate_lecs(lec_so, clst_clmn)

# Identify contaminating B and T cells
# label other CD45+ cells as unassigned
lec_so <- lec_so %>%
  classify_markers(
    feats    = "Ptprc",
    filt     = Ptprc > 0.4,
    type     = "unassigned",
    clst_col = clst_clmn,
    type_col = "cell_type"
  ) %>%
  classify_markers(
    feats    = c("Ptprc", "Cd79a"),
    filt     = Ptprc > 0.4 & Cd79a > 0.25,
    type     = "B cells",
    clst_col = clst_clmn,
    type_col = "cell_type"
  ) %>%
  classify_markers(
    feats    = c("Ptprc", "Cd3e"),
    filt     = Ptprc > 0.4 & Cd3e > 0.25,
    type     = "T cells",
    clst_col = clst_clmn,
    type_col = "cell_type"
  )

# Adjust cell type labels
# this adds BEC label and fixes B and T cell labels
typ_data <- FetchData(lec_so, "cell_type")

cd45neg_so <- cd45neg_so %>%
  AddMetaData(typ_data, col.name = "new_type") %>%
  mutate_meta(~ {
    .x %>%
      mutate(
        cell_type = ifelse(!is.na(new_type), new_type, cell_type),
        subtype   = ifelse(!is.na(new_type), new_type, subtype)
      ) %>%
      dplyr::select(-new_type)
  })

# Re-annotate LECs after removing B and T cells
lec_so <- lec_so %>%
  subset(cell_type %in% c("LEC", "BEC")) %>%
  integrate_sobjs(group_vars = "mouse") %>%
  .annotate_lecs(clst_clmn)

# Adjust cell type labels
typ_data <- lec_so %>%
  FetchData(c("cell_type", "subtype"))

cd45neg_so <- cd45neg_so %>%
  AddMetaData(typ_data, col.name = c("new_type", "new_subtype")) %>%
  mutate_meta(~ {
    .x %>%
      mutate(
        cell_type = ifelse(!is.na(new_type),    new_type,    cell_type),
        subtype   = ifelse(!is.na(new_subtype), new_subtype, subtype)
      ) %>%
      dplyr::select(-c(new_type, new_subtype))
  })
```

```{r "FRC object", eval = create_sobjs}
# Re-cluster and integrate fibroblasts
frc_so <- cd45neg_so %>%
  subset(cell_type == "Fibroblasts")

frc_so <- frc_so %>%
  integrate_sobjs(group_vars = "mouse")

# Annotate FRC subsets
.annotate_frcs <- function(so_in, clst_clmn) {
  res <- so_in %>%
    mutate_meta(dplyr::select, -any_of(c("UMAP_1", "UMAP_2", "type", "r"))) %>%
    clustify(
      ref_mat     = ref_frc,
      cluster_col = clst_clmn,
      n_genes     = 2000
    ) %>%
    mutate_meta(mutate, subtype = type)
  
  res
}

clst_clmn <- "RNA_snn_res.10"

frc_so <- .annotate_frcs(frc_so, clst_clmn)

# Identify contaminating B and T cells
frc_so <- frc_so %>%
  classify_markers(
    feats    = c("Ptprc", "Cd79a"),
    filt     = Ptprc > 0.5 & Cd79a > 0.5,
    type     = "B cells",
    clst_col = clst_clmn,
    type_col = "cell_type"
  ) %>%
  classify_markers(
    feats    = c("Ptprc", "Cd3e"),
    filt     = Ptprc > 0.5 & Cd3e > 0.5,
    type     = "T cells",
    clst_col = clst_clmn,
    type_col = "cell_type"
  )

# Adjust cell type and subtype labels
typ_data <- frc_so %>%
  FetchData("cell_type")

cd45neg_so <- cd45neg_so %>%
  AddMetaData(typ_data, col.name = "new_type") %>%
  mutate_meta(~ {
    .x %>%
      mutate(
        cell_type = ifelse(!is.na(new_type), new_type, cell_type),
        subtype   = ifelse(!is.na(new_type), new_type, subtype)
      ) %>%
      dplyr::select(-new_type)
  })

# Re-annotate fibroblasts after removing B and T cells
frc_so <- frc_so %>%
  subset(cell_type == "Fibroblasts") %>%
  integrate_sobjs(group_vars = "mouse") %>%
  .annotate_frcs(clst_clmn = "RNA_snn_res.5")

# Add FRC subset labels
typ_data <- frc_so %>%
  FetchData(c("cell_type", "subtype"))

cd45neg_so <- cd45neg_so %>%
  AddMetaData(typ_data, col.name = c("new_type", "new_subtype")) %>%
  mutate_meta(~ {
    .x %>%
      mutate(
         cell_type = ifelse(!is.na(new_type),    new_type,    cell_type),
         subtype   = ifelse(!is.na(new_subtype), new_subtype, subtype)
      ) %>%
      dplyr::select(-c(new_type, new_subtype))
  })
```

```{r "DC object", eval = create_sobjs}
# Re-cluster and integrate DCs
# DC subsets were annotated with broad cell types
dc_so <- cd45pos_so %>%
  subset(cell_type == "DC")

dc_so <- dc_so %>%
  integrate_sobjs(group_vars = "mouse")
```

```{r "Ag scores", eval = create_sobjs}
# Calculate Ag scores for CD45- object
cd45neg_so <- cd45neg_so %>%
  AddMetaData(FetchData(., ags, slot = "counts"), col.name = ags)

cd45neg_so <- cd45neg_so %>%
  format_ag_data(
    norm_factors = norm_factors,
    ag_key = ag_key
  )

# Calculate Ag scores for CD45+ object
cd45pos_so <- cd45pos_so %>%
  AddMetaData(FetchData(., ags, slot = "counts"), col.name = ags)

cd45pos_so <- cd45pos_so %>%
  format_ag_data(
    norm_factors = norm_factors,
    ag_key = ag_key
  )

# Add Ag scores to CD45- objects
ag_clmns <- c(
  "ovalbumin", "Ag_3wk", "Ag_6wk",
  "ovalbumin_score", "Ag_3wk_score", "Ag_6wk_score",
  "Ag_score", "Ag_score_3", "Ag_score_6",
  "tm_3", "tm_6"
)

ag_data <- cd45neg_so %>%
  FetchData(ag_clmns)

lec_so <- lec_so %>%
  AddMetaData(ag_data)

frc_so <- frc_so %>%
  AddMetaData(ag_data)

# Add Ag scores to CD45+ objects
ag_data <- cd45pos_so %>%
  FetchData(ag_clmns)

dc_so <- dc_so %>%
  AddMetaData(ag_data)
```

```{r "save seurat objects", eval = create_sobjs}
# Save meta.data for all cells combined
all_meta <- list(cd45neg_so, cd45pos_so) %>%
  map_dfr(~ .x@meta.data) %>%
  as_tibble(rownames = ".cell_id")

all_meta %>%
  write_tsv(here(params$so_dir, "all_meta.tsv.gz"))

# Save objects
save_objs(cd45neg_so, ob_dir = params$so_dir)
save_objs(cd45pos_so, ob_dir = params$so_dir)
save_objs(lec_so,     ob_dir = params$so_dir)
save_objs(frc_so,     ob_dir = params$so_dir)
save_objs(dc_so,      ob_dir = params$so_dir)

rm(all_meta, cd45neg_so, cd45pos_so, lec_so, frc_so, dc_so)
```

```{r "load seurat objects"}
# Load objects
all_meta <- read_tsv(here(params$so_dir, "all_meta.tsv.gz"))
lec_so   <- qread(here(params$so_dir, "lec_so.qs"))
dc_so    <- qread(here(params$so_dir, "dc_so.qs"))

# frc_so   <- qread(here(params$so_dir, "frc_so.qs"))
# cd45neg_so <- qread(here(params$so_dir, "cd45neg_so.qs"))
# cd45pos_so <- qread(here(params$so_dir, "cd45pos_so.qs"))

# Load cell type markers
xlsx_path <- here(params$ref_dir, "cell-type-markers.xlsx")
markers   <- xlsx::read.xlsx(xlsx_path, sheetName = "markers")

markers <- markers %>%
  mutate(
    cell_type = fct_inorder(cell_type),
    subtype   = fct_inorder(subtype)
  ) %>%
  group_by(gene) %>%
  summarize(
    label      = str_c(cell_type, collapse = "/"),
    cell_type  = dplyr::first(cell_type),
    annotation = dplyr::first(subtype),
    .groups    = "drop"
  ) %>%
  arrange(cell_type, annotation) %>%
  mutate(
    gene  = fct_inorder(gene),
    label = fct_inorder(label)
  )
```

```{r "theme"}
# Plot levels
mice <- unique(sam_info$mouse)
tms  <- as.character(na.omit(unique(sam_info$tm)))
vacs <- unique(sam_info$vaccination)

# ggplot2 themes
txt_pt1  <- 10
txt_pt2  <- 12
ttl_pt1  <- 14
ttl_pt2  <- 16
ln_pt    <- 0.5
ln_col   <- "grey85"
cir_size <- 3.5
sqr_size <- 4

txt_theme_1 <- theme(
  strip.text   = element_text(size = ttl_pt2),
  legend.text  = element_text(size = ttl_pt1),
  axis.title   = element_text(size = ttl_pt2),
  axis.text    = element_text(size = txt_pt2, color = "black"),
  legend.title = element_text(size = ttl_pt2),
)

line_theme <- theme(
  axis.line.x  = element_line(linewidth = ln_pt, color = ln_col),
  axis.line.y  = element_line(linewidth = ln_pt, color = ln_col),
  axis.ticks.x = element_line(linewidth = ln_pt, color = ln_col),
  axis.ticks.y = element_line(linewidth = ln_pt, color = ln_col)
)

base_theme <- djvdj_theme() +
  theme(
    plot.title       = element_text(face = "plain", size = ttl_pt2),
    strip.background = element_blank(),
    strip.text       = element_text(face = "plain")
  ) +
  txt_theme_1

umap_theme <- base_theme +
  theme(
    aspect.ratio = 0.9,
    panel.border = element_rect(color = NA, fill = NA),
    axis.title   = element_blank(),
    axis.line.x  = element_blank(),
    axis.line.y  = element_blank(),
    axis.ticks.x = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text    = element_blank()
  )

umap_theme_2 <- umap_theme %+replace%
  theme(panel.border = element_rect(colour = ln_col, linewidth = ln_pt, fill = NA))

fade_0 <- "#FAFAFA"
fade_1 <- "#F0F0F0"
fade_2 <- "#D9D9D9"

# Base colors
# * Okabe Ito color palettes
ito_cols <- c(
  palette_OkabeIto[c(1:3, 5)], "#D7301F",
  palette_OkabeIto[c(4, 6)],   "#6A51A3",
  palette_OkabeIto[7],         "#875C04",
  "#065D43", "#F7AF34", "#00446E"
)

clrs <- ito_cols %>%
  interp_colors(n = 15) %>%
  spruce_colors(
    difference = 15,
    filter = "deutan",
    order = TRUE
  )

# Cell type colors
typs <- unique(all_meta$cell_type)
typs <- typs[typs != "unassigned"]

other_clr <- "grey80"

typ_clrs <- clrs %>%
  assign_colors(typs, filter = "colorblind")

set_typs <- c(
  "LEC", "DC", "Fibroblasts",
  "BEC", "B cells", "unassigned"
)

typ_clrs[set_typs] <- c(
  "#E7301F", "#56B4E9", "#BF73A6",
  "#0072B2", "#F0E442", other_clr
)

typ_clrs <- typ_clrs %>%
  spruce_colors(
    difference = 15,
    exclude_colors = set_typs
  )

# LEC colors
lec_clrs <- lec_so@meta.data %>%
  pull(subtype) %>%
  unique()

lec_clrs <- typ_clrs %>%
  assign_colors(lec_clrs)

set_typs <- c(
  "cLEC", "fLEC", "Ptx3_LEC",
  "Collecting", "tzLEC", "BEC",
  "unassigned"
)

lec_clrs[set_typs] <- c(
  "#E69F00", "#56B4E9", "#F0E442",
  "#E7301F", "#000000", typ_clrs[["BEC"]],
  other_clr
)

lec_clrs <- lec_clrs %>%
  spruce_colors(
    difference = 15,
    exclude_colors = set_typs,
    filter = "deutan"
  )

# DC colors
dc_clrs <- dc_so@meta.data %>%
  pull(subtype) %>%
  unique()

dc_clrs <- typ_clrs[!typ_clrs %in% lec_clrs] %>%
  assign_colors(dc_clrs)

dc_clrs["unassigned"] <- other_clr

dc_clrs <- dc_clrs %>%
  spruce_colors(
    difference = 22,
    adjust_colors = "cDC1"
  )

# Mouse colors
m_clrs <- set_names(clrs[seq_along(mice)], mice)

# Tm colors
tm_clrs <- set_names(
  clrs[seq_along(tms)],
  unname(tms)
)

tm_clrs[3] <- "#6A51A3"
tm_clrs[4] <- "#E7301F"

# Sample keys
mouse_key <- lec_so@meta.data %>%
  distinct(mouse, orig.ident)

mouse_key <- set_names(
  mouse_key$mouse, mouse_key$orig.ident
)

tm_key <- sam_info %>%
  distinct(tm, mouse)

tm_key <- set_names(tm_key$tm, tm_key$mouse)
```



```{r "OLD Ag class", eval = FALSE}
# Format data for clustering based on Ag scores
# * The strategy is to identify Ag-high/low cells for all datasets together
# * One might expect that changes in gene expression will depend on absolute
#   levels of Ag taken up by the cell, and so it would make the most sense to
#   use a consistent cutoff for Ag-high cells across all cell types and
#   datasets.
# * This should work since Ag scores are corrected based on background T/B
#   signal.
all_meta <- list(cd45neg_so, cd45pos_so) %>%
  map_dfr(~ .x@meta.data) %>%
  as_tibble(rownames = ".cell_id")

clust_dat <- all_meta %>%
  filter(Ag_score > 0)

# Downsample samples
# * This ensures that larger samples do not skew the clustering
sam_size <- min(table(clust_dat$mouse))

set.seed(42)
  
clust_dat <- clust_dat %>%
  group_by(mouse) %>%
  slice_sample(n = sam_size) %>%
  ungroup()

# Identify threshold for Ag-high/low cells
ag_threshold <- clust_dat %>%
  mutate(km_class = kmeans(Ag_score, centers = 2)$cluster) %>%
  group_by(km_class) %>%
  summarize(Ag_score = max(Ag_score)) %>%
  pull(Ag_score) %>%
  min()

# Set Ag-high/low classifications
cd45neg_so <- cd45neg_so %>%
  mutate_meta(
    mutate,
    Ag_class = ifelse(
      Ag_score >= ag_threshold,
      "high", "low"
    )
  )

cd45pos_so <- cd45pos_so %>%
  mutate_meta(
    mutate,
    Ag_class = ifelse(
      Ag_score >= ag_threshold,
      "high", "low"
    )
  )
```

```{r "MARKERS TO CHECK ANNOTATIONS", eval = FALSE}
marks <- c(
  "type",
  "Ptprc", "Cd3e", "Cd19", "Nkg7", "Ccr7",
  "Adgre1", "Itgam",                  # Mon
  "Cd14", "Fcgr3", "Fcgr1",           # Mac            
  "Xcr1", "Cd8a", "Cd24a", "Clec9a", "Itgae", "Irf8",   # cDC1
  "Cd4", "Sirpa", "Itgam",                              # cDC2
  "Tbx21", "Dtx1", "Ccr6",                              # cDC2 Tbet+
  "Cx3cr1", "Csf1r", "Clec12a", "Lyz2",                 # cDC2 Tbet-
  "Ly6g", "Itgam"                                       # neutrophils
)

# marks %>%
# c("Dtx1", "Ccr6", "Runx1", "Runx2", "Runx3", "Srebf1", "Srebf2") %>%
# c("Sirpa", "Cd4", "Itgam", "Clec12a", "Cx3cr1", "Cd14", "Il1a", "Lyz2", "Csf1r") %>%
# "Cd3e" %>%
# "Sirpa" %>%
# "type" %>%
# "Cd19" %>%
# "Cd79a" %>%
# "adt_ABPS4" %>%
# "adt_ovalbumin" %>%
"subtype" %>%
  map(~ {
    cd45pos_so %>%
      mutate_meta(mutate, rough_type = str_remove(rough_type, " \\(.+$")) %>%
      mutate_meta(mutate, type = str_remove(type, " \\(.+$")) %>%
      plot_scatter(
        .x,
        "hUMAP_1", "hUMAP_2",
        group_col = "mouse",
        panel_nrow = 1, size = 0.1, top = Inf,
        outline = TRUE,
        plot_colors = c("lightblue", "white", "red"),
        # plot_colors = c(unassigned = "black")
      ) +
      theme(aspect.ratio = 0.9)
  }) %>%
  plot_grid(plotlist = .)

marks[-1] %>%
  map(~ {
    cd45pos_so %>%
      mutate_meta(mutate, type = str_remove(type, " \\(.+$")) %>%
      plot_violin(
        .x,
        cluster_col = "type",
        # group_col = "mouse",
        panel_nrow = 1,
        method = "boxplot"
      )
  }) %>%
  plot_grid(plotlist = .)
```
